// @pankod/refine-inferencer@1.6.1/dist/esm/antd.js
import ne from"react";import{useRouterContext as qr}from"@pankod/refine-core";import*as Br from"@pankod/refine-antd";import V from"react";import{useResource as nr}from"@pankod/refine-core";var or=/(image|photo|avatar|url|logo|cover|thumbnail|icon)/i,se=(e,o)=>{let c=["name","label","title","count","content","username","nickname","login","firstName","lastName","url"];or.test(e)&&c.unshift("url");let s=c.find(u=>u in o&&!!o[u]);if(s)return s==="firstName"&&"lastName"in o?["firstName","lastName"]:s};function ie(e,o,c){var u;if(o)return o;let s=c==null?void 0:c.find(i=>i.name===e);return(u=s==null?void 0:s.options)!=null&&u.dataProviderName?s.options.dataProviderName:"default"}var G=e=>{var o;return(o=e==null?void 0:e.options)==null?void 0:o.dataProviderName};var ce=e=>{let o=[...e].filter(Boolean).sort((c,s)=>(s.priority||0)-(c.priority||0))[0];return o||null};import ar from"prettier/standalone";import sr from"prettier/parser-typescript";var ue=e=>{try{return ar.format(e,{parser:"typescript",plugins:[sr],arrowParens:"always",trailingComma:"all",semi:!0,tabWidth:4,printWidth:80})}catch(o){return console.warn("Formatting error in inferencer",o),e}};var me=e=>e.replace(/ {2,}/g," ").replace(/\n{2,}/g,`
`);var pe=/(-id|-ids|_id|_ids|Id|Ids|ID|IDs)(\[\])?$/,Re=(e,o)=>{let c=pe.test(e),s=typeof o=="string"||typeof o=="number",u=Array.isArray(o)&&o.every(i=>typeof i=="string"||typeof i=="number");return c&&s||c&&u?{key:e,relation:!0,multiple:!!u,type:"relation",priority:1}:!1};var N=e=>e.replace(pe,"");var y=e=>N(e).replace(/([a-z])([A-Z])/g,"$1 $2").replace(/_/g," ").replace(/-/g," ").split(" ").map(m=>m.charAt(0).toUpperCase()+m.slice(1)).join(" ");var le=e=>e.replace(/export\s+(const|let|var|type|interface|function|class)\s+(\w+)\s*(=|:)\s*/g,"$1 $2 = ").split(`
`).filter(u=>!u.trim().startsWith("export default")).join(`
`);var Se=/import(?:(?:(?:[ \n\t]+([^ *\n\t\{\},]+)[ \n\t]*(?:,|[ \n\t]+))?([ \n\t]*\{(?:[ \n\t]*[^ \n\t"'\{\}]+[ \n\t]*,?)+\})?[ \n\t]*)|[ \n\t]*\*[ \n\t]*as[ \n\t]+([^ \n\t\{\}]+)[ \n\t]+)from[ \n\t]*(?:['"])([^'"\n]+)(?:['"])(?:;?)/g,ir=/import[ \n\t](?:['"])([^'"\n]+)(?:['"])(?:;?)/g,cr=/((?:\w|\s|_)*)( as )((?:\w|\s|_)*)( |,)?/g,de=(e,o={})=>{let c=e.matchAll(Se),s=new Set;for(let u of c){let[,i,m,g,d]=u;if(d in o){let $=o[d];if(d==="react")continue;i&&s.add(`const { default: ${i} } = ${$};`),m&&s.add(`const${m.replace(cr,"$1: $3$4")} = ${$};`),g&&s.add(`const ${g} = ${$};`)}}return me(`
    ${Array.from(s).join(`
`)}
    ${e.replace(Se,"").replace(ir,"")}
    `)};import te from"pluralize";var ye=(e,o)=>e?o.find(s=>{let u=N(e.key);return s.name===u||s.name===(te.isPlural(u)?u:te.plural(u))||s.name===(te.isSingular(u)?u:te.singular(u))}):void 0;import ur from"pluralize";var P=(e,o)=>{let c=e.replace(/[^a-zA-Z0-9]/g," "),s=ur.singular(c),u=y(s).replace(/ /g,""),i=o.charAt(0).toUpperCase()+o.slice(1);return/^[a-zA-Z]/.test(e)?`${u}${i}`:`${i}${u}`};var mr=e=>e?Array.isArray(e)?e.join(""):typeof e=="string"?e:typeof e=="object"||typeof e=="function"?e.toString():typeof e=="number"?String(e):typeof e=="boolean"&&e?"true":e:"",l=(e,...o)=>e.reduce((s,u,i)=>`${s}${u}${mr(o[i])}`,"").trim();var fe=(e,o)=>`
    ${e!=null?e:""}
    
    render(typeof ${o} !== "undefined" ? <${o} /> : <></>);
    `;var Ie=e=>e==null?void 0:e.replace(/(\/\* hidden-start \*\/)(.|\n)*?(\/\* hidden-end \*\/)/g,"");var pr=/^[a-zA-Z_$][a-zA-Z_$0-9]*$/,H=e=>pr.test(e);function ge(e,o,c){let s=`${e}`;return o&&(s+="?.",H(o)?s+=o:s+=`['${o}']`),c&&(s+="?.",H(c)?s+=c:s+=`['${c}']`),s}function lr(e,o,c){return(c!=null?c:[]).map(u=>ge(e,o,u))}var p=(e,o,c,s=' + " " + ')=>Array.isArray(c)?s?lr(e,o,c).join(s):ge(e,o,c[0]):ge(e,o,c);var D=e=>{let o=e.reduce((s,[u,i,m])=>(s[i]||(s[i]=[]),s[i].includes(u)||(m?s[i].push([u,!0]):s[i].push(u)),s),{});return Object.entries(o).map(([s,u])=>{let i=u.find($=>Array.isArray($)&&$[1]),m=u.filter($=>typeof $=="string"),g=i?`${i[0]}`:"",d=m.length?`{ ${m.join(", ")} }`:"";return`import ${g}${g&&d?", ":""}${d} from "${s}";`})};import Le from"pluralize";var B=(e="")=>Le.isSingular(e)?e:Le.singular(e);import Ne from"pluralize";var K=(e="")=>Ne.isPlural(e)?e:Ne.plural(e);var dr=/^id$/i,W=e=>dr.test(e);var $e=e=>e.relationInfer&&e.relationInfer.accessor?Array.isArray(e.relationInfer.accessor)?`optionLabel: "${e.relationInfer.accessor[0]}",`:e.relationInfer.accessor!=="title"?`optionLabel: "${e.relationInfer.accessor}",`:"":"";var j=e=>{};var R=(e,o="")=>{if(Number.isInteger(Number(e)))return o.charAt(0).toLowerCase()+o.slice(1)+e;let s=e.replace(/^\d+/,"").replace(/[^a-zA-Z0-9]/g,"-"),u=y(s).replace(/ /g,"");return`${u.charAt(0).toLowerCase()+u.slice(1)}${o.charAt(0).toUpperCase()+o.slice(1)}`};var Pe=e=>{let o=(c,s,u,i=o)=>{let m=e.map(d=>d(c,s,u,i));return ce(m)};return o};var Ee=e=>(c,s,u,i,m)=>e.reduce((g,d)=>d(g,s,u,i,m),c);var Ve=(e,o,c,s)=>{let u=Array.isArray(o),i=Array.isArray(o)&&o.every(m=>typeof m=="string"||typeof m=="number");if(u){if(!i){let g=s(e,o[0],c,s);return g?{...g,key:e,multiple:!0,priority:1}:!1}let m=s(e,o[0],c,s)||{type:"string"};return{key:e,multiple:!0,priority:1,type:m==null?void 0:m.type}}return!1};var De=(e,o)=>typeof o=="boolean"?{key:e,type:"boolean"}:!1;import je from"dayjs";var yr=/(_at|_on|At|On|AT|ON)(\[\])?$/,Be=(e,o)=>{let c=yr.test(e)&&je(o).isValid(),s=typeof o=="string"&&je(o).isValid();return c||s?{key:e,type:"date",priority:1}:!1};var fr=/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,Oe=(e,o)=>typeof o=="string"&&fr.test(o)?{key:e,type:"email",priority:1}:!1;var Ir=/\.(gif|jpe?g|tiff?|png|webp|bmp|svg)$/i,Me=(e,o)=>typeof o=="string"&&Ir.test(o)?{key:e,type:"image",priority:2}:!1;var Ue=(e,o)=>typeof o>"u"||o===null?null:!1;var _e=(e,o)=>typeof o=="number"?{key:e,type:"number"}:!1;var gr=/id$/i,ze=(e,o,c,s)=>{let u=o!==null,i=!Array.isArray(o);if(u&&i&&typeof o=="object"){if(Object.keys(o).length===1&&gr.test(Object.keys(o)[0]))return{key:e,type:"relation",relation:!0,accessor:"id",priority:1};let d=se(e,o);if((d?d.length>0:!1)&&d){let x=d&&Array.isArray(d)?d[0]:d,v=s(x,o[x],o,s);if(v){let k=Array.isArray(d)?d.map(h=>v.accessor?`${h}.${Array.isArray(v.accessor)?v.accessor[0]:v.accessor}`:h):v.accessor?Array.isArray(v.accessor)?`${d}.${v.accessor[0]}`:`${d}.${v.accessor}`:d;return{...v,fieldable:!0,key:e,accessor:k,priority:1}}}return{key:e,fieldable:!1,type:"object",priority:1}}return!1};var qe=(e,o)=>typeof o=="string"&&o.length>100?{key:e,type:"richtext",priority:1}:!1;var He=(e,o)=>typeof o=="string"?{key:e,type:"text"}:!1;var $r=/^(https?|ftp):\/\/(-\.)?([^\s/?\.#-]+\.?)+(\/[^\s]*)?$/i,Ke=(e,o)=>typeof o=="string"&&$r.test(o)?{key:e,type:"url",priority:1}:!1;var We=[Ve,De,Be,Oe,Me,Ue,_e,ze,Re,qe,He,Ke];var Ze=(e,o,c,s)=>e.map(i=>{if(!i.relation&&(i.type==="text"||i.type==="richtext"||i.type==="number")&&!i.canRelation){let m=/^[a-z0-9-]+$/,g=k=>m.test(`${k}`),d=i.key.toLowerCase()!=="id";return!(!i.accessor||typeof i.accessor=="string")||!d?i:(i.multiple?s[i.key]:[s[i.key]]).every(k=>g(i.accessor?k[i.accessor]:k))?{...i,canRelation:!0}:i}return i});var kr=/(image|photo|avatar|cover|thumbnail|icon)/i,Ge=e=>e.map(c=>c.type==="url"&&kr.test(c.key)?{...c,type:"image"}:c);var Xe=(e,o)=>e.map(s=>{let u=ye(s,o);return u?{...s,relation:!0,type:"relation",resource:u}:s});var Je=(e,o,c,s,u)=>e.map(m=>{if(m.relation&&m.type==="relation"&&!m.resource){let g=m.accessor?s[m.key][m.accessor]:s[m.key],d=u(m.key,g,s,u);if(d&&d.type!=="relation")return{...m,fieldable:!0,relation:!1,type:d.type,canRelation:!0}}return m});var Qe=[Ge,Xe,Je,Ze];import X,{useEffect as Fr}from"react";import xr,{defaultProps as vr}from"prism-react-renderer";import hr from"prism-react-renderer/themes/vsDark";var ke=({code:e,wrapperProps:o})=>(Fr(()=>{var s;let c=document.createElement("style");c.appendChild(document.createTextNode(`
                #code-highlight::-webkit-scrollbar {-webkit-appearance: none; height: 7px; width: 7px;}
                #code-highlight::-webkit-scrollbar-thumb {background-color: #666b7a;}
                #code-highlight::-webkit-scrollbar-corner {background-color: rgb(30, 30, 30);}
                `)),(s=document.getElementById("code-highlight"))==null||s.appendChild(c)},[]),X.createElement("div",{id:"code-highlight",style:{maxHeight:"75vh",height:"100%",overflow:"auto",backgroundColor:"rgb(30, 30, 30)"},...o!=null?o:{}},X.createElement(xr,{...vr,theme:hr,code:e,language:"tsx"},({className:c,style:s,tokens:u,getLineProps:i,getTokenProps:m})=>X.createElement("pre",{className:c,style:{...s,padding:"14px 14px 14px 14px",margin:"0",width:"100%",boxSizing:"border-box"}},u.map((g,d)=>X.createElement("div",{key:d,...i({line:g,key:d})},g.map(($,x)=>X.createElement("span",{key:x,...m({token:$,key:x})}))))))));import E from"react";import*as br from"@pankod/refine-core";import{LivePreview as Cr,LiveProvider as Ar}from"@aliemir/react-live";var Ye=[["react","React",E],["@pankod/refine-core","RefineCore",br]],Fe=E.createContext({}),xe=({fetchError:e,code:o,additionalScope:c,errorComponent:s})=>{var g;let u=(g=o&&le(de(o,[...Ye,...c!=null?c:[]].reduce((d,[$,x,v,k])=>k?d:{...d,[$]:x},{}))))!=null?g:"",i=E.useMemo(()=>({React:E,...E,...[...Ye,...c!=null?c:[]].reduce((d,[$,x,v])=>({...d,[x]:v}),{})}),[c]),m=E.useMemo(()=>()=>{let{error:$}=E.useContext(Fe);return s?E.createElement(s,{error:$!=null?$:e?"Something went wrong while fetching the resource data.":void 0}):null},[s,e]);return E.createElement(Ar,{Context:Fe,code:u,scope:i,noInline:!0},!e&&E.createElement(Cr,{Context:Fe}),E.createElement(m,null))};import J from"react";import{useDataProvider as Tr,useResource as wr}from"@pankod/refine-core";var er=(e,o,c)=>{let{resource:s,resourceName:u,id:i,resources:m}=wr({resourceNameOrRouteName:o}),g=c!=null?c:i,d=Tr(),[$,x]=J.useState(void 0),[v,k]=J.useState(!0),[h,w]=J.useState(!1),n=J.useCallback(async C=>{var f,t;let L=(f=G(s))!=null?f:ie(u,void 0,m),a=d(L);w(!0);try{if(e==="list"||e==="create"){let I=await a.getList({resource:u});x((t=I.data)==null?void 0:t[0]),setTimeout(()=>{w(!1)},500)}if(e==="edit"||e==="show"&&C){let I=await a.getOne({resource:u,id:C});x(I.data),setTimeout(()=>{w(!1)},500)}}catch{setTimeout(()=>{w(!1)},500)}},[e,d,s,u,m]);return J.useEffect(()=>{k(!1),!h&&!$&&n(g)},[n,g]),{data:$,loading:h,initial:v}};import Q from"react";import{useDataProvider as Rr}from"@pankod/refine-core";import{get as rr}from"lodash";var tr=({record:e,fields:o,infer:c})=>{let s=Rr(),[u,i]=Q.useState([]),[m,g]=Q.useState(!0),[d,$]=Q.useState(!1),x=Q.useCallback(async v=>{console.groupCollapsed("@pankod/refine-inferencer is trying to detect relations");let k=[];$(!0);try{let h=v.map(async n=>{if(n&&(n.relation||n.canRelation)&&e){let C=G(n.resource),L=s(C),a=Array.isArray(n.accessor)?void 0:n.multiple?e[n.key].map(f=>n.accessor?rr(f,n.accessor):f)[0]:n.accessor?rr(e[n.key],n.accessor):e[n.key];if(a&&n.resource)try{let{data:f}=await L.getOne({resource:n.resource.name,id:a});k.push({status:"success",resource:n.resource.name,field:n.key});let t=c("__",f,{},c);return{...n,relationInfer:t}}catch{return k.push({status:"error",resource:n.resource.name,field:n.key}),{...n,relationInfer:null}}if(a){let f,t;try{let{data:F}=await L.getOne({resource:K(N(n.key)),id:a});k.push({status:"success",resource:K(N(n.key)),field:n.key}),f=F,t=!0}catch{k.push({status:"error",resource:K(N(n.key)),field:n.key});try{let{data:A}=await L.getOne({resource:B(N(n.key)),id:a});k.push({status:"success",resource:B(N(n.key)),field:n.key}),f=A,t=!1}catch{return k.push({status:"error",resource:B(N(n.key)),field:n.key}),{...n,relationInfer:null}}}let I=c("__",f,{},c),r=N(n.key);return{...n,relation:!0,type:"relation",resource:{name:t?K(r):B(r)},fieldable:!1,canRelation:void 0,relationInfer:I}}return{...n,relationInfer:null}}return n}),w=await Promise.all(h);i(w.filter(n=>n)),setTimeout(()=>{$(!1)},500)}catch{setTimeout(()=>{$(!1)},500)}setTimeout(()=>{console.log(`Tried to detect relations with ${k.length} attempts and succeeded with ${k.filter(h=>h.status==="success").length} attempts.`),console.groupEnd()},500)},[s,e]);return Q.useEffect(()=>{g(!1),!d&&o&&o.length>0&&u.length===0&&x(o)},[x,d,o,x]),{fields:u,loading:d,initial:m}};var O=({type:e,additionalScope:o=[],customElements:c=[],fieldTransformers:s=[],renderer:u,loadingComponent:i,errorComponent:m,codeViewerComponent:g})=>{let d=Pe([...We,...c]),$=Ee([...Qe,...s]),x=({resourceName:k,fieldTransformer:h,id:w})=>{var b;let{resource:n,resources:C}=nr({resourceNameOrRouteName:k}),{resource:L}=nr(),{data:a,loading:f,initial:t}=er(e,k!=null?k:n==null?void 0:n.name,w),I=V.useMemo(()=>{if(a){let T=Object.keys(a).map(z=>{let q=a[z];return d(z,q,a,d)}).filter(Boolean),ee=$(T,C,n,a,d);return h?ee.flatMap(z=>{let q=h(z);return q?[q]:[]}):ee}return[]},[a,C,n,h]),{fields:r,loading:F}=tr({record:a,fields:I,infer:d}),A=V.useMemo(()=>!f&&!F&&!t?u({resource:n,resources:C,fields:r,infer:d,isCustomPage:n.name!==L.name,id:w}):"",[n,C,r,f,F]);return V.createElement(V.Fragment,null,i&&(f||F)&&V.createElement(i,null),!f&&!F&&V.createElement(V.Fragment,null,V.createElement(xe,{fetchError:!f&&!t&&!a,code:fe(A,P((b=n.label)!=null?b:n.name,e)),errorComponent:m,additionalScope:o}),g&&V.createElement(g,{code:Ie(A),loading:f||F})))};return({name:k,resource:h,fieldTransformer:w,id:n})=>V.createElement(x,{fieldTransformer:w,resourceName:h!=null?h:k,key:h!=null?h:k,id:n})};import ve from"react";import{Alert as Sr,Row as Lr,Col as Nr}from"@pankod/refine-antd";var M=({error:e})=>e?ve.createElement(Lr,{justify:"center",align:"middle",style:{minHeight:"200px"}},ve.createElement(Nr,null,ve.createElement(Sr,{message:"Error",description:e!=null?e:"",type:"error",showIcon:!0}))):null;import he from"react";import{Spin as Pr,Row as Er,Col as Vr}from"@pankod/refine-antd";var U=()=>he.createElement(Er,{justify:"center",align:"middle",style:{minHeight:"300px"}},he.createElement(Vr,null,he.createElement(Pr,{size:"large"})));import S from"react";import{Button as be,Modal as Dr,Icons as Y,Space as jr}from"@pankod/refine-antd";var _=({code:e,loading:o})=>{let c=S.useMemo(()=>ue(e!=null?e:""),[e]),s=S.useRef(null),[u,i]=S.useState(!1),[m,g]=S.useState(!1);if(o)return null;let d=()=>{var $,x,v,k;($=s==null?void 0:s.current)==null||$.select(),(x=s==null?void 0:s.current)==null||x.setSelectionRange(0,Number.MAX_SAFE_INTEGER),typeof navigator<"u"&&(navigator.clipboard.writeText((k=(v=s==null?void 0:s.current)==null?void 0:v.value)!=null?k:""),g(!0),setTimeout(()=>{g(!1)},1e3))};return c?S.createElement(S.Fragment,null,S.createElement("div",{style:{position:"fixed",bottom:12,right:60}},S.createElement(jr,null,S.createElement(be,{type:"primary",onClick:()=>i(!0),icon:S.createElement(Y.CodeOutlined,null)},"Show Code"),S.createElement(be,{type:"default",target:"_blank",href:"https://github.com/refinedev/refine/discussions/3046",icon:S.createElement(Y.MessageOutlined,null)},"Give Feedback"))),S.createElement("div",{style:{height:"50px"}}),S.createElement(Dr,{width:800,visible:u,onCancel:()=>i(!1),closeIcon:S.createElement(Y.CloseOutlined,{style:{color:"#666b7a"}}),footer:null,bodyStyle:{padding:0}},S.createElement(ke,{code:c}),S.createElement("textarea",{ref:s,defaultValue:(c!=null?c:"").replace(/\\n/g,`\r
`),id:"code-input","aria-hidden":"true",style:{padding:0,margin:0,width:0,height:0,opacity:0,border:"none",display:"block"}}),S.createElement(be,{style:{position:"absolute",bottom:12,right:12,width:100},key:"copy",type:"default",icon:m?S.createElement(Y.CheckOutlined,{style:{color:"green"}}):S.createElement(Y.CopyOutlined,null),onClick:d},m?"Copied":"Copy"))):null};var Ce=O({type:"show",additionalScope:[["@pankod/refine-antd","RefineAntd",Br]],codeViewerComponent:_,loadingComponent:U,errorComponent:M,renderer:({resource:e,fields:o,isCustomPage:c,id:s})=>{var f;let u=P((f=e.label)!=null?f:e.name,"show"),i="record",m=[["React","react",!0],["IResourceComponentsProps","@pankod/refine-core"],["useShow","@pankod/refine-core"],["Show","@pankod/refine-antd"],["Typography","@pankod/refine-antd"]],d=o.filter(t=>(t==null?void 0:t.relation)&&!(t!=null&&t.fieldable)&&(t==null?void 0:t.resource)).filter(Boolean).map(t=>{if((t==null?void 0:t.relation)&&!t.fieldable&&t.resource){if(t.multiple){m.push(["useMany","@pankod/refine-core"]);let I=p(i,t.key);return t.accessor&&(I=`${p(i,t.key)}?.map((item: any) => ${p("item",void 0,t.accessor)})`),`
                    const { data: ${R(t.key,"Data")}, isLoading: ${R(t.key,"IsLoading")} } =
                    useMany({
                        resource: "${t.resource.name}",
                        ids: ${I} || [],
                        queryOptions: {
                            enabled: !!${i},
                        },
                    });
                    `}return m.push(["useOne","@pankod/refine-core"]),`
                    const { data: ${R(t.key,"Data")}, isLoading: ${R(t.key,"IsLoading")} } =
                    useOne({
                        resource: "${t.resource.name}",
                        id: ${p(i,t.key,t.accessor,!1)} || "",
                        queryOptions: {
                            enabled: !!${i},
                        },
                    });
                `}}).filter(Boolean),$=t=>{if(t.relation&&t.resource){let I=R(t.key,"Data"),r=R(t.key,"IsLoading");return t.multiple?(m.push(["TagField","@pankod/refine-antd"]),l`
                    <Title level={5}>${y(t.key)}</Title>
                    {${r} ? <>Loading...</> : (
                        <>
                        ${(()=>{var F,A;if(t.relationInfer)if((F=t.relationInfer)!=null&&F.accessor){if(Array.isArray(t.relationInfer.accessor))return"Not Handled.";{let b=B((A=t.resource)==null?void 0:A.name),T=p(b,void 0,t.relationInfer.accessor);return`{${I}?.data?.map((${b}: any) => <TagField key={${T}} value={${T}} />)}`}}else return;else return})()}
                        </>
                    )}
                    `):l`
                    <Title level={5}>${y(t.key)}</Title>
                    {${r} ? <>Loading...</> : (
                        <>
                        ${(()=>{var F;return t.relationInfer?(F=t.relationInfer)!=null&&F.accessor?Array.isArray(t.relationInfer.accessor)?`{${p(`${I}?.data`,void 0,t.relationInfer.accessor,' + " " + ')}}`:`{${I}?.data?.${t.relationInfer.accessor}}`:`{${I}?.data}`:`{${I}?.data?.id}`})()}
                        </>
                    )}
                    
                    `}},x=t=>{if(t.type==="text"){if(m.push(["TagField","@pankod/refine-antd"],["TextField","@pankod/refine-antd"]),t.multiple){let I=p("item",void 0,t.accessor);return l`
                    <Title level={5}>${y(t.key)}</Title>
                    {${p(i,t.key)}?.map((item: any) => (
                        <TagField value={${I}} key={${I}} />
                    ))}
                `}return l`
                    <Title level={5}>${y(t.key)}</Title>
                    <TextField value={${p(i,t.key,t.accessor)}} />
                `}},v=t=>{if(t.type==="image"){if(m.push(["ImageField","@pankod/refine-antd"]),t.multiple){let I=p("item",void 0,t.accessor);return l`
                    <Title level={5}>${y(t.key)}</Title>
                    {${p(i,t.key)}?.map((item: any) => (
                        <ImageField style={{ maxWidth: 200 }} value={${I}} key={${I}} />
                    ))}
                `}return l`
                    <Title level={5}>${y(t.key)}</Title>
                    <ImageField style={{ maxWidth: 200 }} value={${p(i,t.key,t.accessor," + ")}} />
                    `}},k=t=>{if(t.type==="email"){if(m.push(["TagField","@pankod/refine-antd"],["EmailField","@pankod/refine-antd"]),t.multiple){let I=p("item",void 0,t.accessor);return l`
                    <Title level={5}>${y(t.key)}</Title>
                    <>
                    {${p(i,t.key)}?.map((item: any) => (
                        <TagField value={${I}} key={${I}} />
                    ))}
                    </>
                `}return l`
                    <Title level={5}>${y(t.key)}</Title>
                    <EmailField value={${p(i,t.key,t.accessor," + ")}} />
                `}},h=t=>{if(t.type==="url"){if(m.push(["TagField","@pankod/refine-antd"],["UrlField","@pankod/refine-antd"]),t.multiple){let I=p("item",void 0,t.accessor);return l`
                    <Title level={5}>${y(t.key)}</Title>
                    {${p(i,t.key)}?.map((item: any) => (
                        <TagField value={${I}} key={${I}} />
                    ))}
                `}return l`
                    <Title level={5}>${y(t.key)}</Title>
                    <UrlField value={${p(i,t.key,t.accessor," + ")}} />
                `}},w=t=>{if(t.type==="boolean"){if(m.push(["TagField","@pankod/refine-antd"],["BooleanField","@pankod/refine-antd"]),t.multiple){let I=p("item",void 0,t.accessor);return l`
                    <Title level={5}>${y(t.key)}</Title>
                    {(${p(i,t.key)} as any[])?.map((item, index) => (
                        <BooleanField value={${I}} key={index} />
                    ))}
                `}return l`
                    <Title level={5}>${y(t.key)}</Title>
                    <BooleanField value={${p(i,t.key,t.accessor," && ")}} />
                `}},n=t=>{if(t.type==="date"){if(m.push(["DateField","@pankod/refine-antd"]),t.multiple){let I=p("item",void 0,t.accessor);return l`
                    <Title level={5}>${y(t.key)}</Title>
                    {${p(i,t.key)}?.map((item: any) => (
                        <DateField value={${I}} key={${I}} />
                    ))}
                `}return l`
                    <Title level={5}>${y(t.key)}</Title>
                    <DateField value={${p(i,t.key,t.accessor," + ' ' + ")}} />
                `}},C=t=>{if(t.type==="richtext")return m.push(["MarkdownField","@pankod/refine-antd"]),l`
                    <Title level={5}>${y(t.key)}</Title>
                    <MarkdownField value={${p(i,t.key,t.accessor,' + " " + ')}} />
                `},L=t=>{if(t.type==="number"){if(m.push(["NumberField","@pankod/refine-antd"]),t.multiple){let I=p("item",void 0,t.accessor);return l`
                    <Title level={5}>${y(t.key)}</Title>
                    {${p(i,t.key)}?.map((item: any) => (
                        <TagField value={${I}} key={${I}} />
                    ))}
                `}return l`
                    <Title level={5}>${y(t.key)}</Title>
                    <NumberField value={${p(i,t.key,t.accessor,' + " " + ')} ?? ""} />
                `}},a=o.map(t=>{switch(t==null?void 0:t.type){case"text":return x(t);case"number":return L(t);case"richtext":return C(t);case"email":return k(t);case"image":return v(t);case"date":return n(t);case"boolean":return w(t);case"url":return h(t);case"relation":return $(t);default:return}});return j(m),l`
        ${D(m)}
        
        const { Title } = Typography;

        export const ${u}: React.FC<IResourceComponentsProps> = () => {
            const { queryResult } = useShow(${c?`{ 
                        resource: "${e.name}", 
                        id: ${s}
                    }`:""});
            const { data, isLoading } = queryResult;
        
            const ${i} = data?.data;

            ${d}

            return (
                <Show isLoading={isLoading}>
                    ${a.join("")}
                </Show>
            );
        };
        `}});import*as Or from"@pankod/refine-antd";var Ae=O({type:"list",additionalScope:[["@pankod/refine-antd","RefineAntd",Or]],codeViewerComponent:_,loadingComponent:U,errorComponent:M,renderer:({resource:e,fields:o,isCustomPage:c})=>{var I;let s=P((I=e.label)!=null?I:e.name,"list"),u="tableProps?.dataSource",i=[["React","react",!0],["IResourceComponentsProps","@pankod/refine-core"],["BaseRecord","@pankod/refine-core"],["useTable","@pankod/refine-antd"],["List","@pankod/refine-antd"],["Table","@pankod/refine-antd"],["Space","@pankod/refine-antd"]],g=o.filter(r=>(r==null?void 0:r.relation)&&!(r!=null&&r.fieldable)&&(r==null?void 0:r.resource)).filter(Boolean).map(r=>{if((r==null?void 0:r.relation)&&!r.fieldable&&r.resource){i.push(["useMany","@pankod/refine-core"]);let F="";return r.multiple?F=`[].concat(...(${u}?.map((item) => ${p("item",r.key,r.accessor,!1)}) ?? []))`:F=`${u}?.map((item) => ${p("item",r.key,r.accessor,!1)}) ?? []`,`
                    const { data: ${R(r.key,"Data")}, isLoading: ${R(r.key,"IsLoading")} } =
                    useMany({
                        resource: "${r.resource.name}",
                        ids: ${F},
                        queryOptions: {
                            enabled: !!${u},
                        },
                    });
                    `}}).filter(Boolean),d=r=>{var F,A;if(r.relation&&r.resource){let T=`${R(r.key,"Data")}?.data`,ee=R(r.key,"IsLoading");if(Array.isArray(r.accessor))return;let oe=`${ee} ? <>Loading...</> : `,z=r.multiple?`dataIndex="${r.key}"`:`dataIndex={["${r.key}", ${r.accessor?`"${r.accessor}"`:""}]}`,q=`title="${y(r.key)}"`,Z="";if(r.multiple){i.push(["TagField","@pankod/refine-antd"]);let re="item";if(r!=null&&r.relationInfer){let ae=`${T}?.find((resourceItems) => resourceItems.id === ${p("item",void 0,r.accessor)})`;re=p(ae,void 0,(F=r==null?void 0:r.relationInfer)==null?void 0:F.accessor)}Z=`render={(value: any[]) => ${oe} (
                        <>
                            {${p("value",void 0)}?.map((item, index) => (
                                <TagField key={index} value={${re}} />
                            ))}
                        </>
                    )}`}else if(r!=null&&r.relationInfer){let re=`${T}?.find((item) => item.id === value)`,ae=p(re,void 0,(A=r==null?void 0:r.relationInfer)==null?void 0:A.accessor);Z=`render={(value) => ${oe} ${ae}}`}else Z="";return l`<Table.Column ${z} ${q} ${Z} />`}},$=r=>{if(r.type==="image"){i.push(["ImageField","@pankod/refine-antd"]);let F=Array.isArray(r.accessor)||r.multiple?`dataIndex="${r.key}"`:`dataIndex={["${r.key}", ${r.accessor?`"${r.accessor}"`:""}]}`,A=`title="${y(r.key)}"`,b=l`render={(value: any) => <ImageField style={{ maxWidth: "100px" }} value={${p("value",void 0,Array.isArray(r.accessor)?r.accessor:void 0," + ")}} />}`;if(r.multiple){let T=p("item",void 0,r.accessor," + ");b=l`render={(value: any[]) => (<>{value?.map((item, index) => (
                        <ImageField style={{ maxWidth: "100px" }} value={${T}} key={index} />
                    ))}</>)}`}return l`<Table.Column ${F} ${A} ${b} />`}},x=r=>{if(r.type==="email"){i.push(["TagField","@pankod/refine-antd"],["EmailField","@pankod/refine-antd"]);let F=Array.isArray(r.accessor)||r.multiple?`dataIndex="${r.key}"`:`dataIndex={["${r.key}", ${r.accessor?`"${r.accessor}"`:""}]}`,A=`title="${y(r.key)}"`,b=l`render={(value: any) => <EmailField value={${p("value",void 0,Array.isArray(r.accessor)?r.accessor:void 0,' + " " + ')}} />}`;if(r.multiple){let T=p("item",void 0,r.accessor,' + " " + ');b=l`render={(value: any[]) => (<>{value?.map((item, index) => (
                        <TagField value={${T}} key={index} />
                    ))}</>)}`}return l`<Table.Column ${F} ${A} ${b} />`}},v=r=>{if(r.type==="url"){i.push(["UrlField","@pankod/refine-antd"],["TagField","@pankod/refine-antd"]);let F=Array.isArray(r.accessor)||r.multiple?`dataIndex="${r.key}"`:`dataIndex={["${r.key}", ${r.accessor?`"${r.accessor}"`:""}]}`,A=`title="${y(r.key)}"`,b=l`render={(value: any) => <UrlField value={${p("value",void 0,Array.isArray(r.accessor)?r.accessor:void 0," + ")}} />}`;if(r.multiple){let T=p("item",void 0,r.accessor," + ");b=l`render={(value: any[]) => (<>{value?.map((item, index) => (
                        <TagField value={${T}} key={index} />
                    ))}</>)}`}return l`<Table.Column ${F} ${A} ${b} />`}},k=r=>{if((r==null?void 0:r.type)==="boolean"){i.push(["BooleanField","@pankod/refine-antd"]);let F=Array.isArray(r.accessor)||r.multiple?`dataIndex="${r.key}"`:`dataIndex={["${r.key}", ${r.accessor?`"${r.accessor}"`:""}]}`,A=`title="${y(r.key)}"`,b=l`render={(value: any) => <BooleanField value={${p("value",void 0,Array.isArray(r.accessor)?r.accessor:void 0," && ")}} />}`;if(r.multiple){let T=p("item",void 0,r.accessor," && ");b=l`render={(value: any[]) => (<>{value?.map((item, index) => (
                        <BooleanField value={${T}} key={index} />
                    ))}</>)}`}return l`<Table.Column ${F} ${A} ${b} />`}},h=r=>{if(r.type==="date"){i.push(["DateField","@pankod/refine-antd"]);let F=Array.isArray(r.accessor)||r.multiple?`dataIndex="${r.key}"`:`dataIndex={["${r.key}", ${r.accessor?`"${r.accessor}"`:""}]}`,A=`title="${y(r.key)}"`,b=l`render={(value: any) => <DateField value={${p("value",void 0,Array.isArray(r.accessor)?r.accessor:void 0,' + " " + ')}} />}`;if(r.multiple){let T=p("item",void 0,r.accessor,' + " " + ');b=l`render={(value: any[]) => (<>{value?.map((item, index) => (
                        <DateField value={${T}} key={index} />
                    ))}</>)}`}return l`<Table.Column ${F} ${A} ${b} />`}},w=r=>{if((r==null?void 0:r.type)==="richtext"){i.push(["MarkdownField","@pankod/refine-antd"]);let F=Array.isArray(r.accessor)||r.multiple?`dataIndex="${r.key}"`:r.accessor?`dataIndex={["${r.key}", "${r.accessor}"]}`:`dataIndex="${r.key}"`,A=`title="${y(r.key)}"`,b=l`render={(value: any) => <MarkdownField value={(${p("value",void 0,Array.isArray(r.accessor)?r.accessor:void 0)}).slice(0, 80) + "..."} />}`;if(r.multiple){let T=p("item",void 0,r.accessor,' + " " + ');b=l`render={(value: any[]) => (<>{value?.map((item, index) => (
                        <MarkdownField value={(${T})?.slice(0, 80) + "..."} key={index} />
                    ))}</>)}`}return l`<Table.Column ${F} ${A} ${b} />`}},n=r=>{if(r&&(r.type==="text"||r.type==="number")){let F=r.accessor&&!Array.isArray(r.accessor)&&!r.multiple?`dataIndex={["${r.key}", "${r.accessor}"]}`:`dataIndex="${r.key}"`,A=`title="${y(r.key)}"`,b="";if(r.multiple){i.push(["TagField","@pankod/refine-antd"]);let T=p("item",void 0,r.accessor,' + " " + ');b=`render={(value: any[]) => (<>{value?.map((item) => (
                        <TagField value={${T}} key={${T}} />
                    ))}</>)}`}return!r.multiple&&Array.isArray(r.accessor)&&(b=`render={(value: any) => (<>{${p("value",void 0,r.accessor)}}</>)}`),`<Table.Column ${F} ${A} ${b} />`}},{canEdit:C,canShow:L,canDelete:a}=e!=null?e:{};C&&i.push(["EditButton","@pankod/refine-antd"]),L&&i.push(["ShowButton","@pankod/refine-antd"]),a&&i.push(["DeleteButton","@pankod/refine-antd"]);let f=C||L||a?l`
                <Table.Column
                    title="Actions"
                    dataIndex="actions"
                    render={(_, record: BaseRecord) => (
                        <Space>
                        ${C?l`
                            <EditButton
                                hideText
                                size="small"
                                recordItemId={record.id}
                            />
                            `:""}
                        ${L?l`
                            <ShowButton
                                hideText
                                size="small"
                                recordItemId={record.id}
                            />
                            `:""}
                        ${a?l`
                            <DeleteButton
                                hideText
                                size="small"
                                recordItemId={record.id}
                            />
                            `:""}
                        </Space>
                    )}
                />
            `:"",t=o.map(r=>{switch(r==null?void 0:r.type){case"text":case"number":return n(r);case"richtext":return w(r);case"email":return x(r);case"image":return $(r);case"date":return h(r);case"boolean":return k(r);case"url":return v(r);case"relation":return d(r);default:return}});return j(i),l`
        ${D(i)}
        
        export const ${s}: React.FC<IResourceComponentsProps> = () => {
            const { tableProps } = useTable({
                syncWithLocation: true,
                ${c?` resource: "${e.name}",`:""}
            });
        
            ${g}

            return (
                <List>
                    <Table {...tableProps} rowKey="id">
                        ${t.join(`\r
`)}
                        ${f}
                    </Table>
                </List>
            );
        };
        `}});import*as Mr from"@pankod/refine-antd";import Ur from"dayjs";var Te=O({type:"create",additionalScope:[["@pankod/refine-antd","RefineAntd",Mr],["dayjs","dayjs",Ur,!0]],codeViewerComponent:_,loadingComponent:U,errorComponent:M,renderer:({resource:e,fields:o,isCustomPage:c})=>{var w;let s=P((w=e.label)!=null?w:e.name,"create"),u=[["React","react",!0],["IResourceComponentsProps","@pankod/refine-core"],["Create","@pankod/refine-antd"],["Form","@pankod/refine-antd"],["useForm","@pankod/refine-antd"],["Input","@pankod/refine-antd"]],m=o.filter(n=>(n==null?void 0:n.relation)&&!(n!=null&&n.fieldable)&&(n==null?void 0:n.resource)).filter(Boolean).map(n=>{if((n==null?void 0:n.relation)&&!n.fieldable&&n.resource)return u.push(["useSelect","@pankod/refine-antd"]),`
                    const { selectProps: ${R(n.key,"SelectProps")} } =
                    useSelect({
                        resource: "${n.resource.name}",
                        ${$e(n)}
                    });
                `}).filter(Boolean),g=n=>{if(n.relation&&n.resource){u.push(["Select","@pankod/refine-antd"]);let C=R(n.key,"SelectProps"),L=n.accessor?n.multiple?`"${n.key}"`:`["${n.key}", "${n.accessor}"]`:`"${n.key}"`,a="",f="";return n.accessor&&n.multiple&&(f=`getValueFromEvent={(selected: string[]) => {
                        return selected?.map((item) => ({ ${H(`${n.accessor}`)?n.accessor:`["${n.accessor}"]`}: item }));
                    }}`,a=`getValueProps={(value: any[]) => {
                        return {
                            value: value?.map((item) => ${p("item",void 0,n.accessor)}),
                        };
                    }}`),l`
                    <Form.Item
                        label="${y(n.key)}"
                        name={${L}}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                        ${a}
                        ${f}
                    >
                        <Select ${n.multiple?'mode="multiple"':""} {...${C}} />
                    </Form.Item>             
                    `}},d=n=>{if(n.type==="text"||n.type==="url"||n.type==="email"||n.type==="number")return W(n.key)||n.multiple?void 0:l`
                    <Form.Item
                        label="${y(n.key)}"
                        name={["${n.key}"${n.accessor?', "'+n.accessor+'"':""}]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                    >
                        <Input  />
                    </Form.Item>
                `},$=n=>{if(n.type==="image"){u.push(["Upload","@pankod/refine-antd"],["getValueFromEvent","@pankod/refine-antd"]);let C='valuePropName="fileList"';return n.multiple&&!n.accessor&&(C="getValueProps={(value) => ({ fileList: value?.map((item: any) => ({ url: item, name: item, uid: item }))})}"),n.multiple||(n.accessor?C="getValueProps={(value) => ({ fileList: value ? [value] : [] })}":C="getValueProps={(value) => ({ fileList: [{ url: value, name: value, uid: value }]})}"),l`
                    <Form.Item label="${y(n.key)}">
                        <Form.Item
                            name="${n.key}"
                            ${C}
                            getValueFromEvent={getValueFromEvent}
                            noStyle
                            rules={[
                                {
                                    required: true,
                                },
                            ]}
                        >
                            <Upload.Dragger
                                listType="picture"
                                ${n.multiple?"multiple":""}
                                beforeUpload={() => false}
                            >
                                <p className="ant-upload-text">
                                    Drag & drop a file in this area
                                </p>
                            </Upload.Dragger>
                        </Form.Item>
                    </Form.Item>
                    `}},x=n=>{if(n.type==="boolean")return u.push(["Checkbox","@pankod/refine-antd"]),n.multiple?void 0:l`
                    <Form.Item
                        label="${y(n.key)}"
                        valuePropName="checked"
                        name={["${n.key}"${n.accessor?', "'+n.accessor+'"':""}]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                    >
                        <Checkbox>${y(n.key)}</Checkbox>
                    </Form.Item>
                `},v=n=>{if(n.type==="date")return u.push(["DatePicker","@pankod/refine-antd"],["dayjs","dayjs",!0]),n.multiple?void 0:l`
                    <Form.Item
                        label="${y(n.key)}"
                        name={["${n.key}"${n.accessor?', "'+n.accessor+'"':""}]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                        getValueProps={(value) => ({ value: value ? dayjs(value) : undefined })}
                    >
                        <DatePicker />
                    </Form.Item>
                `},k=n=>{if(n.type==="richtext")return l`
                <Form.Item
                    label="${y(n.key)}"
                    name="${n.key}"
                    rules={[
                        {
                            required: true,
                        },
                    ]}
                >
                    <Input.TextArea rows={5} />
                </Form.Item>
                `},h=o.map(n=>{switch(n==null?void 0:n.type){case"text":case"number":case"email":case"url":return d(n);case"richtext":return k(n);case"image":return $(n);case"date":return v(n);case"boolean":return x(n);case"relation":return g(n);default:return}});return j(u),l`
        ${D(u)}
        
        export const ${s}: React.FC<IResourceComponentsProps> = () => {
            const { formProps, saveButtonProps, queryResult } = useForm(${c?`{
                          resource: "${e.name}",
                          action: "create",
                      }`:""});
        
            ${m}

            return (
                <Create saveButtonProps={saveButtonProps}>
                    <Form {...formProps} layout="vertical">
                        ${h.join("")}
                    </Form>
                </Create>
            );
        };
        `}});import*as _r from"@pankod/refine-antd";import zr from"dayjs";var we=O({type:"edit",additionalScope:[["@pankod/refine-antd","RefineAntd",_r],["dayjs","dayjs",zr,!0]],codeViewerComponent:_,loadingComponent:U,errorComponent:M,renderer:({resource:e,fields:o,isCustomPage:c,id:s})=>{var C,L;let u=P((C=e.label)!=null?C:e.name,"edit"),i=R((L=e.label)!=null?L:e.name,"Data"),m=[["React","react",!0],["IResourceComponentsProps","@pankod/refine-core"],["Edit","@pankod/refine-antd"],["Form","@pankod/refine-antd"],["useForm","@pankod/refine-antd"],["Input","@pankod/refine-antd"]],d=o.filter(a=>(a==null?void 0:a.relation)&&!(a!=null&&a.fieldable)&&(a==null?void 0:a.resource)).filter(Boolean).map(a=>{if((a==null?void 0:a.relation)&&!a.fieldable&&a.resource){m.push(["useSelect","@pankod/refine-antd"]);let f=p(i,a.key,a.accessor,!1);return a.multiple&&a.accessor&&(f=`${p(i,a.key)}?.map((item: any) => ${p("item",void 0,a.accessor)})`),`
                    const { selectProps: ${R(a.key,"SelectProps")} } =
                    useSelect({
                        resource: "${a.resource.name}",
                        defaultValue: ${f},
                        ${a.relationInfer&&a.relationInfer.accessor?typeof a.relationInfer.accessor=="string"?a.relationInfer.accessor!=="title"?`optionLabel: "${a.relationInfer.accessor}",`:"":`optionLabel: "${a.relationInfer.accessor[0]}",`:""}
                    });
                `}}).filter(Boolean),$=a=>{if(a.relation&&a.resource){m.push(["Select","@pankod/refine-antd"]);let f=R(a.key,"SelectProps"),t=a.accessor?a.multiple?`"${a.key}"`:`["${a.key}", "${a.accessor}"]`:`"${a.key}"`,I="",r="";return a.accessor&&a.multiple&&(r=`getValueFromEvent={(selected: string[]) => {
                        return selected?.map((item) => ({ ${H(`${a.accessor}`)?a.accessor:`["${a.accessor}"]`}: item }));
                    }}`,I=`getValueProps={(value: any[]) => {
                        return {
                            value: value?.map((item) => ${p("item",void 0,a.accessor)}),
                        };
                    }}`),l`
                    <Form.Item
                        label="${y(a.key)}"
                        name={${t}}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                        ${I}
                        ${r}
                    >
                        <Select ${a.multiple?'mode="multiple"':""} {...${f}} />
                    </Form.Item>             
                    `}},x=a=>{if(a.type==="text"||a.type==="url"||a.type==="email"||a.type==="number"){if(a.multiple){let f=p(a.key,"index",a.accessor).split("?.").map(t=>`"${t}"`).join(", ").replace(/"index"/,"index");return`
                        <>
                            {(${p(i,a.key)} as any[])?.map((item, index) => (
                                <Form.Item
                                    key={index}
                                    label={\`${y(a.key)} \${index+1}\`}
                                    name={[${f}]}
                                >
                                    <Input
                                        type="${a.type}"
                                    ${W(a.key)?"readOnly disabled":""} />
                                </Form.Item>
                            ))}
                        </>
                    `}return l`
                    <Form.Item
                        label="${y(a.key)}"
                        name={["${a.key}"${a.accessor?', "'+a.accessor+'"':""}]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                    >
                        <Input ${W(a.key)?"readOnly disabled":""} />
                    </Form.Item>
                `}},v=a=>{if(a.type==="image"){m.push(["Upload","@pankod/refine-antd"],["getValueFromEvent","@pankod/refine-antd"]);let f='valuePropName="fileList"';return a.multiple&&!a.accessor&&(f="getValueProps={(value) => ({ fileList: value?.map((item: any) => ({ url: item, name: item, uid: item }))})}"),a.multiple||(a.accessor?f="getValueProps={(value) => ({ fileList: value ? [value] : [] })}":f="getValueProps={(value) => ({ fileList: [{ url: value, name: value, uid: value }]})}"),l`
                    <Form.Item label="${y(a.key)}">
                        <Form.Item
                            name="${a.key}"
                            ${f}
                            getValueFromEvent={getValueFromEvent}
                            noStyle
                            rules={[
                                {
                                    required: true,
                                },
                            ]}
                        >
                            <Upload.Dragger
                                listType="picture"
                                ${a.multiple?"multiple":""}
                                beforeUpload={() => false}
                            >
                                <p className="ant-upload-text">
                                    Drag & drop a file in this area
                                </p>
                            </Upload.Dragger>
                        </Form.Item>
                    </Form.Item>
                    `}},k=a=>{if(a.type==="boolean"){if(m.push(["Checkbox","@pankod/refine-antd"]),a.multiple){let f=p(a.key,"index",a.accessor).split("?.").map(t=>`"${t}"`).join(", ").replace(/"index"/,"index");return`
                        <>
                            {(${p(i,a.key)} as any[])?.map((item, index) => (
                                <Form.Item
                                    key={index}
                                    valuePropName="checked"
                                    label={\`${y(a.key)} \${index+1}\`}
                                    name={[${f}]}
                                >
                                    <Checkbox>${y(a.key)}</Checkbox>
                                </Form.Item>
                            ))}
                        </>
                    `}return l`
                    <Form.Item
                        label="${y(a.key)}"
                        valuePropName="checked"
                        name={["${a.key}"${a.accessor?', "'+a.accessor+'"':""}]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                    >
                        <Checkbox>${y(a.key)}</Checkbox>
                    </Form.Item>
                `}},h=a=>{if(a.type==="date"){if(m.push(["DatePicker","@pankod/refine-antd"],["dayjs","dayjs",!0]),a.multiple){let f=p(a.key,"index",a.accessor).split("?.").map(t=>`"${t}"`).join(", ").replace(/"index"/,"index");return l`
                        <>
                            {(${p(i,a.key)} as any[])?.map((item, index) => (
                                <Form.Item
                                    key={index}
                                    label={\`${y(a.key)} \${index+1}\`}
                                    name={[${f}]}
                                    getValueProps={(value) => ({ value: value ? dayjs(value) : undefined })}
                                >
                                    <DatePicker />
                                </Form.Item>
                            ))}
                        </>
                    `}return l`
                    <Form.Item
                        label="${y(a.key)}"
                        name={["${a.key}"${a.accessor?', "'+a.accessor+'"':""}]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                        getValueProps={(value) => ({ value: value ? dayjs(value) : undefined })}
                    >
                        <DatePicker />
                    </Form.Item>
                `}},w=a=>{if(a.type==="richtext")return l`
                <Form.Item
                    label="${y(a.key)}"
                    name="${a.key}"
                    rules={[
                        {
                            required: true,
                        },
                    ]}
                >
                    <Input.TextArea rows={5} />
                </Form.Item>
                `},n=o.map(a=>{switch(a==null?void 0:a.type){case"text":case"number":case"email":case"url":return x(a);case"richtext":return w(a);case"image":return v(a);case"date":return h(a);case"boolean":return k(a);case"relation":return $(a);default:return}});return j(m),l`
        ${D(m)}
        
        export const ${u}: React.FC<IResourceComponentsProps> = () => {
            const { formProps, saveButtonProps, queryResult } = useForm(${c?`{
                          resource: "${e.name}",
                          id: ${s},
                          action: "edit",
                      }`:""});
        
            const ${i} = queryResult?.data?.data;
        
            ${d}

            return (
                <Edit saveButtonProps={saveButtonProps}>
                    <Form {...formProps} layout="vertical">
                        ${n.join("")}
                    </Form>
                </Edit>
            );
        };
        `}});var ca=({action:e,id:o,...c})=>{let{useParams:s}=qr(),{action:u,id:i}=s();switch(e!=null?e:u){case"show":return ne.createElement(Ce,{...c,id:o!=null?o:i});case"create":return ne.createElement(Te,{...c,id:o!=null?o:i});case"edit":return ne.createElement(we,{...c,id:o!=null?o:i});default:return ne.createElement(Ae,{...c,id:o!=null?o:i})}};export{Te as AntdCreateInferencer,we as AntdEditInferencer,ca as AntdInferencer,Ae as AntdListInferencer,Ce as AntdShowInferencer};
