// @pankod/refine-inferencer@1.6.1/dist/esm/chakra-ui.js
import oe from"react";import{useRouterContext as nn}from"@pankod/refine-core";import*as _r from"@pankod/refine-chakra-ui";import z from"react";import{useResource as er}from"@pankod/refine-core";var nr=/(image|photo|avatar|url|logo|cover|thumbnail|icon)/i,ae=(e,n)=>{let s=["name","label","title","count","content","username","nickname","login","firstName","lastName","url"];nr.test(e)&&s.unshift("url");let o=s.find(u=>u in n&&!!n[u]);if(o)return o==="firstName"&&"lastName"in n?["firstName","lastName"]:o};function se(e,n,s){var u;if(n)return n;let o=s==null?void 0:s.find(a=>a.name===e);return(u=o==null?void 0:o.options)!=null&&u.dataProviderName?o.options.dataProviderName:"default"}var Q=e=>{var n;return(n=e==null?void 0:e.options)==null?void 0:n.dataProviderName};var ie=e=>{let n=[...e].filter(Boolean).sort((s,o)=>(o.priority||0)-(s.priority||0))[0];return n||null};import tr from"prettier/standalone";import or from"prettier/parser-typescript";var ce=e=>{try{return tr.format(e,{parser:"typescript",plugins:[or],arrowParens:"always",trailingComma:"all",semi:!0,tabWidth:4,printWidth:80})}catch(n){return console.warn("Formatting error in inferencer",n),e}};var ue=e=>e.replace(/ {2,}/g," ").replace(/\n{2,}/g,`
`);var pe=/(-id|-ids|_id|_ids|Id|Ids|ID|IDs)(\[\])?$/,ve=(e,n)=>{let s=pe.test(e),o=typeof n=="string"||typeof n=="number",u=Array.isArray(n)&&n.every(a=>typeof a=="string"||typeof a=="number");return s&&o||s&&u?{key:e,relation:!0,multiple:!!u,type:"relation",priority:1}:!1};var N=e=>e.replace(pe,"");var g=e=>N(e).replace(/([a-z])([A-Z])/g,"$1 $2").replace(/_/g," ").replace(/-/g," ").split(" ").map(p=>p.charAt(0).toUpperCase()+p.slice(1)).join(" ");var me=e=>e.replace(/export\s+(const|let|var|type|interface|function|class)\s+(\w+)\s*(=|:)\s*/g,"$1 $2 = ").split(`
`).filter(u=>!u.trim().startsWith("export default")).join(`
`);var Ae=/import(?:(?:(?:[ \n\t]+([^ *\n\t\{\},]+)[ \n\t]*(?:,|[ \n\t]+))?([ \n\t]*\{(?:[ \n\t]*[^ \n\t"'\{\}]+[ \n\t]*,?)+\})?[ \n\t]*)|[ \n\t]*\*[ \n\t]*as[ \n\t]+([^ \n\t\{\}]+)[ \n\t]+)from[ \n\t]*(?:['"])([^'"\n]+)(?:['"])(?:;?)/g,ar=/import[ \n\t](?:['"])([^'"\n]+)(?:['"])(?:;?)/g,sr=/((?:\w|\s|_)*)( as )((?:\w|\s|_)*)( |,)?/g,le=(e,n={})=>{let s=e.matchAll(Ae),o=new Set;for(let u of s){let[,a,p,h,l]=u;if(l in n){let I=n[l];if(l==="react")continue;a&&o.add(`const { default: ${a} } = ${I};`),p&&o.add(`const${p.replace(sr,"$1: $3$4")} = ${I};`),h&&o.add(`const ${h} = ${I};`)}}return ue(`
    ${Array.from(o).join(`
`)}
    ${e.replace(Ae,"").replace(ar,"")}
    `)};import ne from"pluralize";var de=(e,n)=>e?n.find(o=>{let u=N(e.key);return o.name===u||o.name===(ne.isPlural(u)?u:ne.plural(u))||o.name===(ne.isSingular(u)?u:ne.singular(u))}):void 0;import ir from"pluralize";var V=(e,n)=>{let s=e.replace(/[^a-zA-Z0-9]/g," "),o=ir.singular(s),u=g(o).replace(/ /g,""),a=n.charAt(0).toUpperCase()+n.slice(1);return/^[a-zA-Z]/.test(e)?`${u}${a}`:`${a}${u}`};var cr=e=>e?Array.isArray(e)?e.join(""):typeof e=="string"?e:typeof e=="object"||typeof e=="function"?e.toString():typeof e=="number"?String(e):typeof e=="boolean"&&e?"true":e:"",f=(e,...n)=>e.reduce((o,u,a)=>`${o}${u}${cr(n[a])}`,"").trim();var fe=(e,n)=>`
    ${e!=null?e:""}
    
    render(typeof ${n} !== "undefined" ? <${n} /> : <></>);
    `;var ge=e=>e==null?void 0:e.replace(/(\/\* hidden-start \*\/)(.|\n)*?(\/\* hidden-end \*\/)/g,"");var ur=/^[a-zA-Z_$][a-zA-Z_$0-9]*$/,Te=e=>ur.test(e);function ye(e,n,s){let o=`${e}`;return n&&(o+="?.",Te(n)?o+=n:o+=`['${n}']`),s&&(o+="?.",Te(s)?o+=s:o+=`['${s}']`),o}function pr(e,n,s){return(s!=null?s:[]).map(u=>ye(e,n,u))}var m=(e,n,s,o=' + " " + ')=>Array.isArray(s)?o?pr(e,n,s).join(o):ye(e,n,s[0]):ye(e,n,s),E=(e,n,s)=>{let o=e;return n&&(o+=`.${n}`),s&&(Array.isArray(s)?o+=`.${s[0]}`:o+=`.${s}`),o};var P=e=>{let n=e.reduce((o,[u,a,p])=>(o[a]||(o[a]=[]),o[a].includes(u)||(p?o[a].push([u,!0]):o[a].push(u)),o),{});return Object.entries(n).map(([o,u])=>{let a=u.find(I=>Array.isArray(I)&&I[1]),p=u.filter(I=>typeof I=="string"),h=a?`${a[0]}`:"",l=p.length?`{ ${p.join(", ")} }`:"";return`import ${h}${h&&l?", ":""}${l} from "${o}";`})};import Se from"pluralize";var D=(e="")=>Se.isSingular(e)?e:Se.singular(e);import He from"pluralize";var Z=(e="")=>He.isPlural(e)?e:He.plural(e);var mr=/^id$/i,X=e=>mr.test(e);var J=e=>e.relationInfer&&e.relationInfer.accessor?Array.isArray(e.relationInfer.accessor)?`optionLabel: "${e.relationInfer.accessor[0]}",`:e.relationInfer.accessor!=="title"?`optionLabel: "${e.relationInfer.accessor}",`:"":"";var j=e=>{};var C=(e,n="")=>{if(Number.isInteger(Number(e)))return n.charAt(0).toLowerCase()+n.slice(1)+e;let o=e.replace(/^\d+/,"").replace(/[^a-zA-Z0-9]/g,"-"),u=g(o).replace(/ /g,"");return`${u.charAt(0).toLowerCase()+u.slice(1)}${n.charAt(0).toUpperCase()+n.slice(1)}`};var Re=e=>{let n=(s,o,u,a=n)=>{let p=e.map(l=>l(s,o,u,a));return ie(p)};return n};var we=e=>(s,o,u,a,p)=>e.reduce((h,l)=>l(h,o,u,a,p),s);var Le=(e,n,s,o)=>{let u=Array.isArray(n),a=Array.isArray(n)&&n.every(p=>typeof p=="string"||typeof p=="number");if(u){if(!a){let h=o(e,n[0],s,o);return h?{...h,key:e,multiple:!0,priority:1}:!1}let p=o(e,n[0],s,o)||{type:"string"};return{key:e,multiple:!0,priority:1,type:p==null?void 0:p.type}}return!1};var Ee=(e,n)=>typeof n=="boolean"?{key:e,type:"boolean"}:!1;import Ne from"dayjs";var lr=/(_at|_on|At|On|AT|ON)(\[\])?$/,Ve=(e,n)=>{let s=lr.test(e)&&Ne(n).isValid(),o=typeof n=="string"&&Ne(n).isValid();return s||o?{key:e,type:"date",priority:1}:!1};var dr=/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,Be=(e,n)=>typeof n=="string"&&dr.test(n)?{key:e,type:"email",priority:1}:!1;var fr=/\.(gif|jpe?g|tiff?|png|webp|bmp|svg)$/i,Me=(e,n)=>typeof n=="string"&&fr.test(n)?{key:e,type:"image",priority:2}:!1;var De=(e,n)=>typeof n>"u"||n===null?null:!1;var ze=(e,n)=>typeof n=="number"?{key:e,type:"number"}:!1;var gr=/id$/i,Oe=(e,n,s,o)=>{let u=n!==null,a=!Array.isArray(n);if(u&&a&&typeof n=="object"){if(Object.keys(n).length===1&&gr.test(Object.keys(n)[0]))return{key:e,type:"relation",relation:!0,accessor:"id",priority:1};let l=ae(e,n);if((l?l.length>0:!1)&&l){let x=l&&Array.isArray(l)?l[0]:l,b=o(x,n[x],n,o);if(b){let y=Array.isArray(l)?l.map(i=>b.accessor?`${i}.${Array.isArray(b.accessor)?b.accessor[0]:b.accessor}`:i):b.accessor?Array.isArray(b.accessor)?`${l}.${b.accessor[0]}`:`${l}.${b.accessor}`:l;return{...b,fieldable:!0,key:e,accessor:y,priority:1}}}return{key:e,fieldable:!1,type:"object",priority:1}}return!1};var Pe=(e,n)=>typeof n=="string"&&n.length>100?{key:e,type:"richtext",priority:1}:!1;var je=(e,n)=>typeof n=="string"?{key:e,type:"text"}:!1;var yr=/^(https?|ftp):\/\/(-\.)?([^\s/?\.#-]+\.?)+(\/[^\s]*)?$/i,Ue=(e,n)=>typeof n=="string"&&yr.test(n)?{key:e,type:"url",priority:1}:!1;var Ke=[Le,Ee,Ve,Be,Me,De,ze,Oe,ve,Pe,je,Ue];var qe=(e,n,s,o)=>e.map(a=>{if(!a.relation&&(a.type==="text"||a.type==="richtext"||a.type==="number")&&!a.canRelation){let p=/^[a-z0-9-]+$/,h=y=>p.test(`${y}`),l=a.key.toLowerCase()!=="id";return!(!a.accessor||typeof a.accessor=="string")||!l?a:(a.multiple?o[a.key]:[o[a.key]]).every(y=>h(a.accessor?y[a.accessor]:y))?{...a,canRelation:!0}:a}return a});var kr=/(image|photo|avatar|cover|thumbnail|icon)/i,_e=e=>e.map(s=>s.type==="url"&&kr.test(s.key)?{...s,type:"image"}:s);var We=(e,n)=>e.map(o=>{let u=de(o,n);return u?{...o,relation:!0,type:"relation",resource:u}:o});var Ge=(e,n,s,o,u)=>e.map(p=>{if(p.relation&&p.type==="relation"&&!p.resource){let h=p.accessor?o[p.key][p.accessor]:o[p.key],l=u(p.key,h,o,u);if(l&&l.type!=="relation")return{...p,fieldable:!0,relation:!1,type:l.type,canRelation:!0}}return p});var Ze=[_e,We,Ge,qe];import Y,{useEffect as hr}from"react";import Ir,{defaultProps as $r}from"prism-react-renderer";import Fr from"prism-react-renderer/themes/vsDark";var ke=({code:e,wrapperProps:n})=>(hr(()=>{var o;let s=document.createElement("style");s.appendChild(document.createTextNode(`
                #code-highlight::-webkit-scrollbar {-webkit-appearance: none; height: 7px; width: 7px;}
                #code-highlight::-webkit-scrollbar-thumb {background-color: #666b7a;}
                #code-highlight::-webkit-scrollbar-corner {background-color: rgb(30, 30, 30);}
                `)),(o=document.getElementById("code-highlight"))==null||o.appendChild(s)},[]),Y.createElement("div",{id:"code-highlight",style:{maxHeight:"75vh",height:"100%",overflow:"auto",backgroundColor:"rgb(30, 30, 30)"},...n!=null?n:{}},Y.createElement(Ir,{...$r,theme:Fr,code:e,language:"tsx"},({className:s,style:o,tokens:u,getLineProps:a,getTokenProps:p})=>Y.createElement("pre",{className:s,style:{...o,padding:"14px 14px 14px 14px",margin:"0",width:"100%",boxSizing:"border-box"}},u.map((h,l)=>Y.createElement("div",{key:l,...a({line:h,key:l})},h.map((I,x)=>Y.createElement("span",{key:x,...p({token:I,key:x})}))))))));import B from"react";import*as xr from"@pankod/refine-core";import{LivePreview as br,LiveProvider as Cr}from"@aliemir/react-live";var Qe=[["react","React",B],["@pankod/refine-core","RefineCore",xr]],he=B.createContext({}),Ie=({fetchError:e,code:n,additionalScope:s,errorComponent:o})=>{var h;let u=(h=n&&me(le(n,[...Qe,...s!=null?s:[]].reduce((l,[I,x,b,y])=>y?l:{...l,[I]:x},{}))))!=null?h:"",a=B.useMemo(()=>({React:B,...B,...[...Qe,...s!=null?s:[]].reduce((l,[I,x,b])=>({...l,[x]:b}),{})}),[s]),p=B.useMemo(()=>()=>{let{error:I}=B.useContext(he);return o?B.createElement(o,{error:I!=null?I:e?"Something went wrong while fetching the resource data.":void 0}):null},[o,e]);return B.createElement(Cr,{Context:he,code:u,scope:a,noInline:!0},!e&&B.createElement(br,{Context:he}),B.createElement(p,null))};import ee from"react";import{useDataProvider as vr,useResource as Ar}from"@pankod/refine-core";var Xe=(e,n,s)=>{let{resource:o,resourceName:u,id:a,resources:p}=Ar({resourceNameOrRouteName:n}),h=s!=null?s:a,l=vr(),[I,x]=ee.useState(void 0),[b,y]=ee.useState(!0),[i,T]=ee.useState(!1),d=ee.useCallback(async c=>{var A,L;let S=(A=Q(o))!=null?A:se(u,void 0,p),$=l(S);T(!0);try{if(e==="list"||e==="create"){let t=await $.getList({resource:u});x((L=t.data)==null?void 0:L[0]),setTimeout(()=>{T(!1)},500)}if(e==="edit"||e==="show"&&c){let t=await $.getOne({resource:u,id:c});x(t.data),setTimeout(()=>{T(!1)},500)}}catch{setTimeout(()=>{T(!1)},500)}},[e,l,o,u,p]);return ee.useEffect(()=>{y(!1),!i&&!I&&d(h)},[d,h]),{data:I,loading:i,initial:b}};import re from"react";import{useDataProvider as Tr}from"@pankod/refine-core";import{get as Je}from"lodash";var Ye=({record:e,fields:n,infer:s})=>{let o=Tr(),[u,a]=re.useState([]),[p,h]=re.useState(!0),[l,I]=re.useState(!1),x=re.useCallback(async b=>{console.groupCollapsed("@pankod/refine-inferencer is trying to detect relations");let y=[];I(!0);try{let i=b.map(async d=>{if(d&&(d.relation||d.canRelation)&&e){let c=Q(d.resource),S=o(c),$=Array.isArray(d.accessor)?void 0:d.multiple?e[d.key].map(A=>d.accessor?Je(A,d.accessor):A)[0]:d.accessor?Je(e[d.key],d.accessor):e[d.key];if($&&d.resource)try{let{data:A}=await S.getOne({resource:d.resource.name,id:$});y.push({status:"success",resource:d.resource.name,field:d.key});let L=s("__",A,{},s);return{...d,relationInfer:L}}catch{return y.push({status:"error",resource:d.resource.name,field:d.key}),{...d,relationInfer:null}}if($){let A,L;try{let{data:r}=await S.getOne({resource:Z(N(d.key)),id:$});y.push({status:"success",resource:Z(N(d.key)),field:d.key}),A=r,L=!0}catch{y.push({status:"error",resource:Z(N(d.key)),field:d.key});try{let{data:F}=await S.getOne({resource:D(N(d.key)),id:$});y.push({status:"success",resource:D(N(d.key)),field:d.key}),A=F,L=!1}catch{return y.push({status:"error",resource:D(N(d.key)),field:d.key}),{...d,relationInfer:null}}}let t=s("__",A,{},s),k=N(d.key);return{...d,relation:!0,type:"relation",resource:{name:L?Z(k):D(k)},fieldable:!1,canRelation:void 0,relationInfer:t}}return{...d,relationInfer:null}}return d}),T=await Promise.all(i);a(T.filter(d=>d)),setTimeout(()=>{I(!1)},500)}catch{setTimeout(()=>{I(!1)},500)}setTimeout(()=>{console.log(`Tried to detect relations with ${y.length} attempts and succeeded with ${y.filter(i=>i.status==="success").length} attempts.`),console.groupEnd()},500)},[o,e]);return re.useEffect(()=>{h(!1),!l&&n&&n.length>0&&u.length===0&&x(n)},[x,l,n,x]),{fields:u,loading:l,initial:p}};var U=({type:e,additionalScope:n=[],customElements:s=[],fieldTransformers:o=[],renderer:u,loadingComponent:a,errorComponent:p,codeViewerComponent:h})=>{let l=Re([...Ke,...s]),I=we([...Ze,...o]),x=({resourceName:y,fieldTransformer:i,id:T})=>{var R;let{resource:d,resources:c}=er({resourceNameOrRouteName:y}),{resource:S}=er(),{data:$,loading:A,initial:L}=Xe(e,y!=null?y:d==null?void 0:d.name,T),t=z.useMemo(()=>{if($){let w=Object.keys($).map(O=>{let G=$[O];return l(O,G,$,l)}).filter(Boolean),v=I(w,c,d,$,l);return i?v.flatMap(O=>{let G=i(O);return G?[G]:[]}):v}return[]},[$,c,d,i]),{fields:k,loading:r}=Ye({record:$,fields:t,infer:l}),F=z.useMemo(()=>!A&&!r&&!L?u({resource:d,resources:c,fields:k,infer:l,isCustomPage:d.name!==S.name,id:T}):"",[d,c,k,A,r]);return z.createElement(z.Fragment,null,a&&(A||r)&&z.createElement(a,null),!A&&!r&&z.createElement(z.Fragment,null,z.createElement(Ie,{fetchError:!A&&!L&&!$,code:fe(F,V((R=d.label)!=null?R:d.name,e)),errorComponent:p,additionalScope:n}),h&&z.createElement(h,{code:ge(F),loading:A||r})))};return({name:y,resource:i,fieldTransformer:T,id:d})=>z.createElement(x,{fieldTransformer:T,resourceName:i!=null?i:y,key:i!=null?i:y,id:d})};import te from"react";import{Center as Sr,Alert as Hr,AlertIcon as Rr,AlertDescription as wr}from"@pankod/refine-chakra-ui";var K=({error:e})=>e?te.createElement(Sr,{style:{minHeight:300}},te.createElement(Hr,{status:"error",variant:"subtle",flexDirection:"column",alignItems:"center",justifyContent:"center",textAlign:"center",height:"200px"},te.createElement(Rr,{boxSize:"40px",mr:0}),te.createElement(wr,{mt:4,maxWidth:"sm"},e!=null?e:""))):null;import rr from"react";import{Box as Lr,Spinner as Er}from"@pankod/refine-chakra-ui";var q=()=>rr.createElement(Lr,{position:"relative",bg:"chakra-body-bg",minH:120},rr.createElement(Er,{position:"absolute",top:"50%",left:"50%",transform:"translate(-50%, -50%)"}));import H from"react";import{Box as Nr,Button as $e,Modal as Vr,HStack as Br,useDisclosure as Mr,ModalBody as Dr,ModalCloseButton as zr,ModalContent as Or,ModalOverlay as Pr}from"@pankod/refine-chakra-ui";import{IconCode as jr,IconMessageCircle as Ur,IconCopy as Kr,IconCheck as qr}from"@tabler/icons";var _=({code:e,loading:n})=>{let s=H.useMemo(()=>ce(e!=null?e:""),[e]),o=H.useRef(null),[u,a]=H.useState(!1),{isOpen:p,onOpen:h,onClose:l}=Mr();if(n)return null;let I=()=>{var x,b,y,i;(x=o==null?void 0:o.current)==null||x.select(),(b=o==null?void 0:o.current)==null||b.setSelectionRange(0,Number.MAX_SAFE_INTEGER),typeof navigator<"u"&&(navigator.clipboard.writeText((i=(y=o==null?void 0:o.current)==null?void 0:y.value)!=null?i:""),a(!0),setTimeout(()=>{a(!1)},1e3))};return s?H.createElement(H.Fragment,null,H.createElement(Nr,{position:"fixed",bottom:"12px",right:"60px"},H.createElement(Br,{spacing:4},H.createElement($e,{variant:"solid",colorScheme:"green",leftIcon:H.createElement(jr,{size:18}),onClick:h},"Show Code"),H.createElement($e,{colorScheme:"gray",variant:"solid",leftIcon:H.createElement(Ur,{size:18}),as:"a",target:"_blank",href:"https://github.com/refinedev/refine/discussions/3046"},"Give Feedback"))),H.createElement("div",{style:{height:"50px"}}),H.createElement(Vr,{size:"4xl",isOpen:p,onClose:l},H.createElement(Pr,null),H.createElement(Or,null,H.createElement(zr,{sx:{position:"absolute",right:0,top:0,height:"56px",width:"56px",color:"#666b7a"}}),H.createElement(Dr,{sx:{padding:0}},H.createElement(ke,{code:s}),H.createElement("textarea",{ref:o,defaultValue:(s!=null?s:"").replace(/\\n/g,`\r
`),id:"code-input",style:{padding:0,margin:0,width:0,height:0,opacity:0,border:"none",display:"block"}})),H.createElement($e,{sx:{position:"absolute",bottom:"12px",right:"12px",width:100},key:"copy",leftIcon:u?H.createElement(qr,{size:18,style:{color:"green"}}):H.createElement(Kr,{size:18}),onClick:I},u?"Copied":"Copy")))):null};var Fe=U({type:"show",additionalScope:[["@pankod/refine-chakra-ui","RefineChakraUI",_r]],codeViewerComponent:_,loadingComponent:q,errorComponent:K,renderer:({resource:e,fields:n,isCustomPage:s,id:o})=>{var L;let u=V((L=e.label)!=null?L:e.name,"show"),a="record",p=[["useShow","@pankod/refine-core"],["Show","@pankod/refine-chakra-ui"],["Heading","@pankod/refine-chakra-ui"]],l=n.filter(t=>(t==null?void 0:t.relation)&&!(t!=null&&t.fieldable)&&(t==null?void 0:t.resource)).filter(Boolean).map(t=>{if((t==null?void 0:t.relation)&&!t.fieldable&&t.resource){if(t.multiple){p.push(["useMany","@pankod/refine-core"]);let k=m(a,t.key);return t.accessor&&(k=`${m(a,t.key)}?.map((item: any) => ${m("item",void 0,t.accessor)})`),`
                    const { data: ${C(t.key,"Data")}, isLoading: ${C(t.key,"IsLoading")} } =
                    useMany({
                        resource: "${t.resource.name}",
                        ids: ${k} || [],
                        queryOptions: {
                            enabled: !!${a},
                        },
                    });
                    `}return p.push(["useOne","@pankod/refine-core"]),`
                    const { data: ${C(t.key,"Data")}, isLoading: ${C(t.key,"IsLoading")} } =
                    useOne({
                        resource: "${t.resource.name}",
                        id: ${m(a,t.key,t.accessor,!1)} || "",
                        queryOptions: {
                            enabled: !!${a},
                        },
                    });
                `}}).filter(Boolean),I=t=>{if(t.relation&&t.resource){let k=C(t.key,"Data"),r=C(t.key,"IsLoading");return t.multiple?(p.push(["TagField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),f`
                    <Heading as="h5" size="sm" mt={4} >${g(t.key)}</Heading>
                    {${r} ? <>Loading...</> : (
                        <>
                        ${(()=>{var F;if(t.relationInfer)if((F=t.relationInfer)!=null&&F.accessor){if(Array.isArray(t.relationInfer.accessor))return"Not Handled.";{let R=C(t.key),w=m(R,void 0,t.relationInfer.accessor);return`<HStack spacing="12px">{${k}?.data?.map((${R}: any) => <TagField key={${w}} value={${w}} />)}</HStack>`}}else return"Not Handled.";else return"not-handled - relation with multiple but no resource"})()}
                        </>
                    )}
                    `):f`
                    <Heading as="h5" size="sm" mt={4} >${g(t.key)}</Heading>
                    {${r} ? <>Loading...</> : (
                        <>
                        ${(()=>{var F;return t.relationInfer?(F=t.relationInfer)!=null&&F.accessor?Array.isArray(t.relationInfer.accessor)?`{${m(`${k}?.data`,void 0,t.relationInfer.accessor,' + " " + ')}}`:`{${k}?.data?.${t.relationInfer.accessor}}`:`{${k}?.data}`:`{${k}?.data?.id}`})()}
                        </>
                    )}
                    
                    `}},x=t=>{if(t.type==="text"){if(p.push(["TagField","@pankod/refine-chakra-ui"],["TextField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),t.multiple){let k=m("item",void 0,t.accessor);return f`
                    <Heading as="h5" size="sm" mt={4} >${g(t.key)}</Heading>
                    <HStack spacing="12px">
                    {${m(a,t.key)}?.map((item: any) => (
                        <TagField value={${k}} key={${k}} />
                    ))}
                    </HStack>
                `}return f`
                    <Heading as="h5" size="sm" mt={4} >${g(t.key)}</Heading>
                    <TextField value={${m(a,t.key,t.accessor)}} />
                `}},b=t=>{if(t.type==="image"){if(p.push(["Image","@pankod/refine-chakra-ui"]),t.multiple){let k=m("item",void 0,t.accessor);return f`
                    <Heading as="h5" size="sm" mt={4} >${g(t.key)}</Heading>
                    {${m(a,t.key)}?.map((item: any) => (
                        <Image sx={{ maxWidth: 200 }} src={${k}} key={${k}} />
                    ))}
                `}return f`
                    <Heading as="h5" size="sm" mt={4} >${g(t.key)}</Heading>
                    <Image sx={{ maxWidth: 200 }} src={${m(a,t.key,t.accessor," + ")}} />
                    `}},y=t=>{if(t.type==="email"){if(p.push(["TagField","@pankod/refine-chakra-ui"],["EmailField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),t.multiple){let k=m("item",void 0,t.accessor);return f`
                    <Heading as="h5" size="sm" mt={4} >${g(t.key)}</Heading>
                    <HStack spacing="12px">
                    {${m(a,t.key)}?.map((item: any) => (
                        <TagField value={${k}} key={${k}} />
                    ))}
                    </HStack>
                `}return f`
                    <Heading as="h5" size="sm" mt={4} >${g(t.key)}</Heading>
                    <EmailField value={${m(a,t.key,t.accessor," + ")}} />
                `}},i=t=>{if(t.type==="url"){if(p.push(["TagField","@pankod/refine-chakra-ui"],["UrlField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),t.multiple){let k=m("item",void 0,t.accessor);return f`
                    <Heading as="h5" size="sm" mt={4} >${g(t.key)}</Heading>
                    <HStack spacing="12px">
                    {${m(a,t.key)}?.map((item: any) => (
                        <TagField value={${k}} key={${k}} />
                    ))}
                    </HStack>
                `}return f`
                    <Heading as="h5" size="sm" mt={4} >${g(t.key)}</Heading>
                    <UrlField value={${m(a,t.key,t.accessor," + ")}} />
                `}},T=t=>{if(t.type==="boolean"){if(p.push(["TagField","@pankod/refine-chakra-ui"],["BooleanField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),t.multiple){let k=m("item",void 0,t.accessor);return f`
                    <Heading as="h5" size="sm" mt={4} >${g(t.key)}</Heading>
                    <HStack spacing="12px">
                    {${m(a,t.key)}?.map((item: any) => (
                        <TagField value={${k}} key={${k}} />
                    ))}
                    </HStack>
                `}return f`
                    <Heading as="h5" size="sm" mt={4} >${g(t.key)}</Heading>
                    <BooleanField value={${m(a,t.key,t.accessor," && ")}} />
                `}},d=t=>{if(t.type==="date"){if(p.push(["DateField","@pankod/refine-chakra-ui"]),t.multiple){let k=m("item",void 0,t.accessor);return f`
                    <Heading as="h5" size="sm" mt={4} >${g(t.key)}</Heading>
                    {${m(a,t.key)}?.map((item: any) => (
                        <DateField value={${k}} key={${k}} />
                    ))}
                `}return f`
                    <Heading as="h5" size="sm" mt={4} >${g(t.key)}</Heading>
                    <DateField value={${m(a,t.key,t.accessor," + ' ' + ")}} />
                `}},c=t=>{if(t.type==="richtext")return p.push(["MarkdownField","@pankod/refine-chakra-ui"]),f`
                    <Heading as="h5" size="sm" mt={4}>${g(t.key)}</Heading>
                    <MarkdownField value={${m(a,t.key,t.accessor,' + " " + ')}} />
                `},S=t=>{if(t.type==="number"){if(p.push(["NumberField","@pankod/refine-chakra-ui"],["TagField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),t.multiple){let k=m("item",void 0,t.accessor);return f`
                    <Heading as="h5" size="sm" mt={4} >${g(t.key)}</Heading>
                    <HStack spacing="12px">
                    {${m(a,t.key)}?.map((item: any) => (
                        <TagField value={${k}} key={${k}} />
                    ))}
                    </HStack>
                `}return f`
                    <Heading as="h5" size="sm" mt={4} >${g(t.key)}</Heading>
                    <NumberField value={${m(a,t.key,t.accessor,' + " " + ')} ?? ""} />
                `}},$=t=>{if(t)return f`
                    ${t}
                `},A=n.map(t=>{switch(t==null?void 0:t.type){case"text":return $(x(t));case"number":return $(S(t));case"richtext":return $(c(t));case"email":return $(y(t));case"image":return $(b(t));case"date":return $(d(t));case"boolean":return $(T(t));case"url":return $(i(t));case"relation":return $(I(t));default:return}});return j(p),f`
        ${P(p)}
        
        export const ${u} = () => {
            const { queryResult } = useShow(${s?`{ 
                        resource: "${e.name}", 
                        id: ${o}
                    }`:""});
            const { data, isLoading } = queryResult;

            const ${a} = data?.data;
        
            ${l}

            return (
                <Show isLoading={isLoading}>
                    ${A.join("")}
                </Show>
            );
        };
        `}});import*as Wr from"@pankod/refine-chakra-ui";import*as Gr from"@pankod/refine-react-table";import*as Zr from"@tabler/icons";var W=e=>Array.isArray(e.accessor)||e.multiple?`accessorKey: "${e.key}"`:e.accessor?`accessorKey: "${E(e.key,void 0,e.accessor)}"`:`accessorKey: "${e.key}"`,xe=U({type:"list",additionalScope:[["@pankod/refine-chakra-ui","RefineChakraUI",Wr],["@pankod/refine-react-table","RefineReactTable",Gr],["@tabler/icons","TablerIcons",Zr]],codeViewerComponent:_,loadingComponent:q,errorComponent:K,renderer:({resource:e,fields:n,isCustomPage:s})=>{var k;let o=V((k=e.label)!=null?k:e.name,"list"),u="tableData?.data",a=[["IResourceComponentsProps","@pankod/refine-core"],["useTable","@pankod/refine-react-table"],["ColumnDef","@pankod/refine-react-table"],["flexRender","@pankod/refine-react-table"],["List","@pankod/refine-chakra-ui"],["TableContainer","@pankod/refine-chakra-ui"],["Table","@pankod/refine-chakra-ui"],["Thead","@pankod/refine-chakra-ui"],["Tr","@pankod/refine-chakra-ui"],["Th","@pankod/refine-chakra-ui"],["Tbody","@pankod/refine-chakra-ui"],["Tr","@pankod/refine-chakra-ui"],["Td","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"],["Button","@pankod/refine-chakra-ui"],["IconButton","@pankod/refine-chakra-ui"],["usePagination","@pankod/refine-chakra-ui"],["Box","@pankod/refine-chakra-ui"],["IconChevronRight","@tabler/icons"],["IconChevronLeft","@tabler/icons"]],p=n.filter(r=>(r==null?void 0:r.relation)&&!(r!=null&&r.fieldable)&&(r==null?void 0:r.resource)),h=p.filter(Boolean).map(r=>{if((r==null?void 0:r.relation)&&!r.fieldable&&r.resource){a.push(["GetManyResponse","@pankod/refine-core"]),a.push(["useMany","@pankod/refine-core"]);let F="";return r.multiple?F=`[].concat(...(${u}?.map((item) => ${m("item",r.key,r.accessor,!1)}) ?? []))`:F=`${u}?.map((item) => ${m("item",r.key,r.accessor,!1)}) ?? []`,`
                    const { data: ${C(r.key,"Data")} } =
                    useMany({
                        resource: "${r.resource.name}",
                        ids: ${F},
                        queryOptions: {
                            enabled: !!${u},
                        },
                    });
                    `}}).filter(Boolean),l=p==null?void 0:p.map(r=>{if(r&&r.resource)return C(r.key,"Data")}).filter(Boolean),I=r=>{var F;if(r.relation&&r.resource){let R=`${C(r.key,"Data")}?.data`;if(Array.isArray(r.accessor))return;let w=`id: "${r.key}"`,v=`header: "${g(r.key)}"`,M=W(r),O="";if(r.multiple){a.push(["TagField","@pankod/refine-chakra-ui"]);let G="item";r!=null&&r.relationInfer&&(G=m("item",void 0,r.relationInfer.accessor)),O=`cell: function render({ getValue, table }) {
                        const meta = table.options.meta as {
                            ${C(r.key,"Data")}: GetManyResponse;
                        };

                        const ${C(r.key,"")} = getValue<any[]>()?.map((item) => {
                            return meta.${C(r.key,"Data")}?.data?.find(
                                (resourceItems) => resourceItems.id === ${m("item",void 0,r.accessor)}
                            );
                        })


                        return (
                            <HStack>
                                {${C(r.key,"")}?.map((item, index) => (
                                    <TagField key={index} value={${G}} />
                                ))}
                            </HStack>
                        )
                    }
                `}else r!=null&&r.relationInfer?O=`cell: function render({ getValue, table }) {
                            const meta = table.options.meta as {
                                ${C(r.key,"Data")}: GetManyResponse;
                            };

                            const ${C(r.key,"")} = meta.${R}?.find(
                                (item) => item.id === getValue<any>(),
                            );

                            return ${m(C(r.key),void 0,(F=r==null?void 0:r.relationInfer)==null?void 0:F.accessor)} ?? "Loading...";
                        },`:O="";return`
                    {
                        ${w},
                        ${v},
                        ${M},
                        ${O}
                    }
                `}},x=r=>{if(r.type==="image"){a.push(["Image","@pankod/refine-chakra-ui"]);let F=`id: "${r.key}"`,R=W(r),w=`header: "${g(r.key)}"`,v=f`
                    cell: function render({ getValue }) {
                        ${r!=null&&r.accessor?"try {":""}
                            return <Image sx={{ maxWidth: "100px" }} src={${m("getValue<any>()",void 0,Array.isArray(r.accessor)?r.accessor:void 0," + ")}} />
                        ${r!=null&&r.accessor?" } catch (error) { return null; }":""}
                    }
                `;if(r.multiple){let M=m("item",void 0,r.accessor," + ");v=`
                        cell: function render({ getValue }) {
                            ${r!=null&&r.accessor?"try {":""}
                                return (
                                    <HStack>
                                        {getValue<any[]>()?.map((item, index) => (
                                            <Image src={${M}} key={index} sx={{ height: "50px", maxWidth: "100px" }} />
                                        ))}
                                    </HStack>
                                )
                            ${r!=null&&r.accessor?" } catch (error) { return null; }":""}
                        }
                    `}return`
                    {
                        ${F},
                        ${R},
                        ${w},
                        ${v}
                    }
                `}},b=r=>{if(r.type==="email"){a.push(["EmailField","@pankod/refine-chakra-ui"]);let F=`id: "${r.key}"`,R=W(r),w=`header: "${g(r.key)}"`,v=f`
                    cell: function render({ getValue }) {
                        return <EmailField value={${m("getValue<any>()",void 0,Array.isArray(r.accessor)?r.accessor:void 0,' + " " + ')}} />
                    }
                `;return r.multiple&&(a.push(["TagField","@pankod/refine-chakra-ui"]),v=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${m("item",void 0,r.accessor," + ")}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),`
                    {
                        ${F},
                        ${R},
                        ${w},
                        ${v}
                    }
                `}},y=r=>{if(r.type==="url"){a.push(["UrlField","@pankod/refine-chakra-ui"]);let F=`id: "${r.key}"`,R=W(r),w=`header: "${g(r.key)}"`,v=f`
                    cell: function render({ getValue }) {
                        return <UrlField value={${m("getValue<any>()",void 0,Array.isArray(r.accessor)?r.accessor:void 0," + ")}} />
                    }
                `;return r.multiple&&(a.push(["TagField","@pankod/refine-chakra-ui"]),v=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${m("item",void 0,r.accessor," + ")}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),`
                    {
                        ${F},
                        ${R},
                        ${w},
                        ${v}
                    }
                `}},i=r=>{if((r==null?void 0:r.type)==="boolean"){a.push(["BooleanField","@pankod/refine-chakra-ui"]);let F=`id: "${r.key}"`,R=W(r),w=`header: "${g(r.key)}"`,v=f`
                    cell: function render({ getValue }) {
                        return <BooleanField value={${m("getValue<any>()",void 0,Array.isArray(r.accessor)?r.accessor:void 0," + ")}} />
                    }
                `;return r.multiple&&(v=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <BooleanField value={${m("item",void 0,r.accessor," + ")}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),`
                    {
                        ${F},
                        ${R},
                        ${w},
                        ${v}
                    }
                `}},T=r=>{if(r.type==="date"){a.push(["DateField","@pankod/refine-chakra-ui"]);let F=`id: "${r.key}"`,R=W(r),w=`header: "${g(r.key)}"`,v=f`
                    cell: function render({ getValue }) {
                        return <DateField value={${m("getValue<any>()",void 0,Array.isArray(r.accessor)?r.accessor:void 0,' + " " + ')}} />
                    }
                `;return r.multiple&&(v=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <DateField value={${m("item",void 0,r.accessor," + ")}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),`
                    {
                        ${F},
                        ${R},
                        ${w},
                        ${v}
                    }
                `}},d=r=>{if((r==null?void 0:r.type)==="richtext"){a.push(["MarkdownField","@pankod/refine-chakra-ui"]);let F=`id: "${r.key}"`,R=W(r),w=`header: "${g(r.key)}"`,v=f`
                    cell: function render({ getValue }) {
                        return <MarkdownField value={(${m("getValue<string>()",void 0,Array.isArray(r.accessor)?r.accessor:void 0)})?.slice(0, 80) + "..." } />
                    }
                `;return r.multiple&&(v=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<string[]>()?.map((item, index) => (
                                        <MarkdownField value={${m("item",void 0,r.accessor," + ")}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),`
                    {
                        ${F},
                        ${R},
                        ${w},
                        ${v}
                    }
                `}},c=r=>{if(r&&(r.type==="text"||r.type==="number")){let F=`id: "${r.key}"`,R=W(r),w=`header: "${g(r.key)}"`,v="";return r.multiple&&(a.push(["TagField","@pankod/refine-chakra-ui"]),v=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${m("item",void 0,r.accessor,' + " " + ')}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),!r.multiple&&Array.isArray(r.accessor)&&(v=`
                        cell: function render({ getValue }) {
                            return (
                                <>{${m("getValue<any>()",r.key,r.accessor)}}</>
                            );
                        }
                    `),`
                    {
                        ${F},
                        ${R},
                        ${w},
                        ${v}
                    }
                `}},{canEdit:S,canShow:$,canDelete:A}=e!=null?e:{};S&&a.push(["EditButton","@pankod/refine-chakra-ui"]),$&&a.push(["ShowButton","@pankod/refine-chakra-ui"]),A&&a.push(["DeleteButton","@pankod/refine-chakra-ui"]);let L=S||$||A?f`
        {
            id: "actions",
            accessorKey: "id",
            header: "Actions",
            cell: function render({ getValue }) {
                return (
                    <HStack>
                    ${$?f`
                        <ShowButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `:""}
                        ${S?f`
                        <EditButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `:""}
                        ${A?f`
                        <DeleteButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `:""}
                    </HStack>
                );
            },
        },
            `:"",t=n.map(r=>{switch(r==null?void 0:r.type){case"text":case"number":return c(r);case"richtext":return d(r);case"email":return b(r);case"image":return x(r);case"date":return T(r);case"boolean":return i(r);case"url":return y(r);case"relation":return I(r);default:return}});return j(a),f`
        import React from "react";
        ${P(a)}
        
        export const ${o}: React.FC<IResourceComponentsProps> = () => {
            const columns = React.useMemo<ColumnDef<any>[]>(() => [
                ${[...t,L].filter(Boolean).join(",")}
            ], []);

            const {
                getHeaderGroups,
                getRowModel,
                setOptions,
                refineCore: {
                    setCurrent,
                    pageCount,
                    current,
                    tableQueryResult: { data: tableData },
                },
            } = useTable({
                columns,
                ${s?`
                refineCoreProps: {
                    resource: "${e.name}",
                }
                `:""}
                
            });

            ${h}

            setOptions((prev) => ({
                ...prev,
                meta: {
                    ...prev.meta,
                    ${l.join(", ")}
                },
            }));

            return (
                <List>
                    <TableContainer whiteSpace="pre-line">
                        <Table variant="simple">
                            <Thead>
                                {getHeaderGroups().map((headerGroup) => (
                                    <Tr key={headerGroup.id}>
                                        {headerGroup.headers.map((header) => (
                                            <Th key={header.id}>
                                                {!header.isPlaceholder && (
                                                    flexRender(
                                                        header.column.columnDef
                                                            .header,
                                                        header.getContext(),
                                                    )
                                                )}
                                            </Th>
                                        ))}
                                    </Tr>
                                ))}
                            </Thead>
                            <Tbody>
                                {getRowModel().rows.map((row) => (
                                    <Tr key={row.id}>
                                        {row.getVisibleCells().map((cell) => (
                                            <Td key={cell.id}>
                                                {flexRender(
                                                    cell.column.columnDef.cell,
                                                    cell.getContext(),
                                                )}
                                            </Td>
                                        ))}
                                    </Tr>
                                ))}
                            </Tbody>
                        </Table>
                    </TableContainer>
                    <Pagination
                        current={current}
                        pageCount={pageCount}
                        setCurrent={setCurrent}
                    />
                </List>   
            );
        };

        type PaginationProps = {
            current: number;
            pageCount: number;
            setCurrent: (page: number) => void;
        };

        const Pagination: React.FC<PaginationProps> = ({
            current,
            pageCount,
            setCurrent,
        }) => {
            const pagination = usePagination({
                current,
                pageCount,
            });
        
            return (
                <Box display="flex" justifyContent="flex-end">
                    <HStack my="3" spacing="1">
                        {pagination?.prev && (
                            <IconButton
                                aria-label="previous page"
                                onClick={() => setCurrent(current - 1)}
                                disabled={!pagination?.prev}
                                variant="outline"
                            >
                                <IconChevronLeft size="18" />
                            </IconButton>
                        )}
        
                        {pagination?.items.map((page) => {
                            if (typeof page === "string")
                                return <span key={page}>...</span>;
        
                            return (
                                <Button
                                    key={page}
                                    onClick={() => setCurrent(page)}
                                    variant={page === current ? "solid" : "outline"}
                                >
                                    {page}
                                </Button>
                            );
                        })}
                        {pagination?.next && (
                            <IconButton
                                aria-label="next page"
                                onClick={() => setCurrent(current + 1)}
                                variant="outline"
                            >
                                <IconChevronRight size="18" />
                            </IconButton>
                        )}
                    </HStack>
                </Box>
            );
        };
        `}});import*as Qr from"@pankod/refine-core";import*as Xr from"@pankod/refine-chakra-ui";import*as Jr from"@pankod/refine-react-hook-form";var be=U({type:"create",additionalScope:[["@pankod/refine-core","RefineCore",Qr],["@pankod/refine-chakra-ui","RefineChakraUI",Xr],["@pankod/refine-react-hook-form","RefineReactHookForm",Jr]],codeViewerComponent:_,loadingComponent:q,errorComponent:K,renderer:({resource:e,fields:n,isCustomPage:s})=>{var y;let o=V((y=e.label)!=null?y:e.name,"create"),u=[["Create","@pankod/refine-chakra-ui"],["FormControl","@pankod/refine-chakra-ui"],["FormLabel","@pankod/refine-chakra-ui"],["FormErrorMessage","@pankod/refine-chakra-ui"],["useForm","@pankod/refine-react-hook-form"]],p=n.filter(i=>(i==null?void 0:i.relation)&&!(i!=null&&i.fieldable)&&(i==null?void 0:i.resource)).filter(Boolean).map(i=>{if((i==null?void 0:i.relation)&&!i.fieldable&&i.resource)return u.push(["useSelect","@pankod/refine-core"]),`
                    const { options: ${C(i.key,"Options")} } =
                    useSelect({
                        resource: "${i.resource.name}",
                        ${J(i)}
                    });
                `}).filter(Boolean),h=i=>{if(i.relation&&i.resource){u.push(["useSelect","@pankod/refine-core"]),u.push(["Select","@pankod/refine-chakra-ui"]);let T=C(i.key,"Options");return f`
                <FormControl mb="3" isInvalid={!!errors?.${E(i.key,void 0)}}>
                    <FormLabel>${g(i.key)}</FormLabel>
                    <Select
                        placeholder="Select ${D(i.resource.name)}"
                        {...register("${E(i.key,void 0,i.accessor)}", {
                            required: "This field is required",
                        })}
                    >
                        {${T}?.map((option) => (
                            <option value={option.value} key={option.value}>
                                {option.label}
                            </option>
                        ))}
                    </Select>
                    <FormErrorMessage>
                        {${m("(errors as any)",i.key,i.accessor,!1)}?.message as string}
                    </FormErrorMessage>
                </FormControl>
                `}},l=i=>{if(i.type==="text"||i.type==="url"||i.type==="email"||i.type==="number"||i.type==="date"||i.type==="richtext")return X(i.key)||(u.push(["Input","@pankod/refine-chakra-ui"]),i.multiple)?void 0:f`
                    <FormControl mb="3" isInvalid={!!${m("(errors as any)",i.key,i.accessor,!1)}}>
                        <FormLabel>${g(i.key)}</FormLabel>
                        <Input
                            ${i.type!=="date"&&i.type!=="richtext"?`type="${i.type}"`:""}
                            {...register("${E(i.key,void 0,i.accessor)}", {
                                required: "This field is required",
                            })}
                        />
                        <FormErrorMessage>
                            {${m("(errors as any)",i.key,i.accessor,!1)}?.message as string}
                        </FormErrorMessage>
                    </FormControl>
                `},I=i=>{if(i.type==="boolean")return u.push(["Checkbox","@pankod/refine-chakra-ui"]),i.multiple?void 0:f`
                    <FormControl mb="3" isInvalid={!!${m("errors",i.key,i.accessor,!1)}}>
                        <FormLabel>${g(i.key)}</FormLabel>
                        <Checkbox
                            {...register("${E(i.key,void 0,i.accessor)}", {
                                required: "This field is required",
                            })}
                        />
                        <FormErrorMessage>
                            {${m("errors",i.key,i.accessor,!1)}?.message as string}
                        </FormErrorMessage>
                    </FormControl>
                   
                `},x=i=>{if(i.type==="date"){let T=l(i);return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-chakra-ui" package.
                        To use a <DatePicker> component, you can examine the following links:
                        
                        - https://github.com/aboveyunhai/chakra-dayzed-datepicker
                        - https://github.com/wojtekmaj/react-date-picker
                    */}
                    ${T!=null?T:""}
                    `}},b=n.map(i=>{switch(i==null?void 0:i.type){case"text":case"number":case"email":case"url":case"richtext":return l(i);case"date":return x(i);case"boolean":return I(i);case"relation":return h(i);default:return}});return j(u),f`
        ${P(u)}
        
        export const ${o} = () => {
            const {
                refineCore: { formLoading },
                saveButtonProps,
                register,
                formState: { errors },
            } = useForm(
                ${s?`
                { 
                    refineCoreProps: {
                        resource: "${e.name}",
                        action: "create",
                    }
                }`:""}
            );
        
            ${p}

            return (
                <Create isLoading={formLoading} saveButtonProps={saveButtonProps}>
                    ${b.join("")}
                </Create>
            );
        };
        `}});import*as Yr from"@pankod/refine-core";import*as en from"@pankod/refine-chakra-ui";import*as rn from"@pankod/refine-react-hook-form";var Ce=U({type:"edit",additionalScope:[["@pankod/refine-core","RefineCore",Yr],["@pankod/refine-chakra-ui","RefineChakraUI",en],["@pankod/refine-react-hook-form","RefineReactHookForm",rn]],codeViewerComponent:_,loadingComponent:q,errorComponent:K,renderer:({resource:e,fields:n,isCustomPage:s,id:o})=>{var T,d;let u=V((T=e.label)!=null?T:e.name,"edit"),a=C((d=e.label)!=null?d:e.name,"Data"),p=[["Edit","@pankod/refine-chakra-ui"],["FormControl","@pankod/refine-chakra-ui"],["FormLabel","@pankod/refine-chakra-ui"],["FormErrorMessage","@pankod/refine-chakra-ui"],["useForm","@pankod/refine-react-hook-form"]],l=n.filter(c=>(c==null?void 0:c.relation)&&!(c!=null&&c.fieldable)&&(c==null?void 0:c.resource)).filter(Boolean).map(c=>{if((c==null?void 0:c.relation)&&!c.fieldable&&c.resource){p.push(["useSelect","@pankod/refine-core"]);let S=m(a,c.key,c.accessor,!1);return c.multiple&&c.accessor&&(S=`${m(a,c.key)}?.map((item: any) => ${m("item",void 0,c.accessor)})`),`
                    const { options: ${C(c.key,"Options")} } =
                    useSelect({
                        resource: "${c.resource.name}",
                        defaultValue: ${S},
                        ${J(c)}
                    });
                `}}).filter(Boolean),I=c=>{if(c.relation&&c.resource){p.push(["useSelect","@pankod/refine-core"]),p.push(["Select","@pankod/refine-chakra-ui"]);let S=C(c.key,"Options");return f`
                <FormControl mb="3" isInvalid={!!errors?.${E(c.key,void 0)}}>
                    <FormLabel>${g(c.key)}</FormLabel>
                    <Select
                        placeholder="Select ${D(c.resource.name)}"
                        {...register("${E(c.key,void 0,c.accessor)}", {
                            required: "This field is required",
                        })}
                    >
                        {${S}?.map((option) => (
                            <option value={option.value} key={option.value}>
                                {option.label}
                            </option>
                        ))}
                    </Select>
                    <FormErrorMessage>
                        {${m("(errors as any)",c.key,c.accessor,!1)}?.message as string}
                    </FormErrorMessage>
                </FormControl>
                `}},x=c=>{if(c.type==="text"||c.type==="url"||c.type==="email"||c.type==="number"||c.type==="date"||c.type==="richtext"){if(p.push(["Input","@pankod/refine-chakra-ui"]),c.multiple){let S=E(c.key,"${index}",c.accessor),$=m(m("(errors as any)",c.key,void 0,!1)+"?.[index]",void 0,c.accessor);return`
                        <>
                            {${m(a,c.key)}?.map((item: any, index: number) => (
                                <FormControl key={index} mb="3" isInvalid={!!${$}}>
                                    <FormLabel>${g(c.key)} #{index + 1}</FormLabel>
                                    <Input
                                        {...register(\`${S}\`, {
                                            required: "This field is required",
                                        })}
                                    />
                                    <FormErrorMessage>
                                        {${m($,"message")} as string}
                                    </FormErrorMessage>
                                </FormControl>
                            ))}
                        </>
                    `}return f`
                    <FormControl mb="3" isInvalid={!!${m("(errors as any)",c.key,c.accessor,!1)}}>
                        <FormLabel>${g(c.key)}</FormLabel>
                        <Input
                            ${X(c.key)?"disabled":""}
                            ${c.type!=="date"&&c.type!=="richtext"?`type="${c.type}"`:""}
                            {...register("${E(c.key,void 0,c.accessor)}", {
                                required: "This field is required",
                            })}
                        />
                        <FormErrorMessage>
                            {${m("(errors as any)",c.key,c.accessor,!1)}?.message as string}
                        </FormErrorMessage>
                    </FormControl>
                `}},b=c=>{if(c.type==="boolean"){if(p.push(["Checkbox","@pankod/refine-chakra-ui"]),c.multiple){let S=E(c.key,void 0,c.accessor),$=m("(errors as any)",c.key,void 0,!1)+"?.[index]";return`
                        <>
                            {${m(a,c.key)}?.map((item: any, index: number) => (
                                <FormControl key={index} mb="3" isInvalid={!!${$}}>
                                    <FormLabel>${g(c.key)} #{index + 1}</FormLabel>
                                    <Checkbox
                                        {...register(\`${S}.\${index}\`, {
                                            required: "This field is required",
                                        })}
                                    />
                                    <FormErrorMessage>
                                        {${$}?.message as string}
                                    </FormErrorMessage>
                                </FormControl>
                            ))}
                        </>
                    `}return f`
                    <FormControl mb="3" isInvalid={!!${m("errors",c.key,c.accessor,!1)}}>
                        <FormLabel>${g(c.key)}</FormLabel>
                        <Checkbox
                            {...register("${E(c.key,void 0,c.accessor)}", {
                                required: "This field is required",
                            })}
                        />
                        <FormErrorMessage>
                            {${m("errors",c.key,c.accessor,!1)}?.message as string}
                        </FormErrorMessage>
                    </FormControl>
                   
                `}},y=c=>{if(c.type==="date")return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-chakra-ui" package.
                        To use a <DatePicker> component, you can examine the following links:
                        
                        - https://github.com/aboveyunhai/chakra-dayzed-datepicker
                        - https://github.com/wojtekmaj/react-date-picker
                    */}
                    ${x(c)}
                    `},i=n.map(c=>{switch(c==null?void 0:c.type){case"text":case"number":case"email":case"url":case"richtext":return x(c);case"date":return y(c);case"boolean":return b(c);case"relation":return I(c);default:return}});return j(p),f`
        ${P(p)}
        
        export const ${u} = () => {
            const {
                refineCore: { formLoading, queryResult },
                saveButtonProps,
                register,
                formState: { errors },
            } = useForm(
                ${s?`
                { 
                    refineCoreProps: {
                        resource: "${e.name}",
                        id: ${o},
                        action: "edit",
                    }
                }`:""}
            );
        
            const ${a} = queryResult?.data?.data;
        
            ${l}

            return (
                <Edit isLoading={formLoading} saveButtonProps={saveButtonProps}>
                    ${i.join("")}
                </Edit>
            );
        };
        `}});var ya=({action:e,id:n,...s})=>{let{useParams:o}=nn(),{action:u,id:a}=o();switch(e!=null?e:u){case"show":return oe.createElement(Fe,{...s,id:n!=null?n:a});case"create":return oe.createElement(be,{...s,id:n!=null?n:a});case"edit":return oe.createElement(Ce,{...s,id:n!=null?n:a});default:return oe.createElement(xe,{...s,id:n!=null?n:a})}};export{be as ChakraUICreateInferencer,Ce as ChakraUIEditInferencer,ya as ChakraUIInferencer,xe as ChakraUIListInferencer,Fe as ChakraUIShowInferencer};
