// @pankod/refine-inferencer@1.6.1/dist/esm/index.js
import Z from"react";import{useResource as Gr}from"@pankod/refine-core";var zr=/(image|photo|avatar|url|logo|cover|thumbnail|icon)/i,we=(n,u)=>{let $=["name","label","title","count","content","username","nickname","login","firstName","lastName","url"];zr.test(n)&&$.unshift("url");let d=$.find(g=>g in u&&!!u[g]);if(d)return d==="firstName"&&"lastName"in u?["firstName","lastName"]:d};function Le(n,u,$){var g;if(u)return u;let d=$==null?void 0:$.find(o=>o.name===n);return(g=d==null?void 0:d.options)!=null&&g.dataProviderName?d.options.dataProviderName:"default"}var ge=n=>{var u;return(u=n==null?void 0:n.options)==null?void 0:u.dataProviderName};var Ve=n=>{let u=[...n].filter(Boolean).sort(($,d)=>(d.priority||0)-($.priority||0))[0];return u||null};import Ur from"prettier/standalone";import Wr from"prettier/parser-typescript";var J=n=>{try{return Ur.format(n,{parser:"typescript",plugins:[Wr],arrowParens:"always",trailingComma:"all",semi:!0,tabWidth:4,printWidth:80})}catch(u){return console.warn("Formatting error in inferencer",u),n}};var Ee=n=>n.replace(/ {2,}/g," ").replace(/\n{2,}/g,`
`);var Ne=/(-id|-ids|_id|_ids|Id|Ids|ID|IDs)(\[\])?$/,dr=(n,u)=>{let $=Ne.test(n),d=typeof u=="string"||typeof u=="number",g=Array.isArray(u)&&u.every(o=>typeof o=="string"||typeof o=="number");return $&&d||$&&g?{key:n,relation:!0,multiple:!!g,type:"relation",priority:1}:!1};var W=n=>n.replace(Ne,"");var k=n=>W(n).replace(/([a-z])([A-Z])/g,"$1 $2").replace(/_/g," ").replace(/-/g," ").split(" ").map(l=>l.charAt(0).toUpperCase()+l.slice(1)).join(" ");var Be=n=>n.replace(/export\s+(const|let|var|type|interface|function|class)\s+(\w+)\s*(=|:)\s*/g,"$1 $2 = ").split(`
`).filter(g=>!g.trim().startsWith("export default")).join(`
`);var yr=/import(?:(?:(?:[ \n\t]+([^ *\n\t\{\},]+)[ \n\t]*(?:,|[ \n\t]+))?([ \n\t]*\{(?:[ \n\t]*[^ \n\t"'\{\}]+[ \n\t]*,?)+\})?[ \n\t]*)|[ \n\t]*\*[ \n\t]*as[ \n\t]+([^ \n\t\{\}]+)[ \n\t]+)from[ \n\t]*(?:['"])([^'"\n]+)(?:['"])(?:;?)/g,Kr=/import[ \n\t](?:['"])([^'"\n]+)(?:['"])(?:;?)/g,_r=/((?:\w|\s|_)*)( as )((?:\w|\s|_)*)( |,)?/g,Re=(n,u={})=>{let $=n.matchAll(yr),d=new Set;for(let g of $){let[,o,l,C,I]=g;if(I in u){let A=u[I];if(I==="react")continue;o&&d.add(`const { default: ${o} } = ${A};`),l&&d.add(`const${l.replace(_r,"$1: $3$4")} = ${A};`),C&&d.add(`const ${C} = ${A};`)}}return Ee(`
    ${Array.from(d).join(`
`)}
    ${n.replace(yr,"").replace(Kr,"")}
    `)};import be from"pluralize";var Me=(n,u)=>n?u.find(d=>{let g=W(n.key);return d.name===g||d.name===(be.isPlural(g)?g:be.plural(g))||d.name===(be.isSingular(g)?g:be.singular(g))}):void 0;import Zr from"pluralize";var B=(n,u)=>{let $=n.replace(/[^a-zA-Z0-9]/g," "),d=Zr.singular($),g=k(d).replace(/ /g,""),o=u.charAt(0).toUpperCase()+u.slice(1);return/^[a-zA-Z]/.test(n)?`${g}${o}`:`${o}${g}`};var Xr=n=>n?Array.isArray(n)?n.join(""):typeof n=="string"?n:typeof n=="object"||typeof n=="function"?n.toString():typeof n=="number"?String(n):typeof n=="boolean"&&n?"true":n:"",m=(n,...u)=>n.reduce((d,g,o)=>`${d}${g}${Xr(u[o])}`,"").trim();var De=(n,u)=>`
    ${n!=null?n:""}
    
    render(typeof ${u} !== "undefined" ? <${u} /> : <></>);
    `;var Pe=n=>n==null?void 0:n.replace(/(\/\* hidden-start \*\/)(.|\n)*?(\/\* hidden-end \*\/)/g,"");var Jr=/^[a-zA-Z_$][a-zA-Z_$0-9]*$/,ke=n=>Jr.test(n);function He(n,u,$){let d=`${n}`;return u&&(d+="?.",ke(u)?d+=u:d+=`['${u}']`),$&&(d+="?.",ke($)?d+=$:d+=`['${$}']`),d}function Qr(n,u,$){return($!=null?$:[]).map(g=>He(n,u,g))}var s=(n,u,$,d=' + " " + ')=>Array.isArray($)?d?Qr(n,u,$).join(d):He(n,u,$[0]):He(n,u,$),E=(n,u,$)=>{let d=n;return u&&(d+=`.${u}`),$&&(Array.isArray($)?d+=`.${$[0]}`:d+=`.${$}`),d};var R=n=>{let u=n.reduce((d,[g,o,l])=>(d[o]||(d[o]=[]),d[o].includes(g)||(l?d[o].push([g,!0]):d[o].push(g)),d),{});return Object.entries(u).map(([d,g])=>{let o=g.find(A=>Array.isArray(A)&&A[1]),l=g.filter(A=>typeof A=="string"),C=o?`${o[0]}`:"",I=l.length?`{ ${l.join(", ")} }`:"";return`import ${C}${C&&I?", ":""}${I} from "${d}";`})};import kr from"pluralize";var U=(n="")=>kr.isSingular(n)?n:kr.singular(n);import $r from"pluralize";var $e=(n="")=>$r.isPlural(n)?n:$r.plural(n);var Yr=/^id$/i,H=n=>Yr.test(n);var K=n=>n.relationInfer&&n.relationInfer.accessor?Array.isArray(n.relationInfer.accessor)?`optionLabel: "${n.relationInfer.accessor[0]}",`:n.relationInfer.accessor!=="title"?`optionLabel: "${n.relationInfer.accessor}",`:"":"";var M=n=>{};var f=(n,u="")=>{if(Number.isInteger(Number(n)))return u.charAt(0).toLowerCase()+u.slice(1)+n;let d=n.replace(/^\d+/,"").replace(/[^a-zA-Z0-9]/g,"-"),g=k(d).replace(/ /g,"");return`${g.charAt(0).toLowerCase()+g.slice(1)}${u.charAt(0).toUpperCase()+u.slice(1)}`};var gr=n=>{let u=($,d,g,o=u)=>{let l=n.map(I=>I($,d,g,o));return Ve(l)};return u};var hr=n=>($,d,g,o,l)=>n.reduce((C,I)=>I(C,d,g,o,l),$);var Ir=(n,u,$,d)=>{let g=Array.isArray(u),o=Array.isArray(u)&&u.every(l=>typeof l=="string"||typeof l=="number");if(g){if(!o){let C=d(n,u[0],$,d);return C?{...C,key:n,multiple:!0,priority:1}:!1}let l=d(n,u[0],$,d)||{type:"string"};return{key:n,multiple:!0,priority:1,type:l==null?void 0:l.type}}return!1};var fr=(n,u)=>typeof u=="boolean"?{key:n,type:"boolean"}:!1;import xr from"dayjs";var en=/(_at|_on|At|On|AT|ON)(\[\])?$/,Fr=(n,u)=>{let $=en.test(n)&&xr(u).isValid(),d=typeof u=="string"&&xr(u).isValid();return $||d?{key:n,type:"date",priority:1}:!1};var rn=/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,br=(n,u)=>typeof u=="string"&&rn.test(u)?{key:n,type:"email",priority:1}:!1;var nn=/\.(gif|jpe?g|tiff?|png|webp|bmp|svg)$/i,Cr=(n,u)=>typeof u=="string"&&nn.test(u)?{key:n,type:"image",priority:2}:!1;var vr=(n,u)=>typeof u>"u"||u===null?null:!1;var Tr=(n,u)=>typeof u=="number"?{key:n,type:"number"}:!1;var tn=/id$/i,Ar=(n,u,$,d)=>{let g=u!==null,o=!Array.isArray(u);if(g&&o&&typeof u=="object"){if(Object.keys(u).length===1&&tn.test(Object.keys(u)[0]))return{key:n,type:"relation",relation:!0,accessor:"id",priority:1};let I=we(n,u);if((I?I.length>0:!1)&&I){let w=I&&Array.isArray(I)?I[0]:I,L=d(w,u[w],u,d);if(L){let T=Array.isArray(I)?I.map(p=>L.accessor?`${p}.${Array.isArray(L.accessor)?L.accessor[0]:L.accessor}`:p):L.accessor?Array.isArray(L.accessor)?`${I}.${L.accessor[0]}`:`${I}.${L.accessor}`:I;return{...L,fieldable:!0,key:n,accessor:T,priority:1}}}return{key:n,fieldable:!1,type:"object",priority:1}}return!1};var Sr=(n,u)=>typeof u=="string"&&u.length>100?{key:n,type:"richtext",priority:1}:!1;var wr=(n,u)=>typeof u=="string"?{key:n,type:"text"}:!1;var on=/^(https?|ftp):\/\/(-\.)?([^\s/?\.#-]+\.?)+(\/[^\s]*)?$/i,Lr=(n,u)=>typeof u=="string"&&on.test(u)?{key:n,type:"url",priority:1}:!1;var Vr=[Ir,fr,Fr,br,Cr,vr,Tr,Ar,dr,Sr,wr,Lr];var Er=(n,u,$,d)=>n.map(o=>{if(!o.relation&&(o.type==="text"||o.type==="richtext"||o.type==="number")&&!o.canRelation){let l=/^[a-z0-9-]+$/,C=T=>l.test(`${T}`),I=o.key.toLowerCase()!=="id";return!(!o.accessor||typeof o.accessor=="string")||!I?o:(o.multiple?d[o.key]:[d[o.key]]).every(T=>C(o.accessor?T[o.accessor]:T))?{...o,canRelation:!0}:o}return o});var an=/(image|photo|avatar|cover|thumbnail|icon)/i,Nr=n=>n.map($=>$.type==="url"&&an.test($.key)?{...$,type:"image"}:$);var Br=(n,u)=>n.map(d=>{let g=Me(d,u);return g?{...d,relation:!0,type:"relation",resource:g}:d});var Rr=(n,u,$,d,g)=>n.map(l=>{if(l.relation&&l.type==="relation"&&!l.resource){let C=l.accessor?d[l.key][l.accessor]:d[l.key],I=g(l.key,C,d,g);if(I&&I.type!=="relation")return{...l,fieldable:!0,relation:!1,type:I.type,canRelation:!0}}return l});var Mr=[Nr,Br,Rr,Er];import he,{useEffect as sn}from"react";import cn,{defaultProps as un}from"prism-react-renderer";import pn from"prism-react-renderer/themes/vsDark";var Q=({code:n,wrapperProps:u})=>(sn(()=>{var d;let $=document.createElement("style");$.appendChild(document.createTextNode(`
                #code-highlight::-webkit-scrollbar {-webkit-appearance: none; height: 7px; width: 7px;}
                #code-highlight::-webkit-scrollbar-thumb {background-color: #666b7a;}
                #code-highlight::-webkit-scrollbar-corner {background-color: rgb(30, 30, 30);}
                `)),(d=document.getElementById("code-highlight"))==null||d.appendChild($)},[]),he.createElement("div",{id:"code-highlight",style:{maxHeight:"75vh",height:"100%",overflow:"auto",backgroundColor:"rgb(30, 30, 30)"},...u!=null?u:{}},he.createElement(cn,{...un,theme:pn,code:n,language:"tsx"},({className:$,style:d,tokens:g,getLineProps:o,getTokenProps:l})=>he.createElement("pre",{className:$,style:{...d,padding:"14px 14px 14px 14px",margin:"0",width:"100%",boxSizing:"border-box"}},g.map((C,I)=>he.createElement("div",{key:I,...o({line:C,key:I})},C.map((A,w)=>he.createElement("span",{key:w,...l({token:A,key:w})}))))))));import _ from"react";import*as mn from"@pankod/refine-core";import{LivePreview as ln,LiveProvider as dn}from"@aliemir/react-live";var Dr=[["react","React",_],["@pankod/refine-core","RefineCore",mn]],Oe=_.createContext({}),Ge=({fetchError:n,code:u,additionalScope:$,errorComponent:d})=>{var C;let g=(C=u&&Be(Re(u,[...Dr,...$!=null?$:[]].reduce((I,[A,w,L,T])=>T?I:{...I,[A]:w},{}))))!=null?C:"",o=_.useMemo(()=>({React:_,..._,...[...Dr,...$!=null?$:[]].reduce((I,[A,w,L])=>({...I,[w]:L}),{})}),[$]),l=_.useMemo(()=>()=>{let{error:A}=_.useContext(Oe);return d?_.createElement(d,{error:A!=null?A:n?"Something went wrong while fetching the resource data.":void 0}):null},[d,n]);return _.createElement(dn,{Context:Oe,code:g,scope:o,noInline:!0},!n&&_.createElement(ln,{Context:Oe}),_.createElement(l,null))};import Ie from"react";import{useDataProvider as yn,useResource as kn}from"@pankod/refine-core";var Pr=(n,u,$)=>{let{resource:d,resourceName:g,id:o,resources:l}=kn({resourceNameOrRouteName:u}),C=$!=null?$:o,I=yn(),[A,w]=Ie.useState(void 0),[L,T]=Ie.useState(!0),[p,V]=Ie.useState(!1),y=Ie.useCallback(async c=>{var v,t;let S=(v=ge(d))!=null?v:Le(g,void 0,l),a=I(S);V(!0);try{if(n==="list"||n==="create"){let r=await a.getList({resource:g});w((t=r.data)==null?void 0:t[0]),setTimeout(()=>{V(!1)},500)}if(n==="edit"||n==="show"&&c){let r=await a.getOne({resource:g,id:c});w(r.data),setTimeout(()=>{V(!1)},500)}}catch{setTimeout(()=>{V(!1)},500)}},[n,I,d,g,l]);return Ie.useEffect(()=>{T(!1),!p&&!A&&y(C)},[y,C]),{data:A,loading:p,initial:L}};import fe from"react";import{useDataProvider as $n}from"@pankod/refine-core";import{get as Hr}from"lodash";var Or=({record:n,fields:u,infer:$})=>{let d=$n(),[g,o]=fe.useState([]),[l,C]=fe.useState(!0),[I,A]=fe.useState(!1),w=fe.useCallback(async L=>{console.groupCollapsed("@pankod/refine-inferencer is trying to detect relations");let T=[];A(!0);try{let p=L.map(async y=>{if(y&&(y.relation||y.canRelation)&&n){let c=ge(y.resource),S=d(c),a=Array.isArray(y.accessor)?void 0:y.multiple?n[y.key].map(v=>y.accessor?Hr(v,y.accessor):v)[0]:y.accessor?Hr(n[y.key],y.accessor):n[y.key];if(a&&y.resource)try{let{data:v}=await S.getOne({resource:y.resource.name,id:a});T.push({status:"success",resource:y.resource.name,field:y.key});let t=$("__",v,{},$);return{...y,relationInfer:t}}catch{return T.push({status:"error",resource:y.resource.name,field:y.key}),{...y,relationInfer:null}}if(a){let v,t;try{let{data:e}=await S.getOne({resource:$e(W(y.key)),id:a});T.push({status:"success",resource:$e(W(y.key)),field:y.key}),v=e,t=!0}catch{T.push({status:"error",resource:$e(W(y.key)),field:y.key});try{let{data:h}=await S.getOne({resource:U(W(y.key)),id:a});T.push({status:"success",resource:U(W(y.key)),field:y.key}),v=h,t=!1}catch{return T.push({status:"error",resource:U(W(y.key)),field:y.key}),{...y,relationInfer:null}}}let r=$("__",v,{},$),i=W(y.key);return{...y,relation:!0,type:"relation",resource:{name:t?$e(i):U(i)},fieldable:!1,canRelation:void 0,relationInfer:r}}return{...y,relationInfer:null}}return y}),V=await Promise.all(p);o(V.filter(y=>y)),setTimeout(()=>{A(!1)},500)}catch{setTimeout(()=>{A(!1)},500)}setTimeout(()=>{console.log(`Tried to detect relations with ${T.length} attempts and succeeded with ${T.filter(p=>p.status==="success").length} attempts.`),console.groupEnd()},500)},[d,n]);return fe.useEffect(()=>{C(!1),!I&&u&&u.length>0&&g.length===0&&w(u)},[w,I,u,w]),{fields:g,loading:I,initial:l}};var D=({type:n,additionalScope:u=[],customElements:$=[],fieldTransformers:d=[],renderer:g,loadingComponent:o,errorComponent:l,codeViewerComponent:C})=>{let I=gr([...Vr,...$]),A=hr([...Mr,...d]),w=({resourceName:T,fieldTransformer:p,id:V})=>{var x;let{resource:y,resources:c}=Gr({resourceNameOrRouteName:T}),{resource:S}=Gr(),{data:a,loading:v,initial:t}=Pr(n,T!=null?T:y==null?void 0:y.name,V),r=Z.useMemo(()=>{if(a){let b=Object.keys(a).map(q=>{let z=a[q];return I(q,z,a,I)}).filter(Boolean),F=A(b,c,y,a,I);return p?F.flatMap(q=>{let z=p(q);return z?[z]:[]}):F}return[]},[a,c,y,p]),{fields:i,loading:e}=Or({record:a,fields:r,infer:I}),h=Z.useMemo(()=>!v&&!e&&!t?g({resource:y,resources:c,fields:i,infer:I,isCustomPage:y.name!==S.name,id:V}):"",[y,c,i,v,e]);return Z.createElement(Z.Fragment,null,o&&(v||e)&&Z.createElement(o,null),!v&&!e&&Z.createElement(Z.Fragment,null,Z.createElement(Ge,{fetchError:!v&&!t&&!a,code:De(h,B((x=y.label)!=null?x:y.name,n)),errorComponent:l,additionalScope:u}),C&&Z.createElement(C,{code:Pe(h),loading:v||e})))};return({name:T,resource:p,fieldTransformer:V,id:y})=>Z.createElement(w,{fieldTransformer:V,resourceName:p!=null?p:T,key:p!=null?p:T,id:y})};import Ce from"react";import{useRouterContext as Vn}from"@pankod/refine-core";import*as vn from"@pankod/refine-antd";import je from"react";import{Alert as gn,Row as hn,Col as In}from"@pankod/refine-antd";var ee=({error:n})=>n?je.createElement(hn,{justify:"center",align:"middle",style:{minHeight:"200px"}},je.createElement(In,null,je.createElement(gn,{message:"Error",description:n!=null?n:"",type:"error",showIcon:!0}))):null;import qe from"react";import{Spin as fn,Row as xn,Col as Fn}from"@pankod/refine-antd";var re=()=>qe.createElement(xn,{justify:"center",align:"middle",style:{minHeight:"300px"}},qe.createElement(Fn,null,qe.createElement(fn,{size:"large"})));import j from"react";import{Button as ze,Modal as bn,Icons as xe,Space as Cn}from"@pankod/refine-antd";var ne=({code:n,loading:u})=>{let $=j.useMemo(()=>J(n!=null?n:""),[n]),d=j.useRef(null),[g,o]=j.useState(!1),[l,C]=j.useState(!1);if(u)return null;let I=()=>{var A,w,L,T;(A=d==null?void 0:d.current)==null||A.select(),(w=d==null?void 0:d.current)==null||w.setSelectionRange(0,Number.MAX_SAFE_INTEGER),typeof navigator<"u"&&(navigator.clipboard.writeText((T=(L=d==null?void 0:d.current)==null?void 0:L.value)!=null?T:""),C(!0),setTimeout(()=>{C(!1)},1e3))};return $?j.createElement(j.Fragment,null,j.createElement("div",{style:{position:"fixed",bottom:12,right:60}},j.createElement(Cn,null,j.createElement(ze,{type:"primary",onClick:()=>o(!0),icon:j.createElement(xe.CodeOutlined,null)},"Show Code"),j.createElement(ze,{type:"default",target:"_blank",href:"https://github.com/refinedev/refine/discussions/3046",icon:j.createElement(xe.MessageOutlined,null)},"Give Feedback"))),j.createElement("div",{style:{height:"50px"}}),j.createElement(bn,{width:800,visible:g,onCancel:()=>o(!1),closeIcon:j.createElement(xe.CloseOutlined,{style:{color:"#666b7a"}}),footer:null,bodyStyle:{padding:0}},j.createElement(Q,{code:$}),j.createElement("textarea",{ref:d,defaultValue:($!=null?$:"").replace(/\\n/g,`\r
`),id:"code-input","aria-hidden":"true",style:{padding:0,margin:0,width:0,height:0,opacity:0,border:"none",display:"block"}}),j.createElement(ze,{style:{position:"absolute",bottom:12,right:12,width:100},key:"copy",type:"default",icon:l?j.createElement(xe.CheckOutlined,{style:{color:"green"}}):j.createElement(xe.CopyOutlined,null),onClick:I},l?"Copied":"Copy"))):null};var Ue=D({type:"show",additionalScope:[["@pankod/refine-antd","RefineAntd",vn]],codeViewerComponent:ne,loadingComponent:re,errorComponent:ee,renderer:({resource:n,fields:u,isCustomPage:$,id:d})=>{var v;let g=B((v=n.label)!=null?v:n.name,"show"),o="record",l=[["React","react",!0],["IResourceComponentsProps","@pankod/refine-core"],["useShow","@pankod/refine-core"],["Show","@pankod/refine-antd"],["Typography","@pankod/refine-antd"]],I=u.filter(t=>(t==null?void 0:t.relation)&&!(t!=null&&t.fieldable)&&(t==null?void 0:t.resource)).filter(Boolean).map(t=>{if((t==null?void 0:t.relation)&&!t.fieldable&&t.resource){if(t.multiple){l.push(["useMany","@pankod/refine-core"]);let r=s(o,t.key);return t.accessor&&(r=`${s(o,t.key)}?.map((item: any) => ${s("item",void 0,t.accessor)})`),`
                    const { data: ${f(t.key,"Data")}, isLoading: ${f(t.key,"IsLoading")} } =
                    useMany({
                        resource: "${t.resource.name}",
                        ids: ${r} || [],
                        queryOptions: {
                            enabled: !!${o},
                        },
                    });
                    `}return l.push(["useOne","@pankod/refine-core"]),`
                    const { data: ${f(t.key,"Data")}, isLoading: ${f(t.key,"IsLoading")} } =
                    useOne({
                        resource: "${t.resource.name}",
                        id: ${s(o,t.key,t.accessor,!1)} || "",
                        queryOptions: {
                            enabled: !!${o},
                        },
                    });
                `}}).filter(Boolean),A=t=>{if(t.relation&&t.resource){let r=f(t.key,"Data"),i=f(t.key,"IsLoading");return t.multiple?(l.push(["TagField","@pankod/refine-antd"]),m`
                    <Title level={5}>${k(t.key)}</Title>
                    {${i} ? <>Loading...</> : (
                        <>
                        ${(()=>{var e,h;if(t.relationInfer)if((e=t.relationInfer)!=null&&e.accessor){if(Array.isArray(t.relationInfer.accessor))return"Not Handled.";{let x=U((h=t.resource)==null?void 0:h.name),b=s(x,void 0,t.relationInfer.accessor);return`{${r}?.data?.map((${x}: any) => <TagField key={${b}} value={${b}} />)}`}}else return;else return})()}
                        </>
                    )}
                    `):m`
                    <Title level={5}>${k(t.key)}</Title>
                    {${i} ? <>Loading...</> : (
                        <>
                        ${(()=>{var e;return t.relationInfer?(e=t.relationInfer)!=null&&e.accessor?Array.isArray(t.relationInfer.accessor)?`{${s(`${r}?.data`,void 0,t.relationInfer.accessor,' + " " + ')}}`:`{${r}?.data?.${t.relationInfer.accessor}}`:`{${r}?.data}`:`{${r}?.data?.id}`})()}
                        </>
                    )}
                    
                    `}},w=t=>{if(t.type==="text"){if(l.push(["TagField","@pankod/refine-antd"],["TextField","@pankod/refine-antd"]),t.multiple){let r=s("item",void 0,t.accessor);return m`
                    <Title level={5}>${k(t.key)}</Title>
                    {${s(o,t.key)}?.map((item: any) => (
                        <TagField value={${r}} key={${r}} />
                    ))}
                `}return m`
                    <Title level={5}>${k(t.key)}</Title>
                    <TextField value={${s(o,t.key,t.accessor)}} />
                `}},L=t=>{if(t.type==="image"){if(l.push(["ImageField","@pankod/refine-antd"]),t.multiple){let r=s("item",void 0,t.accessor);return m`
                    <Title level={5}>${k(t.key)}</Title>
                    {${s(o,t.key)}?.map((item: any) => (
                        <ImageField style={{ maxWidth: 200 }} value={${r}} key={${r}} />
                    ))}
                `}return m`
                    <Title level={5}>${k(t.key)}</Title>
                    <ImageField style={{ maxWidth: 200 }} value={${s(o,t.key,t.accessor," + ")}} />
                    `}},T=t=>{if(t.type==="email"){if(l.push(["TagField","@pankod/refine-antd"],["EmailField","@pankod/refine-antd"]),t.multiple){let r=s("item",void 0,t.accessor);return m`
                    <Title level={5}>${k(t.key)}</Title>
                    <>
                    {${s(o,t.key)}?.map((item: any) => (
                        <TagField value={${r}} key={${r}} />
                    ))}
                    </>
                `}return m`
                    <Title level={5}>${k(t.key)}</Title>
                    <EmailField value={${s(o,t.key,t.accessor," + ")}} />
                `}},p=t=>{if(t.type==="url"){if(l.push(["TagField","@pankod/refine-antd"],["UrlField","@pankod/refine-antd"]),t.multiple){let r=s("item",void 0,t.accessor);return m`
                    <Title level={5}>${k(t.key)}</Title>
                    {${s(o,t.key)}?.map((item: any) => (
                        <TagField value={${r}} key={${r}} />
                    ))}
                `}return m`
                    <Title level={5}>${k(t.key)}</Title>
                    <UrlField value={${s(o,t.key,t.accessor," + ")}} />
                `}},V=t=>{if(t.type==="boolean"){if(l.push(["TagField","@pankod/refine-antd"],["BooleanField","@pankod/refine-antd"]),t.multiple){let r=s("item",void 0,t.accessor);return m`
                    <Title level={5}>${k(t.key)}</Title>
                    {(${s(o,t.key)} as any[])?.map((item, index) => (
                        <BooleanField value={${r}} key={index} />
                    ))}
                `}return m`
                    <Title level={5}>${k(t.key)}</Title>
                    <BooleanField value={${s(o,t.key,t.accessor," && ")}} />
                `}},y=t=>{if(t.type==="date"){if(l.push(["DateField","@pankod/refine-antd"]),t.multiple){let r=s("item",void 0,t.accessor);return m`
                    <Title level={5}>${k(t.key)}</Title>
                    {${s(o,t.key)}?.map((item: any) => (
                        <DateField value={${r}} key={${r}} />
                    ))}
                `}return m`
                    <Title level={5}>${k(t.key)}</Title>
                    <DateField value={${s(o,t.key,t.accessor," + ' ' + ")}} />
                `}},c=t=>{if(t.type==="richtext")return l.push(["MarkdownField","@pankod/refine-antd"]),m`
                    <Title level={5}>${k(t.key)}</Title>
                    <MarkdownField value={${s(o,t.key,t.accessor,' + " " + ')}} />
                `},S=t=>{if(t.type==="number"){if(l.push(["NumberField","@pankod/refine-antd"]),t.multiple){let r=s("item",void 0,t.accessor);return m`
                    <Title level={5}>${k(t.key)}</Title>
                    {${s(o,t.key)}?.map((item: any) => (
                        <TagField value={${r}} key={${r}} />
                    ))}
                `}return m`
                    <Title level={5}>${k(t.key)}</Title>
                    <NumberField value={${s(o,t.key,t.accessor,' + " " + ')} ?? ""} />
                `}},a=u.map(t=>{switch(t==null?void 0:t.type){case"text":return w(t);case"number":return S(t);case"richtext":return c(t);case"email":return T(t);case"image":return L(t);case"date":return y(t);case"boolean":return V(t);case"url":return p(t);case"relation":return A(t);default:return}});return M(l),m`
        ${R(l)}
        
        const { Title } = Typography;

        export const ${g}: React.FC<IResourceComponentsProps> = () => {
            const { queryResult } = useShow(${$?`{ 
                        resource: "${n.name}", 
                        id: ${d}
                    }`:""});
            const { data, isLoading } = queryResult;
        
            const ${o} = data?.data;

            ${I}

            return (
                <Show isLoading={isLoading}>
                    ${a.join("")}
                </Show>
            );
        };
        `}});import*as Tn from"@pankod/refine-antd";var We=D({type:"list",additionalScope:[["@pankod/refine-antd","RefineAntd",Tn]],codeViewerComponent:ne,loadingComponent:re,errorComponent:ee,renderer:({resource:n,fields:u,isCustomPage:$})=>{var r;let d=B((r=n.label)!=null?r:n.name,"list"),g="tableProps?.dataSource",o=[["React","react",!0],["IResourceComponentsProps","@pankod/refine-core"],["BaseRecord","@pankod/refine-core"],["useTable","@pankod/refine-antd"],["List","@pankod/refine-antd"],["Table","@pankod/refine-antd"],["Space","@pankod/refine-antd"]],C=u.filter(i=>(i==null?void 0:i.relation)&&!(i!=null&&i.fieldable)&&(i==null?void 0:i.resource)).filter(Boolean).map(i=>{if((i==null?void 0:i.relation)&&!i.fieldable&&i.resource){o.push(["useMany","@pankod/refine-core"]);let e="";return i.multiple?e=`[].concat(...(${g}?.map((item) => ${s("item",i.key,i.accessor,!1)}) ?? []))`:e=`${g}?.map((item) => ${s("item",i.key,i.accessor,!1)}) ?? []`,`
                    const { data: ${f(i.key,"Data")}, isLoading: ${f(i.key,"IsLoading")} } =
                    useMany({
                        resource: "${i.resource.name}",
                        ids: ${e},
                        queryOptions: {
                            enabled: !!${g},
                        },
                    });
                    `}}).filter(Boolean),I=i=>{var e,h;if(i.relation&&i.resource){let b=`${f(i.key,"Data")}?.data`,F=f(i.key,"IsLoading");if(Array.isArray(i.accessor))return;let N=`${F} ? <>Loading...</> : `,q=i.multiple?`dataIndex="${i.key}"`:`dataIndex={["${i.key}", ${i.accessor?`"${i.accessor}"`:""}]}`,z=`title="${k(i.key)}"`,ye="";if(i.multiple){o.push(["TagField","@pankod/refine-antd"]);let X="item";if(i!=null&&i.relationInfer){let Y=`${b}?.find((resourceItems) => resourceItems.id === ${s("item",void 0,i.accessor)})`;X=s(Y,void 0,(e=i==null?void 0:i.relationInfer)==null?void 0:e.accessor)}ye=`render={(value: any[]) => ${N} (
                        <>
                            {${s("value",void 0)}?.map((item, index) => (
                                <TagField key={index} value={${X}} />
                            ))}
                        </>
                    )}`}else if(i!=null&&i.relationInfer){let X=`${b}?.find((item) => item.id === value)`,Y=s(X,void 0,(h=i==null?void 0:i.relationInfer)==null?void 0:h.accessor);ye=`render={(value) => ${N} ${Y}}`}else ye="";return m`<Table.Column ${q} ${z} ${ye} />`}},A=i=>{if(i.type==="image"){o.push(["ImageField","@pankod/refine-antd"]);let e=Array.isArray(i.accessor)||i.multiple?`dataIndex="${i.key}"`:`dataIndex={["${i.key}", ${i.accessor?`"${i.accessor}"`:""}]}`,h=`title="${k(i.key)}"`,x=m`render={(value: any) => <ImageField style={{ maxWidth: "100px" }} value={${s("value",void 0,Array.isArray(i.accessor)?i.accessor:void 0," + ")}} />}`;if(i.multiple){let b=s("item",void 0,i.accessor," + ");x=m`render={(value: any[]) => (<>{value?.map((item, index) => (
                        <ImageField style={{ maxWidth: "100px" }} value={${b}} key={index} />
                    ))}</>)}`}return m`<Table.Column ${e} ${h} ${x} />`}},w=i=>{if(i.type==="email"){o.push(["TagField","@pankod/refine-antd"],["EmailField","@pankod/refine-antd"]);let e=Array.isArray(i.accessor)||i.multiple?`dataIndex="${i.key}"`:`dataIndex={["${i.key}", ${i.accessor?`"${i.accessor}"`:""}]}`,h=`title="${k(i.key)}"`,x=m`render={(value: any) => <EmailField value={${s("value",void 0,Array.isArray(i.accessor)?i.accessor:void 0,' + " " + ')}} />}`;if(i.multiple){let b=s("item",void 0,i.accessor,' + " " + ');x=m`render={(value: any[]) => (<>{value?.map((item, index) => (
                        <TagField value={${b}} key={index} />
                    ))}</>)}`}return m`<Table.Column ${e} ${h} ${x} />`}},L=i=>{if(i.type==="url"){o.push(["UrlField","@pankod/refine-antd"],["TagField","@pankod/refine-antd"]);let e=Array.isArray(i.accessor)||i.multiple?`dataIndex="${i.key}"`:`dataIndex={["${i.key}", ${i.accessor?`"${i.accessor}"`:""}]}`,h=`title="${k(i.key)}"`,x=m`render={(value: any) => <UrlField value={${s("value",void 0,Array.isArray(i.accessor)?i.accessor:void 0," + ")}} />}`;if(i.multiple){let b=s("item",void 0,i.accessor," + ");x=m`render={(value: any[]) => (<>{value?.map((item, index) => (
                        <TagField value={${b}} key={index} />
                    ))}</>)}`}return m`<Table.Column ${e} ${h} ${x} />`}},T=i=>{if((i==null?void 0:i.type)==="boolean"){o.push(["BooleanField","@pankod/refine-antd"]);let e=Array.isArray(i.accessor)||i.multiple?`dataIndex="${i.key}"`:`dataIndex={["${i.key}", ${i.accessor?`"${i.accessor}"`:""}]}`,h=`title="${k(i.key)}"`,x=m`render={(value: any) => <BooleanField value={${s("value",void 0,Array.isArray(i.accessor)?i.accessor:void 0," && ")}} />}`;if(i.multiple){let b=s("item",void 0,i.accessor," && ");x=m`render={(value: any[]) => (<>{value?.map((item, index) => (
                        <BooleanField value={${b}} key={index} />
                    ))}</>)}`}return m`<Table.Column ${e} ${h} ${x} />`}},p=i=>{if(i.type==="date"){o.push(["DateField","@pankod/refine-antd"]);let e=Array.isArray(i.accessor)||i.multiple?`dataIndex="${i.key}"`:`dataIndex={["${i.key}", ${i.accessor?`"${i.accessor}"`:""}]}`,h=`title="${k(i.key)}"`,x=m`render={(value: any) => <DateField value={${s("value",void 0,Array.isArray(i.accessor)?i.accessor:void 0,' + " " + ')}} />}`;if(i.multiple){let b=s("item",void 0,i.accessor,' + " " + ');x=m`render={(value: any[]) => (<>{value?.map((item, index) => (
                        <DateField value={${b}} key={index} />
                    ))}</>)}`}return m`<Table.Column ${e} ${h} ${x} />`}},V=i=>{if((i==null?void 0:i.type)==="richtext"){o.push(["MarkdownField","@pankod/refine-antd"]);let e=Array.isArray(i.accessor)||i.multiple?`dataIndex="${i.key}"`:i.accessor?`dataIndex={["${i.key}", "${i.accessor}"]}`:`dataIndex="${i.key}"`,h=`title="${k(i.key)}"`,x=m`render={(value: any) => <MarkdownField value={(${s("value",void 0,Array.isArray(i.accessor)?i.accessor:void 0)}).slice(0, 80) + "..."} />}`;if(i.multiple){let b=s("item",void 0,i.accessor,' + " " + ');x=m`render={(value: any[]) => (<>{value?.map((item, index) => (
                        <MarkdownField value={(${b})?.slice(0, 80) + "..."} key={index} />
                    ))}</>)}`}return m`<Table.Column ${e} ${h} ${x} />`}},y=i=>{if(i&&(i.type==="text"||i.type==="number")){let e=i.accessor&&!Array.isArray(i.accessor)&&!i.multiple?`dataIndex={["${i.key}", "${i.accessor}"]}`:`dataIndex="${i.key}"`,h=`title="${k(i.key)}"`,x="";if(i.multiple){o.push(["TagField","@pankod/refine-antd"]);let b=s("item",void 0,i.accessor,' + " " + ');x=`render={(value: any[]) => (<>{value?.map((item) => (
                        <TagField value={${b}} key={${b}} />
                    ))}</>)}`}return!i.multiple&&Array.isArray(i.accessor)&&(x=`render={(value: any) => (<>{${s("value",void 0,i.accessor)}}</>)}`),`<Table.Column ${e} ${h} ${x} />`}},{canEdit:c,canShow:S,canDelete:a}=n!=null?n:{};c&&o.push(["EditButton","@pankod/refine-antd"]),S&&o.push(["ShowButton","@pankod/refine-antd"]),a&&o.push(["DeleteButton","@pankod/refine-antd"]);let v=c||S||a?m`
                <Table.Column
                    title="Actions"
                    dataIndex="actions"
                    render={(_, record: BaseRecord) => (
                        <Space>
                        ${c?m`
                            <EditButton
                                hideText
                                size="small"
                                recordItemId={record.id}
                            />
                            `:""}
                        ${S?m`
                            <ShowButton
                                hideText
                                size="small"
                                recordItemId={record.id}
                            />
                            `:""}
                        ${a?m`
                            <DeleteButton
                                hideText
                                size="small"
                                recordItemId={record.id}
                            />
                            `:""}
                        </Space>
                    )}
                />
            `:"",t=u.map(i=>{switch(i==null?void 0:i.type){case"text":case"number":return y(i);case"richtext":return V(i);case"email":return w(i);case"image":return A(i);case"date":return p(i);case"boolean":return T(i);case"url":return L(i);case"relation":return I(i);default:return}});return M(o),m`
        ${R(o)}
        
        export const ${d}: React.FC<IResourceComponentsProps> = () => {
            const { tableProps } = useTable({
                syncWithLocation: true,
                ${$?` resource: "${n.name}",`:""}
            });
        
            ${C}

            return (
                <List>
                    <Table {...tableProps} rowKey="id">
                        ${t.join(`\r
`)}
                        ${v}
                    </Table>
                </List>
            );
        };
        `}});import*as An from"@pankod/refine-antd";import Sn from"dayjs";var Ke=D({type:"create",additionalScope:[["@pankod/refine-antd","RefineAntd",An],["dayjs","dayjs",Sn,!0]],codeViewerComponent:ne,loadingComponent:re,errorComponent:ee,renderer:({resource:n,fields:u,isCustomPage:$})=>{var V;let d=B((V=n.label)!=null?V:n.name,"create"),g=[["React","react",!0],["IResourceComponentsProps","@pankod/refine-core"],["Create","@pankod/refine-antd"],["Form","@pankod/refine-antd"],["useForm","@pankod/refine-antd"],["Input","@pankod/refine-antd"]],l=u.filter(y=>(y==null?void 0:y.relation)&&!(y!=null&&y.fieldable)&&(y==null?void 0:y.resource)).filter(Boolean).map(y=>{if((y==null?void 0:y.relation)&&!y.fieldable&&y.resource)return g.push(["useSelect","@pankod/refine-antd"]),`
                    const { selectProps: ${f(y.key,"SelectProps")} } =
                    useSelect({
                        resource: "${y.resource.name}",
                        ${K(y)}
                    });
                `}).filter(Boolean),C=y=>{if(y.relation&&y.resource){g.push(["Select","@pankod/refine-antd"]);let c=f(y.key,"SelectProps"),S=y.accessor?y.multiple?`"${y.key}"`:`["${y.key}", "${y.accessor}"]`:`"${y.key}"`,a="",v="";return y.accessor&&y.multiple&&(v=`getValueFromEvent={(selected: string[]) => {
                        return selected?.map((item) => ({ ${ke(`${y.accessor}`)?y.accessor:`["${y.accessor}"]`}: item }));
                    }}`,a=`getValueProps={(value: any[]) => {
                        return {
                            value: value?.map((item) => ${s("item",void 0,y.accessor)}),
                        };
                    }}`),m`
                    <Form.Item
                        label="${k(y.key)}"
                        name={${S}}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                        ${a}
                        ${v}
                    >
                        <Select ${y.multiple?'mode="multiple"':""} {...${c}} />
                    </Form.Item>             
                    `}},I=y=>{if(y.type==="text"||y.type==="url"||y.type==="email"||y.type==="number")return H(y.key)||y.multiple?void 0:m`
                    <Form.Item
                        label="${k(y.key)}"
                        name={["${y.key}"${y.accessor?', "'+y.accessor+'"':""}]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                    >
                        <Input  />
                    </Form.Item>
                `},A=y=>{if(y.type==="image"){g.push(["Upload","@pankod/refine-antd"],["getValueFromEvent","@pankod/refine-antd"]);let c='valuePropName="fileList"';return y.multiple&&!y.accessor&&(c="getValueProps={(value) => ({ fileList: value?.map((item: any) => ({ url: item, name: item, uid: item }))})}"),y.multiple||(y.accessor?c="getValueProps={(value) => ({ fileList: value ? [value] : [] })}":c="getValueProps={(value) => ({ fileList: [{ url: value, name: value, uid: value }]})}"),m`
                    <Form.Item label="${k(y.key)}">
                        <Form.Item
                            name="${y.key}"
                            ${c}
                            getValueFromEvent={getValueFromEvent}
                            noStyle
                            rules={[
                                {
                                    required: true,
                                },
                            ]}
                        >
                            <Upload.Dragger
                                listType="picture"
                                ${y.multiple?"multiple":""}
                                beforeUpload={() => false}
                            >
                                <p className="ant-upload-text">
                                    Drag & drop a file in this area
                                </p>
                            </Upload.Dragger>
                        </Form.Item>
                    </Form.Item>
                    `}},w=y=>{if(y.type==="boolean")return g.push(["Checkbox","@pankod/refine-antd"]),y.multiple?void 0:m`
                    <Form.Item
                        label="${k(y.key)}"
                        valuePropName="checked"
                        name={["${y.key}"${y.accessor?', "'+y.accessor+'"':""}]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                    >
                        <Checkbox>${k(y.key)}</Checkbox>
                    </Form.Item>
                `},L=y=>{if(y.type==="date")return g.push(["DatePicker","@pankod/refine-antd"],["dayjs","dayjs",!0]),y.multiple?void 0:m`
                    <Form.Item
                        label="${k(y.key)}"
                        name={["${y.key}"${y.accessor?', "'+y.accessor+'"':""}]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                        getValueProps={(value) => ({ value: value ? dayjs(value) : undefined })}
                    >
                        <DatePicker />
                    </Form.Item>
                `},T=y=>{if(y.type==="richtext")return m`
                <Form.Item
                    label="${k(y.key)}"
                    name="${y.key}"
                    rules={[
                        {
                            required: true,
                        },
                    ]}
                >
                    <Input.TextArea rows={5} />
                </Form.Item>
                `},p=u.map(y=>{switch(y==null?void 0:y.type){case"text":case"number":case"email":case"url":return I(y);case"richtext":return T(y);case"image":return A(y);case"date":return L(y);case"boolean":return w(y);case"relation":return C(y);default:return}});return M(g),m`
        ${R(g)}
        
        export const ${d}: React.FC<IResourceComponentsProps> = () => {
            const { formProps, saveButtonProps, queryResult } = useForm(${$?`{
                          resource: "${n.name}",
                          action: "create",
                      }`:""});
        
            ${l}

            return (
                <Create saveButtonProps={saveButtonProps}>
                    <Form {...formProps} layout="vertical">
                        ${p.join("")}
                    </Form>
                </Create>
            );
        };
        `}});import*as wn from"@pankod/refine-antd";import Ln from"dayjs";var _e=D({type:"edit",additionalScope:[["@pankod/refine-antd","RefineAntd",wn],["dayjs","dayjs",Ln,!0]],codeViewerComponent:ne,loadingComponent:re,errorComponent:ee,renderer:({resource:n,fields:u,isCustomPage:$,id:d})=>{var c,S;let g=B((c=n.label)!=null?c:n.name,"edit"),o=f((S=n.label)!=null?S:n.name,"Data"),l=[["React","react",!0],["IResourceComponentsProps","@pankod/refine-core"],["Edit","@pankod/refine-antd"],["Form","@pankod/refine-antd"],["useForm","@pankod/refine-antd"],["Input","@pankod/refine-antd"]],I=u.filter(a=>(a==null?void 0:a.relation)&&!(a!=null&&a.fieldable)&&(a==null?void 0:a.resource)).filter(Boolean).map(a=>{if((a==null?void 0:a.relation)&&!a.fieldable&&a.resource){l.push(["useSelect","@pankod/refine-antd"]);let v=s(o,a.key,a.accessor,!1);return a.multiple&&a.accessor&&(v=`${s(o,a.key)}?.map((item: any) => ${s("item",void 0,a.accessor)})`),`
                    const { selectProps: ${f(a.key,"SelectProps")} } =
                    useSelect({
                        resource: "${a.resource.name}",
                        defaultValue: ${v},
                        ${a.relationInfer&&a.relationInfer.accessor?typeof a.relationInfer.accessor=="string"?a.relationInfer.accessor!=="title"?`optionLabel: "${a.relationInfer.accessor}",`:"":`optionLabel: "${a.relationInfer.accessor[0]}",`:""}
                    });
                `}}).filter(Boolean),A=a=>{if(a.relation&&a.resource){l.push(["Select","@pankod/refine-antd"]);let v=f(a.key,"SelectProps"),t=a.accessor?a.multiple?`"${a.key}"`:`["${a.key}", "${a.accessor}"]`:`"${a.key}"`,r="",i="";return a.accessor&&a.multiple&&(i=`getValueFromEvent={(selected: string[]) => {
                        return selected?.map((item) => ({ ${ke(`${a.accessor}`)?a.accessor:`["${a.accessor}"]`}: item }));
                    }}`,r=`getValueProps={(value: any[]) => {
                        return {
                            value: value?.map((item) => ${s("item",void 0,a.accessor)}),
                        };
                    }}`),m`
                    <Form.Item
                        label="${k(a.key)}"
                        name={${t}}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                        ${r}
                        ${i}
                    >
                        <Select ${a.multiple?'mode="multiple"':""} {...${v}} />
                    </Form.Item>             
                    `}},w=a=>{if(a.type==="text"||a.type==="url"||a.type==="email"||a.type==="number"){if(a.multiple){let v=s(a.key,"index",a.accessor).split("?.").map(t=>`"${t}"`).join(", ").replace(/"index"/,"index");return`
                        <>
                            {(${s(o,a.key)} as any[])?.map((item, index) => (
                                <Form.Item
                                    key={index}
                                    label={\`${k(a.key)} \${index+1}\`}
                                    name={[${v}]}
                                >
                                    <Input
                                        type="${a.type}"
                                    ${H(a.key)?"readOnly disabled":""} />
                                </Form.Item>
                            ))}
                        </>
                    `}return m`
                    <Form.Item
                        label="${k(a.key)}"
                        name={["${a.key}"${a.accessor?', "'+a.accessor+'"':""}]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                    >
                        <Input ${H(a.key)?"readOnly disabled":""} />
                    </Form.Item>
                `}},L=a=>{if(a.type==="image"){l.push(["Upload","@pankod/refine-antd"],["getValueFromEvent","@pankod/refine-antd"]);let v='valuePropName="fileList"';return a.multiple&&!a.accessor&&(v="getValueProps={(value) => ({ fileList: value?.map((item: any) => ({ url: item, name: item, uid: item }))})}"),a.multiple||(a.accessor?v="getValueProps={(value) => ({ fileList: value ? [value] : [] })}":v="getValueProps={(value) => ({ fileList: [{ url: value, name: value, uid: value }]})}"),m`
                    <Form.Item label="${k(a.key)}">
                        <Form.Item
                            name="${a.key}"
                            ${v}
                            getValueFromEvent={getValueFromEvent}
                            noStyle
                            rules={[
                                {
                                    required: true,
                                },
                            ]}
                        >
                            <Upload.Dragger
                                listType="picture"
                                ${a.multiple?"multiple":""}
                                beforeUpload={() => false}
                            >
                                <p className="ant-upload-text">
                                    Drag & drop a file in this area
                                </p>
                            </Upload.Dragger>
                        </Form.Item>
                    </Form.Item>
                    `}},T=a=>{if(a.type==="boolean"){if(l.push(["Checkbox","@pankod/refine-antd"]),a.multiple){let v=s(a.key,"index",a.accessor).split("?.").map(t=>`"${t}"`).join(", ").replace(/"index"/,"index");return`
                        <>
                            {(${s(o,a.key)} as any[])?.map((item, index) => (
                                <Form.Item
                                    key={index}
                                    valuePropName="checked"
                                    label={\`${k(a.key)} \${index+1}\`}
                                    name={[${v}]}
                                >
                                    <Checkbox>${k(a.key)}</Checkbox>
                                </Form.Item>
                            ))}
                        </>
                    `}return m`
                    <Form.Item
                        label="${k(a.key)}"
                        valuePropName="checked"
                        name={["${a.key}"${a.accessor?', "'+a.accessor+'"':""}]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                    >
                        <Checkbox>${k(a.key)}</Checkbox>
                    </Form.Item>
                `}},p=a=>{if(a.type==="date"){if(l.push(["DatePicker","@pankod/refine-antd"],["dayjs","dayjs",!0]),a.multiple){let v=s(a.key,"index",a.accessor).split("?.").map(t=>`"${t}"`).join(", ").replace(/"index"/,"index");return m`
                        <>
                            {(${s(o,a.key)} as any[])?.map((item, index) => (
                                <Form.Item
                                    key={index}
                                    label={\`${k(a.key)} \${index+1}\`}
                                    name={[${v}]}
                                    getValueProps={(value) => ({ value: value ? dayjs(value) : undefined })}
                                >
                                    <DatePicker />
                                </Form.Item>
                            ))}
                        </>
                    `}return m`
                    <Form.Item
                        label="${k(a.key)}"
                        name={["${a.key}"${a.accessor?', "'+a.accessor+'"':""}]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                        getValueProps={(value) => ({ value: value ? dayjs(value) : undefined })}
                    >
                        <DatePicker />
                    </Form.Item>
                `}},V=a=>{if(a.type==="richtext")return m`
                <Form.Item
                    label="${k(a.key)}"
                    name="${a.key}"
                    rules={[
                        {
                            required: true,
                        },
                    ]}
                >
                    <Input.TextArea rows={5} />
                </Form.Item>
                `},y=u.map(a=>{switch(a==null?void 0:a.type){case"text":case"number":case"email":case"url":return w(a);case"richtext":return V(a);case"image":return L(a);case"date":return p(a);case"boolean":return T(a);case"relation":return A(a);default:return}});return M(l),m`
        ${R(l)}
        
        export const ${g}: React.FC<IResourceComponentsProps> = () => {
            const { formProps, saveButtonProps, queryResult } = useForm(${$?`{
                          resource: "${n.name}",
                          id: ${d},
                          action: "edit",
                      }`:""});
        
            const ${o} = queryResult?.data?.data;
        
            ${I}

            return (
                <Edit saveButtonProps={saveButtonProps}>
                    <Form {...formProps} layout="vertical">
                        ${y.join("")}
                    </Form>
                </Edit>
            );
        };
        `}});var ii=({action:n,id:u,...$})=>{let{useParams:d}=Vn(),{action:g,id:o}=d();switch(n!=null?n:g){case"show":return Ce.createElement(Ue,{...$,id:u!=null?u:o});case"create":return Ce.createElement(Ke,{...$,id:u!=null?u:o});case"edit":return Ce.createElement(_e,{...$,id:u!=null?u:o});default:return Ce.createElement(We,{...$,id:u!=null?u:o})}};import ve from"react";import{useRouterContext as Jn}from"@pankod/refine-core";import*as Wn from"@pankod/refine-mantine";import Ze from"react";import{Alert as En,Center as Nn}from"@pankod/refine-mantine";import{IconAlertCircle as Bn}from"@tabler/icons";var te=({error:n})=>n?Ze.createElement(Nn,{style:{minHeight:300}},Ze.createElement(En,{title:"Error",color:"red",icon:Ze.createElement(Bn,null)},n!=null?n:"")):null;import Rn from"react";import{LoadingOverlay as Mn}from"@pankod/refine-mantine";var oe=()=>Rn.createElement(Mn,{visible:!0});import G from"react";import{Affix as Dn,Button as Xe,Modal as Pn,Group as Hn,ActionIcon as On}from"@pankod/refine-mantine";import{IconCode as Gn,IconMessageCircle as jn,IconX as qn,IconCopy as zn,IconCheck as Un}from"@tabler/icons";var ae=({code:n,loading:u})=>{let $=G.useMemo(()=>J(n!=null?n:""),[n]),d=G.useRef(null),[g,o]=G.useState(!1),[l,C]=G.useState(!1);if(u)return null;let I=()=>{var A,w,L,T;(A=d==null?void 0:d.current)==null||A.select(),(w=d==null?void 0:d.current)==null||w.setSelectionRange(0,Number.MAX_SAFE_INTEGER),typeof navigator<"u"&&(navigator.clipboard.writeText((T=(L=d==null?void 0:d.current)==null?void 0:L.value)!=null?T:""),C(!0),setTimeout(()=>{C(!1)},1e3))};return $?G.createElement(G.Fragment,null,G.createElement(Dn,{position:{bottom:12,right:60}},G.createElement(Hn,{spacing:"xs"},G.createElement(Xe,{leftIcon:G.createElement(Gn,{size:18}),variant:"filled",color:"primary",onClick:()=>o(!0)},"Show Code"),G.createElement(Xe,{leftIcon:G.createElement(jn,{size:18}),variant:"default",component:"a",target:"_blank",href:"https://github.com/refinedev/refine/discussions/3046"},"Give Feedback"))),G.createElement("div",{style:{height:"50px"}}),G.createElement(Pn,{size:800,opened:g,onClose:()=>o(!1),styles:{modal:{padding:"0px !important",borderRadius:0},header:{display:"none"}}},G.createElement(On,{onClick:()=>o(!1),variant:"transparent",sx:{position:"absolute",right:0,top:0,height:56,width:56}},G.createElement(qn,{size:18,style:{color:"#666b7a"}})),G.createElement(Q,{code:$}),G.createElement("textarea",{ref:d,defaultValue:($!=null?$:"").replace(/\\n/g,`\r
`),id:"code-input",style:{padding:0,margin:0,width:0,height:0,opacity:0,border:"none",display:"block"}}),G.createElement(Xe,{sx:{position:"absolute",bottom:12,right:12,width:115},key:"copy",variant:"default",leftIcon:l?G.createElement(Un,{size:18,style:{color:"green"}}):G.createElement(zn,{size:18}),onClick:I},l?"Copied":"Copy"))):null};var Je=D({type:"show",additionalScope:[["@pankod/refine-mantine","RefineMantine",Wn]],codeViewerComponent:ae,loadingComponent:oe,errorComponent:te,renderer:({resource:n,fields:u,isCustomPage:$,id:d})=>{var t;let g=B((t=n.label)!=null?t:n.name,"show"),o="record",l=[["useShow","@pankod/refine-core"],["Show","@pankod/refine-mantine"],["Title","@pankod/refine-mantine"]],I=u.filter(r=>(r==null?void 0:r.relation)&&!(r!=null&&r.fieldable)&&(r==null?void 0:r.resource)).filter(Boolean).map(r=>{if((r==null?void 0:r.relation)&&!r.fieldable&&r.resource){if(r.multiple){l.push(["useMany","@pankod/refine-core"]);let i=s(o,r.key);return r.accessor&&(i=`${s(o,r.key)}?.map((item: any) => ${s("item",void 0,r.accessor)})`),`
                    const { data: ${f(r.key,"Data")}, isLoading: ${f(r.key,"IsLoading")} } =
                    useMany({
                        resource: "${r.resource.name}",
                        ids: ${i} || [],
                        queryOptions: {
                            enabled: !!${o},
                        },
                    });
                    `}return l.push(["useOne","@pankod/refine-core"]),`
                    const { data: ${f(r.key,"Data")}, isLoading: ${f(r.key,"IsLoading")} } =
                    useOne({
                        resource: "${r.resource.name}",
                        id: ${s(o,r.key,r.accessor,!1)} || "",
                        queryOptions: {
                            enabled: !!${o},
                        },
                    });
                `}}).filter(Boolean),A=r=>{if(r.relation&&r.resource){let i=f(r.key,"Data"),e=f(r.key,"IsLoading");return r.multiple?(l.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]),m`
                    <Title my="xs" order={5}>${k(r.key)}</Title>
                    {${e} ? <>Loading...</> : (
                        <>
                        ${(()=>{var h,x;if(r.relationInfer)if((h=r.relationInfer)!=null&&h.accessor){if(Array.isArray(r.relationInfer.accessor))return"Not Handled.";{let b=U((x=r.resource)==null?void 0:x.name),F=s(b,void 0,r.relationInfer.accessor);return`
                                        <Group spacing="xs">
                                            {${i}?.data?.map((${b}: any) => <TagField key={${F}} value={${F}} />)}
                                        </Group>
                                        `}}else return"Not Handled.";else return"not-handled - relation with multiple but no resource"})()}
                        </>
                    )}
                    `):m`
                    <Title my="xs" order={5}>${k(r.key)}</Title>
                    {${e} ? <>Loading...</> : (
                        <>
                        ${(()=>{var h;return r.relationInfer?(h=r.relationInfer)!=null&&h.accessor?Array.isArray(r.relationInfer.accessor)?`{${s(`${i}?.data`,void 0,r.relationInfer.accessor,' + " " + ')}}`:`{${i}?.data?.${r.relationInfer.accessor}}`:`{${i}?.data}`:`{${i}?.data?.id}`})()}
                        </>
                    )}
                    
                    `}},w=r=>{if(r.type==="text"){if(l.push(["TextField","@pankod/refine-mantine"]),r.multiple){l.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]);let i=s("item",void 0,r.accessor);return m`
                    <Title my="xs" order={5}>${k(r.key)}</Title>
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any) => (
                            <TagField value={${i}} key={${i}} />
                        ))}
                    </Group>
                `}return m`
                    <Title my="xs" order={5}>${k(r.key)}</Title>
                    <TextField value={${s(o,r.key,r.accessor)}} />
                `}},L=r=>{if(r.type==="image"){if(l.push(["Image","@pankod/refine-mantine"]),r.multiple){l.push(["Group","@pankod/refine-mantine"]);let i=s("item",void 0,r.accessor);return m`
                    <Title my="xs" order={5}>${k(r.key)}</Title>
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any) => (
                            <Image sx={{ maxWidth: 200 }} src={${i}} key={${i}} />
                        ))}
                    </Group>
                `}return m`
                    <Title my="xs" order={5}>${k(r.key)}</Title>
                    <Image sx={{ maxWidth: 200 }} src={${s(o,r.key,r.accessor," + ")}} />
                    `}},T=r=>{if(r.type==="email"){if(l.push(["EmailField","@pankod/refine-mantine"]),r.multiple){l.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]);let i=s("item",void 0,r.accessor);return m`
                    <Title my="xs" order={5}>${k(r.key)}</Title>
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any) => (
                            <TagField value={${i}} key={${i}} />
                        ))}
                    </Group>
                `}return m`
                    <Title my="xs" order={5}>${k(r.key)}</Title>
                    <EmailField value={${s(o,r.key,r.accessor," + ")}} />
                `}},p=r=>{if(r.type==="url"){if(l.push(["UrlField","@pankod/refine-mantine"]),r.multiple){l.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]);let i=s("item",void 0,r.accessor);return m`
                    <Title my="xs" order={5}>${k(r.key)}</Title>
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any) => (
                            <TagField value={${i}} key={${i}} />
                        ))}
                    </Group>
                `}return m`
                    <Title my="xs" order={5}>${k(r.key)}</Title>
                    <UrlField value={${s(o,r.key,r.accessor," + ")}} />
                `}},V=r=>{if(r.type==="boolean"){if(l.push(["BooleanField","@pankod/refine-mantine"]),r.multiple){l.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]);let i=s("item",void 0,r.accessor);return m`
                    <Title my="xs" order={5}>${k(r.key)}</Title>
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any, index: number) => (
                            <TagField value={${i}} key={index} />
                        ))}
                    </Group>
                `}return m`
                    <Title my="xs" order={5}>${k(r.key)}</Title>
                    <BooleanField value={${s(o,r.key,r.accessor," && ")}} />
                `}},y=r=>{if(r.type==="date"){if(l.push(["DateField","@pankod/refine-mantine"]),r.multiple){l.push(["Group","@pankod/refine-mantine"]);let i=s("item",void 0,r.accessor);return m`
                    <Title my="xs" order={5}>${k(r.key)}</Title>
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any) => (
                            <DateField value={${i}} key={${i}} />
                        ))}
                    </Group>
                `}return m`
                    <Title my="xs" order={5}>${k(r.key)}</Title>
                    <DateField value={${s(o,r.key,r.accessor," + ' ' + ")}} />
                `}},c=r=>{if(r.type==="richtext")return l.push(["MarkdownField","@pankod/refine-mantine"]),m`
                    <Title mt="xs" order={5}>${k(r.key)}</Title>
                    <MarkdownField value={${s(o,r.key,r.accessor,' + " " + ')}} />
                `},S=r=>{if(r.type==="number"){if(l.push(["NumberField","@pankod/refine-mantine"]),r.multiple){l.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]);let i=s("item",void 0,r.accessor);return m`
                    <Title my="xs" order={5}>${k(r.key)}</Title>
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any) => (
                            <TagField value={${i}} key={${i}} />
                        ))}
                    </Group>
                `}return m`
                    <Title my="xs" order={5}>${k(r.key)}</Title>
                    <NumberField value={${s(o,r.key,r.accessor,' + " " + ')} ?? ""} />
                `}},a=r=>{if(r)return m`
                    ${r}
                `},v=u.map(r=>{switch(r==null?void 0:r.type){case"text":return a(w(r));case"number":return a(S(r));case"richtext":return a(c(r));case"email":return a(T(r));case"image":return a(L(r));case"date":return a(y(r));case"boolean":return a(V(r));case"url":return a(p(r));case"relation":return a(A(r));default:return}});return M(l),m`
        ${R(l)}
        
        export const ${g} = () => {
            const { queryResult } = useShow(${$?`{ 
                        resource: "${n.name}", 
                        id: ${d}
                    }`:""});
            const { data, isLoading } = queryResult;
        
            const ${o} = data?.data;
        
            ${I}

            return (
                <Show isLoading={isLoading}>
                    ${v.join("")}
                </Show>
            );
        };
        `}});import*as Kn from"@pankod/refine-mantine";import*as _n from"@pankod/refine-react-table";var se=n=>Array.isArray(n.accessor)||n.multiple?`accessorKey: "${n.key}"`:n.accessor?`accessorKey: "${E(n.key,void 0,n.accessor)}"`:`accessorKey: "${n.key}"`,Qe=D({type:"list",additionalScope:[["@pankod/refine-mantine","RefineMantine",Kn],["@pankod/refine-react-table","RefineReactTable",_n]],codeViewerComponent:ae,loadingComponent:oe,errorComponent:te,renderer:({resource:n,fields:u,isCustomPage:$})=>{var i;let d=B((i=n.label)!=null?i:n.name,"list"),g="tableData?.data",o=[["IResourceComponentsProps","@pankod/refine-core"],["useTable","@pankod/refine-react-table"],["ColumnDef","@pankod/refine-react-table"],["flexRender","@pankod/refine-react-table"],["ScrollArea","@pankod/refine-mantine"],["List","@pankod/refine-mantine"],["Table","@pankod/refine-mantine"],["Pagination","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"],["EditButton","@pankod/refine-mantine"],["ShowButton","@pankod/refine-mantine"],["DeleteButton","@pankod/refine-mantine"]],l=u.filter(e=>(e==null?void 0:e.relation)&&!(e!=null&&e.fieldable)&&(e==null?void 0:e.resource)),C=l.filter(Boolean).map(e=>{if((e==null?void 0:e.relation)&&!e.fieldable&&e.resource){o.push(["GetManyResponse","@pankod/refine-core"]),o.push(["useMany","@pankod/refine-core"]);let h="";return e.multiple?h=`[].concat(...(${g}?.map((item) => ${s("item",e.key,e.accessor,!1)}) ?? []))`:h=`${g}?.map((item) => ${s("item",e.key,e.accessor,!1)}) ?? []`,`
                    const { data: ${f(e.key,"Data")} } =
                    useMany({
                        resource: "${e.resource.name}",
                        ids: ${h},
                        queryOptions: {
                            enabled: !!${g},
                        },
                    });
                    `}}).filter(Boolean),I=l==null?void 0:l.map(e=>{if(e&&e.resource)return f(e.key,"Data")}).filter(Boolean),A=e=>{var h;if(e.relation&&e.resource){let x=`${f(e.key,"Data")}?.data`;if(Array.isArray(e.accessor))return;let b=`id: "${e.key}"`,F=`header: "${k(e.key)}"`,N=se(e),q="";if(e.multiple){o.push(["TagField","@pankod/refine-mantine"]);let z="item";e!=null&&e.relationInfer&&(z=s("item",void 0,e.relationInfer.accessor)),q=`cell: function render({ getValue, table }) {
                        const meta = table.options.meta as {
                            ${f(e.key,"Data")}: GetManyResponse;
                        };

                        const ${f(e.key)} = getValue<any[]>()?.map((item) => {
                            return meta.${f(e.key,"Data")}?.data?.find(
                                (resourceItems) => resourceItems.id === ${s("item",void 0,e.accessor)}
                            );
                        })

                        return (
                            <Group spacing="xs">
                                {${f(e.key)}?.map((item, index) => (
                                    <TagField key={index} value={${z}} />
                                ))}
                            </Group>
                        )
                    }
                `}else e!=null&&e.relationInfer?q=`cell: function render({ getValue, table }) {
                            const meta = table.options.meta as {
                                ${f(e.key,"Data")}: GetManyResponse;
                            };

                            const ${f(e.key)} = meta.${x}?.find(
                                (item) => item.id === getValue<any>(),
                            );

                            return ${s(f(e.key),void 0,(h=e==null?void 0:e.relationInfer)==null?void 0:h.accessor)} ?? "Loading...";
                        },`:q="";return`
                    {
                        ${b},
                        ${F},
                        ${N},
                        ${q}
                    }
                `}},w=e=>{if(e.type==="image"){o.push(["Image","@pankod/refine-mantine"]);let h=`id: "${e.key}"`,x=se(e),b=`header: "${k(e.key)}"`,F=m`
                    cell: function render({ getValue }) {
                        ${e!=null&&e.accessor?"try {":""}
                            return <Image sx={{ maxWidth: "100px" }} src={${s("getValue<any>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0," + ")}} />
                        ${e!=null&&e.accessor?" } catch (error) { return null; }":""}
                       
                    }
                `;if(e.multiple){let N=s("item",void 0,e.accessor," + ");F=`
                        cell: function render({ getValue }) {
                            ${e!=null&&e.accessor?"try {":""}
                                return (
                                    <Group spacing="xs">
                                        {getValue<any[]>()?.map((item, index) => (
                                            <Image src={${N}} key={index} sx={{ maxWidth: "100px" }} />
                                        ))}
                                    </Group>
                                )
                            ${e!=null&&e.accessor?" } catch (error) { return null; }":""}
                        }
                    `}return`
                    {
                        ${h},
                        ${x},
                        ${b},
                        ${F}
                    }
                `}},L=e=>{if(e.type==="email"){o.push(["EmailField","@pankod/refine-mantine"]);let h=`id: "${e.key}"`,x=se(e),b=`header: "${k(e.key)}"`,F=m`
                    cell: function render({ getValue }) {
                        return <EmailField value={${s("getValue<any>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0,' + " " + ')}} />
                    }
                `;return e.multiple&&(o.push(["TagField","@pankod/refine-mantine"]),F=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any>()?.map((item, index) => (
                                        <TagField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${b},
                        ${F}
                    }
                `}},T=e=>{if(e.type==="url"){o.push(["UrlField","@pankod/refine-mantine"]);let h=`id: "${e.key}"`,x=se(e),b=`header: "${k(e.key)}"`,F=m`
                    cell: function render({ getValue }) {
                        return <UrlField value={${s("getValue<any>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0," + ")}} />
                    }
                `;return e.multiple&&(o.push(["TagField","@pankod/refine-mantine"]),F=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${b},
                        ${F}
                    }
                `}},p=e=>{if((e==null?void 0:e.type)==="boolean"){o.push(["BooleanField","@pankod/refine-mantine"]);let h=`id: "${e.key}"`,x=se(e),b=`header: "${k(e.key)}"`,F=m`
                    cell: function render({ getValue }) {
                        return <BooleanField value={${s("getValue<any>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0," + ")}} />
                    }
                `;return e.multiple&&(F=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any[]>()?.map((item, index) => (
                                        <BooleanField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${b},
                        ${F}
                    }
                `}},V=e=>{if(e.type==="date"){o.push(["DateField","@pankod/refine-mantine"]);let h=`id: "${e.key}"`,x=se(e),b=`header: "${k(e.key)}"`,F=m`
                    cell: function render({ getValue }) {
                        return <DateField value={${s("getValue<any>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0,' + " " + ')}} />
                    }
                `;return e.multiple&&(F=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any[]>()?.map((item, index) => (
                                        <DateField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${b},
                        ${F}
                    }
                `}},y=e=>{if((e==null?void 0:e.type)==="richtext"){o.push(["MarkdownField","@pankod/refine-mantine"]);let h=`id: "${e.key}"`,x=se(e),b=`header: "${k(e.key)}"`,F=m`
                    cell: function render({ getValue }) {
                        return <MarkdownField value={(${s("getValue<string>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0)})?.slice(0, 80) + "..." } />
                    }
                `;return e.multiple&&(F=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<string[]>()?.map((item, index) => (
                                        <MarkdownField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${b},
                        ${F}
                    }
                `}},c=e=>{if(e&&(e.type==="text"||e.type==="number")){let h=`id: "${e.key}"`,x=se(e),b=`header: "${k(e.key)}"`,F="";return e.multiple&&(o.push(["TagField","@pankod/refine-mantine"]),F=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${s("item",void 0,e.accessor,' + " " + ')}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),!e.multiple&&Array.isArray(e.accessor)&&(F=`
                        cell: function render({ getValue }) {
                            return (
                                <>{${s("getValue()",e.key,e.accessor)}}</>
                            );
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${b},
                        ${F}
                    }
                `}},{canEdit:S,canShow:a,canDelete:v}=n!=null?n:{};S&&o.push(["EditButton","@pankod/refine-mantine"]),a&&o.push(["ShowButton","@pankod/refine-mantine"]),v&&o.push(["DeleteButton","@pankod/refine-mantine"]);let t=S||a||v?m`
        {
            id: "actions",
            accessorKey: "id",
            header: "Actions",
            cell: function render({ getValue }) {
                return (
                    <Group spacing="xs" noWrap>
                        ${a?m`
                        <ShowButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `:""}
                        ${S?m`<EditButton
                            hideText
                            recordItemId={getValue() as string}
                        />`:""}
                        ${v?m`
                        <DeleteButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `:""}
                    </Group>
                );
            },
        },
            `:"",r=u.map(e=>{switch(e==null?void 0:e.type){case"text":case"number":return c(e);case"richtext":return y(e);case"email":return L(e);case"image":return w(e);case"date":return V(e);case"boolean":return p(e);case"url":return T(e);case"relation":return A(e);default:return}});return M(o),m`
        import React from "react";
        ${R(o)}
        
        export const ${d}: React.FC<IResourceComponentsProps> = () => {
            const columns = React.useMemo<ColumnDef<any>[]>(() => [
                ${[...r,t].filter(Boolean).join(",")}
            ], []);

            const {
                getHeaderGroups,
                getRowModel,
                setOptions,
                refineCore: {
                    setCurrent,
                    pageCount,
                    current,
                    tableQueryResult: { data: tableData },
                },
            } = useTable({
                columns,
                ${$?`
                refineCoreProps: {
                    resource: "${n.name}",
                }
                `:""}
            });

            ${C}

            setOptions((prev) => ({
                ...prev,
                meta: {
                    ...prev.meta,
                    ${I.join(", ")}
                },
            }));

            return (
                <List>
                    <ScrollArea>
                        <Table highlightOnHover>
                            <thead>
                                {getHeaderGroups().map((headerGroup) => (
                                    <tr key={headerGroup.id}>
                                        {headerGroup.headers.map((header) => {
                                            return (
                                                <th key={header.id}>
                                                    {!header.isPlaceholder && (
                                                        flexRender(
                                                            header.column
                                                                .columnDef
                                                                .header,
                                                            header.getContext(),
                                                        )
                                                    )}
                                                </th>
                                            );
                                        })}
                                    </tr>
                                ))}
                            </thead>
                            <tbody>
                                {getRowModel().rows.map((row) => {
                                    return (
                                        <tr key={row.id}>
                                            {row.getVisibleCells().map((cell) => {
                                                return (
                                                    <td key={cell.id}>
                                                        {flexRender(
                                                            cell.column.columnDef.cell,
                                                            cell.getContext(),
                                                        )}
                                                    </td>
                                                );
                                            })}
                                        </tr>
                                    );
                                })}
                            </tbody>
                        </Table>
                    </ScrollArea>    
                    <br />
                    <Pagination
                        position="right"
                        total={pageCount}
                        page={current}
                        onChange={setCurrent}
                    />
                </List>
            );
        };
        `}});import*as Zn from"@pankod/refine-mantine";var Ye=D({type:"create",additionalScope:[["@pankod/refine-mantine","RefineMantine",Zn]],codeViewerComponent:ae,loadingComponent:oe,errorComponent:te,renderer:({resource:n,fields:u,isCustomPage:$})=>{var S;let d=B((S=n.label)!=null?S:n.name,"create"),g=[["Create","@pankod/refine-mantine"],["useForm","@pankod/refine-mantine"]],o={},C=u.filter(a=>(a==null?void 0:a.relation)&&!(a!=null&&a.fieldable)&&(a==null?void 0:a.resource)).filter(Boolean).map(a=>{if((a==null?void 0:a.relation)&&!a.fieldable&&a.resource)return g.push(["useSelect","@pankod/refine-mantine"]),`
                    const { selectProps: ${f(a.key,"SelectProps")} } =
                    useSelect({
                        resource: "${a.resource.name}",
                        ${K(a)}
                    });
                `}).filter(Boolean),I=a=>{if(a.relation&&a.resource){o={...o,[a.key]:a.multiple?[]:a.accessor?{[typeof a.accessor=="string"?a.accessor:a.accessor[0]]:""}:""};let v=f(a.key,"SelectProps");return a.multiple?(g.push(["MultiSelect","@pankod/refine-mantine"]),m`
                        <MultiSelect mt="sm" label="${k(a.key)}" {...getInputProps("${E(a.key,void 0)}")} {...${v}} filterDataOnExactSearchMatch={undefined} />
                    `):(g.push(["Select","@pankod/refine-mantine"]),m`
                    <Select mt="sm" label="${k(a.key)}" {...getInputProps("${E(a.key,void 0,a.accessor)}")} {...${v}} />
                `)}},A=a=>{if(a.type==="text"||a.type==="email"||a.type==="date"||a.type==="url")return H(a.key)||(g.push(["TextInput","@pankod/refine-mantine"]),o={...o,[a.key]:a.multiple?[]:""},a.multiple)?void 0:m`
                    <TextInput mt="sm" label="${k(a.key)}" {...getInputProps("${E(a.key,void 0,a.accessor)}")} />
                `},w=a=>{if(a.type==="image")return m`
                {/* 
                    Dropzone component is not included in "@pankod/refine-mantine" package.
                    To use a <Dropzone> component, you can follow the official documentation for Mantine.
                    
                    Docs: https://mantine.dev/others/dropzone/
                */}
                `},L=a=>{if(a.type==="boolean")return g.push(["Checkbox","@pankod/refine-mantine"]),o={...o,[a.key]:a.multiple?[]:""},a.multiple?void 0:m`
                    <Checkbox mt="sm" label="${k(a.key)}" {...getInputProps("${E(a.key,void 0,a.accessor)}", { type: 'checkbox' })} />
                `},T=a=>{if(a.type==="date"){let v=A(a);return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-mantine" package.
                        To use a <DatePicker> component, you can follow the official documentation for Mantine.
                        
                        Docs: https://mantine.dev/dates/date-picker/
                    */}
                    ${v!=null?v:""}
                `}},p=a=>{if(a.type==="richtext")return g.push(["Textarea","@pankod/refine-mantine"]),o={...o,[a.key]:a.multiple?[]:""},a.multiple?void 0:m`
                    <Textarea mt="sm" label="${k(a.key)}" autosize {...getInputProps("${E(a.key,void 0,a.accessor)}")} />
                `},V=a=>{if(a.type==="number")return H(a.key)||(g.push(["NumberInput","@pankod/refine-mantine"]),o={...o,[a.key]:a.multiple?[]:""},a.multiple)?void 0:m`
                    <NumberInput mt="sm" label="${k(a.key)}" {...getInputProps("${E(a.key,void 0,a.accessor)}")}/>
                `},y=a=>{if(a)return m`
                    ${a}
            `},c=u.map(a=>{switch(a==null?void 0:a.type){case"url":case"text":case"email":return y(A(a));case"number":return y(V(a));case"richtext":return y(p(a));case"image":return y(w(a));case"date":return y(T(a));case"boolean":return y(L(a));case"relation":return y(I(a));default:return}});return M(g),m`
        ${R(g)}
        
        export const ${d} = () => {
            const { getInputProps, saveButtonProps, setFieldValue, refineCore: { formLoading } } = useForm({
                initialValues: ${JSON.stringify(o)},
                ${$?`refineCoreProps: {
                            resource: "${n.name}",
                            action: "create",  
                        }`:""}
            });
        
            ${C}

            return (
                <Create isLoading={formLoading} saveButtonProps={saveButtonProps}>
                    ${c.join("")}
                </Create>
            );
        };
        `}});import*as Xn from"@pankod/refine-mantine";var er=D({type:"edit",additionalScope:[["@pankod/refine-mantine","RefineMantine",Xn]],codeViewerComponent:ae,loadingComponent:oe,errorComponent:te,renderer:({resource:n,fields:u,isCustomPage:$,id:d})=>{var v,t;let g=B((v=n.label)!=null?v:n.name,"edit"),o=f((t=n.label)!=null?t:n.name,"Data"),l=[["Edit","@pankod/refine-mantine"],["useForm","@pankod/refine-mantine"]],C={},A=u.filter(r=>(r==null?void 0:r.relation)&&!(r!=null&&r.fieldable)&&(r==null?void 0:r.resource)).filter(Boolean).map(r=>{if((r==null?void 0:r.relation)&&!r.fieldable&&r.resource){l.push(["useSelect","@pankod/refine-mantine"]);let i=s(o,r.key,r.accessor,!1);r.multiple&&r.accessor&&(i=`${s(o,r.key)}?.map((item: any) => ${s("item",void 0,r.accessor)})`);let e="";return r.multiple&&r.accessor&&(e=`React.useEffect(() => {
                            setFieldValue("${r.key}", ${i});
                        }, [${o}]);`),`
                    const { selectProps: ${f(r.key,"SelectProps")} } =
                    useSelect({
                        resource: "${r.resource.name}",
                        defaultValue: ${i},
                        ${K(r)}
                    });

                    ${e}
                `}}).filter(Boolean),w=r=>{if(r.relation&&r.resource){C={...C,[r.key]:r.multiple?[]:r.accessor?{[typeof r.accessor=="string"?r.accessor:r.accessor[0]]:""}:""};let i=f(r.key,"SelectProps");return r.multiple?(l.push(["MultiSelect","@pankod/refine-mantine"]),m`
                        <MultiSelect mt="sm" label="${k(r.key)}" {...getInputProps("${E(r.key,void 0)}")} {...${i}} filterDataOnExactSearchMatch={undefined} />
                    `):(l.push(["Select","@pankod/refine-mantine"]),m`
                    <Select mt="sm" label="${k(r.key)}" {...getInputProps("${E(r.key,void 0,r.accessor)}")} {...${i}} />
                `)}},L=r=>{if(r.type==="text"||r.type==="email"||r.type==="date"||r.type==="url"){if(l.push(["TextInput","@pankod/refine-mantine"]),C={...C,[r.key]:r.multiple?[]:""},r.multiple){l.push(["Group","@pankod/refine-mantine"]);let i=E(r.key,"${index}",r.accessor);return`
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any, index: number) => (
                            <TextInput mt="sm" key={index} label={\`${k(r.key)} \${index + 1} \`} {...getInputProps(\`${i}\`)} />
                        ))}
                    </Group>
                    `}return m`
                    <TextInput mt="sm" ${H(r.key)?"disabled":""} label="${k(r.key)}" {...getInputProps("${E(r.key,void 0,r.accessor)}")} />
                `}},T=r=>{if(r.type==="image")return m`
                {/* 
                    Dropzone component is not included in "@pankod/refine-mantine" package.
                    To use a <Dropzone> component, you can follow the official documentation for Mantine.
                    
                    Docs: https://mantine.dev/others/dropzone/
                */}
                `},p=r=>{if(r.type==="boolean"){if(l.push(["Checkbox","@pankod/refine-mantine"]),C={...C,[r.key]:r.multiple?[]:""},r.multiple){l.push(["Group","@pankod/refine-mantine"]);let i=E(r.key,"${index}",r.accessor);return`
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any, index: number) => (
                            <Checkbox mt="sm" key={index} label={\`${k(r.key)} \${index + 1} \`} {...getInputProps(\`${i}\`, { type: 'checkbox' })} />
                        ))}
                    </Group>
                    `}return m`
                    <Checkbox mt="sm" label="${k(r.key)}" {...getInputProps("${E(r.key,void 0,r.accessor)}", { type: 'checkbox' })} />
                `}},V=r=>{if(r.type==="date")return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-mantine" package.
                        To use a <DatePicker> component, you can follow the official documentation for Mantine.
                        
                        Docs: https://mantine.dev/dates/date-picker/
                    */}
                    ${L(r)}
                `},y=r=>{if(r.type==="richtext"){if(l.push(["Textarea","@pankod/refine-mantine"]),C={...C,[r.key]:r.multiple?[]:""},r.multiple){l.push(["Group","@pankod/refine-mantine"]);let i=E(r.key,"${index}",r.accessor);return`
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any, index: number) => (
                            <Textarea mt="sm" key={index} label={\`${k(r.key)} \${index + 1} \`} {...getInputProps(\`${i}\`)} />
                        ))}
                    </Group>
                    `}return m`
                    <Textarea mt="sm" label="${k(r.key)}" autosize {...getInputProps("${E(r.key,void 0,r.accessor)}")} />
                `}},c=r=>{if(r.type==="number"){if(l.push(["NumberInput","@pankod/refine-mantine"]),C={...C,[r.key]:r.multiple?[]:""},r.multiple){l.push(["Group","@pankod/refine-mantine"]);let i=E(r.key,"${index}",r.accessor);return`
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any, index: number) => (
                            <NumberInput mt="sm" key={index} label={\`${k(r.key)} \${index + 1} \`} {...getInputProps(\`${i}\`)} />
                        ))}
                    </Group>
                    `}return m`
                    <NumberInput mt="sm" ${H(r.key)?"disabled":""} label="${k(r.key)}" {...getInputProps("${E(r.key,void 0,r.accessor)}")}/>
                `}},S=r=>{if(r)return m`
                    ${r}
            `},a=u.map(r=>{switch(r==null?void 0:r.type){case"url":case"text":case"email":return S(L(r));case"number":return S(c(r));case"richtext":return S(y(r));case"image":return S(T(r));case"date":return S(V(r));case"boolean":return S(p(r));case"relation":return S(w(r));default:return}});return M(l),m`
        ${R(l)}
        
        export const ${g} = () => {
            const { getInputProps, saveButtonProps, setFieldValue, refineCore: { queryResult } } = useForm({
                initialValues: ${JSON.stringify(C)},
                ${$?`refineCoreProps: {
                            resource: "${n.name}",
                            id: ${d},
                            action: "edit",  
                        }`:""}
            });
        
            const ${o} = queryResult?.data?.data;
        
            ${A}

            return (
                <Edit saveButtonProps={saveButtonProps}>
                    ${a.join("")}
                </Edit>
            );
        };
        `}});var oc=({action:n,id:u,...$})=>{let{useParams:d}=Jn(),{action:g,id:o}=d();switch(n!=null?n:g){case"show":return ve.createElement(Je,{...$,id:u!=null?u:o});case"create":return ve.createElement(Ye,{...$,id:u!=null?u:o});case"edit":return ve.createElement(er,{...$,id:u!=null?u:o});default:return ve.createElement(Qe,{...$,id:u!=null?u:o})}};import Te from"react";import{useRouterContext as $t}from"@pankod/refine-core";import*as pt from"@pankod/refine-mui";import rr from"react";import{Alert as Qn,AlertTitle as Yn,Box as et}from"@pankod/refine-mui";var ie=({error:n})=>n?rr.createElement(et,{style:{display:"flex",justifyContent:"center",alignItems:"center",minHeight:"200px"}},rr.createElement(Qn,{severity:"error"},rr.createElement(Yn,null,"Error"),n!=null?n:"")):null;import jr from"react";import{Box as rt,CircularProgress as nt}from"@pankod/refine-mui";var ce=()=>jr.createElement(rt,{style:{display:"flex",justifyContent:"center",alignItems:"center",minHeight:"300px"}},jr.createElement(nt,{size:"large"}));import P from"react";import{Button as nr,Box as tr,Modal as tt,IconButton as ot}from"@pankod/refine-mui";import{IconCode as at,IconMessageCircle as st,IconX as it,IconCopy as ct,IconCheck as ut}from"@tabler/icons";var ue=({code:n,loading:u})=>{let $=P.useMemo(()=>J(n!=null?n:""),[n]),d=P.useRef(null),[g,o]=P.useState(!1),[l,C]=P.useState(!1);if(u)return null;let I=()=>{var A,w,L,T;(A=d==null?void 0:d.current)==null||A.select(),(w=d==null?void 0:d.current)==null||w.setSelectionRange(0,99999),typeof navigator<"u"&&(navigator.clipboard.writeText((T=(L=d==null?void 0:d.current)==null?void 0:L.value)!=null?T:""),C(!0),setTimeout(()=>{C(!1)},1e3))};return $?P.createElement(P.Fragment,null,P.createElement(tr,{sx:{position:"fixed",bottom:12,right:60}},P.createElement(tr,{sx:{display:"flex",gap:2}},P.createElement(nr,{variant:"contained",color:"primary",onClick:()=>o(!0),startIcon:P.createElement(at,{size:18}),sx:{whiteSpace:"nowrap"}},"Show Code"),P.createElement(nr,{variant:"contained",color:"secondary",component:"a",target:"_blank",href:"https://github.com/refinedev/refine/discussions/3046",startIcon:P.createElement(st,{size:18}),sx:{whiteSpace:"nowrap"}},"Give Feedback"))),P.createElement("div",{style:{height:"50px"}}),P.createElement(tt,{open:g,onClose:()=>o(!1)},P.createElement(tr,{sx:{position:"absolute",maxWidth:"800px",width:"100%",top:"50%",left:"50%",transform:"translate(-50%, -50%)",boxShadow:24}},P.createElement(ot,{onClick:()=>o(!1),disableRipple:!0,sx:{position:"absolute",right:0,top:0,height:56,width:56,display:"flex",alignItems:"center",justifyContent:"center"}},P.createElement(it,{size:18,style:{color:"#666b7a"}})),P.createElement(Q,{code:$}),P.createElement("textarea",{ref:d,defaultValue:($!=null?$:"").replace(/\\n/g,`\r
`),id:"code-input",style:{display:"none"}}),P.createElement("div",{style:{position:"absolute",bottom:12,right:12}},P.createElement(nr,{key:"copy",sx:{textTransform:"capitalize",width:100},startIcon:l?P.createElement(ut,{size:18,style:{color:"green"}}):P.createElement(ct,{size:18}),color:"inherit",variant:"contained",onClick:I},l?"Copied":"Copy"))))):null};var or=D({type:"show",additionalScope:[["@pankod/refine-mui","RefineMui",pt]],codeViewerComponent:ue,loadingComponent:ce,errorComponent:ie,renderer:({resource:n,fields:u,isCustomPage:$,id:d})=>{var v;let g=B((v=n.label)!=null?v:n.name,"show"),o="record",l=[["useShow","@pankod/refine-core"],["Show","@pankod/refine-mui"],["Typography","@pankod/refine-mui"],["Stack","@pankod/refine-mui"]],I=u.filter(t=>(t==null?void 0:t.relation)&&!(t!=null&&t.fieldable)&&(t==null?void 0:t.resource)).filter(Boolean).map(t=>{if((t==null?void 0:t.relation)&&!t.fieldable&&t.resource){if(t.multiple){l.push(["useMany","@pankod/refine-core"]);let r=s(o,t.key);return t.accessor&&(r=`${s(o,t.key)}?.map((item: any) => ${s("item",void 0,t.accessor)})`),`
                    const { data: ${f(t.key,"Data")}, isLoading: ${f(t.key,"IsLoading")} } =
                    useMany({
                        resource: "${t.resource.name}",
                        ids: ${r} || [],
                        queryOptions: {
                            enabled: !!${o},
                        },
                    });
                    `}return l.push(["useOne","@pankod/refine-core"]),`
                    const { data: ${f(t.key,"Data")}, isLoading: ${f(t.key,"IsLoading")} } =
                    useOne({
                        resource: "${t.resource.name}",
                        id: ${s(o,t.key,t.accessor,!1)} || "",
                        queryOptions: {
                            enabled: !!${o},
                        },
                    });
                `}}).filter(Boolean),A=t=>{if(t.relation&&t.resource){let r=f(t.key,"Data"),i=f(t.key,"IsLoading");return t.multiple?(l.push(["TagField","@pankod/refine-mui"]),m`
                    <Typography variant="body1" fontWeight="bold">
                        ${k(t.key)}
                    </Typography>
                    {${i} ? (
                        <>
                            Loading...
                        </>
                        ) : (
                        <>
                        ${(()=>{var e;if(t.relationInfer)if((e=t.relationInfer)!=null&&e.accessor){if(Array.isArray(t.relationInfer.accessor))return"Not Handled.";{let h=f(t.key),x=s(h,void 0,t.relationInfer.accessor);return`
                                            <Stack direction="row" spacing={1}>
                                                {${r}?.data?.map((${h}: any) => (
                                                    <TagField key={${x}} value={${x}} />
                                                ))}
                                            </Stack>
                                        `}}else return"Not Handled.";else return"not-handled - relation with multiple but no resource"})()}
                        </>
                    )}
                    `):m`
                    <Typography variant="body1" fontWeight="bold">
                        ${k(t.key)}
                    </Typography>

                    {${i} ? (
                        <>Loading...</>
                    ) : (
                        <>
                        ${(()=>{var e;return t.relationInfer?(e=t.relationInfer)!=null&&e.accessor?Array.isArray(t.relationInfer.accessor)?`{${s(`${r}?.data`,void 0,t.relationInfer.accessor,' + " " + ')}}`:`{${r}?.data?.${t.relationInfer.accessor}}`:`{${r}?.data}`:`{${r}?.data?.id}`})()}
                        </>
                    )}
                    `}},w=t=>{if(t.type==="text"){if(l.push(["TextFieldComponent as TextField","@pankod/refine-mui"]),t.multiple){l.push(["TagField","@pankod/refine-mui"]);let r=s("item",void 0,t.accessor);return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${k(t.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${s(o,t.key)}?.map((item: any) => (
                        <TagField value={${r}} key={${r}} />
                    ))}
                    </Stack>
                `}return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${k(t.key)}
                    </Typography>
                    <TextField value={${s(o,t.key,t.accessor)}} />
                `}},L=t=>{if(t.type==="image"){if(t.multiple){let r=s("item",void 0,t.accessor);return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${k(t.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${s(o,t.key)}?.map((item: any) => (
                        <img style={{ maxWidth: 200, width: "100%", height: 200 }} src={${r}} key={${r}} />
                    ))}
                    </Stack>
                `}return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${k(t.key)}
                    </Typography>
                    <img style={{ maxWidth: 200, width: "100%", height: 200 }} src={${s(o,t.key,t.accessor," + ")}} />
                    `}},T=t=>{if(t.type==="email"){if(l.push(["EmailField","@pankod/refine-mui"]),t.multiple){l.push(["TagField","@pankod/refine-mui"]);let r=s("item",void 0,t.accessor);return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${k(t.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${s(o,t.key)}?.map((item: any) => (
                        <TagField value={${r}} key={${r}} />
                    ))}
                    </Stack>
                `}return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${k(t.key)}
                    </Typography>
                    <EmailField value={${s(o,t.key,t.accessor," + ")}} />
                `}},p=t=>{if(t.type==="url"){if(l.push(["UrlField","@pankod/refine-mui"]),t.multiple){l.push(["TagField","@pankod/refine-mui"]);let r=s("item",void 0,t.accessor);return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${k(t.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${s(o,t.key)}?.map((item: any) => (
                        <TagField value={${r}} key={${r}} />
                    ))}
                    </Stack>
                `}return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${k(t.key)}
                    </Typography>
                    <UrlField value={${s(o,t.key,t.accessor," + ")}} />
                `}},V=t=>{if(t.type==="boolean"){if(l.push(["BooleanField","@pankod/refine-mui"]),t.multiple){l.push(["TagField","@pankod/refine-mui"]);let r=s("item",void 0,t.accessor);return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${k(t.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${s(o,t.key)}?.map((item: any, index: number) => (
                        <TagField value={${r}} key={index} />
                    ))}
                    </Stack>
                `}return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${k(t.key)}
                    </Typography>
                    <BooleanField value={${s(o,t.key,t.accessor," && ")}} />
                `}},y=t=>{if(t.type==="date"){if(l.push(["DateField","@pankod/refine-mui"]),t.multiple){let r=s("item",void 0,t.accessor);return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${k(t.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${s(o,t.key)}?.map((item: any) => (
                        <DateField value={${r}} key={${r}} />
                    ))}
                    </Stack>
                `}return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${k(t.key)}
                    </Typography>
                    <DateField value={${s(o,t.key,t.accessor," + ' ' + ")}} />
                `}},c=t=>{if(t.type==="richtext")return l.push(["MarkdownField","@pankod/refine-mui"]),m`
                    <Typography variant="body1" fontWeight="bold">
                        ${k(t.key)}
                    </Typography>
                    <MarkdownField value={${s(o,t.key,t.accessor,' + " " + ')}} />
                `},S=t=>{if(t.type==="number"){if(l.push(["NumberField","@pankod/refine-mui"]),t.multiple){l.push(["TagField","@pankod/refine-mui"]);let r=s("item",void 0,t.accessor);return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${k(t.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${s(o,t.key)}?.map((item: any) => (
                        <TagField value={${r}} key={${r}} />
                    ))}
                    </Stack>
                `}return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${k(t.key)}
                    </Typography>
                    <NumberField value={${s(o,t.key,t.accessor,' + " " + ')} ?? ""} />
                `}},a=u.map(t=>{switch(t==null?void 0:t.type){case"text":return w(t);case"number":return S(t);case"richtext":return c(t);case"email":return T(t);case"image":return L(t);case"date":return y(t);case"boolean":return V(t);case"url":return p(t);case"relation":return A(t);default:return}});return M(l),m`
        ${R(l)}

        export const ${g} = () => {
            const { queryResult } = useShow(${$?`{ 
                        resource: "${n.name}", 
                        id: ${d}
                    }`:""});
            const { data, isLoading } = queryResult;
        
            const ${o} = data?.data;
        
            ${I}

            return (
                <Show isLoading={isLoading}>
                    <Stack gap={1}>
                        ${a.join("")}
                    </Stack>
                </Show>
            );
        };
        `}});import*as mt from"@pankod/refine-mui";var ar=D({type:"list",additionalScope:[["@pankod/refine-mui","RefineMui",mt]],codeViewerComponent:ue,loadingComponent:ce,errorComponent:ie,renderer:({resource:n,fields:u,isCustomPage:$})=>{var i;let d=B((i=n.label)!=null?i:n.name,"list"),g="dataGridProps?.rows",o=[["React","react",!0],["useDataGrid","@pankod/refine-mui"],["DataGrid","@pankod/refine-mui"],["GridColumns","@pankod/refine-mui"],["EditButton","@pankod/refine-mui"],["ShowButton","@pankod/refine-mui"],["DeleteButton","@pankod/refine-mui"],["List","@pankod/refine-mui"]],l=u.filter(e=>(e==null?void 0:e.relation)&&!(e!=null&&e.fieldable)&&(e==null?void 0:e.resource)),C=l.filter(Boolean).map(e=>{if((e==null?void 0:e.relation)&&!e.fieldable&&e.resource){o.push(["useMany","@pankod/refine-core"]);let h="";return e.multiple?h=`[].concat(...(${g}?.map((item: any) => ${s("item",e.key,e.accessor,!1)}) ?? []))`:h=`${g}?.map((item: any) => ${s("item",e.key,e.accessor,!1)}) ?? []`,`
                    const { data: ${f(e.key,"Data")}, isLoading: ${f(e.key,"IsLoading")} } =
                    useMany({
                        resource: "${e.resource.name}",
                        ids: ${h},
                        queryOptions: {
                            enabled: !!${g},
                        },
                    });
                    `}}).filter(Boolean),I=l==null?void 0:l.map(e=>{if(e&&e.resource)return`${f(e.key,"Data")}?.data`}).filter(Boolean),A=e=>{var h,x;if(e.relation&&e.resource){let b=`${f(e.key,"Data")}?.data`,F=f(e.key,"IsLoading");if(Array.isArray(e.accessor))return;let N=`${F} ? <>Loading...</> : `,q=`field: "${e.key}"`,z=e.accessor&&!e.multiple&&!Array.isArray(e.accessor)?`valueGetter: ({ row }) => {
                const value = ${s("row",e.key,e.accessor,!1)};

                return value;
            },`:"",ye=`headerName: "${k(e.key)}"`,X="";if(e.multiple){o.push(["TagField","@pankod/refine-mui"]);let Y="item";if(e!=null&&e.relationInfer){let Fe=`${b}?.find((resourceItems) => resourceItems.id === ${s("item",void 0,e.accessor)})`;Y=s(Fe,void 0,(h=e==null?void 0:e.relationInfer)==null?void 0:h.accessor)}X=`
                    renderCell: function render({ value }) {
                        return ${N} (
                            <>
                                {${s("value",void 0)}?.map((item: any, index: number) => (
                                    <TagField key={index} value={${Y}} />
                                ))}
                            </>
                        )
                    }
                    `}else if(e!=null&&e.relationInfer){let Y=`${b}?.find((item) => item.id === value)`,Fe=s(Y,void 0,(x=e==null?void 0:e.relationInfer)==null?void 0:x.accessor);X=`
                        renderCell: function render({ value }) {
                            return ${N} ${Fe};
                        }
                        `}else X="";return`
                    {
                        ${q},
                        ${ye},${z}
                        minWidth: 300,${X}
                    }
                `}},w=e=>{if(e.type==="image"){let h=`field: "${e.key}"`,x=`headerName: "${k(e.key)}"`,b=e.accessor&&!e.multiple&&!Array.isArray(e.accessor)?`valueGetter: ({ row }) => {
                const value = ${s("row",e.key,e.accessor,!1)};

                return value;
            },`:"",F=`
                    renderCell: function render({ value }) {
                        return (
                            <img src={${s("value",void 0,Array.isArray(e.accessor)?e.accessor:void 0," + ")}} style={{ height: "50px", maxWidth: "100px" }} />
                        )
                    }
                `;return e.multiple&&(F=`
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item: any, index: number) => (
                                    <img src={${s("item",void 0,e.accessor," + ")}} key={index} style={{ height: "50px", maxWidth: "100px" }} />
                                ))}
                                </>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},${b}
                        minWidth: 100,${F}
                    }
                `}},L=e=>{if(e.type==="email"){o.push(["EmailField","@pankod/refine-mui"]);let h=`field: "${e.key}"`,x=`headerName: "${k(e.key)}"`,b=e.accessor&&!e.multiple&&!Array.isArray(e.accessor)?`valueGetter: ({ row }) => {
                const value = ${s("row",e.key,e.accessor,!1)};

                return value;
            },`:"",F=`
                renderCell: function render({ value }) {
                    return (
                        <EmailField value={${s("value",void 0,Array.isArray(e.accessor)?e.accessor:void 0," + ")}} />
                    )
                }
            `;return e.multiple&&(o.push(["TagField","@pankod/refine-mui"]),F=`
                    renderCell: function render({ value }) {
                        return (
                            <>
                            {value?.map((item: any, index: number) => (
                                <TagField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                            ))}
                            </>
                        )
                    }
                `),`
                {
                    ${h},
                    ${x},${b}
                    minWidth: 250,${F}
                }
            `}},T=e=>{if(e.type==="url"){o.push(["UrlField","@pankod/refine-mui"]);let h=`field: "${e.key}"`,x=`headerName: "${k(e.key)}"`,b=e.accessor&&!e.multiple&&!Array.isArray(e.accessor)?`valueGetter: ({ row }) => {
                const value = ${s("row",e.key,e.accessor,!1)};

                return value;
            },`:"",F=`
                    renderCell: function render({ value }) {
                        return (
                            <UrlField value={${s("value",void 0,Array.isArray(e.accessor)?e.accessor:void 0," + ")}} />
                        )
                    }
                `;return e.multiple&&(o.push(["TagField","@pankod/refine-mui"]),F=`
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item: any, index: any) => (
                                    <TagField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                ))}
                                </>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},${b}
                        minWidth: 250,${F}
                    }
                `}},p=e=>{if(e!=null&&e.type){o.push(["Checkbox","@pankod/refine-mui"]);let h=`field: "${e.key}"`,x=`headerName: "${k(e.key)}"`,b=e.accessor&&!e.multiple&&!Array.isArray(e.accessor)?`valueGetter: ({ row }) => {
                const value = ${s("row",e.key,e.accessor,!1)};

                return value;
            },`:"",F=`
                    renderCell: function render({ value }) {
                        return (
                            <Checkbox checked={!!${s("value",void 0,Array.isArray(e.accessor)?e.accessor:void 0," && ")}} />
                        );
                    }
                `;return e.multiple&&(F=`
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item: any, index: number) => (
                                    <Checkbox checked={!!${s("item",void 0,e.accessor," && ")}} key={index} />
                                ))}
                                </>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},${b}
                        minWidth: 250,${F}
                    }
                `}},V=e=>{if(e.type==="date"){o.push(["DateField","@pankod/refine-mui"]);let h=`field: "${e.key}"`,x=`headerName: "${k(e.key)}"`,b=e.accessor&&!e.multiple&&!Array.isArray(e.accessor)?`valueGetter: ({ row }) => {
                const value = ${s("row",e.key,e.accessor,!1)};

                return value;
            },`:"",F=`
                    renderCell: function render({ value }) {
                        return <DateField value={value} />;
                    }
                `;return e.multiple&&(F=`
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item: any, index: number) => (
                                    <DateField value={${s("item",void 0,e.accessor,' + " " + ')}} key={index} />
                                ))}
                                </>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},${b}
                        minWidth: 250,${F}
                    }
                `}},y=e=>{if((e==null?void 0:e.type)==="richtext"){o.push(["MarkdownField","@pankod/refine-mui"]);let h=`field: "${e.key}"`,x=`headerName: "${k(e.key)}"`,b=e.accessor&&!e.multiple&&!Array.isArray(e.accessor)?`valueGetter: ({ row }) => {
                const value = ${s("row",e.key,e.accessor,!1)};

                return value;
            },`:"",F=`
                    renderCell: function render({ value }) {
                        return <MarkdownField value={(value ?? "").slice(0, 80) + "..."} />;
                    }
                `;return e.multiple&&(F=`
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item: any, index: number) => (
                                    <MarkdownField value={(${s("item",void 0,e.accessor,' + " " + ')}).slice(0, 80) + "..."} key={index} />
                                ))}
                                </>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},${b}
                        minWidth: 250,${F}
                    }
                `}},c=e=>{if(e&&(e.type==="text"||e.type==="number")){let h=`field: "${e.key}"`,x=`headerName: "${k(e.key)}"`,b=e.accessor&&!e.multiple&&!Array.isArray(e.accessor)?`valueGetter: ({ row }) => {
                const value = ${s("row",e.key,e.accessor,!1)};

                return value;
            },`:"",F="";if(e.multiple){o.push(["TagField","@pankod/refine-mui"]);let N=s("item",void 0,e.accessor,' + " " + ');F=`
                    renderCell: function render({ row }) {
                        return (
                            <>
                                {(${s("row",e.key)})?.map((item: any) => (
                                    <TagField value={${N}} key={${N}} />
                                ))}
                            </>
                        )
                    }
                    `}return!e.multiple&&Array.isArray(e.accessor)&&(F=`
                    renderCell: function render({ row }) {
                        return 
                            <>{${s("row",e.key,e.accessor)}}</>
                        );
                    }
                    `),`
                {
                    ${h},
                    ${x},${b}${e.type==="number"?"type: 'number',":""}
                    minWidth: ${H(e.key)?50:200},${F}
                }
                `}},{canEdit:S,canShow:a,canDelete:v}=n!=null?n:{};S&&o.push(["EditButton","@pankod/refine-mui"]),a&&o.push(["ShowButton","@pankod/refine-mui"]),v&&o.push(["DeleteButton","@pankod/refine-mui"]);let t=S||a||v?m`
                {
                    field: "actions",
                    headerName: "Actions",
                    renderCell: function render({ row }) {
                        return (
                            <>
                                ${S?m`<EditButton hideText recordItemId={row.id} />`:""}
                                ${a?m`<ShowButton hideText recordItemId={row.id} />`:""}
                                ${v?m`<DeleteButton hideText recordItemId={row.id} />`:""}
                            </>
                        );
                    },
                    align: "center",
                    headerAlign: "center",
                    minWidth: 80,
                },
        `:"",r=u.map(e=>{switch(e==null?void 0:e.type){case"text":case"number":return c(e);case"richtext":return y(e);case"email":return L(e);case"image":return w(e);case"date":return V(e);case"boolean":return p(e);case"url":return T(e);case"relation":return A(e);default:return}});return M(o),m`
        ${R(o)}
        
        export const ${d} = () => {
            const { dataGridProps } = useDataGrid(
                ${$?`{ resource: "${n.name}" }`:""} 
            );
        
            ${C}

            const columns = React.useMemo<GridColumns<any>>(() => [
                ${[...r,t].filter(Boolean).join(`,\r
`)}
            ], [${I.join(",")}]);

            return (
                <List>
                    <DataGrid {...dataGridProps} columns={columns} autoHeight />
                </List>
            );
        };
        `}});import*as lt from"@pankod/refine-mui";import*as dt from"@pankod/refine-react-hook-form";var sr=D({type:"create",additionalScope:[["@pankod/refine-mui","RefineMui",lt],["@pankod/refine-react-hook-form","RefineReactHookForm",dt]],codeViewerComponent:ue,loadingComponent:ce,errorComponent:ie,renderer:({resource:n,fields:u,isCustomPage:$})=>{var T;let d=B((T=n.label)!=null?T:n.name,"create"),g=[["Create","@pankod/refine-mui"],["Box","@pankod/refine-mui"],["useForm","@pankod/refine-react-hook-form"]],l=u.filter(p=>(p==null?void 0:p.relation)&&!(p!=null&&p.fieldable)&&(p==null?void 0:p.resource)).filter(Boolean).map(p=>{if((p==null?void 0:p.relation)&&!p.fieldable&&p.resource)return g.push(["useAutocomplete","@pankod/refine-mui"]),`
                    const { autocompleteProps: ${f(p.key,"AutocompleteProps")} } =
                    useAutocomplete({
                        resource: "${p.resource.name}",
                    });
                `}).filter(Boolean),C=p=>{if(p.relation&&p.resource){g.push(["Autocomplete","@pankod/refine-mui"],["Controller","@pankod/refine-react-hook-form"]);let V=f(p.key,"AutocompleteProps"),y=p.relationInfer&&p.relationInfer.accessor?typeof p.relationInfer.accessor=="string"?p.relationInfer.accessor:p.relationInfer.accessor[0]:"title",c=y.includes(".")||y.includes("[")||y.includes("]")||y.includes("-"),S=p.accessor?s("item",void 0,p.accessor,!1):"(item?.id ?? item)",a=p.accessor?s("value",void 0,p.accessor,!1):"(value?.id ?? value)",v=p.accessor?"value":p.multiple?"value?.map((item: any) => item?.id ?? item)":"value?.id ?? value";return m`
                    <Controller
                        control={control}
                        name="${E(p.key,void 0)}"
                        rules={{ required: "This field is required" }}
                        // eslint-disable-next-line
                        ${p.multiple?"defaultValue={[] as any}":"defaultValue={null as any}"} 
                        render={({ field }) => (
                            <Autocomplete
                                {...${V}}
                                {...field}
                                ${p.multiple?"multiple":""}
                                onChange={(_, value) => {
                                    field.onChange(${v});
                                }}
                                getOptionLabel={(item) => {
                                    return (
                                        ${V}?.options?.find(
                                            (p) =>
                                                p?.id?.toString() ===
                                                ${S}?.toString(),
                                        )?.${c?`["${y}"]`:y} ?? ""
                                    );
                                }}
                                isOptionEqualToValue={(option, value) =>
                                    value === undefined ||
                                    option.id.toString() === ${a}?.toString()
                                }
                                renderInput={(params) => (
                                    <TextField
                                        {...params}
                                        label="${k(p.key)}"
                                        margin="normal"
                                        variant="outlined"
                                        error={!!${s("(errors as any)",p.key,p.accessor,!1)}}
                                        helperText={${s("(errors as any)",p.key,p.accessor,!1)}?.message}
                                        required
                                    />
                                )}
                            />
                        )}
                    />
                `}},I=p=>{if(p.type==="text"||p.type==="url"||p.type==="email"||p.type==="number"||p.type==="date"||p.type==="richtext")return H(p.key)||(g.push(["TextField","@pankod/refine-mui"]),p.multiple)?void 0:m`
                    <TextField
                        {...register("${E(p.key,void 0,p.accessor)}", {
                            required: "This field is required",
                        })}
                        error={!!${s("(errors as any)",p.key,p.accessor,!1)}}
                        helperText={${s("(errors as any)",p.key,p.accessor,!1)}?.message}
                        margin="normal"
                        fullWidth
                        InputLabelProps={{ shrink: true }}
                        ${p.type!=="date"&&p.type!=="richtext"?`type="${p.type}"`:""}
                        ${p.type==="richtext"?"multiline":""}
                        label="${k(p.key)}"
                        name="${E(p.key,void 0,p.accessor)}"
                    />
                `},A=p=>{if(p.type==="boolean")return g.push(["Checkbox","@pankod/refine-mui"],["FormControlLabel","@pankod/refine-mui"],["Controller","@pankod/refine-react-hook-form"]),p.multiple?void 0:m`
                    <Controller
                        control={control}
                        name="${E(p.key,void 0,p.accessor)}"
                        // eslint-disable-next-line
                        defaultValue={null as any}
                        render={({ field }) => (
                            <FormControlLabel label="${k(p.key)}" control={
                                <Checkbox
                                    {...field}
                                    checked={field.value}
                                    onChange={(event) => {
                                        field.onChange(event.target.checked);
                                    }}
                                />
                            } />
                        )}
                    />
                `},w=p=>{if(p.type==="date"){let V=I(p);return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-mui" package.
                        To use a <DatePicker> component, you can follow the official documentation for Material UI.
                        
                        Docs: https://mui.com/x/react-date-pickers/date-picker/#basic-usage
                    */}
                    ${V!=null?V:""}
                    `}},L=u.map(p=>{switch(p==null?void 0:p.type){case"text":case"number":case"email":case"url":case"richtext":return I(p);case"date":return w(p);case"boolean":return A(p);case"relation":return C(p);default:return}});return M(g),m`
        ${R(g)}
        
        export const ${d} = () => {
            const {
                saveButtonProps,
                refineCore: { formLoading },
                register,
                control,
                formState: { errors },
            } = useForm(
                ${$?`{ 
                    refineCoreProps: {
                        resource: "${n.name}",
                        action: "create",
                    }
                }`:""}
            );
        
            ${l}

            return (
                <Create isLoading={formLoading} saveButtonProps={saveButtonProps}>
                    <Box
                        component="form"
                        sx={{ display: "flex", flexDirection: "column" }}
                        autoComplete="off"
                    >
                        ${L.join("")}
                    </Box>
                </Create>
            );
        };
        `}});import*as yt from"@pankod/refine-mui";import*as kt from"@pankod/refine-react-hook-form";var ir=D({type:"edit",additionalScope:[["@pankod/refine-mui","RefineMui",yt],["@pankod/refine-react-hook-form","RefineReactHookForm",kt]],codeViewerComponent:ue,loadingComponent:ce,errorComponent:ie,renderer:({resource:n,fields:u,isCustomPage:$,id:d})=>{var V,y;let g=B((V=n.label)!=null?V:n.name,"edit"),o=f((y=n.label)!=null?y:n.name,"Data"),l=[["Edit","@pankod/refine-mui"],["Box","@pankod/refine-mui"],["useForm","@pankod/refine-react-hook-form"]],I=u.filter(c=>(c==null?void 0:c.relation)&&!(c!=null&&c.fieldable)&&(c==null?void 0:c.resource)).filter(Boolean).map(c=>{if((c==null?void 0:c.relation)&&!c.fieldable&&c.resource){l.push(["useAutocomplete","@pankod/refine-mui"]);let S=s(o,c.key,c.accessor,!1);return c.multiple&&c.accessor&&(S=`${s(o,c.key)}?.map((item: any) => ${s("item",void 0,c.accessor)})`),`
                    const { autocompleteProps: ${f(c.key,"AutocompleteProps")} } =
                    useAutocomplete({
                        resource: "${c.resource.name}",
                        defaultValue: ${S},
                    });
                `}}).filter(Boolean),A=c=>{if(c.relation&&c.resource){l.push(["Autocomplete","@pankod/refine-mui"],["Controller","@pankod/refine-react-hook-form"]);let S=f(c.key,"AutocompleteProps"),a=c.relationInfer&&c.relationInfer.accessor?typeof c.relationInfer.accessor=="string"?c.relationInfer.accessor:c.relationInfer.accessor[0]:"title",v=a.includes(".")||a.includes("[")||a.includes("]")||a.includes("-"),t=c.accessor?s("item",void 0,c.accessor,!1):"(item?.id ?? item)",r=c.accessor?s("value",void 0,c.accessor,!1):"(value?.id ?? value)",i=c.accessor?"value":c.multiple?"value?.map((item: any) => item?.id ?? item)":"value?.id ?? value";return m`
                    <Controller
                        control={control}
                        name="${E(c.key,void 0)}"
                        rules={{ required: "This field is required" }}
                        // eslint-disable-next-line
                        ${c.multiple?"defaultValue={[] as any}":"defaultValue={null as any}"} 
                        render={({ field }) => (
                            <Autocomplete
                                {...${S}}
                                {...field}
                                ${c.multiple?"multiple":""}
                                onChange={(_, value) => {
                                    field.onChange(${i});
                                }}
                                getOptionLabel={(item) => {
                                    return (
                                        ${S}?.options?.find(
                                            (p) =>
                                                p?.id?.toString() ===
                                                ${t}?.toString(),
                                        )?.${v?`["${a}"]`:a} ?? ""
                                    );
                                }}
                                isOptionEqualToValue={(option, value) =>
                                    value === undefined ||
                                    option?.id?.toString() === ${r}?.toString()
                                }
                                renderInput={(params) => (
                                    <TextField
                                        {...params}
                                        label="${k(c.key)}"
                                        margin="normal"
                                        variant="outlined"
                                        error={!!${s("(errors as any)",c.key,c.accessor,!1)}}
                                        helperText={${s("(errors as any)",c.key,c.accessor,!1)}?.message}
                                        required
                                    />
                                )}
                            />
                        )}
                    />
                `}},w=c=>{if(c.type==="text"||c.type==="url"||c.type==="email"||c.type==="number"||c.type==="date"||c.type==="richtext"){if(l.push(["TextField","@pankod/refine-mui"]),c.multiple){l.push(["Box","@pankod/refine-mui"]);let S=E(c.key,"${index}",c.accessor),a=s("(errors as any)",c.key,void 0,!1)+"?.[index]";return`
                        <Box sx={{display: "flex", gap: 1}}>
                            {${s(o,c.key)}?.map((item: any, index: number) => (
                                <TextField
                                    key={index}
                                    {...register(\`${S}\`, {
                                        required: "This field is required",
                                    })}
                                    error={!!${a}}
                                    helperText={${a}?.message as string}
                                    margin="normal"
                                    fullWidth
                                    InputLabelProps={{ shrink: true }}
                                    ${c.type!=="date"&&c.type!=="richtext"?`type="${c.type}"`:""}
                                    ${c.type==="richtext"?"multiline":""}
                                    label={\`${k(c.key)} \${index+1}\`}
                                    name={\`${E(c.key,void 0,c.accessor)}.\${index}\`}
                                />
                            ))}
                        </Box>
                    `}return m`
                    <TextField
                        {...register("${E(c.key,void 0,c.accessor)}", {
                            required: "This field is required",
                        })}
                        error={!!${s("(errors as any)",c.key,c.accessor,!1)}}
                        helperText={${s("(errors as any)",c.key,c.accessor,!1)}?.message}
                        margin="normal"
                        fullWidth
                        InputLabelProps={{ shrink: true }}
                        ${c.type!=="date"&&c.type!=="richtext"?`type="${c.type}"`:""}
                        ${c.type==="richtext"?"multiline":""}
                        label="${k(c.key)}"
                        name="${E(c.key,void 0,c.accessor)}"
                        ${H(c.key)?"disabled":""}
                    />
                `}},L=c=>{if(c.type==="boolean"){if(l.push(["Checkbox","@pankod/refine-mui"],["FormControlLabel","@pankod/refine-mui"],["Controller","@pankod/refine-react-hook-form"]),c.multiple){l.push(["Box","@pankod/refine-mui"]);let S=E(c.key,"${index}",c.accessor);return`
                        <Box sx={{display: "flex", gap: 1}}>
                            {${s(o,c.key)}?.map((item: any, index: number) => (
                                <Controller
                                    key={index}
                                    control={control}
                                    name={\`${S}\`}
                                    // eslint-disable-next-line
                                    defaultValue={null as any}
                                    render={({ field }) => (
                                        <FormControlLabel label={\`${k(c.key)} \${index+1}\`} control={
                                            <Checkbox
                                                {...field}
                                                checked={field.value}
                                                onChange={(event) => {
                                                    field.onChange(event.target.checked);
                                                }}
                                            />
                                        } />
                                    )}
                                />
                            ))}
                        </Box>
                    `}return m`
                    <Controller
                        control={control}
                        name="${E(c.key,void 0,c.accessor)}"
                        // eslint-disable-next-line
                        defaultValue={null as any}
                        render={({ field }) => (
                            <FormControlLabel label="${k(c.key)}" control={
                                <Checkbox
                                    {...field}
                                    checked={field.value}
                                    onChange={(event) => {
                                        field.onChange(event.target.checked);
                                    }}
                                />
                            } />
                        )}
                    />
                `}},T=c=>{if(c.type==="date")return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-mui" package.
                        To use a <DatePicker> component, you can follow the official documentation for Material UI.
                        
                        Docs: https://mui.com/x/react-date-pickers/date-picker/#basic-usage
                    */}
                    ${w(c)}
                    `},p=u.map(c=>{switch(c==null?void 0:c.type){case"text":case"number":case"email":case"url":case"richtext":return w(c);case"date":return T(c);case"boolean":return L(c);case"relation":return A(c);default:return}});return M(l),m`
        ${R(l)}
        
        export const ${g} = () => {
            const {
                saveButtonProps,
                refineCore: { queryResult },
                register,
                control,
                formState: { errors },
            } = useForm(
                ${$?`{ 
                    refineCoreProps: {
                        resource: "${n.name}",
                        id: ${d},
                        action: "edit",
                    }
                }`:""}
            );
        
            const ${o} = queryResult?.data?.data;
        
            ${I}

            return (
                <Edit saveButtonProps={saveButtonProps}>
                    <Box
                        component="form"
                        sx={{ display: "flex", flexDirection: "column" }}
                        autoComplete="off"
                    >
                        ${p.join("")}
                    </Box>
                </Edit>
            );
        };
        `}});var eu=({action:n,id:u,...$})=>{let{useParams:d}=$t(),{action:g,id:o}=d();switch(n!=null?n:g){case"show":return Te.createElement(or,{...$,id:u!=null?u:o});case"create":return Te.createElement(sr,{...$,id:u!=null?u:o});case"edit":return Te.createElement(ir,{...$,id:u!=null?u:o});default:return Te.createElement(ar,{...$,id:u!=null?u:o})}};import Se from"react";import{useRouterContext as Ut}from"@pankod/refine-core";import*as Rt from"@pankod/refine-chakra-ui";import Ae from"react";import{Center as gt,Alert as ht,AlertIcon as It,AlertDescription as ft}from"@pankod/refine-chakra-ui";var pe=({error:n})=>n?Ae.createElement(gt,{style:{minHeight:300}},Ae.createElement(ht,{status:"error",variant:"subtle",flexDirection:"column",alignItems:"center",justifyContent:"center",textAlign:"center",height:"200px"},Ae.createElement(It,{boxSize:"40px",mr:0}),Ae.createElement(ft,{mt:4,maxWidth:"sm"},n!=null?n:""))):null;import qr from"react";import{Box as xt,Spinner as Ft}from"@pankod/refine-chakra-ui";var me=()=>qr.createElement(xt,{position:"relative",bg:"chakra-body-bg",minH:120},qr.createElement(Ft,{position:"absolute",top:"50%",left:"50%",transform:"translate(-50%, -50%)"}));import O from"react";import{Box as bt,Button as cr,Modal as Ct,HStack as vt,useDisclosure as Tt,ModalBody as At,ModalCloseButton as St,ModalContent as wt,ModalOverlay as Lt}from"@pankod/refine-chakra-ui";import{IconCode as Vt,IconMessageCircle as Et,IconCopy as Nt,IconCheck as Bt}from"@tabler/icons";var le=({code:n,loading:u})=>{let $=O.useMemo(()=>J(n!=null?n:""),[n]),d=O.useRef(null),[g,o]=O.useState(!1),{isOpen:l,onOpen:C,onClose:I}=Tt();if(u)return null;let A=()=>{var w,L,T,p;(w=d==null?void 0:d.current)==null||w.select(),(L=d==null?void 0:d.current)==null||L.setSelectionRange(0,Number.MAX_SAFE_INTEGER),typeof navigator<"u"&&(navigator.clipboard.writeText((p=(T=d==null?void 0:d.current)==null?void 0:T.value)!=null?p:""),o(!0),setTimeout(()=>{o(!1)},1e3))};return $?O.createElement(O.Fragment,null,O.createElement(bt,{position:"fixed",bottom:"12px",right:"60px"},O.createElement(vt,{spacing:4},O.createElement(cr,{variant:"solid",colorScheme:"green",leftIcon:O.createElement(Vt,{size:18}),onClick:C},"Show Code"),O.createElement(cr,{colorScheme:"gray",variant:"solid",leftIcon:O.createElement(Et,{size:18}),as:"a",target:"_blank",href:"https://github.com/refinedev/refine/discussions/3046"},"Give Feedback"))),O.createElement("div",{style:{height:"50px"}}),O.createElement(Ct,{size:"4xl",isOpen:l,onClose:I},O.createElement(Lt,null),O.createElement(wt,null,O.createElement(St,{sx:{position:"absolute",right:0,top:0,height:"56px",width:"56px",color:"#666b7a"}}),O.createElement(At,{sx:{padding:0}},O.createElement(Q,{code:$}),O.createElement("textarea",{ref:d,defaultValue:($!=null?$:"").replace(/\\n/g,`\r
`),id:"code-input",style:{padding:0,margin:0,width:0,height:0,opacity:0,border:"none",display:"block"}})),O.createElement(cr,{sx:{position:"absolute",bottom:"12px",right:"12px",width:100},key:"copy",leftIcon:g?O.createElement(Bt,{size:18,style:{color:"green"}}):O.createElement(Nt,{size:18}),onClick:A},g?"Copied":"Copy")))):null};var ur=D({type:"show",additionalScope:[["@pankod/refine-chakra-ui","RefineChakraUI",Rt]],codeViewerComponent:le,loadingComponent:me,errorComponent:pe,renderer:({resource:n,fields:u,isCustomPage:$,id:d})=>{var t;let g=B((t=n.label)!=null?t:n.name,"show"),o="record",l=[["useShow","@pankod/refine-core"],["Show","@pankod/refine-chakra-ui"],["Heading","@pankod/refine-chakra-ui"]],I=u.filter(r=>(r==null?void 0:r.relation)&&!(r!=null&&r.fieldable)&&(r==null?void 0:r.resource)).filter(Boolean).map(r=>{if((r==null?void 0:r.relation)&&!r.fieldable&&r.resource){if(r.multiple){l.push(["useMany","@pankod/refine-core"]);let i=s(o,r.key);return r.accessor&&(i=`${s(o,r.key)}?.map((item: any) => ${s("item",void 0,r.accessor)})`),`
                    const { data: ${f(r.key,"Data")}, isLoading: ${f(r.key,"IsLoading")} } =
                    useMany({
                        resource: "${r.resource.name}",
                        ids: ${i} || [],
                        queryOptions: {
                            enabled: !!${o},
                        },
                    });
                    `}return l.push(["useOne","@pankod/refine-core"]),`
                    const { data: ${f(r.key,"Data")}, isLoading: ${f(r.key,"IsLoading")} } =
                    useOne({
                        resource: "${r.resource.name}",
                        id: ${s(o,r.key,r.accessor,!1)} || "",
                        queryOptions: {
                            enabled: !!${o},
                        },
                    });
                `}}).filter(Boolean),A=r=>{if(r.relation&&r.resource){let i=f(r.key,"Data"),e=f(r.key,"IsLoading");return r.multiple?(l.push(["TagField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),m`
                    <Heading as="h5" size="sm" mt={4} >${k(r.key)}</Heading>
                    {${e} ? <>Loading...</> : (
                        <>
                        ${(()=>{var h;if(r.relationInfer)if((h=r.relationInfer)!=null&&h.accessor){if(Array.isArray(r.relationInfer.accessor))return"Not Handled.";{let x=f(r.key),b=s(x,void 0,r.relationInfer.accessor);return`<HStack spacing="12px">{${i}?.data?.map((${x}: any) => <TagField key={${b}} value={${b}} />)}</HStack>`}}else return"Not Handled.";else return"not-handled - relation with multiple but no resource"})()}
                        </>
                    )}
                    `):m`
                    <Heading as="h5" size="sm" mt={4} >${k(r.key)}</Heading>
                    {${e} ? <>Loading...</> : (
                        <>
                        ${(()=>{var h;return r.relationInfer?(h=r.relationInfer)!=null&&h.accessor?Array.isArray(r.relationInfer.accessor)?`{${s(`${i}?.data`,void 0,r.relationInfer.accessor,' + " " + ')}}`:`{${i}?.data?.${r.relationInfer.accessor}}`:`{${i}?.data}`:`{${i}?.data?.id}`})()}
                        </>
                    )}
                    
                    `}},w=r=>{if(r.type==="text"){if(l.push(["TagField","@pankod/refine-chakra-ui"],["TextField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),r.multiple){let i=s("item",void 0,r.accessor);return m`
                    <Heading as="h5" size="sm" mt={4} >${k(r.key)}</Heading>
                    <HStack spacing="12px">
                    {${s(o,r.key)}?.map((item: any) => (
                        <TagField value={${i}} key={${i}} />
                    ))}
                    </HStack>
                `}return m`
                    <Heading as="h5" size="sm" mt={4} >${k(r.key)}</Heading>
                    <TextField value={${s(o,r.key,r.accessor)}} />
                `}},L=r=>{if(r.type==="image"){if(l.push(["Image","@pankod/refine-chakra-ui"]),r.multiple){let i=s("item",void 0,r.accessor);return m`
                    <Heading as="h5" size="sm" mt={4} >${k(r.key)}</Heading>
                    {${s(o,r.key)}?.map((item: any) => (
                        <Image sx={{ maxWidth: 200 }} src={${i}} key={${i}} />
                    ))}
                `}return m`
                    <Heading as="h5" size="sm" mt={4} >${k(r.key)}</Heading>
                    <Image sx={{ maxWidth: 200 }} src={${s(o,r.key,r.accessor," + ")}} />
                    `}},T=r=>{if(r.type==="email"){if(l.push(["TagField","@pankod/refine-chakra-ui"],["EmailField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),r.multiple){let i=s("item",void 0,r.accessor);return m`
                    <Heading as="h5" size="sm" mt={4} >${k(r.key)}</Heading>
                    <HStack spacing="12px">
                    {${s(o,r.key)}?.map((item: any) => (
                        <TagField value={${i}} key={${i}} />
                    ))}
                    </HStack>
                `}return m`
                    <Heading as="h5" size="sm" mt={4} >${k(r.key)}</Heading>
                    <EmailField value={${s(o,r.key,r.accessor," + ")}} />
                `}},p=r=>{if(r.type==="url"){if(l.push(["TagField","@pankod/refine-chakra-ui"],["UrlField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),r.multiple){let i=s("item",void 0,r.accessor);return m`
                    <Heading as="h5" size="sm" mt={4} >${k(r.key)}</Heading>
                    <HStack spacing="12px">
                    {${s(o,r.key)}?.map((item: any) => (
                        <TagField value={${i}} key={${i}} />
                    ))}
                    </HStack>
                `}return m`
                    <Heading as="h5" size="sm" mt={4} >${k(r.key)}</Heading>
                    <UrlField value={${s(o,r.key,r.accessor," + ")}} />
                `}},V=r=>{if(r.type==="boolean"){if(l.push(["TagField","@pankod/refine-chakra-ui"],["BooleanField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),r.multiple){let i=s("item",void 0,r.accessor);return m`
                    <Heading as="h5" size="sm" mt={4} >${k(r.key)}</Heading>
                    <HStack spacing="12px">
                    {${s(o,r.key)}?.map((item: any) => (
                        <TagField value={${i}} key={${i}} />
                    ))}
                    </HStack>
                `}return m`
                    <Heading as="h5" size="sm" mt={4} >${k(r.key)}</Heading>
                    <BooleanField value={${s(o,r.key,r.accessor," && ")}} />
                `}},y=r=>{if(r.type==="date"){if(l.push(["DateField","@pankod/refine-chakra-ui"]),r.multiple){let i=s("item",void 0,r.accessor);return m`
                    <Heading as="h5" size="sm" mt={4} >${k(r.key)}</Heading>
                    {${s(o,r.key)}?.map((item: any) => (
                        <DateField value={${i}} key={${i}} />
                    ))}
                `}return m`
                    <Heading as="h5" size="sm" mt={4} >${k(r.key)}</Heading>
                    <DateField value={${s(o,r.key,r.accessor," + ' ' + ")}} />
                `}},c=r=>{if(r.type==="richtext")return l.push(["MarkdownField","@pankod/refine-chakra-ui"]),m`
                    <Heading as="h5" size="sm" mt={4}>${k(r.key)}</Heading>
                    <MarkdownField value={${s(o,r.key,r.accessor,' + " " + ')}} />
                `},S=r=>{if(r.type==="number"){if(l.push(["NumberField","@pankod/refine-chakra-ui"],["TagField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),r.multiple){let i=s("item",void 0,r.accessor);return m`
                    <Heading as="h5" size="sm" mt={4} >${k(r.key)}</Heading>
                    <HStack spacing="12px">
                    {${s(o,r.key)}?.map((item: any) => (
                        <TagField value={${i}} key={${i}} />
                    ))}
                    </HStack>
                `}return m`
                    <Heading as="h5" size="sm" mt={4} >${k(r.key)}</Heading>
                    <NumberField value={${s(o,r.key,r.accessor,' + " " + ')} ?? ""} />
                `}},a=r=>{if(r)return m`
                    ${r}
                `},v=u.map(r=>{switch(r==null?void 0:r.type){case"text":return a(w(r));case"number":return a(S(r));case"richtext":return a(c(r));case"email":return a(T(r));case"image":return a(L(r));case"date":return a(y(r));case"boolean":return a(V(r));case"url":return a(p(r));case"relation":return a(A(r));default:return}});return M(l),m`
        ${R(l)}
        
        export const ${g} = () => {
            const { queryResult } = useShow(${$?`{ 
                        resource: "${n.name}", 
                        id: ${d}
                    }`:""});
            const { data, isLoading } = queryResult;

            const ${o} = data?.data;
        
            ${I}

            return (
                <Show isLoading={isLoading}>
                    ${v.join("")}
                </Show>
            );
        };
        `}});import*as Mt from"@pankod/refine-chakra-ui";import*as Dt from"@pankod/refine-react-table";import*as Pt from"@tabler/icons";var de=n=>Array.isArray(n.accessor)||n.multiple?`accessorKey: "${n.key}"`:n.accessor?`accessorKey: "${E(n.key,void 0,n.accessor)}"`:`accessorKey: "${n.key}"`,pr=D({type:"list",additionalScope:[["@pankod/refine-chakra-ui","RefineChakraUI",Mt],["@pankod/refine-react-table","RefineReactTable",Dt],["@tabler/icons","TablerIcons",Pt]],codeViewerComponent:le,loadingComponent:me,errorComponent:pe,renderer:({resource:n,fields:u,isCustomPage:$})=>{var i;let d=B((i=n.label)!=null?i:n.name,"list"),g="tableData?.data",o=[["IResourceComponentsProps","@pankod/refine-core"],["useTable","@pankod/refine-react-table"],["ColumnDef","@pankod/refine-react-table"],["flexRender","@pankod/refine-react-table"],["List","@pankod/refine-chakra-ui"],["TableContainer","@pankod/refine-chakra-ui"],["Table","@pankod/refine-chakra-ui"],["Thead","@pankod/refine-chakra-ui"],["Tr","@pankod/refine-chakra-ui"],["Th","@pankod/refine-chakra-ui"],["Tbody","@pankod/refine-chakra-ui"],["Tr","@pankod/refine-chakra-ui"],["Td","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"],["Button","@pankod/refine-chakra-ui"],["IconButton","@pankod/refine-chakra-ui"],["usePagination","@pankod/refine-chakra-ui"],["Box","@pankod/refine-chakra-ui"],["IconChevronRight","@tabler/icons"],["IconChevronLeft","@tabler/icons"]],l=u.filter(e=>(e==null?void 0:e.relation)&&!(e!=null&&e.fieldable)&&(e==null?void 0:e.resource)),C=l.filter(Boolean).map(e=>{if((e==null?void 0:e.relation)&&!e.fieldable&&e.resource){o.push(["GetManyResponse","@pankod/refine-core"]),o.push(["useMany","@pankod/refine-core"]);let h="";return e.multiple?h=`[].concat(...(${g}?.map((item) => ${s("item",e.key,e.accessor,!1)}) ?? []))`:h=`${g}?.map((item) => ${s("item",e.key,e.accessor,!1)}) ?? []`,`
                    const { data: ${f(e.key,"Data")} } =
                    useMany({
                        resource: "${e.resource.name}",
                        ids: ${h},
                        queryOptions: {
                            enabled: !!${g},
                        },
                    });
                    `}}).filter(Boolean),I=l==null?void 0:l.map(e=>{if(e&&e.resource)return f(e.key,"Data")}).filter(Boolean),A=e=>{var h;if(e.relation&&e.resource){let x=`${f(e.key,"Data")}?.data`;if(Array.isArray(e.accessor))return;let b=`id: "${e.key}"`,F=`header: "${k(e.key)}"`,N=de(e),q="";if(e.multiple){o.push(["TagField","@pankod/refine-chakra-ui"]);let z="item";e!=null&&e.relationInfer&&(z=s("item",void 0,e.relationInfer.accessor)),q=`cell: function render({ getValue, table }) {
                        const meta = table.options.meta as {
                            ${f(e.key,"Data")}: GetManyResponse;
                        };

                        const ${f(e.key,"")} = getValue<any[]>()?.map((item) => {
                            return meta.${f(e.key,"Data")}?.data?.find(
                                (resourceItems) => resourceItems.id === ${s("item",void 0,e.accessor)}
                            );
                        })


                        return (
                            <HStack>
                                {${f(e.key,"")}?.map((item, index) => (
                                    <TagField key={index} value={${z}} />
                                ))}
                            </HStack>
                        )
                    }
                `}else e!=null&&e.relationInfer?q=`cell: function render({ getValue, table }) {
                            const meta = table.options.meta as {
                                ${f(e.key,"Data")}: GetManyResponse;
                            };

                            const ${f(e.key,"")} = meta.${x}?.find(
                                (item) => item.id === getValue<any>(),
                            );

                            return ${s(f(e.key),void 0,(h=e==null?void 0:e.relationInfer)==null?void 0:h.accessor)} ?? "Loading...";
                        },`:q="";return`
                    {
                        ${b},
                        ${F},
                        ${N},
                        ${q}
                    }
                `}},w=e=>{if(e.type==="image"){o.push(["Image","@pankod/refine-chakra-ui"]);let h=`id: "${e.key}"`,x=de(e),b=`header: "${k(e.key)}"`,F=m`
                    cell: function render({ getValue }) {
                        ${e!=null&&e.accessor?"try {":""}
                            return <Image sx={{ maxWidth: "100px" }} src={${s("getValue<any>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0," + ")}} />
                        ${e!=null&&e.accessor?" } catch (error) { return null; }":""}
                    }
                `;if(e.multiple){let N=s("item",void 0,e.accessor," + ");F=`
                        cell: function render({ getValue }) {
                            ${e!=null&&e.accessor?"try {":""}
                                return (
                                    <HStack>
                                        {getValue<any[]>()?.map((item, index) => (
                                            <Image src={${N}} key={index} sx={{ height: "50px", maxWidth: "100px" }} />
                                        ))}
                                    </HStack>
                                )
                            ${e!=null&&e.accessor?" } catch (error) { return null; }":""}
                        }
                    `}return`
                    {
                        ${h},
                        ${x},
                        ${b},
                        ${F}
                    }
                `}},L=e=>{if(e.type==="email"){o.push(["EmailField","@pankod/refine-chakra-ui"]);let h=`id: "${e.key}"`,x=de(e),b=`header: "${k(e.key)}"`,F=m`
                    cell: function render({ getValue }) {
                        return <EmailField value={${s("getValue<any>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0,' + " " + ')}} />
                    }
                `;return e.multiple&&(o.push(["TagField","@pankod/refine-chakra-ui"]),F=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${b},
                        ${F}
                    }
                `}},T=e=>{if(e.type==="url"){o.push(["UrlField","@pankod/refine-chakra-ui"]);let h=`id: "${e.key}"`,x=de(e),b=`header: "${k(e.key)}"`,F=m`
                    cell: function render({ getValue }) {
                        return <UrlField value={${s("getValue<any>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0," + ")}} />
                    }
                `;return e.multiple&&(o.push(["TagField","@pankod/refine-chakra-ui"]),F=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${b},
                        ${F}
                    }
                `}},p=e=>{if((e==null?void 0:e.type)==="boolean"){o.push(["BooleanField","@pankod/refine-chakra-ui"]);let h=`id: "${e.key}"`,x=de(e),b=`header: "${k(e.key)}"`,F=m`
                    cell: function render({ getValue }) {
                        return <BooleanField value={${s("getValue<any>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0," + ")}} />
                    }
                `;return e.multiple&&(F=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <BooleanField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${b},
                        ${F}
                    }
                `}},V=e=>{if(e.type==="date"){o.push(["DateField","@pankod/refine-chakra-ui"]);let h=`id: "${e.key}"`,x=de(e),b=`header: "${k(e.key)}"`,F=m`
                    cell: function render({ getValue }) {
                        return <DateField value={${s("getValue<any>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0,' + " " + ')}} />
                    }
                `;return e.multiple&&(F=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <DateField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${b},
                        ${F}
                    }
                `}},y=e=>{if((e==null?void 0:e.type)==="richtext"){o.push(["MarkdownField","@pankod/refine-chakra-ui"]);let h=`id: "${e.key}"`,x=de(e),b=`header: "${k(e.key)}"`,F=m`
                    cell: function render({ getValue }) {
                        return <MarkdownField value={(${s("getValue<string>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0)})?.slice(0, 80) + "..." } />
                    }
                `;return e.multiple&&(F=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<string[]>()?.map((item, index) => (
                                        <MarkdownField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${b},
                        ${F}
                    }
                `}},c=e=>{if(e&&(e.type==="text"||e.type==="number")){let h=`id: "${e.key}"`,x=de(e),b=`header: "${k(e.key)}"`,F="";return e.multiple&&(o.push(["TagField","@pankod/refine-chakra-ui"]),F=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${s("item",void 0,e.accessor,' + " " + ')}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),!e.multiple&&Array.isArray(e.accessor)&&(F=`
                        cell: function render({ getValue }) {
                            return (
                                <>{${s("getValue<any>()",e.key,e.accessor)}}</>
                            );
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${b},
                        ${F}
                    }
                `}},{canEdit:S,canShow:a,canDelete:v}=n!=null?n:{};S&&o.push(["EditButton","@pankod/refine-chakra-ui"]),a&&o.push(["ShowButton","@pankod/refine-chakra-ui"]),v&&o.push(["DeleteButton","@pankod/refine-chakra-ui"]);let t=S||a||v?m`
        {
            id: "actions",
            accessorKey: "id",
            header: "Actions",
            cell: function render({ getValue }) {
                return (
                    <HStack>
                    ${a?m`
                        <ShowButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `:""}
                        ${S?m`
                        <EditButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `:""}
                        ${v?m`
                        <DeleteButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `:""}
                    </HStack>
                );
            },
        },
            `:"",r=u.map(e=>{switch(e==null?void 0:e.type){case"text":case"number":return c(e);case"richtext":return y(e);case"email":return L(e);case"image":return w(e);case"date":return V(e);case"boolean":return p(e);case"url":return T(e);case"relation":return A(e);default:return}});return M(o),m`
        import React from "react";
        ${R(o)}
        
        export const ${d}: React.FC<IResourceComponentsProps> = () => {
            const columns = React.useMemo<ColumnDef<any>[]>(() => [
                ${[...r,t].filter(Boolean).join(",")}
            ], []);

            const {
                getHeaderGroups,
                getRowModel,
                setOptions,
                refineCore: {
                    setCurrent,
                    pageCount,
                    current,
                    tableQueryResult: { data: tableData },
                },
            } = useTable({
                columns,
                ${$?`
                refineCoreProps: {
                    resource: "${n.name}",
                }
                `:""}
                
            });

            ${C}

            setOptions((prev) => ({
                ...prev,
                meta: {
                    ...prev.meta,
                    ${I.join(", ")}
                },
            }));

            return (
                <List>
                    <TableContainer whiteSpace="pre-line">
                        <Table variant="simple">
                            <Thead>
                                {getHeaderGroups().map((headerGroup) => (
                                    <Tr key={headerGroup.id}>
                                        {headerGroup.headers.map((header) => (
                                            <Th key={header.id}>
                                                {!header.isPlaceholder && (
                                                    flexRender(
                                                        header.column.columnDef
                                                            .header,
                                                        header.getContext(),
                                                    )
                                                )}
                                            </Th>
                                        ))}
                                    </Tr>
                                ))}
                            </Thead>
                            <Tbody>
                                {getRowModel().rows.map((row) => (
                                    <Tr key={row.id}>
                                        {row.getVisibleCells().map((cell) => (
                                            <Td key={cell.id}>
                                                {flexRender(
                                                    cell.column.columnDef.cell,
                                                    cell.getContext(),
                                                )}
                                            </Td>
                                        ))}
                                    </Tr>
                                ))}
                            </Tbody>
                        </Table>
                    </TableContainer>
                    <Pagination
                        current={current}
                        pageCount={pageCount}
                        setCurrent={setCurrent}
                    />
                </List>   
            );
        };

        type PaginationProps = {
            current: number;
            pageCount: number;
            setCurrent: (page: number) => void;
        };

        const Pagination: React.FC<PaginationProps> = ({
            current,
            pageCount,
            setCurrent,
        }) => {
            const pagination = usePagination({
                current,
                pageCount,
            });
        
            return (
                <Box display="flex" justifyContent="flex-end">
                    <HStack my="3" spacing="1">
                        {pagination?.prev && (
                            <IconButton
                                aria-label="previous page"
                                onClick={() => setCurrent(current - 1)}
                                disabled={!pagination?.prev}
                                variant="outline"
                            >
                                <IconChevronLeft size="18" />
                            </IconButton>
                        )}
        
                        {pagination?.items.map((page) => {
                            if (typeof page === "string")
                                return <span key={page}>...</span>;
        
                            return (
                                <Button
                                    key={page}
                                    onClick={() => setCurrent(page)}
                                    variant={page === current ? "solid" : "outline"}
                                >
                                    {page}
                                </Button>
                            );
                        })}
                        {pagination?.next && (
                            <IconButton
                                aria-label="next page"
                                onClick={() => setCurrent(current + 1)}
                                variant="outline"
                            >
                                <IconChevronRight size="18" />
                            </IconButton>
                        )}
                    </HStack>
                </Box>
            );
        };
        `}});import*as Ht from"@pankod/refine-core";import*as Ot from"@pankod/refine-chakra-ui";import*as Gt from"@pankod/refine-react-hook-form";var mr=D({type:"create",additionalScope:[["@pankod/refine-core","RefineCore",Ht],["@pankod/refine-chakra-ui","RefineChakraUI",Ot],["@pankod/refine-react-hook-form","RefineReactHookForm",Gt]],codeViewerComponent:le,loadingComponent:me,errorComponent:pe,renderer:({resource:n,fields:u,isCustomPage:$})=>{var T;let d=B((T=n.label)!=null?T:n.name,"create"),g=[["Create","@pankod/refine-chakra-ui"],["FormControl","@pankod/refine-chakra-ui"],["FormLabel","@pankod/refine-chakra-ui"],["FormErrorMessage","@pankod/refine-chakra-ui"],["useForm","@pankod/refine-react-hook-form"]],l=u.filter(p=>(p==null?void 0:p.relation)&&!(p!=null&&p.fieldable)&&(p==null?void 0:p.resource)).filter(Boolean).map(p=>{if((p==null?void 0:p.relation)&&!p.fieldable&&p.resource)return g.push(["useSelect","@pankod/refine-core"]),`
                    const { options: ${f(p.key,"Options")} } =
                    useSelect({
                        resource: "${p.resource.name}",
                        ${K(p)}
                    });
                `}).filter(Boolean),C=p=>{if(p.relation&&p.resource){g.push(["useSelect","@pankod/refine-core"]),g.push(["Select","@pankod/refine-chakra-ui"]);let V=f(p.key,"Options");return m`
                <FormControl mb="3" isInvalid={!!errors?.${E(p.key,void 0)}}>
                    <FormLabel>${k(p.key)}</FormLabel>
                    <Select
                        placeholder="Select ${U(p.resource.name)}"
                        {...register("${E(p.key,void 0,p.accessor)}", {
                            required: "This field is required",
                        })}
                    >
                        {${V}?.map((option) => (
                            <option value={option.value} key={option.value}>
                                {option.label}
                            </option>
                        ))}
                    </Select>
                    <FormErrorMessage>
                        {${s("(errors as any)",p.key,p.accessor,!1)}?.message as string}
                    </FormErrorMessage>
                </FormControl>
                `}},I=p=>{if(p.type==="text"||p.type==="url"||p.type==="email"||p.type==="number"||p.type==="date"||p.type==="richtext")return H(p.key)||(g.push(["Input","@pankod/refine-chakra-ui"]),p.multiple)?void 0:m`
                    <FormControl mb="3" isInvalid={!!${s("(errors as any)",p.key,p.accessor,!1)}}>
                        <FormLabel>${k(p.key)}</FormLabel>
                        <Input
                            ${p.type!=="date"&&p.type!=="richtext"?`type="${p.type}"`:""}
                            {...register("${E(p.key,void 0,p.accessor)}", {
                                required: "This field is required",
                            })}
                        />
                        <FormErrorMessage>
                            {${s("(errors as any)",p.key,p.accessor,!1)}?.message as string}
                        </FormErrorMessage>
                    </FormControl>
                `},A=p=>{if(p.type==="boolean")return g.push(["Checkbox","@pankod/refine-chakra-ui"]),p.multiple?void 0:m`
                    <FormControl mb="3" isInvalid={!!${s("errors",p.key,p.accessor,!1)}}>
                        <FormLabel>${k(p.key)}</FormLabel>
                        <Checkbox
                            {...register("${E(p.key,void 0,p.accessor)}", {
                                required: "This field is required",
                            })}
                        />
                        <FormErrorMessage>
                            {${s("errors",p.key,p.accessor,!1)}?.message as string}
                        </FormErrorMessage>
                    </FormControl>
                   
                `},w=p=>{if(p.type==="date"){let V=I(p);return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-chakra-ui" package.
                        To use a <DatePicker> component, you can examine the following links:
                        
                        - https://github.com/aboveyunhai/chakra-dayzed-datepicker
                        - https://github.com/wojtekmaj/react-date-picker
                    */}
                    ${V!=null?V:""}
                    `}},L=u.map(p=>{switch(p==null?void 0:p.type){case"text":case"number":case"email":case"url":case"richtext":return I(p);case"date":return w(p);case"boolean":return A(p);case"relation":return C(p);default:return}});return M(g),m`
        ${R(g)}
        
        export const ${d} = () => {
            const {
                refineCore: { formLoading },
                saveButtonProps,
                register,
                formState: { errors },
            } = useForm(
                ${$?`
                { 
                    refineCoreProps: {
                        resource: "${n.name}",
                        action: "create",
                    }
                }`:""}
            );
        
            ${l}

            return (
                <Create isLoading={formLoading} saveButtonProps={saveButtonProps}>
                    ${L.join("")}
                </Create>
            );
        };
        `}});import*as jt from"@pankod/refine-core";import*as qt from"@pankod/refine-chakra-ui";import*as zt from"@pankod/refine-react-hook-form";var lr=D({type:"edit",additionalScope:[["@pankod/refine-core","RefineCore",jt],["@pankod/refine-chakra-ui","RefineChakraUI",qt],["@pankod/refine-react-hook-form","RefineReactHookForm",zt]],codeViewerComponent:le,loadingComponent:me,errorComponent:pe,renderer:({resource:n,fields:u,isCustomPage:$,id:d})=>{var V,y;let g=B((V=n.label)!=null?V:n.name,"edit"),o=f((y=n.label)!=null?y:n.name,"Data"),l=[["Edit","@pankod/refine-chakra-ui"],["FormControl","@pankod/refine-chakra-ui"],["FormLabel","@pankod/refine-chakra-ui"],["FormErrorMessage","@pankod/refine-chakra-ui"],["useForm","@pankod/refine-react-hook-form"]],I=u.filter(c=>(c==null?void 0:c.relation)&&!(c!=null&&c.fieldable)&&(c==null?void 0:c.resource)).filter(Boolean).map(c=>{if((c==null?void 0:c.relation)&&!c.fieldable&&c.resource){l.push(["useSelect","@pankod/refine-core"]);let S=s(o,c.key,c.accessor,!1);return c.multiple&&c.accessor&&(S=`${s(o,c.key)}?.map((item: any) => ${s("item",void 0,c.accessor)})`),`
                    const { options: ${f(c.key,"Options")} } =
                    useSelect({
                        resource: "${c.resource.name}",
                        defaultValue: ${S},
                        ${K(c)}
                    });
                `}}).filter(Boolean),A=c=>{if(c.relation&&c.resource){l.push(["useSelect","@pankod/refine-core"]),l.push(["Select","@pankod/refine-chakra-ui"]);let S=f(c.key,"Options");return m`
                <FormControl mb="3" isInvalid={!!errors?.${E(c.key,void 0)}}>
                    <FormLabel>${k(c.key)}</FormLabel>
                    <Select
                        placeholder="Select ${U(c.resource.name)}"
                        {...register("${E(c.key,void 0,c.accessor)}", {
                            required: "This field is required",
                        })}
                    >
                        {${S}?.map((option) => (
                            <option value={option.value} key={option.value}>
                                {option.label}
                            </option>
                        ))}
                    </Select>
                    <FormErrorMessage>
                        {${s("(errors as any)",c.key,c.accessor,!1)}?.message as string}
                    </FormErrorMessage>
                </FormControl>
                `}},w=c=>{if(c.type==="text"||c.type==="url"||c.type==="email"||c.type==="number"||c.type==="date"||c.type==="richtext"){if(l.push(["Input","@pankod/refine-chakra-ui"]),c.multiple){let S=E(c.key,"${index}",c.accessor),a=s(s("(errors as any)",c.key,void 0,!1)+"?.[index]",void 0,c.accessor);return`
                        <>
                            {${s(o,c.key)}?.map((item: any, index: number) => (
                                <FormControl key={index} mb="3" isInvalid={!!${a}}>
                                    <FormLabel>${k(c.key)} #{index + 1}</FormLabel>
                                    <Input
                                        {...register(\`${S}\`, {
                                            required: "This field is required",
                                        })}
                                    />
                                    <FormErrorMessage>
                                        {${s(a,"message")} as string}
                                    </FormErrorMessage>
                                </FormControl>
                            ))}
                        </>
                    `}return m`
                    <FormControl mb="3" isInvalid={!!${s("(errors as any)",c.key,c.accessor,!1)}}>
                        <FormLabel>${k(c.key)}</FormLabel>
                        <Input
                            ${H(c.key)?"disabled":""}
                            ${c.type!=="date"&&c.type!=="richtext"?`type="${c.type}"`:""}
                            {...register("${E(c.key,void 0,c.accessor)}", {
                                required: "This field is required",
                            })}
                        />
                        <FormErrorMessage>
                            {${s("(errors as any)",c.key,c.accessor,!1)}?.message as string}
                        </FormErrorMessage>
                    </FormControl>
                `}},L=c=>{if(c.type==="boolean"){if(l.push(["Checkbox","@pankod/refine-chakra-ui"]),c.multiple){let S=E(c.key,void 0,c.accessor),a=s("(errors as any)",c.key,void 0,!1)+"?.[index]";return`
                        <>
                            {${s(o,c.key)}?.map((item: any, index: number) => (
                                <FormControl key={index} mb="3" isInvalid={!!${a}}>
                                    <FormLabel>${k(c.key)} #{index + 1}</FormLabel>
                                    <Checkbox
                                        {...register(\`${S}.\${index}\`, {
                                            required: "This field is required",
                                        })}
                                    />
                                    <FormErrorMessage>
                                        {${a}?.message as string}
                                    </FormErrorMessage>
                                </FormControl>
                            ))}
                        </>
                    `}return m`
                    <FormControl mb="3" isInvalid={!!${s("errors",c.key,c.accessor,!1)}}>
                        <FormLabel>${k(c.key)}</FormLabel>
                        <Checkbox
                            {...register("${E(c.key,void 0,c.accessor)}", {
                                required: "This field is required",
                            })}
                        />
                        <FormErrorMessage>
                            {${s("errors",c.key,c.accessor,!1)}?.message as string}
                        </FormErrorMessage>
                    </FormControl>
                   
                `}},T=c=>{if(c.type==="date")return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-chakra-ui" package.
                        To use a <DatePicker> component, you can examine the following links:
                        
                        - https://github.com/aboveyunhai/chakra-dayzed-datepicker
                        - https://github.com/wojtekmaj/react-date-picker
                    */}
                    ${w(c)}
                    `},p=u.map(c=>{switch(c==null?void 0:c.type){case"text":case"number":case"email":case"url":case"richtext":return w(c);case"date":return T(c);case"boolean":return L(c);case"relation":return A(c);default:return}});return M(l),m`
        ${R(l)}
        
        export const ${g} = () => {
            const {
                refineCore: { formLoading, queryResult },
                saveButtonProps,
                register,
                formState: { errors },
            } = useForm(
                ${$?`
                { 
                    refineCoreProps: {
                        resource: "${n.name}",
                        id: ${d},
                        action: "edit",
                    }
                }`:""}
            );
        
            const ${o} = queryResult?.data?.data;
        
            ${I}

            return (
                <Edit isLoading={formLoading} saveButtonProps={saveButtonProps}>
                    ${p.join("")}
                </Edit>
            );
        };
        `}});var Xu=({action:n,id:u,...$})=>{let{useParams:d}=Ut(),{action:g,id:o}=d();switch(n!=null?n:g){case"show":return Se.createElement(ur,{...$,id:u!=null?u:o});case"create":return Se.createElement(mr,{...$,id:u!=null?u:o});case"edit":return Se.createElement(lr,{...$,id:u!=null?u:o});default:return Se.createElement(pr,{...$,id:u!=null?u:o})}};export{Ke as AntdCreateInferencer,_e as AntdEditInferencer,ii as AntdInferencer,We as AntdListInferencer,Ue as AntdShowInferencer,mr as ChakraUICreateInferencer,lr as ChakraUIEditInferencer,Xu as ChakraUIInferencer,pr as ChakraUIListInferencer,ur as ChakraUIShowInferencer,Ye as MantineCreateInferencer,er as MantineEditInferencer,oc as MantineInferencer,Qe as MantineListInferencer,Je as MantineShowInferencer,sr as MuiCreateInferencer,ir as MuiEditInferencer,eu as MuiInferencer,ar as MuiListInferencer,or as MuiShowInferencer,D as createInferencer};
