// @pankod/refine-inferencer@1.6.1/dist/esm/mantine.js
import te from"react";import{useRouterContext as _r}from"@pankod/refine-core";import*as zr from"@pankod/refine-mantine";import E from"react";import{useResource as er}from"@pankod/refine-core";var rr=/(image|photo|avatar|url|logo|cover|thumbnail|icon)/i,oe=(r,t)=>{let i=["name","label","title","count","content","username","nickname","login","firstName","lastName","url"];rr.test(r)&&i.unshift("url");let s=i.find(u=>u in t&&!!t[u]);if(s)return s==="firstName"&&"lastName"in t?["firstName","lastName"]:s};function ae(r,t,i){var u;if(t)return t;let s=i==null?void 0:i.find(o=>o.name===r);return(u=s==null?void 0:s.options)!=null&&u.dataProviderName?s.options.dataProviderName:"default"}var X=r=>{var t;return(t=r==null?void 0:r.options)==null?void 0:t.dataProviderName};var se=r=>{let t=[...r].filter(Boolean).sort((i,s)=>(s.priority||0)-(i.priority||0))[0];return t||null};import nr from"prettier/standalone";import tr from"prettier/parser-typescript";var ie=r=>{try{return nr.format(r,{parser:"typescript",plugins:[tr],arrowParens:"always",trailingComma:"all",semi:!0,tabWidth:4,printWidth:80})}catch(t){return console.warn("Formatting error in inferencer",t),r}};var ce=r=>r.replace(/ {2,}/g," ").replace(/\n{2,}/g,`
`);var ue=/(-id|-ids|_id|_ids|Id|Ids|ID|IDs)(\[\])?$/,Te=(r,t)=>{let i=ue.test(r),s=typeof t=="string"||typeof t=="number",u=Array.isArray(t)&&t.every(o=>typeof o=="string"||typeof o=="number");return i&&s||i&&u?{key:r,relation:!0,multiple:!!u,type:"relation",priority:1}:!1};var G=r=>r.replace(ue,"");var g=r=>G(r).replace(/([a-z])([A-Z])/g,"$1 $2").replace(/_/g," ").replace(/-/g," ").split(" ").map(c=>c.charAt(0).toUpperCase()+c.slice(1)).join(" ");var pe=r=>r.replace(/export\s+(const|let|var|type|interface|function|class)\s+(\w+)\s*(=|:)\s*/g,"$1 $2 = ").split(`
`).filter(u=>!u.trim().startsWith("export default")).join(`
`);var ve=/import(?:(?:(?:[ \n\t]+([^ *\n\t\{\},]+)[ \n\t]*(?:,|[ \n\t]+))?([ \n\t]*\{(?:[ \n\t]*[^ \n\t"'\{\}]+[ \n\t]*,?)+\})?[ \n\t]*)|[ \n\t]*\*[ \n\t]*as[ \n\t]+([^ \n\t\{\}]+)[ \n\t]+)from[ \n\t]*(?:['"])([^'"\n]+)(?:['"])(?:;?)/g,or=/import[ \n\t](?:['"])([^'"\n]+)(?:['"])(?:;?)/g,ar=/((?:\w|\s|_)*)( as )((?:\w|\s|_)*)( |,)?/g,me=(r,t={})=>{let i=r.matchAll(ve),s=new Set;for(let u of i){let[,o,c,f,m]=u;if(m in t){let k=t[m];if(m==="react")continue;o&&s.add(`const { default: ${o} } = ${k};`),c&&s.add(`const${c.replace(ar,"$1: $3$4")} = ${k};`),f&&s.add(`const ${f} = ${k};`)}}return ce(`
    ${Array.from(s).join(`
`)}
    ${r.replace(ve,"").replace(or,"")}
    `)};import ne from"pluralize";var le=(r,t)=>r?t.find(s=>{let u=G(r.key);return s.name===u||s.name===(ne.isPlural(u)?u:ne.plural(u))||s.name===(ne.isSingular(u)?u:ne.singular(u))}):void 0;import sr from"pluralize";var D=(r,t)=>{let i=r.replace(/[^a-zA-Z0-9]/g," "),s=sr.singular(i),u=g(s).replace(/ /g,""),o=t.charAt(0).toUpperCase()+t.slice(1);return/^[a-zA-Z]/.test(r)?`${u}${o}`:`${o}${u}`};var ir=r=>r?Array.isArray(r)?r.join(""):typeof r=="string"?r:typeof r=="object"||typeof r=="function"?r.toString():typeof r=="number"?String(r):typeof r=="boolean"&&r?"true":r:"",y=(r,...t)=>r.reduce((s,u,o)=>`${s}${u}${ir(t[o])}`,"").trim();var de=(r,t)=>`
    ${r!=null?r:""}
    
    render(typeof ${t} !== "undefined" ? <${t} /> : <></>);
    `;var ye=r=>r==null?void 0:r.replace(/(\/\* hidden-start \*\/)(.|\n)*?(\/\* hidden-end \*\/)/g,"");var cr=/^[a-zA-Z_$][a-zA-Z_$0-9]*$/,Ae=r=>cr.test(r);function fe(r,t,i){let s=`${r}`;return t&&(s+="?.",Ae(t)?s+=t:s+=`['${t}']`),i&&(s+="?.",Ae(i)?s+=i:s+=`['${i}']`),s}function ur(r,t,i){return(i!=null?i:[]).map(u=>fe(r,t,u))}var l=(r,t,i,s=' + " " + ')=>Array.isArray(i)?s?ur(r,t,i).join(s):fe(r,t,i[0]):fe(r,t,i),w=(r,t,i)=>{let s=r;return t&&(s+=`.${t}`),i&&(Array.isArray(i)?s+=`.${i[0]}`:s+=`.${i}`),s};var O=r=>{let t=r.reduce((s,[u,o,c])=>(s[o]||(s[o]=[]),s[o].includes(u)||(c?s[o].push([u,!0]):s[o].push(u)),s),{});return Object.entries(t).map(([s,u])=>{let o=u.find(k=>Array.isArray(k)&&k[1]),c=u.filter(k=>typeof k=="string"),f=o?`${o[0]}`:"",m=c.length?`{ ${c.join(", ")} }`:"";return`import ${f}${f&&m?", ":""}${m} from "${s}";`})};import we from"pluralize";var j=(r="")=>we.isSingular(r)?r:we.singular(r);import Re from"pluralize";var J=(r="")=>Re.isPlural(r)?r:Re.plural(r);var pr=/^id$/i,Z=r=>pr.test(r);var Q=r=>r.relationInfer&&r.relationInfer.accessor?Array.isArray(r.relationInfer.accessor)?`optionLabel: "${r.relationInfer.accessor[0]}",`:r.relationInfer.accessor!=="title"?`optionLabel: "${r.relationInfer.accessor}",`:"":"";var z=r=>{};var T=(r,t="")=>{if(Number.isInteger(Number(r)))return t.charAt(0).toLowerCase()+t.slice(1)+r;let s=r.replace(/^\d+/,"").replace(/[^a-zA-Z0-9]/g,"-"),u=g(s).replace(/ /g,"");return`${u.charAt(0).toLowerCase()+u.slice(1)}${t.charAt(0).toUpperCase()+t.slice(1)}`};var Se=r=>{let t=(i,s,u,o=t)=>{let c=r.map(m=>m(i,s,u,o));return se(c)};return t};var Ne=r=>(i,s,u,o,c)=>r.reduce((f,m)=>m(f,s,u,o,c),i);var Ve=(r,t,i,s)=>{let u=Array.isArray(t),o=Array.isArray(t)&&t.every(c=>typeof c=="string"||typeof c=="number");if(u){if(!o){let f=s(r,t[0],i,s);return f?{...f,key:r,multiple:!0,priority:1}:!1}let c=s(r,t[0],i,s)||{type:"string"};return{key:r,multiple:!0,priority:1,type:c==null?void 0:c.type}}return!1};var Le=(r,t)=>typeof t=="boolean"?{key:r,type:"boolean"}:!1;import Ge from"dayjs";var mr=/(_at|_on|At|On|AT|ON)(\[\])?$/,De=(r,t)=>{let i=mr.test(r)&&Ge(t).isValid(),s=typeof t=="string"&&Ge(t).isValid();return i||s?{key:r,type:"date",priority:1}:!1};var lr=/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,Pe=(r,t)=>typeof t=="string"&&lr.test(t)?{key:r,type:"email",priority:1}:!1;var dr=/\.(gif|jpe?g|tiff?|png|webp|bmp|svg)$/i,Me=(r,t)=>typeof t=="string"&&dr.test(t)?{key:r,type:"image",priority:2}:!1;var Ee=(r,t)=>typeof t>"u"||t===null?null:!1;var Be=(r,t)=>typeof t=="number"?{key:r,type:"number"}:!1;var yr=/id$/i,Oe=(r,t,i,s)=>{let u=t!==null,o=!Array.isArray(t);if(u&&o&&typeof t=="object"){if(Object.keys(t).length===1&&yr.test(Object.keys(t)[0]))return{key:r,type:"relation",relation:!0,accessor:"id",priority:1};let m=oe(r,t);if((m?m.length>0:!1)&&m){let b=m&&Array.isArray(m)?m[0]:m,x=s(b,t[b],t,s);if(x){let $=Array.isArray(m)?m.map(F=>x.accessor?`${F}.${Array.isArray(x.accessor)?x.accessor[0]:x.accessor}`:F):x.accessor?Array.isArray(x.accessor)?`${m}.${x.accessor[0]}`:`${m}.${x.accessor}`:m;return{...x,fieldable:!0,key:r,accessor:$,priority:1}}}return{key:r,fieldable:!1,type:"object",priority:1}}return!1};var ze=(r,t)=>typeof t=="string"&&t.length>100?{key:r,type:"richtext",priority:1}:!1;var je=(r,t)=>typeof t=="string"?{key:r,type:"text"}:!1;var fr=/^(https?|ftp):\/\/(-\.)?([^\s/?\.#-]+\.?)+(\/[^\s]*)?$/i,Ke=(r,t)=>typeof t=="string"&&fr.test(t)?{key:r,type:"url",priority:1}:!1;var He=[Ve,Le,De,Pe,Me,Ee,Be,Oe,Te,ze,je,Ke];var Ue=(r,t,i,s)=>r.map(o=>{if(!o.relation&&(o.type==="text"||o.type==="richtext"||o.type==="number")&&!o.canRelation){let c=/^[a-z0-9-]+$/,f=$=>c.test(`${$}`),m=o.key.toLowerCase()!=="id";return!(!o.accessor||typeof o.accessor=="string")||!m?o:(o.multiple?s[o.key]:[s[o.key]]).every($=>f(o.accessor?$[o.accessor]:$))?{...o,canRelation:!0}:o}return o});var gr=/(image|photo|avatar|cover|thumbnail|icon)/i,_e=r=>r.map(i=>i.type==="url"&&gr.test(i.key)?{...i,type:"image"}:i);var We=(r,t)=>r.map(s=>{let u=le(s,t);return u?{...s,relation:!0,type:"relation",resource:u}:s});var qe=(r,t,i,s,u)=>r.map(c=>{if(c.relation&&c.type==="relation"&&!c.resource){let f=c.accessor?s[c.key][c.accessor]:s[c.key],m=u(c.key,f,s,u);if(m&&m.type!=="relation")return{...c,fieldable:!0,relation:!1,type:m.type,canRelation:!0}}return c});var Ze=[_e,We,qe,Ue];import Y,{useEffect as kr}from"react";import $r,{defaultProps as Ir}from"prism-react-renderer";import hr from"prism-react-renderer/themes/vsDark";var ge=({code:r,wrapperProps:t})=>(kr(()=>{var s;let i=document.createElement("style");i.appendChild(document.createTextNode(`
                #code-highlight::-webkit-scrollbar {-webkit-appearance: none; height: 7px; width: 7px;}
                #code-highlight::-webkit-scrollbar-thumb {background-color: #666b7a;}
                #code-highlight::-webkit-scrollbar-corner {background-color: rgb(30, 30, 30);}
                `)),(s=document.getElementById("code-highlight"))==null||s.appendChild(i)},[]),Y.createElement("div",{id:"code-highlight",style:{maxHeight:"75vh",height:"100%",overflow:"auto",backgroundColor:"rgb(30, 30, 30)"},...t!=null?t:{}},Y.createElement($r,{...Ir,theme:hr,code:r,language:"tsx"},({className:i,style:s,tokens:u,getLineProps:o,getTokenProps:c})=>Y.createElement("pre",{className:i,style:{...s,padding:"14px 14px 14px 14px",margin:"0",width:"100%",boxSizing:"border-box"}},u.map((f,m)=>Y.createElement("div",{key:m,...o({line:f,key:m})},f.map((k,b)=>Y.createElement("span",{key:b,...c({token:k,key:b})}))))))));import P from"react";import*as xr from"@pankod/refine-core";import{LivePreview as br,LiveProvider as Fr}from"@aliemir/react-live";var Je=[["react","React",P],["@pankod/refine-core","RefineCore",xr]],ke=P.createContext({}),$e=({fetchError:r,code:t,additionalScope:i,errorComponent:s})=>{var f;let u=(f=t&&pe(me(t,[...Je,...i!=null?i:[]].reduce((m,[k,b,x,$])=>$?m:{...m,[k]:b},{}))))!=null?f:"",o=P.useMemo(()=>({React:P,...P,...[...Je,...i!=null?i:[]].reduce((m,[k,b,x])=>({...m,[b]:x}),{})}),[i]),c=P.useMemo(()=>()=>{let{error:k}=P.useContext(ke);return s?P.createElement(s,{error:k!=null?k:r?"Something went wrong while fetching the resource data.":void 0}):null},[s,r]);return P.createElement(Fr,{Context:ke,code:u,scope:o,noInline:!0},!r&&P.createElement(br,{Context:ke}),P.createElement(c,null))};import ee from"react";import{useDataProvider as Cr,useResource as Tr}from"@pankod/refine-core";var Xe=(r,t,i)=>{let{resource:s,resourceName:u,id:o,resources:c}=Tr({resourceNameOrRouteName:t}),f=i!=null?i:o,m=Cr(),[k,b]=ee.useState(void 0),[x,$]=ee.useState(!0),[F,N]=ee.useState(!1),p=ee.useCallback(async L=>{var I,V;let v=(I=X(s))!=null?I:ae(u,void 0,c),a=m(v);N(!0);try{if(r==="list"||r==="create"){let e=await a.getList({resource:u});b((V=e.data)==null?void 0:V[0]),setTimeout(()=>{N(!1)},500)}if(r==="edit"||r==="show"&&L){let e=await a.getOne({resource:u,id:L});b(e.data),setTimeout(()=>{N(!1)},500)}}catch{setTimeout(()=>{N(!1)},500)}},[r,m,s,u,c]);return ee.useEffect(()=>{$(!1),!F&&!k&&p(f)},[p,f]),{data:k,loading:F,initial:x}};import re from"react";import{useDataProvider as vr}from"@pankod/refine-core";import{get as Qe}from"lodash";var Ye=({record:r,fields:t,infer:i})=>{let s=vr(),[u,o]=re.useState([]),[c,f]=re.useState(!0),[m,k]=re.useState(!1),b=re.useCallback(async x=>{console.groupCollapsed("@pankod/refine-inferencer is trying to detect relations");let $=[];k(!0);try{let F=x.map(async p=>{if(p&&(p.relation||p.canRelation)&&r){let L=X(p.resource),v=s(L),a=Array.isArray(p.accessor)?void 0:p.multiple?r[p.key].map(I=>p.accessor?Qe(I,p.accessor):I)[0]:p.accessor?Qe(r[p.key],p.accessor):r[p.key];if(a&&p.resource)try{let{data:I}=await v.getOne({resource:p.resource.name,id:a});$.push({status:"success",resource:p.resource.name,field:p.key});let V=i("__",I,{},i);return{...p,relationInfer:V}}catch{return $.push({status:"error",resource:p.resource.name,field:p.key}),{...p,relationInfer:null}}if(a){let I,V;try{let{data:n}=await v.getOne({resource:J(G(p.key)),id:a});$.push({status:"success",resource:J(G(p.key)),field:p.key}),I=n,V=!0}catch{$.push({status:"error",resource:J(G(p.key)),field:p.key});try{let{data:h}=await v.getOne({resource:j(G(p.key)),id:a});$.push({status:"success",resource:j(G(p.key)),field:p.key}),I=h,V=!1}catch{return $.push({status:"error",resource:j(G(p.key)),field:p.key}),{...p,relationInfer:null}}}let e=i("__",I,{},i),d=G(p.key);return{...p,relation:!0,type:"relation",resource:{name:V?J(d):j(d)},fieldable:!1,canRelation:void 0,relationInfer:e}}return{...p,relationInfer:null}}return p}),N=await Promise.all(F);o(N.filter(p=>p)),setTimeout(()=>{k(!1)},500)}catch{setTimeout(()=>{k(!1)},500)}setTimeout(()=>{console.log(`Tried to detect relations with ${$.length} attempts and succeeded with ${$.filter(F=>F.status==="success").length} attempts.`),console.groupEnd()},500)},[s,r]);return re.useEffect(()=>{f(!1),!m&&t&&t.length>0&&u.length===0&&b(t)},[b,m,t,b]),{fields:u,loading:m,initial:c}};var K=({type:r,additionalScope:t=[],customElements:i=[],fieldTransformers:s=[],renderer:u,loadingComponent:o,errorComponent:c,codeViewerComponent:f})=>{let m=Se([...He,...i]),k=Ne([...Ze,...s]),b=({resourceName:$,fieldTransformer:F,id:N})=>{var A;let{resource:p,resources:L}=er({resourceNameOrRouteName:$}),{resource:v}=er(),{data:a,loading:I,initial:V}=Xe(r,$!=null?$:p==null?void 0:p.name,N),e=E.useMemo(()=>{if(a){let S=Object.keys(a).map(B=>{let q=a[B];return m(B,q,a,m)}).filter(Boolean),C=k(S,L,p,a,m);return F?C.flatMap(B=>{let q=F(B);return q?[q]:[]}):C}return[]},[a,L,p,F]),{fields:d,loading:n}=Ye({record:a,fields:e,infer:m}),h=E.useMemo(()=>!I&&!n&&!V?u({resource:p,resources:L,fields:d,infer:m,isCustomPage:p.name!==v.name,id:N}):"",[p,L,d,I,n]);return E.createElement(E.Fragment,null,o&&(I||n)&&E.createElement(o,null),!I&&!n&&E.createElement(E.Fragment,null,E.createElement($e,{fetchError:!I&&!V&&!a,code:de(h,D((A=p.label)!=null?A:p.name,r)),errorComponent:c,additionalScope:t}),f&&E.createElement(f,{code:ye(h),loading:I||n})))};return({name:$,resource:F,fieldTransformer:N,id:p})=>E.createElement(b,{fieldTransformer:N,resourceName:F!=null?F:$,key:F!=null?F:$,id:p})};import Ie from"react";import{Alert as Ar,Center as wr}from"@pankod/refine-mantine";import{IconAlertCircle as Rr}from"@tabler/icons";var H=({error:r})=>r?Ie.createElement(wr,{style:{minHeight:300}},Ie.createElement(Ar,{title:"Error",color:"red",icon:Ie.createElement(Rr,null)},r!=null?r:"")):null;import Sr from"react";import{LoadingOverlay as Nr}from"@pankod/refine-mantine";var U=()=>Sr.createElement(Nr,{visible:!0});import R from"react";import{Affix as Vr,Button as he,Modal as Lr,Group as Gr,ActionIcon as Dr}from"@pankod/refine-mantine";import{IconCode as Pr,IconMessageCircle as Mr,IconX as Er,IconCopy as Br,IconCheck as Or}from"@tabler/icons";var _=({code:r,loading:t})=>{let i=R.useMemo(()=>ie(r!=null?r:""),[r]),s=R.useRef(null),[u,o]=R.useState(!1),[c,f]=R.useState(!1);if(t)return null;let m=()=>{var k,b,x,$;(k=s==null?void 0:s.current)==null||k.select(),(b=s==null?void 0:s.current)==null||b.setSelectionRange(0,Number.MAX_SAFE_INTEGER),typeof navigator<"u"&&(navigator.clipboard.writeText(($=(x=s==null?void 0:s.current)==null?void 0:x.value)!=null?$:""),f(!0),setTimeout(()=>{f(!1)},1e3))};return i?R.createElement(R.Fragment,null,R.createElement(Vr,{position:{bottom:12,right:60}},R.createElement(Gr,{spacing:"xs"},R.createElement(he,{leftIcon:R.createElement(Pr,{size:18}),variant:"filled",color:"primary",onClick:()=>o(!0)},"Show Code"),R.createElement(he,{leftIcon:R.createElement(Mr,{size:18}),variant:"default",component:"a",target:"_blank",href:"https://github.com/refinedev/refine/discussions/3046"},"Give Feedback"))),R.createElement("div",{style:{height:"50px"}}),R.createElement(Lr,{size:800,opened:u,onClose:()=>o(!1),styles:{modal:{padding:"0px !important",borderRadius:0},header:{display:"none"}}},R.createElement(Dr,{onClick:()=>o(!1),variant:"transparent",sx:{position:"absolute",right:0,top:0,height:56,width:56}},R.createElement(Er,{size:18,style:{color:"#666b7a"}})),R.createElement(ge,{code:i}),R.createElement("textarea",{ref:s,defaultValue:(i!=null?i:"").replace(/\\n/g,`\r
`),id:"code-input",style:{padding:0,margin:0,width:0,height:0,opacity:0,border:"none",display:"block"}}),R.createElement(he,{sx:{position:"absolute",bottom:12,right:12,width:115},key:"copy",variant:"default",leftIcon:c?R.createElement(Or,{size:18,style:{color:"green"}}):R.createElement(Br,{size:18}),onClick:m},c?"Copied":"Copy"))):null};var xe=K({type:"show",additionalScope:[["@pankod/refine-mantine","RefineMantine",zr]],codeViewerComponent:_,loadingComponent:U,errorComponent:H,renderer:({resource:r,fields:t,isCustomPage:i,id:s})=>{var V;let u=D((V=r.label)!=null?V:r.name,"show"),o="record",c=[["useShow","@pankod/refine-core"],["Show","@pankod/refine-mantine"],["Title","@pankod/refine-mantine"]],m=t.filter(e=>(e==null?void 0:e.relation)&&!(e!=null&&e.fieldable)&&(e==null?void 0:e.resource)).filter(Boolean).map(e=>{if((e==null?void 0:e.relation)&&!e.fieldable&&e.resource){if(e.multiple){c.push(["useMany","@pankod/refine-core"]);let d=l(o,e.key);return e.accessor&&(d=`${l(o,e.key)}?.map((item: any) => ${l("item",void 0,e.accessor)})`),`
                    const { data: ${T(e.key,"Data")}, isLoading: ${T(e.key,"IsLoading")} } =
                    useMany({
                        resource: "${e.resource.name}",
                        ids: ${d} || [],
                        queryOptions: {
                            enabled: !!${o},
                        },
                    });
                    `}return c.push(["useOne","@pankod/refine-core"]),`
                    const { data: ${T(e.key,"Data")}, isLoading: ${T(e.key,"IsLoading")} } =
                    useOne({
                        resource: "${e.resource.name}",
                        id: ${l(o,e.key,e.accessor,!1)} || "",
                        queryOptions: {
                            enabled: !!${o},
                        },
                    });
                `}}).filter(Boolean),k=e=>{if(e.relation&&e.resource){let d=T(e.key,"Data"),n=T(e.key,"IsLoading");return e.multiple?(c.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]),y`
                    <Title my="xs" order={5}>${g(e.key)}</Title>
                    {${n} ? <>Loading...</> : (
                        <>
                        ${(()=>{var h,A;if(e.relationInfer)if((h=e.relationInfer)!=null&&h.accessor){if(Array.isArray(e.relationInfer.accessor))return"Not Handled.";{let S=j((A=e.resource)==null?void 0:A.name),C=l(S,void 0,e.relationInfer.accessor);return`
                                        <Group spacing="xs">
                                            {${d}?.data?.map((${S}: any) => <TagField key={${C}} value={${C}} />)}
                                        </Group>
                                        `}}else return"Not Handled.";else return"not-handled - relation with multiple but no resource"})()}
                        </>
                    )}
                    `):y`
                    <Title my="xs" order={5}>${g(e.key)}</Title>
                    {${n} ? <>Loading...</> : (
                        <>
                        ${(()=>{var h;return e.relationInfer?(h=e.relationInfer)!=null&&h.accessor?Array.isArray(e.relationInfer.accessor)?`{${l(`${d}?.data`,void 0,e.relationInfer.accessor,' + " " + ')}}`:`{${d}?.data?.${e.relationInfer.accessor}}`:`{${d}?.data}`:`{${d}?.data?.id}`})()}
                        </>
                    )}
                    
                    `}},b=e=>{if(e.type==="text"){if(c.push(["TextField","@pankod/refine-mantine"]),e.multiple){c.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]);let d=l("item",void 0,e.accessor);return y`
                    <Title my="xs" order={5}>${g(e.key)}</Title>
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any) => (
                            <TagField value={${d}} key={${d}} />
                        ))}
                    </Group>
                `}return y`
                    <Title my="xs" order={5}>${g(e.key)}</Title>
                    <TextField value={${l(o,e.key,e.accessor)}} />
                `}},x=e=>{if(e.type==="image"){if(c.push(["Image","@pankod/refine-mantine"]),e.multiple){c.push(["Group","@pankod/refine-mantine"]);let d=l("item",void 0,e.accessor);return y`
                    <Title my="xs" order={5}>${g(e.key)}</Title>
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any) => (
                            <Image sx={{ maxWidth: 200 }} src={${d}} key={${d}} />
                        ))}
                    </Group>
                `}return y`
                    <Title my="xs" order={5}>${g(e.key)}</Title>
                    <Image sx={{ maxWidth: 200 }} src={${l(o,e.key,e.accessor," + ")}} />
                    `}},$=e=>{if(e.type==="email"){if(c.push(["EmailField","@pankod/refine-mantine"]),e.multiple){c.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]);let d=l("item",void 0,e.accessor);return y`
                    <Title my="xs" order={5}>${g(e.key)}</Title>
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any) => (
                            <TagField value={${d}} key={${d}} />
                        ))}
                    </Group>
                `}return y`
                    <Title my="xs" order={5}>${g(e.key)}</Title>
                    <EmailField value={${l(o,e.key,e.accessor," + ")}} />
                `}},F=e=>{if(e.type==="url"){if(c.push(["UrlField","@pankod/refine-mantine"]),e.multiple){c.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]);let d=l("item",void 0,e.accessor);return y`
                    <Title my="xs" order={5}>${g(e.key)}</Title>
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any) => (
                            <TagField value={${d}} key={${d}} />
                        ))}
                    </Group>
                `}return y`
                    <Title my="xs" order={5}>${g(e.key)}</Title>
                    <UrlField value={${l(o,e.key,e.accessor," + ")}} />
                `}},N=e=>{if(e.type==="boolean"){if(c.push(["BooleanField","@pankod/refine-mantine"]),e.multiple){c.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]);let d=l("item",void 0,e.accessor);return y`
                    <Title my="xs" order={5}>${g(e.key)}</Title>
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any, index: number) => (
                            <TagField value={${d}} key={index} />
                        ))}
                    </Group>
                `}return y`
                    <Title my="xs" order={5}>${g(e.key)}</Title>
                    <BooleanField value={${l(o,e.key,e.accessor," && ")}} />
                `}},p=e=>{if(e.type==="date"){if(c.push(["DateField","@pankod/refine-mantine"]),e.multiple){c.push(["Group","@pankod/refine-mantine"]);let d=l("item",void 0,e.accessor);return y`
                    <Title my="xs" order={5}>${g(e.key)}</Title>
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any) => (
                            <DateField value={${d}} key={${d}} />
                        ))}
                    </Group>
                `}return y`
                    <Title my="xs" order={5}>${g(e.key)}</Title>
                    <DateField value={${l(o,e.key,e.accessor," + ' ' + ")}} />
                `}},L=e=>{if(e.type==="richtext")return c.push(["MarkdownField","@pankod/refine-mantine"]),y`
                    <Title mt="xs" order={5}>${g(e.key)}</Title>
                    <MarkdownField value={${l(o,e.key,e.accessor,' + " " + ')}} />
                `},v=e=>{if(e.type==="number"){if(c.push(["NumberField","@pankod/refine-mantine"]),e.multiple){c.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]);let d=l("item",void 0,e.accessor);return y`
                    <Title my="xs" order={5}>${g(e.key)}</Title>
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any) => (
                            <TagField value={${d}} key={${d}} />
                        ))}
                    </Group>
                `}return y`
                    <Title my="xs" order={5}>${g(e.key)}</Title>
                    <NumberField value={${l(o,e.key,e.accessor,' + " " + ')} ?? ""} />
                `}},a=e=>{if(e)return y`
                    ${e}
                `},I=t.map(e=>{switch(e==null?void 0:e.type){case"text":return a(b(e));case"number":return a(v(e));case"richtext":return a(L(e));case"email":return a($(e));case"image":return a(x(e));case"date":return a(p(e));case"boolean":return a(N(e));case"url":return a(F(e));case"relation":return a(k(e));default:return}});return z(c),y`
        ${O(c)}
        
        export const ${u} = () => {
            const { queryResult } = useShow(${i?`{ 
                        resource: "${r.name}", 
                        id: ${s}
                    }`:""});
            const { data, isLoading } = queryResult;
        
            const ${o} = data?.data;
        
            ${m}

            return (
                <Show isLoading={isLoading}>
                    ${I.join("")}
                </Show>
            );
        };
        `}});import*as jr from"@pankod/refine-mantine";import*as Kr from"@pankod/refine-react-table";var W=r=>Array.isArray(r.accessor)||r.multiple?`accessorKey: "${r.key}"`:r.accessor?`accessorKey: "${w(r.key,void 0,r.accessor)}"`:`accessorKey: "${r.key}"`,be=K({type:"list",additionalScope:[["@pankod/refine-mantine","RefineMantine",jr],["@pankod/refine-react-table","RefineReactTable",Kr]],codeViewerComponent:_,loadingComponent:U,errorComponent:H,renderer:({resource:r,fields:t,isCustomPage:i})=>{var d;let s=D((d=r.label)!=null?d:r.name,"list"),u="tableData?.data",o=[["IResourceComponentsProps","@pankod/refine-core"],["useTable","@pankod/refine-react-table"],["ColumnDef","@pankod/refine-react-table"],["flexRender","@pankod/refine-react-table"],["ScrollArea","@pankod/refine-mantine"],["List","@pankod/refine-mantine"],["Table","@pankod/refine-mantine"],["Pagination","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"],["EditButton","@pankod/refine-mantine"],["ShowButton","@pankod/refine-mantine"],["DeleteButton","@pankod/refine-mantine"]],c=t.filter(n=>(n==null?void 0:n.relation)&&!(n!=null&&n.fieldable)&&(n==null?void 0:n.resource)),f=c.filter(Boolean).map(n=>{if((n==null?void 0:n.relation)&&!n.fieldable&&n.resource){o.push(["GetManyResponse","@pankod/refine-core"]),o.push(["useMany","@pankod/refine-core"]);let h="";return n.multiple?h=`[].concat(...(${u}?.map((item) => ${l("item",n.key,n.accessor,!1)}) ?? []))`:h=`${u}?.map((item) => ${l("item",n.key,n.accessor,!1)}) ?? []`,`
                    const { data: ${T(n.key,"Data")} } =
                    useMany({
                        resource: "${n.resource.name}",
                        ids: ${h},
                        queryOptions: {
                            enabled: !!${u},
                        },
                    });
                    `}}).filter(Boolean),m=c==null?void 0:c.map(n=>{if(n&&n.resource)return T(n.key,"Data")}).filter(Boolean),k=n=>{var h;if(n.relation&&n.resource){let A=`${T(n.key,"Data")}?.data`;if(Array.isArray(n.accessor))return;let S=`id: "${n.key}"`,C=`header: "${g(n.key)}"`,M=W(n),B="";if(n.multiple){o.push(["TagField","@pankod/refine-mantine"]);let q="item";n!=null&&n.relationInfer&&(q=l("item",void 0,n.relationInfer.accessor)),B=`cell: function render({ getValue, table }) {
                        const meta = table.options.meta as {
                            ${T(n.key,"Data")}: GetManyResponse;
                        };

                        const ${T(n.key)} = getValue<any[]>()?.map((item) => {
                            return meta.${T(n.key,"Data")}?.data?.find(
                                (resourceItems) => resourceItems.id === ${l("item",void 0,n.accessor)}
                            );
                        })

                        return (
                            <Group spacing="xs">
                                {${T(n.key)}?.map((item, index) => (
                                    <TagField key={index} value={${q}} />
                                ))}
                            </Group>
                        )
                    }
                `}else n!=null&&n.relationInfer?B=`cell: function render({ getValue, table }) {
                            const meta = table.options.meta as {
                                ${T(n.key,"Data")}: GetManyResponse;
                            };

                            const ${T(n.key)} = meta.${A}?.find(
                                (item) => item.id === getValue<any>(),
                            );

                            return ${l(T(n.key),void 0,(h=n==null?void 0:n.relationInfer)==null?void 0:h.accessor)} ?? "Loading...";
                        },`:B="";return`
                    {
                        ${S},
                        ${C},
                        ${M},
                        ${B}
                    }
                `}},b=n=>{if(n.type==="image"){o.push(["Image","@pankod/refine-mantine"]);let h=`id: "${n.key}"`,A=W(n),S=`header: "${g(n.key)}"`,C=y`
                    cell: function render({ getValue }) {
                        ${n!=null&&n.accessor?"try {":""}
                            return <Image sx={{ maxWidth: "100px" }} src={${l("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0," + ")}} />
                        ${n!=null&&n.accessor?" } catch (error) { return null; }":""}
                       
                    }
                `;if(n.multiple){let M=l("item",void 0,n.accessor," + ");C=`
                        cell: function render({ getValue }) {
                            ${n!=null&&n.accessor?"try {":""}
                                return (
                                    <Group spacing="xs">
                                        {getValue<any[]>()?.map((item, index) => (
                                            <Image src={${M}} key={index} sx={{ maxWidth: "100px" }} />
                                        ))}
                                    </Group>
                                )
                            ${n!=null&&n.accessor?" } catch (error) { return null; }":""}
                        }
                    `}return`
                    {
                        ${h},
                        ${A},
                        ${S},
                        ${C}
                    }
                `}},x=n=>{if(n.type==="email"){o.push(["EmailField","@pankod/refine-mantine"]);let h=`id: "${n.key}"`,A=W(n),S=`header: "${g(n.key)}"`,C=y`
                    cell: function render({ getValue }) {
                        return <EmailField value={${l("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0,' + " " + ')}} />
                    }
                `;return n.multiple&&(o.push(["TagField","@pankod/refine-mantine"]),C=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any>()?.map((item, index) => (
                                        <TagField value={${l("item",void 0,n.accessor," + ")}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${A},
                        ${S},
                        ${C}
                    }
                `}},$=n=>{if(n.type==="url"){o.push(["UrlField","@pankod/refine-mantine"]);let h=`id: "${n.key}"`,A=W(n),S=`header: "${g(n.key)}"`,C=y`
                    cell: function render({ getValue }) {
                        return <UrlField value={${l("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0," + ")}} />
                    }
                `;return n.multiple&&(o.push(["TagField","@pankod/refine-mantine"]),C=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${l("item",void 0,n.accessor," + ")}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${A},
                        ${S},
                        ${C}
                    }
                `}},F=n=>{if((n==null?void 0:n.type)==="boolean"){o.push(["BooleanField","@pankod/refine-mantine"]);let h=`id: "${n.key}"`,A=W(n),S=`header: "${g(n.key)}"`,C=y`
                    cell: function render({ getValue }) {
                        return <BooleanField value={${l("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0," + ")}} />
                    }
                `;return n.multiple&&(C=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any[]>()?.map((item, index) => (
                                        <BooleanField value={${l("item",void 0,n.accessor," + ")}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${A},
                        ${S},
                        ${C}
                    }
                `}},N=n=>{if(n.type==="date"){o.push(["DateField","@pankod/refine-mantine"]);let h=`id: "${n.key}"`,A=W(n),S=`header: "${g(n.key)}"`,C=y`
                    cell: function render({ getValue }) {
                        return <DateField value={${l("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0,' + " " + ')}} />
                    }
                `;return n.multiple&&(C=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any[]>()?.map((item, index) => (
                                        <DateField value={${l("item",void 0,n.accessor," + ")}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${A},
                        ${S},
                        ${C}
                    }
                `}},p=n=>{if((n==null?void 0:n.type)==="richtext"){o.push(["MarkdownField","@pankod/refine-mantine"]);let h=`id: "${n.key}"`,A=W(n),S=`header: "${g(n.key)}"`,C=y`
                    cell: function render({ getValue }) {
                        return <MarkdownField value={(${l("getValue<string>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0)})?.slice(0, 80) + "..." } />
                    }
                `;return n.multiple&&(C=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<string[]>()?.map((item, index) => (
                                        <MarkdownField value={${l("item",void 0,n.accessor," + ")}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${A},
                        ${S},
                        ${C}
                    }
                `}},L=n=>{if(n&&(n.type==="text"||n.type==="number")){let h=`id: "${n.key}"`,A=W(n),S=`header: "${g(n.key)}"`,C="";return n.multiple&&(o.push(["TagField","@pankod/refine-mantine"]),C=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${l("item",void 0,n.accessor,' + " " + ')}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),!n.multiple&&Array.isArray(n.accessor)&&(C=`
                        cell: function render({ getValue }) {
                            return (
                                <>{${l("getValue()",n.key,n.accessor)}}</>
                            );
                        }
                    `),`
                    {
                        ${h},
                        ${A},
                        ${S},
                        ${C}
                    }
                `}},{canEdit:v,canShow:a,canDelete:I}=r!=null?r:{};v&&o.push(["EditButton","@pankod/refine-mantine"]),a&&o.push(["ShowButton","@pankod/refine-mantine"]),I&&o.push(["DeleteButton","@pankod/refine-mantine"]);let V=v||a||I?y`
        {
            id: "actions",
            accessorKey: "id",
            header: "Actions",
            cell: function render({ getValue }) {
                return (
                    <Group spacing="xs" noWrap>
                        ${a?y`
                        <ShowButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `:""}
                        ${v?y`<EditButton
                            hideText
                            recordItemId={getValue() as string}
                        />`:""}
                        ${I?y`
                        <DeleteButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `:""}
                    </Group>
                );
            },
        },
            `:"",e=t.map(n=>{switch(n==null?void 0:n.type){case"text":case"number":return L(n);case"richtext":return p(n);case"email":return x(n);case"image":return b(n);case"date":return N(n);case"boolean":return F(n);case"url":return $(n);case"relation":return k(n);default:return}});return z(o),y`
        import React from "react";
        ${O(o)}
        
        export const ${s}: React.FC<IResourceComponentsProps> = () => {
            const columns = React.useMemo<ColumnDef<any>[]>(() => [
                ${[...e,V].filter(Boolean).join(",")}
            ], []);

            const {
                getHeaderGroups,
                getRowModel,
                setOptions,
                refineCore: {
                    setCurrent,
                    pageCount,
                    current,
                    tableQueryResult: { data: tableData },
                },
            } = useTable({
                columns,
                ${i?`
                refineCoreProps: {
                    resource: "${r.name}",
                }
                `:""}
            });

            ${f}

            setOptions((prev) => ({
                ...prev,
                meta: {
                    ...prev.meta,
                    ${m.join(", ")}
                },
            }));

            return (
                <List>
                    <ScrollArea>
                        <Table highlightOnHover>
                            <thead>
                                {getHeaderGroups().map((headerGroup) => (
                                    <tr key={headerGroup.id}>
                                        {headerGroup.headers.map((header) => {
                                            return (
                                                <th key={header.id}>
                                                    {!header.isPlaceholder && (
                                                        flexRender(
                                                            header.column
                                                                .columnDef
                                                                .header,
                                                            header.getContext(),
                                                        )
                                                    )}
                                                </th>
                                            );
                                        })}
                                    </tr>
                                ))}
                            </thead>
                            <tbody>
                                {getRowModel().rows.map((row) => {
                                    return (
                                        <tr key={row.id}>
                                            {row.getVisibleCells().map((cell) => {
                                                return (
                                                    <td key={cell.id}>
                                                        {flexRender(
                                                            cell.column.columnDef.cell,
                                                            cell.getContext(),
                                                        )}
                                                    </td>
                                                );
                                            })}
                                        </tr>
                                    );
                                })}
                            </tbody>
                        </Table>
                    </ScrollArea>    
                    <br />
                    <Pagination
                        position="right"
                        total={pageCount}
                        page={current}
                        onChange={setCurrent}
                    />
                </List>
            );
        };
        `}});import*as Hr from"@pankod/refine-mantine";var Fe=K({type:"create",additionalScope:[["@pankod/refine-mantine","RefineMantine",Hr]],codeViewerComponent:_,loadingComponent:U,errorComponent:H,renderer:({resource:r,fields:t,isCustomPage:i})=>{var v;let s=D((v=r.label)!=null?v:r.name,"create"),u=[["Create","@pankod/refine-mantine"],["useForm","@pankod/refine-mantine"]],o={},f=t.filter(a=>(a==null?void 0:a.relation)&&!(a!=null&&a.fieldable)&&(a==null?void 0:a.resource)).filter(Boolean).map(a=>{if((a==null?void 0:a.relation)&&!a.fieldable&&a.resource)return u.push(["useSelect","@pankod/refine-mantine"]),`
                    const { selectProps: ${T(a.key,"SelectProps")} } =
                    useSelect({
                        resource: "${a.resource.name}",
                        ${Q(a)}
                    });
                `}).filter(Boolean),m=a=>{if(a.relation&&a.resource){o={...o,[a.key]:a.multiple?[]:a.accessor?{[typeof a.accessor=="string"?a.accessor:a.accessor[0]]:""}:""};let I=T(a.key,"SelectProps");return a.multiple?(u.push(["MultiSelect","@pankod/refine-mantine"]),y`
                        <MultiSelect mt="sm" label="${g(a.key)}" {...getInputProps("${w(a.key,void 0)}")} {...${I}} filterDataOnExactSearchMatch={undefined} />
                    `):(u.push(["Select","@pankod/refine-mantine"]),y`
                    <Select mt="sm" label="${g(a.key)}" {...getInputProps("${w(a.key,void 0,a.accessor)}")} {...${I}} />
                `)}},k=a=>{if(a.type==="text"||a.type==="email"||a.type==="date"||a.type==="url")return Z(a.key)||(u.push(["TextInput","@pankod/refine-mantine"]),o={...o,[a.key]:a.multiple?[]:""},a.multiple)?void 0:y`
                    <TextInput mt="sm" label="${g(a.key)}" {...getInputProps("${w(a.key,void 0,a.accessor)}")} />
                `},b=a=>{if(a.type==="image")return y`
                {/* 
                    Dropzone component is not included in "@pankod/refine-mantine" package.
                    To use a <Dropzone> component, you can follow the official documentation for Mantine.
                    
                    Docs: https://mantine.dev/others/dropzone/
                */}
                `},x=a=>{if(a.type==="boolean")return u.push(["Checkbox","@pankod/refine-mantine"]),o={...o,[a.key]:a.multiple?[]:""},a.multiple?void 0:y`
                    <Checkbox mt="sm" label="${g(a.key)}" {...getInputProps("${w(a.key,void 0,a.accessor)}", { type: 'checkbox' })} />
                `},$=a=>{if(a.type==="date"){let I=k(a);return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-mantine" package.
                        To use a <DatePicker> component, you can follow the official documentation for Mantine.
                        
                        Docs: https://mantine.dev/dates/date-picker/
                    */}
                    ${I!=null?I:""}
                `}},F=a=>{if(a.type==="richtext")return u.push(["Textarea","@pankod/refine-mantine"]),o={...o,[a.key]:a.multiple?[]:""},a.multiple?void 0:y`
                    <Textarea mt="sm" label="${g(a.key)}" autosize {...getInputProps("${w(a.key,void 0,a.accessor)}")} />
                `},N=a=>{if(a.type==="number")return Z(a.key)||(u.push(["NumberInput","@pankod/refine-mantine"]),o={...o,[a.key]:a.multiple?[]:""},a.multiple)?void 0:y`
                    <NumberInput mt="sm" label="${g(a.key)}" {...getInputProps("${w(a.key,void 0,a.accessor)}")}/>
                `},p=a=>{if(a)return y`
                    ${a}
            `},L=t.map(a=>{switch(a==null?void 0:a.type){case"url":case"text":case"email":return p(k(a));case"number":return p(N(a));case"richtext":return p(F(a));case"image":return p(b(a));case"date":return p($(a));case"boolean":return p(x(a));case"relation":return p(m(a));default:return}});return z(u),y`
        ${O(u)}
        
        export const ${s} = () => {
            const { getInputProps, saveButtonProps, setFieldValue, refineCore: { formLoading } } = useForm({
                initialValues: ${JSON.stringify(o)},
                ${i?`refineCoreProps: {
                            resource: "${r.name}",
                            action: "create",  
                        }`:""}
            });
        
            ${f}

            return (
                <Create isLoading={formLoading} saveButtonProps={saveButtonProps}>
                    ${L.join("")}
                </Create>
            );
        };
        `}});import*as Ur from"@pankod/refine-mantine";var Ce=K({type:"edit",additionalScope:[["@pankod/refine-mantine","RefineMantine",Ur]],codeViewerComponent:_,loadingComponent:U,errorComponent:H,renderer:({resource:r,fields:t,isCustomPage:i,id:s})=>{var I,V;let u=D((I=r.label)!=null?I:r.name,"edit"),o=T((V=r.label)!=null?V:r.name,"Data"),c=[["Edit","@pankod/refine-mantine"],["useForm","@pankod/refine-mantine"]],f={},k=t.filter(e=>(e==null?void 0:e.relation)&&!(e!=null&&e.fieldable)&&(e==null?void 0:e.resource)).filter(Boolean).map(e=>{if((e==null?void 0:e.relation)&&!e.fieldable&&e.resource){c.push(["useSelect","@pankod/refine-mantine"]);let d=l(o,e.key,e.accessor,!1);e.multiple&&e.accessor&&(d=`${l(o,e.key)}?.map((item: any) => ${l("item",void 0,e.accessor)})`);let n="";return e.multiple&&e.accessor&&(n=`React.useEffect(() => {
                            setFieldValue("${e.key}", ${d});
                        }, [${o}]);`),`
                    const { selectProps: ${T(e.key,"SelectProps")} } =
                    useSelect({
                        resource: "${e.resource.name}",
                        defaultValue: ${d},
                        ${Q(e)}
                    });

                    ${n}
                `}}).filter(Boolean),b=e=>{if(e.relation&&e.resource){f={...f,[e.key]:e.multiple?[]:e.accessor?{[typeof e.accessor=="string"?e.accessor:e.accessor[0]]:""}:""};let d=T(e.key,"SelectProps");return e.multiple?(c.push(["MultiSelect","@pankod/refine-mantine"]),y`
                        <MultiSelect mt="sm" label="${g(e.key)}" {...getInputProps("${w(e.key,void 0)}")} {...${d}} filterDataOnExactSearchMatch={undefined} />
                    `):(c.push(["Select","@pankod/refine-mantine"]),y`
                    <Select mt="sm" label="${g(e.key)}" {...getInputProps("${w(e.key,void 0,e.accessor)}")} {...${d}} />
                `)}},x=e=>{if(e.type==="text"||e.type==="email"||e.type==="date"||e.type==="url"){if(c.push(["TextInput","@pankod/refine-mantine"]),f={...f,[e.key]:e.multiple?[]:""},e.multiple){c.push(["Group","@pankod/refine-mantine"]);let d=w(e.key,"${index}",e.accessor);return`
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any, index: number) => (
                            <TextInput mt="sm" key={index} label={\`${g(e.key)} \${index + 1} \`} {...getInputProps(\`${d}\`)} />
                        ))}
                    </Group>
                    `}return y`
                    <TextInput mt="sm" ${Z(e.key)?"disabled":""} label="${g(e.key)}" {...getInputProps("${w(e.key,void 0,e.accessor)}")} />
                `}},$=e=>{if(e.type==="image")return y`
                {/* 
                    Dropzone component is not included in "@pankod/refine-mantine" package.
                    To use a <Dropzone> component, you can follow the official documentation for Mantine.
                    
                    Docs: https://mantine.dev/others/dropzone/
                */}
                `},F=e=>{if(e.type==="boolean"){if(c.push(["Checkbox","@pankod/refine-mantine"]),f={...f,[e.key]:e.multiple?[]:""},e.multiple){c.push(["Group","@pankod/refine-mantine"]);let d=w(e.key,"${index}",e.accessor);return`
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any, index: number) => (
                            <Checkbox mt="sm" key={index} label={\`${g(e.key)} \${index + 1} \`} {...getInputProps(\`${d}\`, { type: 'checkbox' })} />
                        ))}
                    </Group>
                    `}return y`
                    <Checkbox mt="sm" label="${g(e.key)}" {...getInputProps("${w(e.key,void 0,e.accessor)}", { type: 'checkbox' })} />
                `}},N=e=>{if(e.type==="date")return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-mantine" package.
                        To use a <DatePicker> component, you can follow the official documentation for Mantine.
                        
                        Docs: https://mantine.dev/dates/date-picker/
                    */}
                    ${x(e)}
                `},p=e=>{if(e.type==="richtext"){if(c.push(["Textarea","@pankod/refine-mantine"]),f={...f,[e.key]:e.multiple?[]:""},e.multiple){c.push(["Group","@pankod/refine-mantine"]);let d=w(e.key,"${index}",e.accessor);return`
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any, index: number) => (
                            <Textarea mt="sm" key={index} label={\`${g(e.key)} \${index + 1} \`} {...getInputProps(\`${d}\`)} />
                        ))}
                    </Group>
                    `}return y`
                    <Textarea mt="sm" label="${g(e.key)}" autosize {...getInputProps("${w(e.key,void 0,e.accessor)}")} />
                `}},L=e=>{if(e.type==="number"){if(c.push(["NumberInput","@pankod/refine-mantine"]),f={...f,[e.key]:e.multiple?[]:""},e.multiple){c.push(["Group","@pankod/refine-mantine"]);let d=w(e.key,"${index}",e.accessor);return`
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any, index: number) => (
                            <NumberInput mt="sm" key={index} label={\`${g(e.key)} \${index + 1} \`} {...getInputProps(\`${d}\`)} />
                        ))}
                    </Group>
                    `}return y`
                    <NumberInput mt="sm" ${Z(e.key)?"disabled":""} label="${g(e.key)}" {...getInputProps("${w(e.key,void 0,e.accessor)}")}/>
                `}},v=e=>{if(e)return y`
                    ${e}
            `},a=t.map(e=>{switch(e==null?void 0:e.type){case"url":case"text":case"email":return v(x(e));case"number":return v(L(e));case"richtext":return v(p(e));case"image":return v($(e));case"date":return v(N(e));case"boolean":return v(F(e));case"relation":return v(b(e));default:return}});return z(c),y`
        ${O(c)}
        
        export const ${u} = () => {
            const { getInputProps, saveButtonProps, setFieldValue, refineCore: { queryResult } } = useForm({
                initialValues: ${JSON.stringify(f)},
                ${i?`refineCoreProps: {
                            resource: "${r.name}",
                            id: ${s},
                            action: "edit",  
                        }`:""}
            });
        
            const ${o} = queryResult?.data?.data;
        
            ${k}

            return (
                <Edit saveButtonProps={saveButtonProps}>
                    ${a.join("")}
                </Edit>
            );
        };
        `}});var ia=({action:r,id:t,...i})=>{let{useParams:s}=_r(),{action:u,id:o}=s();switch(r!=null?r:u){case"show":return te.createElement(xe,{...i,id:t!=null?t:o});case"create":return te.createElement(Fe,{...i,id:t!=null?t:o});case"edit":return te.createElement(Ce,{...i,id:t!=null?t:o});default:return te.createElement(be,{...i,id:t!=null?t:o})}};export{Fe as MantineCreateInferencer,Ce as MantineEditInferencer,ia as MantineInferencer,be as MantineListInferencer,xe as MantineShowInferencer};
