// @pankod/refine-inferencer@1.6.1/dist/esm/mui.js
import ne from"react";import{useRouterContext as Zr}from"@pankod/refine-core";import*as zr from"@pankod/refine-mui";import M from"react";import{useResource as nr}from"@pankod/refine-core";var ar=/(image|photo|avatar|url|logo|cover|thumbnail|icon)/i,ae=(e,t)=>{let i=["name","label","title","count","content","username","nickname","login","firstName","lastName","url"];ar.test(e)&&i.unshift("url");let o=i.find(u=>u in t&&!!t[u]);if(o)return o==="firstName"&&"lastName"in t?["firstName","lastName"]:o};function se(e,t,i){var u;if(t)return t;let o=i==null?void 0:i.find(s=>s.name===e);return(u=o==null?void 0:o.options)!=null&&u.dataProviderName?o.options.dataProviderName:"default"}var X=e=>{var t;return(t=e==null?void 0:e.options)==null?void 0:t.dataProviderName};var ie=e=>{let t=[...e].filter(Boolean).sort((i,o)=>(o.priority||0)-(i.priority||0))[0];return t||null};import sr from"prettier/standalone";import ir from"prettier/parser-typescript";var ce=e=>{try{return sr.format(e,{parser:"typescript",plugins:[ir],arrowParens:"always",trailingComma:"all",semi:!0,tabWidth:4,printWidth:80})}catch(t){return console.warn("Formatting error in inferencer",t),e}};var ue=e=>e.replace(/ {2,}/g," ").replace(/\n{2,}/g,`
`);var pe=/(-id|-ids|_id|_ids|Id|Ids|ID|IDs)(\[\])?$/,we=(e,t)=>{let i=pe.test(e),o=typeof t=="string"||typeof t=="number",u=Array.isArray(t)&&t.every(s=>typeof s=="string"||typeof s=="number");return i&&o||i&&u?{key:e,relation:!0,multiple:!!u,type:"relation",priority:1}:!1};var B=e=>e.replace(pe,"");var y=e=>B(e).replace(/([a-z])([A-Z])/g,"$1 $2").replace(/_/g," ").replace(/-/g," ").split(" ").map(p=>p.charAt(0).toUpperCase()+p.slice(1)).join(" ");var me=e=>e.replace(/export\s+(const|let|var|type|interface|function|class)\s+(\w+)\s*(=|:)\s*/g,"$1 $2 = ").split(`
`).filter(u=>!u.trim().startsWith("export default")).join(`
`);var Se=/import(?:(?:(?:[ \n\t]+([^ *\n\t\{\},]+)[ \n\t]*(?:,|[ \n\t]+))?([ \n\t]*\{(?:[ \n\t]*[^ \n\t"'\{\}]+[ \n\t]*,?)+\})?[ \n\t]*)|[ \n\t]*\*[ \n\t]*as[ \n\t]+([^ \n\t\{\}]+)[ \n\t]+)from[ \n\t]*(?:['"])([^'"\n]+)(?:['"])(?:;?)/g,cr=/import[ \n\t](?:['"])([^'"\n]+)(?:['"])(?:;?)/g,ur=/((?:\w|\s|_)*)( as )((?:\w|\s|_)*)( |,)?/g,le=(e,t={})=>{let i=e.matchAll(Se),o=new Set;for(let u of i){let[,s,p,g,d]=u;if(d in t){let h=t[d];if(d==="react")continue;s&&o.add(`const { default: ${s} } = ${h};`),p&&o.add(`const${p.replace(ur,"$1: $3$4")} = ${h};`),g&&o.add(`const ${g} = ${h};`)}}return ue(`
    ${Array.from(o).join(`
`)}
    ${e.replace(Se,"").replace(cr,"")}
    `)};import te from"pluralize";var de=(e,t)=>e?t.find(o=>{let u=B(e.key);return o.name===u||o.name===(te.isPlural(u)?u:te.plural(u))||o.name===(te.isSingular(u)?u:te.singular(u))}):void 0;import pr from"pluralize";var V=(e,t)=>{let i=e.replace(/[^a-zA-Z0-9]/g," "),o=pr.singular(i),u=y(o).replace(/ /g,""),s=t.charAt(0).toUpperCase()+t.slice(1);return/^[a-zA-Z]/.test(e)?`${u}${s}`:`${s}${u}`};var mr=e=>e?Array.isArray(e)?e.join(""):typeof e=="string"?e:typeof e=="object"||typeof e=="function"?e.toString():typeof e=="number"?String(e):typeof e=="boolean"&&e?"true":e:"",$=(e,...t)=>e.reduce((o,u,s)=>`${o}${u}${mr(t[s])}`,"").trim();var ye=(e,t)=>`
    ${e!=null?e:""}
    
    render(typeof ${t} !== "undefined" ? <${t} /> : <></>);
    `;var fe=e=>e==null?void 0:e.replace(/(\/\* hidden-start \*\/)(.|\n)*?(\/\* hidden-end \*\/)/g,"");var lr=/^[a-zA-Z_$][a-zA-Z_$0-9]*$/,Le=e=>lr.test(e);function ge(e,t,i){let o=`${e}`;return t&&(o+="?.",Le(t)?o+=t:o+=`['${t}']`),i&&(o+="?.",Le(i)?o+=i:o+=`['${i}']`),o}function dr(e,t,i){return(i!=null?i:[]).map(u=>ge(e,t,u))}var m=(e,t,i,o=' + " " + ')=>Array.isArray(i)?o?dr(e,t,i).join(o):ge(e,t,i[0]):ge(e,t,i),R=(e,t,i)=>{let o=e;return t&&(o+=`.${t}`),i&&(Array.isArray(i)?o+=`.${i[0]}`:o+=`.${i}`),o};var W=e=>{let t=e.reduce((o,[u,s,p])=>(o[s]||(o[s]=[]),o[s].includes(u)||(p?o[s].push([u,!0]):o[s].push(u)),o),{});return Object.entries(t).map(([o,u])=>{let s=u.find(h=>Array.isArray(h)&&h[1]),p=u.filter(h=>typeof h=="string"),g=s?`${s[0]}`:"",d=p.length?`{ ${p.join(", ")} }`:"";return`import ${g}${g&&d?", ":""}${d} from "${o}";`})};import Ne from"pluralize";var K=(e="")=>Ne.isSingular(e)?e:Ne.singular(e);import Re from"pluralize";var Z=(e="")=>Re.isPlural(e)?e:Re.plural(e);var yr=/^id$/i,q=e=>yr.test(e);var O=e=>{};var L=(e,t="")=>{if(Number.isInteger(Number(e)))return t.charAt(0).toLowerCase()+t.slice(1)+e;let o=e.replace(/^\d+/,"").replace(/[^a-zA-Z0-9]/g,"-"),u=y(o).replace(/ /g,"");return`${u.charAt(0).toLowerCase()+u.slice(1)}${t.charAt(0).toUpperCase()+t.slice(1)}`};var Pe=e=>{let t=(i,o,u,s=t)=>{let p=e.map(d=>d(i,o,u,s));return ie(p)};return t};var Be=e=>(i,o,u,s,p)=>e.reduce((g,d)=>d(g,o,u,s,p),i);var Ee=(e,t,i,o)=>{let u=Array.isArray(t),s=Array.isArray(t)&&t.every(p=>typeof p=="string"||typeof p=="number");if(u){if(!s){let g=o(e,t[0],i,o);return g?{...g,key:e,multiple:!0,priority:1}:!1}let p=o(e,t[0],i,o)||{type:"string"};return{key:e,multiple:!0,priority:1,type:p==null?void 0:p.type}}return!1};var Ve=(e,t)=>typeof t=="boolean"?{key:e,type:"boolean"}:!1;import De from"dayjs";var fr=/(_at|_on|At|On|AT|ON)(\[\])?$/,Me=(e,t)=>{let i=fr.test(e)&&De(t).isValid(),o=typeof t=="string"&&De(t).isValid();return i||o?{key:e,type:"date",priority:1}:!1};var gr=/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,We=(e,t)=>typeof t=="string"&&gr.test(t)?{key:e,type:"email",priority:1}:!1;var hr=/\.(gif|jpe?g|tiff?|png|webp|bmp|svg)$/i,Oe=(e,t)=>typeof t=="string"&&hr.test(t)?{key:e,type:"image",priority:2}:!1;var je=(e,t)=>typeof t>"u"||t===null?null:!1;var Ge=(e,t)=>typeof t=="number"?{key:e,type:"number"}:!1;var kr=/id$/i,ze=(e,t,i,o)=>{let u=t!==null,s=!Array.isArray(t);if(u&&s&&typeof t=="object"){if(Object.keys(t).length===1&&kr.test(Object.keys(t)[0]))return{key:e,type:"relation",relation:!0,accessor:"id",priority:1};let d=ae(e,t);if((d?d.length>0:!1)&&d){let x=d&&Array.isArray(d)?d[0]:d,b=o(x,t[x],t,o);if(b){let k=Array.isArray(d)?d.map(c=>b.accessor?`${c}.${Array.isArray(b.accessor)?b.accessor[0]:b.accessor}`:c):b.accessor?Array.isArray(b.accessor)?`${d}.${b.accessor[0]}`:`${d}.${b.accessor}`:d;return{...b,fieldable:!0,key:e,accessor:k,priority:1}}}return{key:e,fieldable:!1,type:"object",priority:1}}return!1};var He=(e,t)=>typeof t=="string"&&t.length>100?{key:e,type:"richtext",priority:1}:!1;var qe=(e,t)=>typeof t=="string"?{key:e,type:"text"}:!1;var $r=/^(https?|ftp):\/\/(-\.)?([^\s/?\.#-]+\.?)+(\/[^\s]*)?$/i,Ue=(e,t)=>typeof t=="string"&&$r.test(t)?{key:e,type:"url",priority:1}:!1;var _e=[Ee,Ve,Me,We,Oe,je,Ge,ze,we,He,qe,Ue];var Ke=(e,t,i,o)=>e.map(s=>{if(!s.relation&&(s.type==="text"||s.type==="richtext"||s.type==="number")&&!s.canRelation){let p=/^[a-z0-9-]+$/,g=k=>p.test(`${k}`),d=s.key.toLowerCase()!=="id";return!(!s.accessor||typeof s.accessor=="string")||!d?s:(s.multiple?o[s.key]:[o[s.key]]).every(k=>g(s.accessor?k[s.accessor]:k))?{...s,canRelation:!0}:s}return s});var Ir=/(image|photo|avatar|cover|thumbnail|icon)/i,Ze=e=>e.map(i=>i.type==="url"&&Ir.test(i.key)?{...i,type:"image"}:i);var Xe=(e,t)=>e.map(o=>{let u=de(o,t);return u?{...o,relation:!0,type:"relation",resource:u}:o});var Je=(e,t,i,o,u)=>e.map(p=>{if(p.relation&&p.type==="relation"&&!p.resource){let g=p.accessor?o[p.key][p.accessor]:o[p.key],d=u(p.key,g,o,u);if(d&&d.type!=="relation")return{...p,fieldable:!0,relation:!1,type:d.type,canRelation:!0}}return p});var Qe=[Ze,Xe,Je,Ke];import J,{useEffect as xr}from"react";import br,{defaultProps as vr}from"prism-react-renderer";import Cr from"prism-react-renderer/themes/vsDark";var he=({code:e,wrapperProps:t})=>(xr(()=>{var o;let i=document.createElement("style");i.appendChild(document.createTextNode(`
                #code-highlight::-webkit-scrollbar {-webkit-appearance: none; height: 7px; width: 7px;}
                #code-highlight::-webkit-scrollbar-thumb {background-color: #666b7a;}
                #code-highlight::-webkit-scrollbar-corner {background-color: rgb(30, 30, 30);}
                `)),(o=document.getElementById("code-highlight"))==null||o.appendChild(i)},[]),J.createElement("div",{id:"code-highlight",style:{maxHeight:"75vh",height:"100%",overflow:"auto",backgroundColor:"rgb(30, 30, 30)"},...t!=null?t:{}},J.createElement(br,{...vr,theme:Cr,code:e,language:"tsx"},({className:i,style:o,tokens:u,getLineProps:s,getTokenProps:p})=>J.createElement("pre",{className:i,style:{...o,padding:"14px 14px 14px 14px",margin:"0",width:"100%",boxSizing:"border-box"}},u.map((g,d)=>J.createElement("div",{key:d,...s({line:g,key:d})},g.map((h,x)=>J.createElement("span",{key:x,...p({token:h,key:x})}))))))));import D from"react";import*as Fr from"@pankod/refine-core";import{LivePreview as Tr,LiveProvider as Ar}from"@aliemir/react-live";var Ye=[["react","React",D],["@pankod/refine-core","RefineCore",Fr]],ke=D.createContext({}),$e=({fetchError:e,code:t,additionalScope:i,errorComponent:o})=>{var g;let u=(g=t&&me(le(t,[...Ye,...i!=null?i:[]].reduce((d,[h,x,b,k])=>k?d:{...d,[h]:x},{}))))!=null?g:"",s=D.useMemo(()=>({React:D,...D,...[...Ye,...i!=null?i:[]].reduce((d,[h,x,b])=>({...d,[x]:b}),{})}),[i]),p=D.useMemo(()=>()=>{let{error:h}=D.useContext(ke);return o?D.createElement(o,{error:h!=null?h:e?"Something went wrong while fetching the resource data.":void 0}):null},[o,e]);return D.createElement(Ar,{Context:ke,code:u,scope:s,noInline:!0},!e&&D.createElement(Tr,{Context:ke}),D.createElement(p,null))};import Q from"react";import{useDataProvider as wr,useResource as Sr}from"@pankod/refine-core";var er=(e,t,i)=>{let{resource:o,resourceName:u,id:s,resources:p}=Sr({resourceNameOrRouteName:t}),g=i!=null?i:s,d=wr(),[h,x]=Q.useState(void 0),[b,k]=Q.useState(!0),[c,A]=Q.useState(!1),l=Q.useCallback(async a=>{var v,n;let F=(v=X(o))!=null?v:se(u,void 0,p),I=d(F);A(!0);try{if(e==="list"||e==="create"){let f=await I.getList({resource:u});x((n=f.data)==null?void 0:n[0]),setTimeout(()=>{A(!1)},500)}if(e==="edit"||e==="show"&&a){let f=await I.getOne({resource:u,id:a});x(f.data),setTimeout(()=>{A(!1)},500)}}catch{setTimeout(()=>{A(!1)},500)}},[e,d,o,u,p]);return Q.useEffect(()=>{k(!1),!c&&!h&&l(g)},[l,g]),{data:h,loading:c,initial:b}};import Y from"react";import{useDataProvider as Lr}from"@pankod/refine-core";import{get as rr}from"lodash";var tr=({record:e,fields:t,infer:i})=>{let o=Lr(),[u,s]=Y.useState([]),[p,g]=Y.useState(!0),[d,h]=Y.useState(!1),x=Y.useCallback(async b=>{console.groupCollapsed("@pankod/refine-inferencer is trying to detect relations");let k=[];h(!0);try{let c=b.map(async l=>{if(l&&(l.relation||l.canRelation)&&e){let a=X(l.resource),F=o(a),I=Array.isArray(l.accessor)?void 0:l.multiple?e[l.key].map(v=>l.accessor?rr(v,l.accessor):v)[0]:l.accessor?rr(e[l.key],l.accessor):e[l.key];if(I&&l.resource)try{let{data:v}=await F.getOne({resource:l.resource.name,id:I});k.push({status:"success",resource:l.resource.name,field:l.key});let n=i("__",v,{},i);return{...l,relationInfer:n}}catch{return k.push({status:"error",resource:l.resource.name,field:l.key}),{...l,relationInfer:null}}if(I){let v,n;try{let{data:r}=await F.getOne({resource:Z(B(l.key)),id:I});k.push({status:"success",resource:Z(B(l.key)),field:l.key}),v=r,n=!0}catch{k.push({status:"error",resource:Z(B(l.key)),field:l.key});try{let{data:C}=await F.getOne({resource:K(B(l.key)),id:I});k.push({status:"success",resource:K(B(l.key)),field:l.key}),v=C,n=!1}catch{return k.push({status:"error",resource:K(B(l.key)),field:l.key}),{...l,relationInfer:null}}}let f=i("__",v,{},i),P=B(l.key);return{...l,relation:!0,type:"relation",resource:{name:n?Z(P):K(P)},fieldable:!1,canRelation:void 0,relationInfer:f}}return{...l,relationInfer:null}}return l}),A=await Promise.all(c);s(A.filter(l=>l)),setTimeout(()=>{h(!1)},500)}catch{setTimeout(()=>{h(!1)},500)}setTimeout(()=>{console.log(`Tried to detect relations with ${k.length} attempts and succeeded with ${k.filter(c=>c.status==="success").length} attempts.`),console.groupEnd()},500)},[o,e]);return Y.useEffect(()=>{g(!1),!d&&t&&t.length>0&&u.length===0&&x(t)},[x,d,t,x]),{fields:u,loading:d,initial:p}};var j=({type:e,additionalScope:t=[],customElements:i=[],fieldTransformers:o=[],renderer:u,loadingComponent:s,errorComponent:p,codeViewerComponent:g})=>{let d=Pe([..._e,...i]),h=Be([...Qe,...o]),x=({resourceName:k,fieldTransformer:c,id:A})=>{var S;let{resource:l,resources:a}=nr({resourceNameOrRouteName:k}),{resource:F}=nr(),{data:I,loading:v,initial:n}=er(e,k!=null?k:l==null?void 0:l.name,A),f=M.useMemo(()=>{if(I){let N=Object.keys(I).map(U=>{let _=I[U];return d(U,_,I,d)}).filter(Boolean),T=h(N,a,l,I,d);return c?T.flatMap(U=>{let _=c(U);return _?[_]:[]}):T}return[]},[I,a,l,c]),{fields:P,loading:r}=tr({record:I,fields:f,infer:d}),C=M.useMemo(()=>!v&&!r&&!n?u({resource:l,resources:a,fields:P,infer:d,isCustomPage:l.name!==F.name,id:A}):"",[l,a,P,v,r]);return M.createElement(M.Fragment,null,s&&(v||r)&&M.createElement(s,null),!v&&!r&&M.createElement(M.Fragment,null,M.createElement($e,{fetchError:!v&&!n&&!I,code:ye(C,V((S=l.label)!=null?S:l.name,e)),errorComponent:p,additionalScope:t}),g&&M.createElement(g,{code:fe(C),loading:v||r})))};return({name:k,resource:c,fieldTransformer:A,id:l})=>M.createElement(x,{fieldTransformer:A,resourceName:c!=null?c:k,key:c!=null?c:k,id:l})};import Ie from"react";import{Alert as Nr,AlertTitle as Rr,Box as Pr}from"@pankod/refine-mui";var G=({error:e})=>e?Ie.createElement(Pr,{style:{display:"flex",justifyContent:"center",alignItems:"center",minHeight:"200px"}},Ie.createElement(Nr,{severity:"error"},Ie.createElement(Rr,null,"Error"),e!=null?e:"")):null;import or from"react";import{Box as Br,CircularProgress as Er}from"@pankod/refine-mui";var z=()=>or.createElement(Br,{style:{display:"flex",justifyContent:"center",alignItems:"center",minHeight:"300px"}},or.createElement(Er,{size:"large"}));import w from"react";import{Button as xe,Box as be,Modal as Vr,IconButton as Dr}from"@pankod/refine-mui";import{IconCode as Mr,IconMessageCircle as Wr,IconX as Or,IconCopy as jr,IconCheck as Gr}from"@tabler/icons";var H=({code:e,loading:t})=>{let i=w.useMemo(()=>ce(e!=null?e:""),[e]),o=w.useRef(null),[u,s]=w.useState(!1),[p,g]=w.useState(!1);if(t)return null;let d=()=>{var h,x,b,k;(h=o==null?void 0:o.current)==null||h.select(),(x=o==null?void 0:o.current)==null||x.setSelectionRange(0,99999),typeof navigator<"u"&&(navigator.clipboard.writeText((k=(b=o==null?void 0:o.current)==null?void 0:b.value)!=null?k:""),g(!0),setTimeout(()=>{g(!1)},1e3))};return i?w.createElement(w.Fragment,null,w.createElement(be,{sx:{position:"fixed",bottom:12,right:60}},w.createElement(be,{sx:{display:"flex",gap:2}},w.createElement(xe,{variant:"contained",color:"primary",onClick:()=>s(!0),startIcon:w.createElement(Mr,{size:18}),sx:{whiteSpace:"nowrap"}},"Show Code"),w.createElement(xe,{variant:"contained",color:"secondary",component:"a",target:"_blank",href:"https://github.com/refinedev/refine/discussions/3046",startIcon:w.createElement(Wr,{size:18}),sx:{whiteSpace:"nowrap"}},"Give Feedback"))),w.createElement("div",{style:{height:"50px"}}),w.createElement(Vr,{open:u,onClose:()=>s(!1)},w.createElement(be,{sx:{position:"absolute",maxWidth:"800px",width:"100%",top:"50%",left:"50%",transform:"translate(-50%, -50%)",boxShadow:24}},w.createElement(Dr,{onClick:()=>s(!1),disableRipple:!0,sx:{position:"absolute",right:0,top:0,height:56,width:56,display:"flex",alignItems:"center",justifyContent:"center"}},w.createElement(Or,{size:18,style:{color:"#666b7a"}})),w.createElement(he,{code:i}),w.createElement("textarea",{ref:o,defaultValue:(i!=null?i:"").replace(/\\n/g,`\r
`),id:"code-input",style:{display:"none"}}),w.createElement("div",{style:{position:"absolute",bottom:12,right:12}},w.createElement(xe,{key:"copy",sx:{textTransform:"capitalize",width:100},startIcon:p?w.createElement(Gr,{size:18,style:{color:"green"}}):w.createElement(jr,{size:18}),color:"inherit",variant:"contained",onClick:d},p?"Copied":"Copy"))))):null};var ve=j({type:"show",additionalScope:[["@pankod/refine-mui","RefineMui",zr]],codeViewerComponent:H,loadingComponent:z,errorComponent:G,renderer:({resource:e,fields:t,isCustomPage:i,id:o})=>{var v;let u=V((v=e.label)!=null?v:e.name,"show"),s="record",p=[["useShow","@pankod/refine-core"],["Show","@pankod/refine-mui"],["Typography","@pankod/refine-mui"],["Stack","@pankod/refine-mui"]],d=t.filter(n=>(n==null?void 0:n.relation)&&!(n!=null&&n.fieldable)&&(n==null?void 0:n.resource)).filter(Boolean).map(n=>{if((n==null?void 0:n.relation)&&!n.fieldable&&n.resource){if(n.multiple){p.push(["useMany","@pankod/refine-core"]);let f=m(s,n.key);return n.accessor&&(f=`${m(s,n.key)}?.map((item: any) => ${m("item",void 0,n.accessor)})`),`
                    const { data: ${L(n.key,"Data")}, isLoading: ${L(n.key,"IsLoading")} } =
                    useMany({
                        resource: "${n.resource.name}",
                        ids: ${f} || [],
                        queryOptions: {
                            enabled: !!${s},
                        },
                    });
                    `}return p.push(["useOne","@pankod/refine-core"]),`
                    const { data: ${L(n.key,"Data")}, isLoading: ${L(n.key,"IsLoading")} } =
                    useOne({
                        resource: "${n.resource.name}",
                        id: ${m(s,n.key,n.accessor,!1)} || "",
                        queryOptions: {
                            enabled: !!${s},
                        },
                    });
                `}}).filter(Boolean),h=n=>{if(n.relation&&n.resource){let f=L(n.key,"Data"),P=L(n.key,"IsLoading");return n.multiple?(p.push(["TagField","@pankod/refine-mui"]),$`
                    <Typography variant="body1" fontWeight="bold">
                        ${y(n.key)}
                    </Typography>
                    {${P} ? (
                        <>
                            Loading...
                        </>
                        ) : (
                        <>
                        ${(()=>{var r;if(n.relationInfer)if((r=n.relationInfer)!=null&&r.accessor){if(Array.isArray(n.relationInfer.accessor))return"Not Handled.";{let C=L(n.key),S=m(C,void 0,n.relationInfer.accessor);return`
                                            <Stack direction="row" spacing={1}>
                                                {${f}?.data?.map((${C}: any) => (
                                                    <TagField key={${S}} value={${S}} />
                                                ))}
                                            </Stack>
                                        `}}else return"Not Handled.";else return"not-handled - relation with multiple but no resource"})()}
                        </>
                    )}
                    `):$`
                    <Typography variant="body1" fontWeight="bold">
                        ${y(n.key)}
                    </Typography>

                    {${P} ? (
                        <>Loading...</>
                    ) : (
                        <>
                        ${(()=>{var r;return n.relationInfer?(r=n.relationInfer)!=null&&r.accessor?Array.isArray(n.relationInfer.accessor)?`{${m(`${f}?.data`,void 0,n.relationInfer.accessor,' + " " + ')}}`:`{${f}?.data?.${n.relationInfer.accessor}}`:`{${f}?.data}`:`{${f}?.data?.id}`})()}
                        </>
                    )}
                    `}},x=n=>{if(n.type==="text"){if(p.push(["TextFieldComponent as TextField","@pankod/refine-mui"]),n.multiple){p.push(["TagField","@pankod/refine-mui"]);let f=m("item",void 0,n.accessor);return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${y(n.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${m(s,n.key)}?.map((item: any) => (
                        <TagField value={${f}} key={${f}} />
                    ))}
                    </Stack>
                `}return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${y(n.key)}
                    </Typography>
                    <TextField value={${m(s,n.key,n.accessor)}} />
                `}},b=n=>{if(n.type==="image"){if(n.multiple){let f=m("item",void 0,n.accessor);return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${y(n.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${m(s,n.key)}?.map((item: any) => (
                        <img style={{ maxWidth: 200, width: "100%", height: 200 }} src={${f}} key={${f}} />
                    ))}
                    </Stack>
                `}return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${y(n.key)}
                    </Typography>
                    <img style={{ maxWidth: 200, width: "100%", height: 200 }} src={${m(s,n.key,n.accessor," + ")}} />
                    `}},k=n=>{if(n.type==="email"){if(p.push(["EmailField","@pankod/refine-mui"]),n.multiple){p.push(["TagField","@pankod/refine-mui"]);let f=m("item",void 0,n.accessor);return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${y(n.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${m(s,n.key)}?.map((item: any) => (
                        <TagField value={${f}} key={${f}} />
                    ))}
                    </Stack>
                `}return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${y(n.key)}
                    </Typography>
                    <EmailField value={${m(s,n.key,n.accessor," + ")}} />
                `}},c=n=>{if(n.type==="url"){if(p.push(["UrlField","@pankod/refine-mui"]),n.multiple){p.push(["TagField","@pankod/refine-mui"]);let f=m("item",void 0,n.accessor);return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${y(n.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${m(s,n.key)}?.map((item: any) => (
                        <TagField value={${f}} key={${f}} />
                    ))}
                    </Stack>
                `}return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${y(n.key)}
                    </Typography>
                    <UrlField value={${m(s,n.key,n.accessor," + ")}} />
                `}},A=n=>{if(n.type==="boolean"){if(p.push(["BooleanField","@pankod/refine-mui"]),n.multiple){p.push(["TagField","@pankod/refine-mui"]);let f=m("item",void 0,n.accessor);return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${y(n.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${m(s,n.key)}?.map((item: any, index: number) => (
                        <TagField value={${f}} key={index} />
                    ))}
                    </Stack>
                `}return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${y(n.key)}
                    </Typography>
                    <BooleanField value={${m(s,n.key,n.accessor," && ")}} />
                `}},l=n=>{if(n.type==="date"){if(p.push(["DateField","@pankod/refine-mui"]),n.multiple){let f=m("item",void 0,n.accessor);return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${y(n.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${m(s,n.key)}?.map((item: any) => (
                        <DateField value={${f}} key={${f}} />
                    ))}
                    </Stack>
                `}return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${y(n.key)}
                    </Typography>
                    <DateField value={${m(s,n.key,n.accessor," + ' ' + ")}} />
                `}},a=n=>{if(n.type==="richtext")return p.push(["MarkdownField","@pankod/refine-mui"]),$`
                    <Typography variant="body1" fontWeight="bold">
                        ${y(n.key)}
                    </Typography>
                    <MarkdownField value={${m(s,n.key,n.accessor,' + " " + ')}} />
                `},F=n=>{if(n.type==="number"){if(p.push(["NumberField","@pankod/refine-mui"]),n.multiple){p.push(["TagField","@pankod/refine-mui"]);let f=m("item",void 0,n.accessor);return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${y(n.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${m(s,n.key)}?.map((item: any) => (
                        <TagField value={${f}} key={${f}} />
                    ))}
                    </Stack>
                `}return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${y(n.key)}
                    </Typography>
                    <NumberField value={${m(s,n.key,n.accessor,' + " " + ')} ?? ""} />
                `}},I=t.map(n=>{switch(n==null?void 0:n.type){case"text":return x(n);case"number":return F(n);case"richtext":return a(n);case"email":return k(n);case"image":return b(n);case"date":return l(n);case"boolean":return A(n);case"url":return c(n);case"relation":return h(n);default:return}});return O(p),$`
        ${W(p)}

        export const ${u} = () => {
            const { queryResult } = useShow(${i?`{ 
                        resource: "${e.name}", 
                        id: ${o}
                    }`:""});
            const { data, isLoading } = queryResult;
        
            const ${s} = data?.data;
        
            ${d}

            return (
                <Show isLoading={isLoading}>
                    <Stack gap={1}>
                        ${I.join("")}
                    </Stack>
                </Show>
            );
        };
        `}});import*as Hr from"@pankod/refine-mui";var Ce=j({type:"list",additionalScope:[["@pankod/refine-mui","RefineMui",Hr]],codeViewerComponent:H,loadingComponent:z,errorComponent:G,renderer:({resource:e,fields:t,isCustomPage:i})=>{var P;let o=V((P=e.label)!=null?P:e.name,"list"),u="dataGridProps?.rows",s=[["React","react",!0],["useDataGrid","@pankod/refine-mui"],["DataGrid","@pankod/refine-mui"],["GridColumns","@pankod/refine-mui"],["EditButton","@pankod/refine-mui"],["ShowButton","@pankod/refine-mui"],["DeleteButton","@pankod/refine-mui"],["List","@pankod/refine-mui"]],p=t.filter(r=>(r==null?void 0:r.relation)&&!(r!=null&&r.fieldable)&&(r==null?void 0:r.resource)),g=p.filter(Boolean).map(r=>{if((r==null?void 0:r.relation)&&!r.fieldable&&r.resource){s.push(["useMany","@pankod/refine-core"]);let C="";return r.multiple?C=`[].concat(...(${u}?.map((item: any) => ${m("item",r.key,r.accessor,!1)}) ?? []))`:C=`${u}?.map((item: any) => ${m("item",r.key,r.accessor,!1)}) ?? []`,`
                    const { data: ${L(r.key,"Data")}, isLoading: ${L(r.key,"IsLoading")} } =
                    useMany({
                        resource: "${r.resource.name}",
                        ids: ${C},
                        queryOptions: {
                            enabled: !!${u},
                        },
                    });
                    `}}).filter(Boolean),d=p==null?void 0:p.map(r=>{if(r&&r.resource)return`${L(r.key,"Data")}?.data`}).filter(Boolean),h=r=>{var C,S;if(r.relation&&r.resource){let N=`${L(r.key,"Data")}?.data`,T=L(r.key,"IsLoading");if(Array.isArray(r.accessor))return;let E=`${T} ? <>Loading...</> : `,U=`field: "${r.key}"`,_=r.accessor&&!r.multiple&&!Array.isArray(r.accessor)?`valueGetter: ({ row }) => {
                const value = ${m("row",r.key,r.accessor,!1)};

                return value;
            },`:"",Ae=`headerName: "${y(r.key)}"`,ee="";if(r.multiple){s.push(["TagField","@pankod/refine-mui"]);let re="item";if(r!=null&&r.relationInfer){let oe=`${N}?.find((resourceItems) => resourceItems.id === ${m("item",void 0,r.accessor)})`;re=m(oe,void 0,(C=r==null?void 0:r.relationInfer)==null?void 0:C.accessor)}ee=`
                    renderCell: function render({ value }) {
                        return ${E} (
                            <>
                                {${m("value",void 0)}?.map((item: any, index: number) => (
                                    <TagField key={index} value={${re}} />
                                ))}
                            </>
                        )
                    }
                    `}else if(r!=null&&r.relationInfer){let re=`${N}?.find((item) => item.id === value)`,oe=m(re,void 0,(S=r==null?void 0:r.relationInfer)==null?void 0:S.accessor);ee=`
                        renderCell: function render({ value }) {
                            return ${E} ${oe};
                        }
                        `}else ee="";return`
                    {
                        ${U},
                        ${Ae},${_}
                        minWidth: 300,${ee}
                    }
                `}},x=r=>{if(r.type==="image"){let C=`field: "${r.key}"`,S=`headerName: "${y(r.key)}"`,N=r.accessor&&!r.multiple&&!Array.isArray(r.accessor)?`valueGetter: ({ row }) => {
                const value = ${m("row",r.key,r.accessor,!1)};

                return value;
            },`:"",T=`
                    renderCell: function render({ value }) {
                        return (
                            <img src={${m("value",void 0,Array.isArray(r.accessor)?r.accessor:void 0," + ")}} style={{ height: "50px", maxWidth: "100px" }} />
                        )
                    }
                `;return r.multiple&&(T=`
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item: any, index: number) => (
                                    <img src={${m("item",void 0,r.accessor," + ")}} key={index} style={{ height: "50px", maxWidth: "100px" }} />
                                ))}
                                </>
                            )
                        }
                    `),`
                    {
                        ${C},
                        ${S},${N}
                        minWidth: 100,${T}
                    }
                `}},b=r=>{if(r.type==="email"){s.push(["EmailField","@pankod/refine-mui"]);let C=`field: "${r.key}"`,S=`headerName: "${y(r.key)}"`,N=r.accessor&&!r.multiple&&!Array.isArray(r.accessor)?`valueGetter: ({ row }) => {
                const value = ${m("row",r.key,r.accessor,!1)};

                return value;
            },`:"",T=`
                renderCell: function render({ value }) {
                    return (
                        <EmailField value={${m("value",void 0,Array.isArray(r.accessor)?r.accessor:void 0," + ")}} />
                    )
                }
            `;return r.multiple&&(s.push(["TagField","@pankod/refine-mui"]),T=`
                    renderCell: function render({ value }) {
                        return (
                            <>
                            {value?.map((item: any, index: number) => (
                                <TagField value={${m("item",void 0,r.accessor," + ")}} key={index} />
                            ))}
                            </>
                        )
                    }
                `),`
                {
                    ${C},
                    ${S},${N}
                    minWidth: 250,${T}
                }
            `}},k=r=>{if(r.type==="url"){s.push(["UrlField","@pankod/refine-mui"]);let C=`field: "${r.key}"`,S=`headerName: "${y(r.key)}"`,N=r.accessor&&!r.multiple&&!Array.isArray(r.accessor)?`valueGetter: ({ row }) => {
                const value = ${m("row",r.key,r.accessor,!1)};

                return value;
            },`:"",T=`
                    renderCell: function render({ value }) {
                        return (
                            <UrlField value={${m("value",void 0,Array.isArray(r.accessor)?r.accessor:void 0," + ")}} />
                        )
                    }
                `;return r.multiple&&(s.push(["TagField","@pankod/refine-mui"]),T=`
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item: any, index: any) => (
                                    <TagField value={${m("item",void 0,r.accessor," + ")}} key={index} />
                                ))}
                                </>
                            )
                        }
                    `),`
                    {
                        ${C},
                        ${S},${N}
                        minWidth: 250,${T}
                    }
                `}},c=r=>{if(r!=null&&r.type){s.push(["Checkbox","@pankod/refine-mui"]);let C=`field: "${r.key}"`,S=`headerName: "${y(r.key)}"`,N=r.accessor&&!r.multiple&&!Array.isArray(r.accessor)?`valueGetter: ({ row }) => {
                const value = ${m("row",r.key,r.accessor,!1)};

                return value;
            },`:"",T=`
                    renderCell: function render({ value }) {
                        return (
                            <Checkbox checked={!!${m("value",void 0,Array.isArray(r.accessor)?r.accessor:void 0," && ")}} />
                        );
                    }
                `;return r.multiple&&(T=`
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item: any, index: number) => (
                                    <Checkbox checked={!!${m("item",void 0,r.accessor," && ")}} key={index} />
                                ))}
                                </>
                            )
                        }
                    `),`
                    {
                        ${C},
                        ${S},${N}
                        minWidth: 250,${T}
                    }
                `}},A=r=>{if(r.type==="date"){s.push(["DateField","@pankod/refine-mui"]);let C=`field: "${r.key}"`,S=`headerName: "${y(r.key)}"`,N=r.accessor&&!r.multiple&&!Array.isArray(r.accessor)?`valueGetter: ({ row }) => {
                const value = ${m("row",r.key,r.accessor,!1)};

                return value;
            },`:"",T=`
                    renderCell: function render({ value }) {
                        return <DateField value={value} />;
                    }
                `;return r.multiple&&(T=`
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item: any, index: number) => (
                                    <DateField value={${m("item",void 0,r.accessor,' + " " + ')}} key={index} />
                                ))}
                                </>
                            )
                        }
                    `),`
                    {
                        ${C},
                        ${S},${N}
                        minWidth: 250,${T}
                    }
                `}},l=r=>{if((r==null?void 0:r.type)==="richtext"){s.push(["MarkdownField","@pankod/refine-mui"]);let C=`field: "${r.key}"`,S=`headerName: "${y(r.key)}"`,N=r.accessor&&!r.multiple&&!Array.isArray(r.accessor)?`valueGetter: ({ row }) => {
                const value = ${m("row",r.key,r.accessor,!1)};

                return value;
            },`:"",T=`
                    renderCell: function render({ value }) {
                        return <MarkdownField value={(value ?? "").slice(0, 80) + "..."} />;
                    }
                `;return r.multiple&&(T=`
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item: any, index: number) => (
                                    <MarkdownField value={(${m("item",void 0,r.accessor,' + " " + ')}).slice(0, 80) + "..."} key={index} />
                                ))}
                                </>
                            )
                        }
                    `),`
                    {
                        ${C},
                        ${S},${N}
                        minWidth: 250,${T}
                    }
                `}},a=r=>{if(r&&(r.type==="text"||r.type==="number")){let C=`field: "${r.key}"`,S=`headerName: "${y(r.key)}"`,N=r.accessor&&!r.multiple&&!Array.isArray(r.accessor)?`valueGetter: ({ row }) => {
                const value = ${m("row",r.key,r.accessor,!1)};

                return value;
            },`:"",T="";if(r.multiple){s.push(["TagField","@pankod/refine-mui"]);let E=m("item",void 0,r.accessor,' + " " + ');T=`
                    renderCell: function render({ row }) {
                        return (
                            <>
                                {(${m("row",r.key)})?.map((item: any) => (
                                    <TagField value={${E}} key={${E}} />
                                ))}
                            </>
                        )
                    }
                    `}return!r.multiple&&Array.isArray(r.accessor)&&(T=`
                    renderCell: function render({ row }) {
                        return 
                            <>{${m("row",r.key,r.accessor)}}</>
                        );
                    }
                    `),`
                {
                    ${C},
                    ${S},${N}${r.type==="number"?"type: 'number',":""}
                    minWidth: ${q(r.key)?50:200},${T}
                }
                `}},{canEdit:F,canShow:I,canDelete:v}=e!=null?e:{};F&&s.push(["EditButton","@pankod/refine-mui"]),I&&s.push(["ShowButton","@pankod/refine-mui"]),v&&s.push(["DeleteButton","@pankod/refine-mui"]);let n=F||I||v?$`
                {
                    field: "actions",
                    headerName: "Actions",
                    renderCell: function render({ row }) {
                        return (
                            <>
                                ${F?$`<EditButton hideText recordItemId={row.id} />`:""}
                                ${I?$`<ShowButton hideText recordItemId={row.id} />`:""}
                                ${v?$`<DeleteButton hideText recordItemId={row.id} />`:""}
                            </>
                        );
                    },
                    align: "center",
                    headerAlign: "center",
                    minWidth: 80,
                },
        `:"",f=t.map(r=>{switch(r==null?void 0:r.type){case"text":case"number":return a(r);case"richtext":return l(r);case"email":return b(r);case"image":return x(r);case"date":return A(r);case"boolean":return c(r);case"url":return k(r);case"relation":return h(r);default:return}});return O(s),$`
        ${W(s)}
        
        export const ${o} = () => {
            const { dataGridProps } = useDataGrid(
                ${i?`{ resource: "${e.name}" }`:""} 
            );
        
            ${g}

            const columns = React.useMemo<GridColumns<any>>(() => [
                ${[...f,n].filter(Boolean).join(`,\r
`)}
            ], [${d.join(",")}]);

            return (
                <List>
                    <DataGrid {...dataGridProps} columns={columns} autoHeight />
                </List>
            );
        };
        `}});import*as qr from"@pankod/refine-mui";import*as Ur from"@pankod/refine-react-hook-form";var Fe=j({type:"create",additionalScope:[["@pankod/refine-mui","RefineMui",qr],["@pankod/refine-react-hook-form","RefineReactHookForm",Ur]],codeViewerComponent:H,loadingComponent:z,errorComponent:G,renderer:({resource:e,fields:t,isCustomPage:i})=>{var k;let o=V((k=e.label)!=null?k:e.name,"create"),u=[["Create","@pankod/refine-mui"],["Box","@pankod/refine-mui"],["useForm","@pankod/refine-react-hook-form"]],p=t.filter(c=>(c==null?void 0:c.relation)&&!(c!=null&&c.fieldable)&&(c==null?void 0:c.resource)).filter(Boolean).map(c=>{if((c==null?void 0:c.relation)&&!c.fieldable&&c.resource)return u.push(["useAutocomplete","@pankod/refine-mui"]),`
                    const { autocompleteProps: ${L(c.key,"AutocompleteProps")} } =
                    useAutocomplete({
                        resource: "${c.resource.name}",
                    });
                `}).filter(Boolean),g=c=>{if(c.relation&&c.resource){u.push(["Autocomplete","@pankod/refine-mui"],["Controller","@pankod/refine-react-hook-form"]);let A=L(c.key,"AutocompleteProps"),l=c.relationInfer&&c.relationInfer.accessor?typeof c.relationInfer.accessor=="string"?c.relationInfer.accessor:c.relationInfer.accessor[0]:"title",a=l.includes(".")||l.includes("[")||l.includes("]")||l.includes("-"),F=c.accessor?m("item",void 0,c.accessor,!1):"(item?.id ?? item)",I=c.accessor?m("value",void 0,c.accessor,!1):"(value?.id ?? value)",v=c.accessor?"value":c.multiple?"value?.map((item: any) => item?.id ?? item)":"value?.id ?? value";return $`
                    <Controller
                        control={control}
                        name="${R(c.key,void 0)}"
                        rules={{ required: "This field is required" }}
                        // eslint-disable-next-line
                        ${c.multiple?"defaultValue={[] as any}":"defaultValue={null as any}"} 
                        render={({ field }) => (
                            <Autocomplete
                                {...${A}}
                                {...field}
                                ${c.multiple?"multiple":""}
                                onChange={(_, value) => {
                                    field.onChange(${v});
                                }}
                                getOptionLabel={(item) => {
                                    return (
                                        ${A}?.options?.find(
                                            (p) =>
                                                p?.id?.toString() ===
                                                ${F}?.toString(),
                                        )?.${a?`["${l}"]`:l} ?? ""
                                    );
                                }}
                                isOptionEqualToValue={(option, value) =>
                                    value === undefined ||
                                    option.id.toString() === ${I}?.toString()
                                }
                                renderInput={(params) => (
                                    <TextField
                                        {...params}
                                        label="${y(c.key)}"
                                        margin="normal"
                                        variant="outlined"
                                        error={!!${m("(errors as any)",c.key,c.accessor,!1)}}
                                        helperText={${m("(errors as any)",c.key,c.accessor,!1)}?.message}
                                        required
                                    />
                                )}
                            />
                        )}
                    />
                `}},d=c=>{if(c.type==="text"||c.type==="url"||c.type==="email"||c.type==="number"||c.type==="date"||c.type==="richtext")return q(c.key)||(u.push(["TextField","@pankod/refine-mui"]),c.multiple)?void 0:$`
                    <TextField
                        {...register("${R(c.key,void 0,c.accessor)}", {
                            required: "This field is required",
                        })}
                        error={!!${m("(errors as any)",c.key,c.accessor,!1)}}
                        helperText={${m("(errors as any)",c.key,c.accessor,!1)}?.message}
                        margin="normal"
                        fullWidth
                        InputLabelProps={{ shrink: true }}
                        ${c.type!=="date"&&c.type!=="richtext"?`type="${c.type}"`:""}
                        ${c.type==="richtext"?"multiline":""}
                        label="${y(c.key)}"
                        name="${R(c.key,void 0,c.accessor)}"
                    />
                `},h=c=>{if(c.type==="boolean")return u.push(["Checkbox","@pankod/refine-mui"],["FormControlLabel","@pankod/refine-mui"],["Controller","@pankod/refine-react-hook-form"]),c.multiple?void 0:$`
                    <Controller
                        control={control}
                        name="${R(c.key,void 0,c.accessor)}"
                        // eslint-disable-next-line
                        defaultValue={null as any}
                        render={({ field }) => (
                            <FormControlLabel label="${y(c.key)}" control={
                                <Checkbox
                                    {...field}
                                    checked={field.value}
                                    onChange={(event) => {
                                        field.onChange(event.target.checked);
                                    }}
                                />
                            } />
                        )}
                    />
                `},x=c=>{if(c.type==="date"){let A=d(c);return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-mui" package.
                        To use a <DatePicker> component, you can follow the official documentation for Material UI.
                        
                        Docs: https://mui.com/x/react-date-pickers/date-picker/#basic-usage
                    */}
                    ${A!=null?A:""}
                    `}},b=t.map(c=>{switch(c==null?void 0:c.type){case"text":case"number":case"email":case"url":case"richtext":return d(c);case"date":return x(c);case"boolean":return h(c);case"relation":return g(c);default:return}});return O(u),$`
        ${W(u)}
        
        export const ${o} = () => {
            const {
                saveButtonProps,
                refineCore: { formLoading },
                register,
                control,
                formState: { errors },
            } = useForm(
                ${i?`{ 
                    refineCoreProps: {
                        resource: "${e.name}",
                        action: "create",
                    }
                }`:""}
            );
        
            ${p}

            return (
                <Create isLoading={formLoading} saveButtonProps={saveButtonProps}>
                    <Box
                        component="form"
                        sx={{ display: "flex", flexDirection: "column" }}
                        autoComplete="off"
                    >
                        ${b.join("")}
                    </Box>
                </Create>
            );
        };
        `}});import*as _r from"@pankod/refine-mui";import*as Kr from"@pankod/refine-react-hook-form";var Te=j({type:"edit",additionalScope:[["@pankod/refine-mui","RefineMui",_r],["@pankod/refine-react-hook-form","RefineReactHookForm",Kr]],codeViewerComponent:H,loadingComponent:z,errorComponent:G,renderer:({resource:e,fields:t,isCustomPage:i,id:o})=>{var A,l;let u=V((A=e.label)!=null?A:e.name,"edit"),s=L((l=e.label)!=null?l:e.name,"Data"),p=[["Edit","@pankod/refine-mui"],["Box","@pankod/refine-mui"],["useForm","@pankod/refine-react-hook-form"]],d=t.filter(a=>(a==null?void 0:a.relation)&&!(a!=null&&a.fieldable)&&(a==null?void 0:a.resource)).filter(Boolean).map(a=>{if((a==null?void 0:a.relation)&&!a.fieldable&&a.resource){p.push(["useAutocomplete","@pankod/refine-mui"]);let F=m(s,a.key,a.accessor,!1);return a.multiple&&a.accessor&&(F=`${m(s,a.key)}?.map((item: any) => ${m("item",void 0,a.accessor)})`),`
                    const { autocompleteProps: ${L(a.key,"AutocompleteProps")} } =
                    useAutocomplete({
                        resource: "${a.resource.name}",
                        defaultValue: ${F},
                    });
                `}}).filter(Boolean),h=a=>{if(a.relation&&a.resource){p.push(["Autocomplete","@pankod/refine-mui"],["Controller","@pankod/refine-react-hook-form"]);let F=L(a.key,"AutocompleteProps"),I=a.relationInfer&&a.relationInfer.accessor?typeof a.relationInfer.accessor=="string"?a.relationInfer.accessor:a.relationInfer.accessor[0]:"title",v=I.includes(".")||I.includes("[")||I.includes("]")||I.includes("-"),n=a.accessor?m("item",void 0,a.accessor,!1):"(item?.id ?? item)",f=a.accessor?m("value",void 0,a.accessor,!1):"(value?.id ?? value)",P=a.accessor?"value":a.multiple?"value?.map((item: any) => item?.id ?? item)":"value?.id ?? value";return $`
                    <Controller
                        control={control}
                        name="${R(a.key,void 0)}"
                        rules={{ required: "This field is required" }}
                        // eslint-disable-next-line
                        ${a.multiple?"defaultValue={[] as any}":"defaultValue={null as any}"} 
                        render={({ field }) => (
                            <Autocomplete
                                {...${F}}
                                {...field}
                                ${a.multiple?"multiple":""}
                                onChange={(_, value) => {
                                    field.onChange(${P});
                                }}
                                getOptionLabel={(item) => {
                                    return (
                                        ${F}?.options?.find(
                                            (p) =>
                                                p?.id?.toString() ===
                                                ${n}?.toString(),
                                        )?.${v?`["${I}"]`:I} ?? ""
                                    );
                                }}
                                isOptionEqualToValue={(option, value) =>
                                    value === undefined ||
                                    option?.id?.toString() === ${f}?.toString()
                                }
                                renderInput={(params) => (
                                    <TextField
                                        {...params}
                                        label="${y(a.key)}"
                                        margin="normal"
                                        variant="outlined"
                                        error={!!${m("(errors as any)",a.key,a.accessor,!1)}}
                                        helperText={${m("(errors as any)",a.key,a.accessor,!1)}?.message}
                                        required
                                    />
                                )}
                            />
                        )}
                    />
                `}},x=a=>{if(a.type==="text"||a.type==="url"||a.type==="email"||a.type==="number"||a.type==="date"||a.type==="richtext"){if(p.push(["TextField","@pankod/refine-mui"]),a.multiple){p.push(["Box","@pankod/refine-mui"]);let F=R(a.key,"${index}",a.accessor),I=m("(errors as any)",a.key,void 0,!1)+"?.[index]";return`
                        <Box sx={{display: "flex", gap: 1}}>
                            {${m(s,a.key)}?.map((item: any, index: number) => (
                                <TextField
                                    key={index}
                                    {...register(\`${F}\`, {
                                        required: "This field is required",
                                    })}
                                    error={!!${I}}
                                    helperText={${I}?.message as string}
                                    margin="normal"
                                    fullWidth
                                    InputLabelProps={{ shrink: true }}
                                    ${a.type!=="date"&&a.type!=="richtext"?`type="${a.type}"`:""}
                                    ${a.type==="richtext"?"multiline":""}
                                    label={\`${y(a.key)} \${index+1}\`}
                                    name={\`${R(a.key,void 0,a.accessor)}.\${index}\`}
                                />
                            ))}
                        </Box>
                    `}return $`
                    <TextField
                        {...register("${R(a.key,void 0,a.accessor)}", {
                            required: "This field is required",
                        })}
                        error={!!${m("(errors as any)",a.key,a.accessor,!1)}}
                        helperText={${m("(errors as any)",a.key,a.accessor,!1)}?.message}
                        margin="normal"
                        fullWidth
                        InputLabelProps={{ shrink: true }}
                        ${a.type!=="date"&&a.type!=="richtext"?`type="${a.type}"`:""}
                        ${a.type==="richtext"?"multiline":""}
                        label="${y(a.key)}"
                        name="${R(a.key,void 0,a.accessor)}"
                        ${q(a.key)?"disabled":""}
                    />
                `}},b=a=>{if(a.type==="boolean"){if(p.push(["Checkbox","@pankod/refine-mui"],["FormControlLabel","@pankod/refine-mui"],["Controller","@pankod/refine-react-hook-form"]),a.multiple){p.push(["Box","@pankod/refine-mui"]);let F=R(a.key,"${index}",a.accessor);return`
                        <Box sx={{display: "flex", gap: 1}}>
                            {${m(s,a.key)}?.map((item: any, index: number) => (
                                <Controller
                                    key={index}
                                    control={control}
                                    name={\`${F}\`}
                                    // eslint-disable-next-line
                                    defaultValue={null as any}
                                    render={({ field }) => (
                                        <FormControlLabel label={\`${y(a.key)} \${index+1}\`} control={
                                            <Checkbox
                                                {...field}
                                                checked={field.value}
                                                onChange={(event) => {
                                                    field.onChange(event.target.checked);
                                                }}
                                            />
                                        } />
                                    )}
                                />
                            ))}
                        </Box>
                    `}return $`
                    <Controller
                        control={control}
                        name="${R(a.key,void 0,a.accessor)}"
                        // eslint-disable-next-line
                        defaultValue={null as any}
                        render={({ field }) => (
                            <FormControlLabel label="${y(a.key)}" control={
                                <Checkbox
                                    {...field}
                                    checked={field.value}
                                    onChange={(event) => {
                                        field.onChange(event.target.checked);
                                    }}
                                />
                            } />
                        )}
                    />
                `}},k=a=>{if(a.type==="date")return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-mui" package.
                        To use a <DatePicker> component, you can follow the official documentation for Material UI.
                        
                        Docs: https://mui.com/x/react-date-pickers/date-picker/#basic-usage
                    */}
                    ${x(a)}
                    `},c=t.map(a=>{switch(a==null?void 0:a.type){case"text":case"number":case"email":case"url":case"richtext":return x(a);case"date":return k(a);case"boolean":return b(a);case"relation":return h(a);default:return}});return O(p),$`
        ${W(p)}
        
        export const ${u} = () => {
            const {
                saveButtonProps,
                refineCore: { queryResult },
                register,
                control,
                formState: { errors },
            } = useForm(
                ${i?`{ 
                    refineCoreProps: {
                        resource: "${e.name}",
                        id: ${o},
                        action: "edit",
                    }
                }`:""}
            );
        
            const ${s} = queryResult?.data?.data;
        
            ${d}

            return (
                <Edit saveButtonProps={saveButtonProps}>
                    <Box
                        component="form"
                        sx={{ display: "flex", flexDirection: "column" }}
                        autoComplete="off"
                    >
                        ${c.join("")}
                    </Box>
                </Edit>
            );
        };
        `}});var pa=({action:e,id:t,...i})=>{let{useParams:o}=Zr(),{action:u,id:s}=o();switch(e!=null?e:u){case"show":return ne.createElement(ve,{...i,id:t!=null?t:s});case"create":return ne.createElement(Fe,{...i,id:t!=null?t:s});case"edit":return ne.createElement(Te,{...i,id:t!=null?t:s});default:return ne.createElement(Ce,{...i,id:t!=null?t:s})}};export{Fe as MuiCreateInferencer,Te as MuiEditInferencer,pa as MuiInferencer,Ce as MuiListInferencer,ve as MuiShowInferencer};
