// @pankod/refine-inferencer@1.6.1/dist/chakra-ui.js
var hr=Object.create;var pe=Object.defineProperty;var Ir=Object.getOwnPropertyDescriptor;var $r=Object.getOwnPropertyNames;var Fr=Object.getPrototypeOf,xr=Object.prototype.hasOwnProperty;var br=(e,r)=>{for(var a in r)pe(e,a,{get:r[a],enumerable:!0})},ze=(e,r,a,o)=>{if(r&&typeof r=="object"||typeof r=="function")for(let u of $r(r))!xr.call(e,u)&&u!==a&&pe(e,u,{get:()=>r[u],enumerable:!(o=Ir(r,u))||o.enumerable});return e};var v=(e,r,a)=>(a=e!=null?hr(Fr(e)):{},ze(r||!e||!e.__esModule?pe(a,"default",{value:e,enumerable:!0}):a,e)),Cr=e=>ze(pe({},"__esModule",{value:!0}),e);var Qr={};br(Qr,{ChakraUICreateInferencer:()=>ke,ChakraUIEditInferencer:()=>he,ChakraUIInferencer:()=>Zr,ChakraUIListInferencer:()=>ye,ChakraUIShowInferencer:()=>ge});module.exports=Cr(Qr);var ue=v(require("react")),kr=require("@pankod/refine-core");var Pr=v(require("@pankod/refine-chakra-ui"));var z=v(require("react")),De=require("@pankod/refine-core");var vr=/(image|photo|avatar|url|logo|cover|thumbnail|icon)/i,Ie=(e,r)=>{let a=["name","label","title","count","content","username","nickname","login","firstName","lastName","url"];vr.test(e)&&a.unshift("url");let o=a.find(u=>u in r&&!!r[u]);if(o)return o==="firstName"&&"lastName"in r?["firstName","lastName"]:o};function $e(e,r,a){var u;if(r)return r;let o=a==null?void 0:a.find(s=>s.name===e);return(u=o==null?void 0:o.options)!=null&&u.dataProviderName?o.options.dataProviderName:"default"}var oe=e=>{var r;return(r=e==null?void 0:e.options)==null?void 0:r.dataProviderName};var Fe=e=>{let r=[...e].filter(Boolean).sort((a,o)=>(o.priority||0)-(a.priority||0))[0];return r||null};var Oe=v(require("prettier/standalone")),je=v(require("prettier/parser-typescript")),xe=e=>{try{return Oe.default.format(e,{parser:"typescript",plugins:[je.default],arrowParens:"always",trailingComma:"all",semi:!0,tabWidth:4,printWidth:80})}catch(r){return console.warn("Formatting error in inferencer",r),e}};var be=e=>e.replace(/ {2,}/g," ").replace(/\n{2,}/g,`
`);var Ce=/(-id|-ids|_id|_ids|Id|Ids|ID|IDs)(\[\])?$/,Ue=(e,r)=>{let a=Ce.test(e),o=typeof r=="string"||typeof r=="number",u=Array.isArray(r)&&r.every(s=>typeof s=="string"||typeof s=="number");return a&&o||a&&u?{key:e,relation:!0,multiple:!!u,type:"relation",priority:1}:!1};var M=e=>e.replace(Ce,"");var y=e=>M(e).replace(/([a-z])([A-Z])/g,"$1 $2").replace(/_/g," ").replace(/-/g," ").split(" ").map(p=>p.charAt(0).toUpperCase()+p.slice(1)).join(" ");var ve=e=>e.replace(/export\s+(const|let|var|type|interface|function|class)\s+(\w+)\s*(=|:)\s*/g,"$1 $2 = ").split(`
`).filter(u=>!u.trim().startsWith("export default")).join(`
`);var Ke=/import(?:(?:(?:[ \n\t]+([^ *\n\t\{\},]+)[ \n\t]*(?:,|[ \n\t]+))?([ \n\t]*\{(?:[ \n\t]*[^ \n\t"'\{\}]+[ \n\t]*,?)+\})?[ \n\t]*)|[ \n\t]*\*[ \n\t]*as[ \n\t]+([^ \n\t\{\}]+)[ \n\t]+)from[ \n\t]*(?:['"])([^'"\n]+)(?:['"])(?:;?)/g,Ar=/import[ \n\t](?:['"])([^'"\n]+)(?:['"])(?:;?)/g,Tr=/((?:\w|\s|_)*)( as )((?:\w|\s|_)*)( |,)?/g,Ae=(e,r={})=>{let a=e.matchAll(Ke),o=new Set;for(let u of a){let[,s,p,I,l]=u;if(l in r){let $=r[l];if(l==="react")continue;s&&o.add(`const { default: ${s} } = ${$};`),p&&o.add(`const${p.replace(Tr,"$1: $3$4")} = ${$};`),I&&o.add(`const ${I} = ${$};`)}}return be(`
    ${Array.from(o).join(`
`)}
    ${e.replace(Ke,"").replace(Ar,"")}
    `)};var ae=v(require("pluralize"));var Te=(e,r)=>e?r.find(o=>{let u=M(e.key);return o.name===u||o.name===(ae.default.isPlural(u)?u:ae.default.plural(u))||o.name===(ae.default.isSingular(u)?u:ae.default.singular(u))}):void 0;var qe=v(require("pluralize"));var P=(e,r)=>{let a=e.replace(/[^a-zA-Z0-9]/g," "),o=qe.default.singular(a),u=y(o).replace(/ /g,""),s=r.charAt(0).toUpperCase()+r.slice(1);return/^[a-zA-Z]/.test(e)?`${u}${s}`:`${s}${u}`};var Sr=e=>e?Array.isArray(e)?e.join(""):typeof e=="string"?e:typeof e=="object"||typeof e=="function"?e.toString():typeof e=="number"?String(e):typeof e=="boolean"&&e?"true":e:"",g=(e,...r)=>e.reduce((o,u,s)=>`${o}${u}${Sr(r[s])}`,"").trim();var Se=(e,r)=>`
    ${e!=null?e:""}
    
    render(typeof ${r} !== "undefined" ? <${r} /> : <></>);
    `;var Re=e=>e==null?void 0:e.replace(/(\/\* hidden-start \*\/)(.|\n)*?(\/\* hidden-end \*\/)/g,"");var Rr=/^[a-zA-Z_$][a-zA-Z_$0-9]*$/,_e=e=>Rr.test(e);function we(e,r,a){let o=`${e}`;return r&&(o+="?.",_e(r)?o+=r:o+=`['${r}']`),a&&(o+="?.",_e(a)?o+=a:o+=`['${a}']`),o}function wr(e,r,a){return(a!=null?a:[]).map(u=>we(e,r,u))}var m=(e,r,a,o=' + " " + ')=>Array.isArray(a)?o?wr(e,r,a).join(o):we(e,r,a[0]):we(e,r,a),B=(e,r,a)=>{let o=e;return r&&(o+=`.${r}`),a&&(Array.isArray(a)?o+=`.${a[0]}`:o+=`.${a}`),o};var K=e=>{let r=e.reduce((o,[u,s,p])=>(o[s]||(o[s]=[]),o[s].includes(u)||(p?o[s].push([u,!0]):o[s].push(u)),o),{});return Object.entries(r).map(([o,u])=>{let s=u.find($=>Array.isArray($)&&$[1]),p=u.filter($=>typeof $=="string"),I=s?`${s[0]}`:"",l=p.length?`{ ${p.join(", ")} }`:"";return`import ${I}${I&&l?", ":""}${l} from "${o}";`})};var He=v(require("pluralize")),j=(e="")=>He.default.isSingular(e)?e:He.default.singular(e);var Le=v(require("pluralize")),re=(e="")=>Le.default.isPlural(e)?e:Le.default.plural(e);var Hr=/^id$/i,se=e=>Hr.test(e);var ie=e=>e.relationInfer&&e.relationInfer.accessor?Array.isArray(e.relationInfer.accessor)?`optionLabel: "${e.relationInfer.accessor[0]}",`:e.relationInfer.accessor!=="title"?`optionLabel: "${e.relationInfer.accessor}",`:"":"";var q=e=>{};var A=(e,r="")=>{if(Number.isInteger(Number(e)))return r.charAt(0).toLowerCase()+r.slice(1)+e;let o=e.replace(/^\d+/,"").replace(/[^a-zA-Z0-9]/g,"-"),u=y(o).replace(/ /g,"");return`${u.charAt(0).toLowerCase()+u.slice(1)}${r.charAt(0).toUpperCase()+r.slice(1)}`};var We=e=>{let r=(a,o,u,s=r)=>{let p=e.map(l=>l(a,o,u,s));return Fe(p)};return r};var Ge=e=>(a,o,u,s,p)=>e.reduce((I,l)=>l(I,o,u,s,p),a);var Ze=(e,r,a,o)=>{let u=Array.isArray(r),s=Array.isArray(r)&&r.every(p=>typeof p=="string"||typeof p=="number");if(u){if(!s){let I=o(e,r[0],a,o);return I?{...I,key:e,multiple:!0,priority:1}:!1}let p=o(e,r[0],a,o)||{type:"string"};return{key:e,multiple:!0,priority:1,type:p==null?void 0:p.type}}return!1};var Qe=(e,r)=>typeof r=="boolean"?{key:e,type:"boolean"}:!1;var Ee=v(require("dayjs")),Lr=/(_at|_on|At|On|AT|ON)(\[\])?$/,Xe=(e,r)=>{let a=Lr.test(e)&&(0,Ee.default)(r).isValid(),o=typeof r=="string"&&(0,Ee.default)(r).isValid();return a||o?{key:e,type:"date",priority:1}:!1};var Er=/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,Je=(e,r)=>typeof r=="string"&&Er.test(r)?{key:e,type:"email",priority:1}:!1;var Nr=/\.(gif|jpe?g|tiff?|png|webp|bmp|svg)$/i,Ye=(e,r)=>typeof r=="string"&&Nr.test(r)?{key:e,type:"image",priority:2}:!1;var er=(e,r)=>typeof r>"u"||r===null?null:!1;var rr=(e,r)=>typeof r=="number"?{key:e,type:"number"}:!1;var Vr=/id$/i,nr=(e,r,a,o)=>{let u=r!==null,s=!Array.isArray(r);if(u&&s&&typeof r=="object"){if(Object.keys(r).length===1&&Vr.test(Object.keys(r)[0]))return{key:e,type:"relation",relation:!0,accessor:"id",priority:1};let l=Ie(e,r);if((l?l.length>0:!1)&&l){let b=l&&Array.isArray(l)?l[0]:l,C=o(b,r[b],r,o);if(C){let k=Array.isArray(l)?l.map(i=>C.accessor?`${i}.${Array.isArray(C.accessor)?C.accessor[0]:C.accessor}`:i):C.accessor?Array.isArray(C.accessor)?`${l}.${C.accessor[0]}`:`${l}.${C.accessor}`:l;return{...C,fieldable:!0,key:e,accessor:k,priority:1}}}return{key:e,fieldable:!1,type:"object",priority:1}}return!1};var tr=(e,r)=>typeof r=="string"&&r.length>100?{key:e,type:"richtext",priority:1}:!1;var or=(e,r)=>typeof r=="string"?{key:e,type:"text"}:!1;var Br=/^(https?|ftp):\/\/(-\.)?([^\s/?\.#-]+\.?)+(\/[^\s]*)?$/i,ar=(e,r)=>typeof r=="string"&&Br.test(r)?{key:e,type:"url",priority:1}:!1;var sr=[Ze,Qe,Xe,Je,Ye,er,rr,nr,Ue,tr,or,ar];var ir=(e,r,a,o)=>e.map(s=>{if(!s.relation&&(s.type==="text"||s.type==="richtext"||s.type==="number")&&!s.canRelation){let p=/^[a-z0-9-]+$/,I=k=>p.test(`${k}`),l=s.key.toLowerCase()!=="id";return!(!s.accessor||typeof s.accessor=="string")||!l?s:(s.multiple?o[s.key]:[o[s.key]]).every(k=>I(s.accessor?k[s.accessor]:k))?{...s,canRelation:!0}:s}return s});var Mr=/(image|photo|avatar|cover|thumbnail|icon)/i,cr=e=>e.map(a=>a.type==="url"&&Mr.test(a.key)?{...a,type:"image"}:a);var ur=(e,r)=>e.map(o=>{let u=Te(o,r);return u?{...o,relation:!0,type:"relation",resource:u}:o});var pr=(e,r,a,o,u)=>e.map(p=>{if(p.relation&&p.type==="relation"&&!p.resource){let I=p.accessor?o[p.key][p.accessor]:o[p.key],l=u(p.key,I,o,u);if(l&&l.type!=="relation")return{...p,fieldable:!0,relation:!1,type:l.type,canRelation:!0}}return p});var mr=[cr,ur,pr,ir];var _=v(require("react")),me=v(require("prism-react-renderer")),lr=v(require("prism-react-renderer/themes/vsDark")),Ne=({code:e,wrapperProps:r})=>((0,_.useEffect)(()=>{var o;let a=document.createElement("style");a.appendChild(document.createTextNode(`
                #code-highlight::-webkit-scrollbar {-webkit-appearance: none; height: 7px; width: 7px;}
                #code-highlight::-webkit-scrollbar-thumb {background-color: #666b7a;}
                #code-highlight::-webkit-scrollbar-corner {background-color: rgb(30, 30, 30);}
                `)),(o=document.getElementById("code-highlight"))==null||o.appendChild(a)},[]),_.default.createElement("div",{id:"code-highlight",style:{maxHeight:"75vh",height:"100%",overflow:"auto",backgroundColor:"rgb(30, 30, 30)"},...r!=null?r:{}},_.default.createElement(me.default,{...me.defaultProps,theme:lr.default,code:e,language:"tsx"},({className:a,style:o,tokens:u,getLineProps:s,getTokenProps:p})=>_.default.createElement("pre",{className:a,style:{...o,padding:"14px 14px 14px 14px",margin:"0",width:"100%",boxSizing:"border-box"}},u.map((I,l)=>_.default.createElement("div",{key:l,...s({line:I,key:l})},I.map(($,b)=>_.default.createElement("span",{key:b,...p({token:$,key:b})}))))))));var D=v(require("react")),Dr=v(require("@pankod/refine-core")),le=require("@aliemir/react-live");var dr=[["react","React",D.default],["@pankod/refine-core","RefineCore",Dr]],Ve=D.default.createContext({}),Be=({fetchError:e,code:r,additionalScope:a,errorComponent:o})=>{var I;let u=(I=r&&ve(Ae(r,[...dr,...a!=null?a:[]].reduce((l,[$,b,C,k])=>k?l:{...l,[$]:b},{}))))!=null?I:"",s=D.default.useMemo(()=>({React:D.default,...D.default,...[...dr,...a!=null?a:[]].reduce((l,[$,b,C])=>({...l,[b]:C}),{})}),[a]),p=D.default.useMemo(()=>()=>{let{error:$}=D.default.useContext(Ve);return o?D.default.createElement(o,{error:$!=null?$:e?"Something went wrong while fetching the resource data.":void 0}):null},[o,e]);return D.default.createElement(le.LiveProvider,{Context:Ve,code:u,scope:s,noInline:!0},!e&&D.default.createElement(le.LivePreview,{Context:Ve}),D.default.createElement(p,null))};var ne=v(require("react")),de=require("@pankod/refine-core");var fr=(e,r,a)=>{let{resource:o,resourceName:u,id:s,resources:p}=(0,de.useResource)({resourceNameOrRouteName:r}),I=a!=null?a:s,l=(0,de.useDataProvider)(),[$,b]=ne.default.useState(void 0),[C,k]=ne.default.useState(!0),[i,R]=ne.default.useState(!1),d=ne.default.useCallback(async c=>{var S,V;let w=(S=oe(o))!=null?S:$e(u,void 0,p),F=l(w);R(!0);try{if(e==="list"||e==="create"){let t=await F.getList({resource:u});b((V=t.data)==null?void 0:V[0]),setTimeout(()=>{R(!1)},500)}if(e==="edit"||e==="show"&&c){let t=await F.getOne({resource:u,id:c});b(t.data),setTimeout(()=>{R(!1)},500)}}catch{setTimeout(()=>{R(!1)},500)}},[e,l,o,u,p]);return ne.default.useEffect(()=>{k(!1),!i&&!$&&d(I)},[d,I]),{data:$,loading:i,initial:C}};var te=v(require("react")),gr=require("@pankod/refine-core");var Me=require("lodash"),yr=({record:e,fields:r,infer:a})=>{let o=(0,gr.useDataProvider)(),[u,s]=te.default.useState([]),[p,I]=te.default.useState(!0),[l,$]=te.default.useState(!1),b=te.default.useCallback(async C=>{console.groupCollapsed("@pankod/refine-inferencer is trying to detect relations");let k=[];$(!0);try{let i=C.map(async d=>{if(d&&(d.relation||d.canRelation)&&e){let c=oe(d.resource),w=o(c),F=Array.isArray(d.accessor)?void 0:d.multiple?e[d.key].map(S=>d.accessor?(0,Me.get)(S,d.accessor):S)[0]:d.accessor?(0,Me.get)(e[d.key],d.accessor):e[d.key];if(F&&d.resource)try{let{data:S}=await w.getOne({resource:d.resource.name,id:F});k.push({status:"success",resource:d.resource.name,field:d.key});let V=a("__",S,{},a);return{...d,relationInfer:V}}catch{return k.push({status:"error",resource:d.resource.name,field:d.key}),{...d,relationInfer:null}}if(F){let S,V;try{let{data:n}=await w.getOne({resource:re(M(d.key)),id:F});k.push({status:"success",resource:re(M(d.key)),field:d.key}),S=n,V=!0}catch{k.push({status:"error",resource:re(M(d.key)),field:d.key});try{let{data:x}=await w.getOne({resource:j(M(d.key)),id:F});k.push({status:"success",resource:j(M(d.key)),field:d.key}),S=x,V=!1}catch{return k.push({status:"error",resource:j(M(d.key)),field:d.key}),{...d,relationInfer:null}}}let t=a("__",S,{},a),h=M(d.key);return{...d,relation:!0,type:"relation",resource:{name:V?re(h):j(h)},fieldable:!1,canRelation:void 0,relationInfer:t}}return{...d,relationInfer:null}}return d}),R=await Promise.all(i);s(R.filter(d=>d)),setTimeout(()=>{$(!1)},500)}catch{setTimeout(()=>{$(!1)},500)}setTimeout(()=>{console.log(`Tried to detect relations with ${k.length} attempts and succeeded with ${k.filter(i=>i.status==="success").length} attempts.`),console.groupEnd()},500)},[o,e]);return te.default.useEffect(()=>{I(!1),!l&&r&&r.length>0&&u.length===0&&b(r)},[b,l,r,b]),{fields:u,loading:l,initial:p}};var W=({type:e,additionalScope:r=[],customElements:a=[],fieldTransformers:o=[],renderer:u,loadingComponent:s,errorComponent:p,codeViewerComponent:I})=>{let l=We([...sr,...a]),$=Ge([...mr,...o]),b=({resourceName:k,fieldTransformer:i,id:R})=>{var L;let{resource:d,resources:c}=(0,De.useResource)({resourceNameOrRouteName:k}),{resource:w}=(0,De.useResource)(),{data:F,loading:S,initial:V}=fr(e,k!=null?k:d==null?void 0:d.name,R),t=z.default.useMemo(()=>{if(F){let E=Object.keys(F).map(U=>{let ee=F[U];return l(U,ee,F,l)}).filter(Boolean),T=$(E,c,d,F,l);return i?T.flatMap(U=>{let ee=i(U);return ee?[ee]:[]}):T}return[]},[F,c,d,i]),{fields:h,loading:n}=yr({record:F,fields:t,infer:l}),x=z.default.useMemo(()=>!S&&!n&&!V?u({resource:d,resources:c,fields:h,infer:l,isCustomPage:d.name!==w.name,id:R}):"",[d,c,h,S,n]);return z.default.createElement(z.default.Fragment,null,s&&(S||n)&&z.default.createElement(s,null),!S&&!n&&z.default.createElement(z.default.Fragment,null,z.default.createElement(Be,{fetchError:!S&&!V&&!F,code:Se(x,P((L=d.label)!=null?L:d.name,e)),errorComponent:p,additionalScope:r}),I&&z.default.createElement(I,{code:Re(x),loading:S||n})))};return({name:k,resource:i,fieldTransformer:R,id:d})=>z.default.createElement(b,{fieldTransformer:R,resourceName:i!=null?i:k,key:i!=null?i:k,id:d})};var ce=v(require("react")),G=require("@pankod/refine-chakra-ui"),Z=({error:e})=>e?ce.default.createElement(G.Center,{style:{minHeight:300}},ce.default.createElement(G.Alert,{status:"error",variant:"subtle",flexDirection:"column",alignItems:"center",justifyContent:"center",textAlign:"center",height:"200px"},ce.default.createElement(G.AlertIcon,{boxSize:"40px",mr:0}),ce.default.createElement(G.AlertDescription,{mt:4,maxWidth:"sm"},e!=null?e:""))):null;var Pe=v(require("react")),fe=require("@pankod/refine-chakra-ui"),Q=()=>Pe.default.createElement(fe.Box,{position:"relative",bg:"chakra-body-bg",minH:120},Pe.default.createElement(fe.Spinner,{position:"absolute",top:"50%",left:"50%",transform:"translate(-50%, -50%)"}));var H=v(require("react")),N=require("@pankod/refine-chakra-ui"),X=require("@tabler/icons");var J=({code:e,loading:r})=>{let a=H.default.useMemo(()=>xe(e!=null?e:""),[e]),o=H.default.useRef(null),[u,s]=H.default.useState(!1),{isOpen:p,onOpen:I,onClose:l}=(0,N.useDisclosure)();if(r)return null;let $=()=>{var b,C,k,i;(b=o==null?void 0:o.current)==null||b.select(),(C=o==null?void 0:o.current)==null||C.setSelectionRange(0,Number.MAX_SAFE_INTEGER),typeof navigator<"u"&&(navigator.clipboard.writeText((i=(k=o==null?void 0:o.current)==null?void 0:k.value)!=null?i:""),s(!0),setTimeout(()=>{s(!1)},1e3))};return a?H.default.createElement(H.default.Fragment,null,H.default.createElement(N.Box,{position:"fixed",bottom:"12px",right:"60px"},H.default.createElement(N.HStack,{spacing:4},H.default.createElement(N.Button,{variant:"solid",colorScheme:"green",leftIcon:H.default.createElement(X.IconCode,{size:18}),onClick:I},"Show Code"),H.default.createElement(N.Button,{colorScheme:"gray",variant:"solid",leftIcon:H.default.createElement(X.IconMessageCircle,{size:18}),as:"a",target:"_blank",href:"https://github.com/refinedev/refine/discussions/3046"},"Give Feedback"))),H.default.createElement("div",{style:{height:"50px"}}),H.default.createElement(N.Modal,{size:"4xl",isOpen:p,onClose:l},H.default.createElement(N.ModalOverlay,null),H.default.createElement(N.ModalContent,null,H.default.createElement(N.ModalCloseButton,{sx:{position:"absolute",right:0,top:0,height:"56px",width:"56px",color:"#666b7a"}}),H.default.createElement(N.ModalBody,{sx:{padding:0}},H.default.createElement(Ne,{code:a}),H.default.createElement("textarea",{ref:o,defaultValue:(a!=null?a:"").replace(/\\n/g,`\r
`),id:"code-input",style:{padding:0,margin:0,width:0,height:0,opacity:0,border:"none",display:"block"}})),H.default.createElement(N.Button,{sx:{position:"absolute",bottom:"12px",right:"12px",width:100},key:"copy",leftIcon:u?H.default.createElement(X.IconCheck,{size:18,style:{color:"green"}}):H.default.createElement(X.IconCopy,{size:18}),onClick:$},u?"Copied":"Copy")))):null};var ge=W({type:"show",additionalScope:[["@pankod/refine-chakra-ui","RefineChakraUI",Pr]],codeViewerComponent:J,loadingComponent:Q,errorComponent:Z,renderer:({resource:e,fields:r,isCustomPage:a,id:o})=>{var V;let u=P((V=e.label)!=null?V:e.name,"show"),s="record",p=[["useShow","@pankod/refine-core"],["Show","@pankod/refine-chakra-ui"],["Heading","@pankod/refine-chakra-ui"]],l=r.filter(t=>(t==null?void 0:t.relation)&&!(t!=null&&t.fieldable)&&(t==null?void 0:t.resource)).filter(Boolean).map(t=>{if((t==null?void 0:t.relation)&&!t.fieldable&&t.resource){if(t.multiple){p.push(["useMany","@pankod/refine-core"]);let h=m(s,t.key);return t.accessor&&(h=`${m(s,t.key)}?.map((item: any) => ${m("item",void 0,t.accessor)})`),`
                    const { data: ${A(t.key,"Data")}, isLoading: ${A(t.key,"IsLoading")} } =
                    useMany({
                        resource: "${t.resource.name}",
                        ids: ${h} || [],
                        queryOptions: {
                            enabled: !!${s},
                        },
                    });
                    `}return p.push(["useOne","@pankod/refine-core"]),`
                    const { data: ${A(t.key,"Data")}, isLoading: ${A(t.key,"IsLoading")} } =
                    useOne({
                        resource: "${t.resource.name}",
                        id: ${m(s,t.key,t.accessor,!1)} || "",
                        queryOptions: {
                            enabled: !!${s},
                        },
                    });
                `}}).filter(Boolean),$=t=>{if(t.relation&&t.resource){let h=A(t.key,"Data"),n=A(t.key,"IsLoading");return t.multiple?(p.push(["TagField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),g`
                    <Heading as="h5" size="sm" mt={4} >${y(t.key)}</Heading>
                    {${n} ? <>Loading...</> : (
                        <>
                        ${(()=>{var x;if(t.relationInfer)if((x=t.relationInfer)!=null&&x.accessor){if(Array.isArray(t.relationInfer.accessor))return"Not Handled.";{let L=A(t.key),E=m(L,void 0,t.relationInfer.accessor);return`<HStack spacing="12px">{${h}?.data?.map((${L}: any) => <TagField key={${E}} value={${E}} />)}</HStack>`}}else return"Not Handled.";else return"not-handled - relation with multiple but no resource"})()}
                        </>
                    )}
                    `):g`
                    <Heading as="h5" size="sm" mt={4} >${y(t.key)}</Heading>
                    {${n} ? <>Loading...</> : (
                        <>
                        ${(()=>{var x;return t.relationInfer?(x=t.relationInfer)!=null&&x.accessor?Array.isArray(t.relationInfer.accessor)?`{${m(`${h}?.data`,void 0,t.relationInfer.accessor,' + " " + ')}}`:`{${h}?.data?.${t.relationInfer.accessor}}`:`{${h}?.data}`:`{${h}?.data?.id}`})()}
                        </>
                    )}
                    
                    `}},b=t=>{if(t.type==="text"){if(p.push(["TagField","@pankod/refine-chakra-ui"],["TextField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),t.multiple){let h=m("item",void 0,t.accessor);return g`
                    <Heading as="h5" size="sm" mt={4} >${y(t.key)}</Heading>
                    <HStack spacing="12px">
                    {${m(s,t.key)}?.map((item: any) => (
                        <TagField value={${h}} key={${h}} />
                    ))}
                    </HStack>
                `}return g`
                    <Heading as="h5" size="sm" mt={4} >${y(t.key)}</Heading>
                    <TextField value={${m(s,t.key,t.accessor)}} />
                `}},C=t=>{if(t.type==="image"){if(p.push(["Image","@pankod/refine-chakra-ui"]),t.multiple){let h=m("item",void 0,t.accessor);return g`
                    <Heading as="h5" size="sm" mt={4} >${y(t.key)}</Heading>
                    {${m(s,t.key)}?.map((item: any) => (
                        <Image sx={{ maxWidth: 200 }} src={${h}} key={${h}} />
                    ))}
                `}return g`
                    <Heading as="h5" size="sm" mt={4} >${y(t.key)}</Heading>
                    <Image sx={{ maxWidth: 200 }} src={${m(s,t.key,t.accessor," + ")}} />
                    `}},k=t=>{if(t.type==="email"){if(p.push(["TagField","@pankod/refine-chakra-ui"],["EmailField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),t.multiple){let h=m("item",void 0,t.accessor);return g`
                    <Heading as="h5" size="sm" mt={4} >${y(t.key)}</Heading>
                    <HStack spacing="12px">
                    {${m(s,t.key)}?.map((item: any) => (
                        <TagField value={${h}} key={${h}} />
                    ))}
                    </HStack>
                `}return g`
                    <Heading as="h5" size="sm" mt={4} >${y(t.key)}</Heading>
                    <EmailField value={${m(s,t.key,t.accessor," + ")}} />
                `}},i=t=>{if(t.type==="url"){if(p.push(["TagField","@pankod/refine-chakra-ui"],["UrlField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),t.multiple){let h=m("item",void 0,t.accessor);return g`
                    <Heading as="h5" size="sm" mt={4} >${y(t.key)}</Heading>
                    <HStack spacing="12px">
                    {${m(s,t.key)}?.map((item: any) => (
                        <TagField value={${h}} key={${h}} />
                    ))}
                    </HStack>
                `}return g`
                    <Heading as="h5" size="sm" mt={4} >${y(t.key)}</Heading>
                    <UrlField value={${m(s,t.key,t.accessor," + ")}} />
                `}},R=t=>{if(t.type==="boolean"){if(p.push(["TagField","@pankod/refine-chakra-ui"],["BooleanField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),t.multiple){let h=m("item",void 0,t.accessor);return g`
                    <Heading as="h5" size="sm" mt={4} >${y(t.key)}</Heading>
                    <HStack spacing="12px">
                    {${m(s,t.key)}?.map((item: any) => (
                        <TagField value={${h}} key={${h}} />
                    ))}
                    </HStack>
                `}return g`
                    <Heading as="h5" size="sm" mt={4} >${y(t.key)}</Heading>
                    <BooleanField value={${m(s,t.key,t.accessor," && ")}} />
                `}},d=t=>{if(t.type==="date"){if(p.push(["DateField","@pankod/refine-chakra-ui"]),t.multiple){let h=m("item",void 0,t.accessor);return g`
                    <Heading as="h5" size="sm" mt={4} >${y(t.key)}</Heading>
                    {${m(s,t.key)}?.map((item: any) => (
                        <DateField value={${h}} key={${h}} />
                    ))}
                `}return g`
                    <Heading as="h5" size="sm" mt={4} >${y(t.key)}</Heading>
                    <DateField value={${m(s,t.key,t.accessor," + ' ' + ")}} />
                `}},c=t=>{if(t.type==="richtext")return p.push(["MarkdownField","@pankod/refine-chakra-ui"]),g`
                    <Heading as="h5" size="sm" mt={4}>${y(t.key)}</Heading>
                    <MarkdownField value={${m(s,t.key,t.accessor,' + " " + ')}} />
                `},w=t=>{if(t.type==="number"){if(p.push(["NumberField","@pankod/refine-chakra-ui"],["TagField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),t.multiple){let h=m("item",void 0,t.accessor);return g`
                    <Heading as="h5" size="sm" mt={4} >${y(t.key)}</Heading>
                    <HStack spacing="12px">
                    {${m(s,t.key)}?.map((item: any) => (
                        <TagField value={${h}} key={${h}} />
                    ))}
                    </HStack>
                `}return g`
                    <Heading as="h5" size="sm" mt={4} >${y(t.key)}</Heading>
                    <NumberField value={${m(s,t.key,t.accessor,' + " " + ')} ?? ""} />
                `}},F=t=>{if(t)return g`
                    ${t}
                `},S=r.map(t=>{switch(t==null?void 0:t.type){case"text":return F(b(t));case"number":return F(w(t));case"richtext":return F(c(t));case"email":return F(k(t));case"image":return F(C(t));case"date":return F(d(t));case"boolean":return F(R(t));case"url":return F(i(t));case"relation":return F($(t));default:return}});return q(p),g`
        ${K(p)}
        
        export const ${u} = () => {
            const { queryResult } = useShow(${a?`{ 
                        resource: "${e.name}", 
                        id: ${o}
                    }`:""});
            const { data, isLoading } = queryResult;

            const ${s} = data?.data;
        
            ${l}

            return (
                <Show isLoading={isLoading}>
                    ${S.join("")}
                </Show>
            );
        };
        `}});var zr=v(require("@pankod/refine-chakra-ui")),Or=v(require("@pankod/refine-react-table")),jr=v(require("@tabler/icons"));var Y=e=>Array.isArray(e.accessor)||e.multiple?`accessorKey: "${e.key}"`:e.accessor?`accessorKey: "${B(e.key,void 0,e.accessor)}"`:`accessorKey: "${e.key}"`,ye=W({type:"list",additionalScope:[["@pankod/refine-chakra-ui","RefineChakraUI",zr],["@pankod/refine-react-table","RefineReactTable",Or],["@tabler/icons","TablerIcons",jr]],codeViewerComponent:J,loadingComponent:Q,errorComponent:Z,renderer:({resource:e,fields:r,isCustomPage:a})=>{var h;let o=P((h=e.label)!=null?h:e.name,"list"),u="tableData?.data",s=[["IResourceComponentsProps","@pankod/refine-core"],["useTable","@pankod/refine-react-table"],["ColumnDef","@pankod/refine-react-table"],["flexRender","@pankod/refine-react-table"],["List","@pankod/refine-chakra-ui"],["TableContainer","@pankod/refine-chakra-ui"],["Table","@pankod/refine-chakra-ui"],["Thead","@pankod/refine-chakra-ui"],["Tr","@pankod/refine-chakra-ui"],["Th","@pankod/refine-chakra-ui"],["Tbody","@pankod/refine-chakra-ui"],["Tr","@pankod/refine-chakra-ui"],["Td","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"],["Button","@pankod/refine-chakra-ui"],["IconButton","@pankod/refine-chakra-ui"],["usePagination","@pankod/refine-chakra-ui"],["Box","@pankod/refine-chakra-ui"],["IconChevronRight","@tabler/icons"],["IconChevronLeft","@tabler/icons"]],p=r.filter(n=>(n==null?void 0:n.relation)&&!(n!=null&&n.fieldable)&&(n==null?void 0:n.resource)),I=p.filter(Boolean).map(n=>{if((n==null?void 0:n.relation)&&!n.fieldable&&n.resource){s.push(["GetManyResponse","@pankod/refine-core"]),s.push(["useMany","@pankod/refine-core"]);let x="";return n.multiple?x=`[].concat(...(${u}?.map((item) => ${m("item",n.key,n.accessor,!1)}) ?? []))`:x=`${u}?.map((item) => ${m("item",n.key,n.accessor,!1)}) ?? []`,`
                    const { data: ${A(n.key,"Data")} } =
                    useMany({
                        resource: "${n.resource.name}",
                        ids: ${x},
                        queryOptions: {
                            enabled: !!${u},
                        },
                    });
                    `}}).filter(Boolean),l=p==null?void 0:p.map(n=>{if(n&&n.resource)return A(n.key,"Data")}).filter(Boolean),$=n=>{var x;if(n.relation&&n.resource){let L=`${A(n.key,"Data")}?.data`;if(Array.isArray(n.accessor))return;let E=`id: "${n.key}"`,T=`header: "${y(n.key)}"`,O=Y(n),U="";if(n.multiple){s.push(["TagField","@pankod/refine-chakra-ui"]);let ee="item";n!=null&&n.relationInfer&&(ee=m("item",void 0,n.relationInfer.accessor)),U=`cell: function render({ getValue, table }) {
                        const meta = table.options.meta as {
                            ${A(n.key,"Data")}: GetManyResponse;
                        };

                        const ${A(n.key,"")} = getValue<any[]>()?.map((item) => {
                            return meta.${A(n.key,"Data")}?.data?.find(
                                (resourceItems) => resourceItems.id === ${m("item",void 0,n.accessor)}
                            );
                        })


                        return (
                            <HStack>
                                {${A(n.key,"")}?.map((item, index) => (
                                    <TagField key={index} value={${ee}} />
                                ))}
                            </HStack>
                        )
                    }
                `}else n!=null&&n.relationInfer?U=`cell: function render({ getValue, table }) {
                            const meta = table.options.meta as {
                                ${A(n.key,"Data")}: GetManyResponse;
                            };

                            const ${A(n.key,"")} = meta.${L}?.find(
                                (item) => item.id === getValue<any>(),
                            );

                            return ${m(A(n.key),void 0,(x=n==null?void 0:n.relationInfer)==null?void 0:x.accessor)} ?? "Loading...";
                        },`:U="";return`
                    {
                        ${E},
                        ${T},
                        ${O},
                        ${U}
                    }
                `}},b=n=>{if(n.type==="image"){s.push(["Image","@pankod/refine-chakra-ui"]);let x=`id: "${n.key}"`,L=Y(n),E=`header: "${y(n.key)}"`,T=g`
                    cell: function render({ getValue }) {
                        ${n!=null&&n.accessor?"try {":""}
                            return <Image sx={{ maxWidth: "100px" }} src={${m("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0," + ")}} />
                        ${n!=null&&n.accessor?" } catch (error) { return null; }":""}
                    }
                `;if(n.multiple){let O=m("item",void 0,n.accessor," + ");T=`
                        cell: function render({ getValue }) {
                            ${n!=null&&n.accessor?"try {":""}
                                return (
                                    <HStack>
                                        {getValue<any[]>()?.map((item, index) => (
                                            <Image src={${O}} key={index} sx={{ height: "50px", maxWidth: "100px" }} />
                                        ))}
                                    </HStack>
                                )
                            ${n!=null&&n.accessor?" } catch (error) { return null; }":""}
                        }
                    `}return`
                    {
                        ${x},
                        ${L},
                        ${E},
                        ${T}
                    }
                `}},C=n=>{if(n.type==="email"){s.push(["EmailField","@pankod/refine-chakra-ui"]);let x=`id: "${n.key}"`,L=Y(n),E=`header: "${y(n.key)}"`,T=g`
                    cell: function render({ getValue }) {
                        return <EmailField value={${m("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0,' + " " + ')}} />
                    }
                `;return n.multiple&&(s.push(["TagField","@pankod/refine-chakra-ui"]),T=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${m("item",void 0,n.accessor," + ")}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),`
                    {
                        ${x},
                        ${L},
                        ${E},
                        ${T}
                    }
                `}},k=n=>{if(n.type==="url"){s.push(["UrlField","@pankod/refine-chakra-ui"]);let x=`id: "${n.key}"`,L=Y(n),E=`header: "${y(n.key)}"`,T=g`
                    cell: function render({ getValue }) {
                        return <UrlField value={${m("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0," + ")}} />
                    }
                `;return n.multiple&&(s.push(["TagField","@pankod/refine-chakra-ui"]),T=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${m("item",void 0,n.accessor," + ")}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),`
                    {
                        ${x},
                        ${L},
                        ${E},
                        ${T}
                    }
                `}},i=n=>{if((n==null?void 0:n.type)==="boolean"){s.push(["BooleanField","@pankod/refine-chakra-ui"]);let x=`id: "${n.key}"`,L=Y(n),E=`header: "${y(n.key)}"`,T=g`
                    cell: function render({ getValue }) {
                        return <BooleanField value={${m("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0," + ")}} />
                    }
                `;return n.multiple&&(T=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <BooleanField value={${m("item",void 0,n.accessor," + ")}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),`
                    {
                        ${x},
                        ${L},
                        ${E},
                        ${T}
                    }
                `}},R=n=>{if(n.type==="date"){s.push(["DateField","@pankod/refine-chakra-ui"]);let x=`id: "${n.key}"`,L=Y(n),E=`header: "${y(n.key)}"`,T=g`
                    cell: function render({ getValue }) {
                        return <DateField value={${m("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0,' + " " + ')}} />
                    }
                `;return n.multiple&&(T=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <DateField value={${m("item",void 0,n.accessor," + ")}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),`
                    {
                        ${x},
                        ${L},
                        ${E},
                        ${T}
                    }
                `}},d=n=>{if((n==null?void 0:n.type)==="richtext"){s.push(["MarkdownField","@pankod/refine-chakra-ui"]);let x=`id: "${n.key}"`,L=Y(n),E=`header: "${y(n.key)}"`,T=g`
                    cell: function render({ getValue }) {
                        return <MarkdownField value={(${m("getValue<string>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0)})?.slice(0, 80) + "..." } />
                    }
                `;return n.multiple&&(T=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<string[]>()?.map((item, index) => (
                                        <MarkdownField value={${m("item",void 0,n.accessor," + ")}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),`
                    {
                        ${x},
                        ${L},
                        ${E},
                        ${T}
                    }
                `}},c=n=>{if(n&&(n.type==="text"||n.type==="number")){let x=`id: "${n.key}"`,L=Y(n),E=`header: "${y(n.key)}"`,T="";return n.multiple&&(s.push(["TagField","@pankod/refine-chakra-ui"]),T=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${m("item",void 0,n.accessor,' + " " + ')}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),!n.multiple&&Array.isArray(n.accessor)&&(T=`
                        cell: function render({ getValue }) {
                            return (
                                <>{${m("getValue<any>()",n.key,n.accessor)}}</>
                            );
                        }
                    `),`
                    {
                        ${x},
                        ${L},
                        ${E},
                        ${T}
                    }
                `}},{canEdit:w,canShow:F,canDelete:S}=e!=null?e:{};w&&s.push(["EditButton","@pankod/refine-chakra-ui"]),F&&s.push(["ShowButton","@pankod/refine-chakra-ui"]),S&&s.push(["DeleteButton","@pankod/refine-chakra-ui"]);let V=w||F||S?g`
        {
            id: "actions",
            accessorKey: "id",
            header: "Actions",
            cell: function render({ getValue }) {
                return (
                    <HStack>
                    ${F?g`
                        <ShowButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `:""}
                        ${w?g`
                        <EditButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `:""}
                        ${S?g`
                        <DeleteButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `:""}
                    </HStack>
                );
            },
        },
            `:"",t=r.map(n=>{switch(n==null?void 0:n.type){case"text":case"number":return c(n);case"richtext":return d(n);case"email":return C(n);case"image":return b(n);case"date":return R(n);case"boolean":return i(n);case"url":return k(n);case"relation":return $(n);default:return}});return q(s),g`
        import React from "react";
        ${K(s)}
        
        export const ${o}: React.FC<IResourceComponentsProps> = () => {
            const columns = React.useMemo<ColumnDef<any>[]>(() => [
                ${[...t,V].filter(Boolean).join(",")}
            ], []);

            const {
                getHeaderGroups,
                getRowModel,
                setOptions,
                refineCore: {
                    setCurrent,
                    pageCount,
                    current,
                    tableQueryResult: { data: tableData },
                },
            } = useTable({
                columns,
                ${a?`
                refineCoreProps: {
                    resource: "${e.name}",
                }
                `:""}
                
            });

            ${I}

            setOptions((prev) => ({
                ...prev,
                meta: {
                    ...prev.meta,
                    ${l.join(", ")}
                },
            }));

            return (
                <List>
                    <TableContainer whiteSpace="pre-line">
                        <Table variant="simple">
                            <Thead>
                                {getHeaderGroups().map((headerGroup) => (
                                    <Tr key={headerGroup.id}>
                                        {headerGroup.headers.map((header) => (
                                            <Th key={header.id}>
                                                {!header.isPlaceholder && (
                                                    flexRender(
                                                        header.column.columnDef
                                                            .header,
                                                        header.getContext(),
                                                    )
                                                )}
                                            </Th>
                                        ))}
                                    </Tr>
                                ))}
                            </Thead>
                            <Tbody>
                                {getRowModel().rows.map((row) => (
                                    <Tr key={row.id}>
                                        {row.getVisibleCells().map((cell) => (
                                            <Td key={cell.id}>
                                                {flexRender(
                                                    cell.column.columnDef.cell,
                                                    cell.getContext(),
                                                )}
                                            </Td>
                                        ))}
                                    </Tr>
                                ))}
                            </Tbody>
                        </Table>
                    </TableContainer>
                    <Pagination
                        current={current}
                        pageCount={pageCount}
                        setCurrent={setCurrent}
                    />
                </List>   
            );
        };

        type PaginationProps = {
            current: number;
            pageCount: number;
            setCurrent: (page: number) => void;
        };

        const Pagination: React.FC<PaginationProps> = ({
            current,
            pageCount,
            setCurrent,
        }) => {
            const pagination = usePagination({
                current,
                pageCount,
            });
        
            return (
                <Box display="flex" justifyContent="flex-end">
                    <HStack my="3" spacing="1">
                        {pagination?.prev && (
                            <IconButton
                                aria-label="previous page"
                                onClick={() => setCurrent(current - 1)}
                                disabled={!pagination?.prev}
                                variant="outline"
                            >
                                <IconChevronLeft size="18" />
                            </IconButton>
                        )}
        
                        {pagination?.items.map((page) => {
                            if (typeof page === "string")
                                return <span key={page}>...</span>;
        
                            return (
                                <Button
                                    key={page}
                                    onClick={() => setCurrent(page)}
                                    variant={page === current ? "solid" : "outline"}
                                >
                                    {page}
                                </Button>
                            );
                        })}
                        {pagination?.next && (
                            <IconButton
                                aria-label="next page"
                                onClick={() => setCurrent(current + 1)}
                                variant="outline"
                            >
                                <IconChevronRight size="18" />
                            </IconButton>
                        )}
                    </HStack>
                </Box>
            );
        };
        `}});var Ur=v(require("@pankod/refine-core")),Kr=v(require("@pankod/refine-chakra-ui")),qr=v(require("@pankod/refine-react-hook-form"));var ke=W({type:"create",additionalScope:[["@pankod/refine-core","RefineCore",Ur],["@pankod/refine-chakra-ui","RefineChakraUI",Kr],["@pankod/refine-react-hook-form","RefineReactHookForm",qr]],codeViewerComponent:J,loadingComponent:Q,errorComponent:Z,renderer:({resource:e,fields:r,isCustomPage:a})=>{var k;let o=P((k=e.label)!=null?k:e.name,"create"),u=[["Create","@pankod/refine-chakra-ui"],["FormControl","@pankod/refine-chakra-ui"],["FormLabel","@pankod/refine-chakra-ui"],["FormErrorMessage","@pankod/refine-chakra-ui"],["useForm","@pankod/refine-react-hook-form"]],p=r.filter(i=>(i==null?void 0:i.relation)&&!(i!=null&&i.fieldable)&&(i==null?void 0:i.resource)).filter(Boolean).map(i=>{if((i==null?void 0:i.relation)&&!i.fieldable&&i.resource)return u.push(["useSelect","@pankod/refine-core"]),`
                    const { options: ${A(i.key,"Options")} } =
                    useSelect({
                        resource: "${i.resource.name}",
                        ${ie(i)}
                    });
                `}).filter(Boolean),I=i=>{if(i.relation&&i.resource){u.push(["useSelect","@pankod/refine-core"]),u.push(["Select","@pankod/refine-chakra-ui"]);let R=A(i.key,"Options");return g`
                <FormControl mb="3" isInvalid={!!errors?.${B(i.key,void 0)}}>
                    <FormLabel>${y(i.key)}</FormLabel>
                    <Select
                        placeholder="Select ${j(i.resource.name)}"
                        {...register("${B(i.key,void 0,i.accessor)}", {
                            required: "This field is required",
                        })}
                    >
                        {${R}?.map((option) => (
                            <option value={option.value} key={option.value}>
                                {option.label}
                            </option>
                        ))}
                    </Select>
                    <FormErrorMessage>
                        {${m("(errors as any)",i.key,i.accessor,!1)}?.message as string}
                    </FormErrorMessage>
                </FormControl>
                `}},l=i=>{if(i.type==="text"||i.type==="url"||i.type==="email"||i.type==="number"||i.type==="date"||i.type==="richtext")return se(i.key)||(u.push(["Input","@pankod/refine-chakra-ui"]),i.multiple)?void 0:g`
                    <FormControl mb="3" isInvalid={!!${m("(errors as any)",i.key,i.accessor,!1)}}>
                        <FormLabel>${y(i.key)}</FormLabel>
                        <Input
                            ${i.type!=="date"&&i.type!=="richtext"?`type="${i.type}"`:""}
                            {...register("${B(i.key,void 0,i.accessor)}", {
                                required: "This field is required",
                            })}
                        />
                        <FormErrorMessage>
                            {${m("(errors as any)",i.key,i.accessor,!1)}?.message as string}
                        </FormErrorMessage>
                    </FormControl>
                `},$=i=>{if(i.type==="boolean")return u.push(["Checkbox","@pankod/refine-chakra-ui"]),i.multiple?void 0:g`
                    <FormControl mb="3" isInvalid={!!${m("errors",i.key,i.accessor,!1)}}>
                        <FormLabel>${y(i.key)}</FormLabel>
                        <Checkbox
                            {...register("${B(i.key,void 0,i.accessor)}", {
                                required: "This field is required",
                            })}
                        />
                        <FormErrorMessage>
                            {${m("errors",i.key,i.accessor,!1)}?.message as string}
                        </FormErrorMessage>
                    </FormControl>
                   
                `},b=i=>{if(i.type==="date"){let R=l(i);return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-chakra-ui" package.
                        To use a <DatePicker> component, you can examine the following links:
                        
                        - https://github.com/aboveyunhai/chakra-dayzed-datepicker
                        - https://github.com/wojtekmaj/react-date-picker
                    */}
                    ${R!=null?R:""}
                    `}},C=r.map(i=>{switch(i==null?void 0:i.type){case"text":case"number":case"email":case"url":case"richtext":return l(i);case"date":return b(i);case"boolean":return $(i);case"relation":return I(i);default:return}});return q(u),g`
        ${K(u)}
        
        export const ${o} = () => {
            const {
                refineCore: { formLoading },
                saveButtonProps,
                register,
                formState: { errors },
            } = useForm(
                ${a?`
                { 
                    refineCoreProps: {
                        resource: "${e.name}",
                        action: "create",
                    }
                }`:""}
            );
        
            ${p}

            return (
                <Create isLoading={formLoading} saveButtonProps={saveButtonProps}>
                    ${C.join("")}
                </Create>
            );
        };
        `}});var _r=v(require("@pankod/refine-core")),Wr=v(require("@pankod/refine-chakra-ui")),Gr=v(require("@pankod/refine-react-hook-form"));var he=W({type:"edit",additionalScope:[["@pankod/refine-core","RefineCore",_r],["@pankod/refine-chakra-ui","RefineChakraUI",Wr],["@pankod/refine-react-hook-form","RefineReactHookForm",Gr]],codeViewerComponent:J,loadingComponent:Q,errorComponent:Z,renderer:({resource:e,fields:r,isCustomPage:a,id:o})=>{var R,d;let u=P((R=e.label)!=null?R:e.name,"edit"),s=A((d=e.label)!=null?d:e.name,"Data"),p=[["Edit","@pankod/refine-chakra-ui"],["FormControl","@pankod/refine-chakra-ui"],["FormLabel","@pankod/refine-chakra-ui"],["FormErrorMessage","@pankod/refine-chakra-ui"],["useForm","@pankod/refine-react-hook-form"]],l=r.filter(c=>(c==null?void 0:c.relation)&&!(c!=null&&c.fieldable)&&(c==null?void 0:c.resource)).filter(Boolean).map(c=>{if((c==null?void 0:c.relation)&&!c.fieldable&&c.resource){p.push(["useSelect","@pankod/refine-core"]);let w=m(s,c.key,c.accessor,!1);return c.multiple&&c.accessor&&(w=`${m(s,c.key)}?.map((item: any) => ${m("item",void 0,c.accessor)})`),`
                    const { options: ${A(c.key,"Options")} } =
                    useSelect({
                        resource: "${c.resource.name}",
                        defaultValue: ${w},
                        ${ie(c)}
                    });
                `}}).filter(Boolean),$=c=>{if(c.relation&&c.resource){p.push(["useSelect","@pankod/refine-core"]),p.push(["Select","@pankod/refine-chakra-ui"]);let w=A(c.key,"Options");return g`
                <FormControl mb="3" isInvalid={!!errors?.${B(c.key,void 0)}}>
                    <FormLabel>${y(c.key)}</FormLabel>
                    <Select
                        placeholder="Select ${j(c.resource.name)}"
                        {...register("${B(c.key,void 0,c.accessor)}", {
                            required: "This field is required",
                        })}
                    >
                        {${w}?.map((option) => (
                            <option value={option.value} key={option.value}>
                                {option.label}
                            </option>
                        ))}
                    </Select>
                    <FormErrorMessage>
                        {${m("(errors as any)",c.key,c.accessor,!1)}?.message as string}
                    </FormErrorMessage>
                </FormControl>
                `}},b=c=>{if(c.type==="text"||c.type==="url"||c.type==="email"||c.type==="number"||c.type==="date"||c.type==="richtext"){if(p.push(["Input","@pankod/refine-chakra-ui"]),c.multiple){let w=B(c.key,"${index}",c.accessor),F=m(m("(errors as any)",c.key,void 0,!1)+"?.[index]",void 0,c.accessor);return`
                        <>
                            {${m(s,c.key)}?.map((item: any, index: number) => (
                                <FormControl key={index} mb="3" isInvalid={!!${F}}>
                                    <FormLabel>${y(c.key)} #{index + 1}</FormLabel>
                                    <Input
                                        {...register(\`${w}\`, {
                                            required: "This field is required",
                                        })}
                                    />
                                    <FormErrorMessage>
                                        {${m(F,"message")} as string}
                                    </FormErrorMessage>
                                </FormControl>
                            ))}
                        </>
                    `}return g`
                    <FormControl mb="3" isInvalid={!!${m("(errors as any)",c.key,c.accessor,!1)}}>
                        <FormLabel>${y(c.key)}</FormLabel>
                        <Input
                            ${se(c.key)?"disabled":""}
                            ${c.type!=="date"&&c.type!=="richtext"?`type="${c.type}"`:""}
                            {...register("${B(c.key,void 0,c.accessor)}", {
                                required: "This field is required",
                            })}
                        />
                        <FormErrorMessage>
                            {${m("(errors as any)",c.key,c.accessor,!1)}?.message as string}
                        </FormErrorMessage>
                    </FormControl>
                `}},C=c=>{if(c.type==="boolean"){if(p.push(["Checkbox","@pankod/refine-chakra-ui"]),c.multiple){let w=B(c.key,void 0,c.accessor),F=m("(errors as any)",c.key,void 0,!1)+"?.[index]";return`
                        <>
                            {${m(s,c.key)}?.map((item: any, index: number) => (
                                <FormControl key={index} mb="3" isInvalid={!!${F}}>
                                    <FormLabel>${y(c.key)} #{index + 1}</FormLabel>
                                    <Checkbox
                                        {...register(\`${w}.\${index}\`, {
                                            required: "This field is required",
                                        })}
                                    />
                                    <FormErrorMessage>
                                        {${F}?.message as string}
                                    </FormErrorMessage>
                                </FormControl>
                            ))}
                        </>
                    `}return g`
                    <FormControl mb="3" isInvalid={!!${m("errors",c.key,c.accessor,!1)}}>
                        <FormLabel>${y(c.key)}</FormLabel>
                        <Checkbox
                            {...register("${B(c.key,void 0,c.accessor)}", {
                                required: "This field is required",
                            })}
                        />
                        <FormErrorMessage>
                            {${m("errors",c.key,c.accessor,!1)}?.message as string}
                        </FormErrorMessage>
                    </FormControl>
                   
                `}},k=c=>{if(c.type==="date")return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-chakra-ui" package.
                        To use a <DatePicker> component, you can examine the following links:
                        
                        - https://github.com/aboveyunhai/chakra-dayzed-datepicker
                        - https://github.com/wojtekmaj/react-date-picker
                    */}
                    ${b(c)}
                    `},i=r.map(c=>{switch(c==null?void 0:c.type){case"text":case"number":case"email":case"url":case"richtext":return b(c);case"date":return k(c);case"boolean":return C(c);case"relation":return $(c);default:return}});return q(p),g`
        ${K(p)}
        
        export const ${u} = () => {
            const {
                refineCore: { formLoading, queryResult },
                saveButtonProps,
                register,
                formState: { errors },
            } = useForm(
                ${a?`
                { 
                    refineCoreProps: {
                        resource: "${e.name}",
                        id: ${o},
                        action: "edit",
                    }
                }`:""}
            );
        
            const ${s} = queryResult?.data?.data;
        
            ${l}

            return (
                <Edit isLoading={formLoading} saveButtonProps={saveButtonProps}>
                    ${i.join("")}
                </Edit>
            );
        };
        `}});var Zr=({action:e,id:r,...a})=>{let{useParams:o}=(0,kr.useRouterContext)(),{action:u,id:s}=o();switch(e!=null?e:u){case"show":return ue.default.createElement(ge,{...a,id:r!=null?r:s});case"create":return ue.default.createElement(ke,{...a,id:r!=null?r:s});case"edit":return ue.default.createElement(he,{...a,id:r!=null?r:s});default:return ue.default.createElement(ye,{...a,id:r!=null?r:s})}};
