// @pankod/refine-inferencer@1.6.1/dist/index.js
var fn=Object.create;var Ge=Object.defineProperty;var xn=Object.getOwnPropertyDescriptor;var Fn=Object.getOwnPropertyNames;var bn=Object.getPrototypeOf,Cn=Object.prototype.hasOwnProperty;var vn=(n,c)=>{for(var k in c)Ge(n,k,{get:c[k],enumerable:!0})},Rr=(n,c,k,d)=>{if(c&&typeof c=="object"||typeof c=="function")for(let $ of Fn(c))!Cn.call(n,$)&&$!==k&&Ge(n,$,{get:()=>c[$],enumerable:!(d=xn(c,$))||d.enumerable});return n};var E=(n,c,k)=>(k=n!=null?fn(bn(n)):{},Rr(c||!n||!n.__esModule?Ge(k,"default",{value:n,enumerable:!0}):k,n)),Tn=n=>Rr(Ge({},"__esModule",{value:!0}),n);var gt={};vn(gt,{AntdCreateInferencer:()=>Ze,AntdEditInferencer:()=>Xe,AntdInferencer:()=>Un,AntdListInferencer:()=>_e,AntdShowInferencer:()=>Ke,ChakraUICreateInferencer:()=>lr,ChakraUIEditInferencer:()=>dr,ChakraUIInferencer:()=>kt,ChakraUIListInferencer:()=>mr,ChakraUIShowInferencer:()=>pr,MantineCreateInferencer:()=>rr,MantineEditInferencer:()=>nr,MantineInferencer:()=>Jn,MantineListInferencer:()=>er,MantineShowInferencer:()=>Ye,MuiCreateInferencer:()=>ir,MuiEditInferencer:()=>cr,MuiInferencer:()=>ot,MuiListInferencer:()=>sr,MuiShowInferencer:()=>ar,createInferencer:()=>D});module.exports=Tn(gt);var re=E(require("react")),Lr=require("@pankod/refine-core");var An=/(image|photo|avatar|url|logo|cover|thumbnail|icon)/i,yr=(n,c)=>{let k=["name","label","title","count","content","username","nickname","login","firstName","lastName","url"];An.test(n)&&k.unshift("url");let d=k.find($=>$ in c&&!!c[$]);if(d)return d==="firstName"&&"lastName"in c?["firstName","lastName"]:d};function kr(n,c,k){var $;if(c)return c;let d=k==null?void 0:k.find(o=>o.name===n);return($=d==null?void 0:d.options)!=null&&$.dataProviderName?d.options.dataProviderName:"default"}var Ne=n=>{var c;return(c=n==null?void 0:n.options)==null?void 0:c.dataProviderName};var gr=n=>{let c=[...n].filter(Boolean).sort((k,d)=>(d.priority||0)-(k.priority||0))[0];return c||null};var Br=E(require("prettier/standalone")),Mr=E(require("prettier/parser-typescript")),ae=n=>{try{return Br.default.format(n,{parser:"typescript",plugins:[Mr.default],arrowParens:"always",trailingComma:"all",semi:!0,tabWidth:4,printWidth:80})}catch(c){return console.warn("Formatting error in inferencer",c),n}};var $r=n=>n.replace(/ {2,}/g," ").replace(/\n{2,}/g,`
`);var hr=/(-id|-ids|_id|_ids|Id|Ids|ID|IDs)(\[\])?$/,Dr=(n,c)=>{let k=hr.test(n),d=typeof c=="string"||typeof c=="number",$=Array.isArray(c)&&c.every(o=>typeof o=="string"||typeof o=="number");return k&&d||k&&$?{key:n,relation:!0,multiple:!!$,type:"relation",priority:1}:!1};var X=n=>n.replace(hr,"");var g=n=>X(n).replace(/([a-z])([A-Z])/g,"$1 $2").replace(/_/g," ").replace(/-/g," ").split(" ").map(l=>l.charAt(0).toUpperCase()+l.slice(1)).join(" ");var Ir=n=>n.replace(/export\s+(const|let|var|type|interface|function|class)\s+(\w+)\s*(=|:)\s*/g,"$1 $2 = ").split(`
`).filter($=>!$.trim().startsWith("export default")).join(`
`);var Pr=/import(?:(?:(?:[ \n\t]+([^ *\n\t\{\},]+)[ \n\t]*(?:,|[ \n\t]+))?([ \n\t]*\{(?:[ \n\t]*[^ \n\t"'\{\}]+[ \n\t]*,?)+\})?[ \n\t]*)|[ \n\t]*\*[ \n\t]*as[ \n\t]+([^ \n\t\{\}]+)[ \n\t]+)from[ \n\t]*(?:['"])([^'"\n]+)(?:['"])(?:;?)/g,Sn=/import[ \n\t](?:['"])([^'"\n]+)(?:['"])(?:;?)/g,wn=/((?:\w|\s|_)*)( as )((?:\w|\s|_)*)( |,)?/g,fr=(n,c={})=>{let k=n.matchAll(Pr),d=new Set;for(let $ of k){let[,o,l,v,I]=$;if(I in c){let S=c[I];if(I==="react")continue;o&&d.add(`const { default: ${o} } = ${S};`),l&&d.add(`const${l.replace(wn,"$1: $3$4")} = ${S};`),v&&d.add(`const ${v} = ${S};`)}}return $r(`
    ${Array.from(d).join(`
`)}
    ${n.replace(Pr,"").replace(Sn,"")}
    `)};var Re=E(require("pluralize"));var xr=(n,c)=>n?c.find(d=>{let $=X(n.key);return d.name===$||d.name===(Re.default.isPlural($)?$:Re.default.plural($))||d.name===(Re.default.isSingular($)?$:Re.default.singular($))}):void 0;var Hr=E(require("pluralize"));var M=(n,c)=>{let k=n.replace(/[^a-zA-Z0-9]/g," "),d=Hr.default.singular(k),$=g(d).replace(/ /g,""),o=c.charAt(0).toUpperCase()+c.slice(1);return/^[a-zA-Z]/.test(n)?`${$}${o}`:`${o}${$}`};var Vn=n=>n?Array.isArray(n)?n.join(""):typeof n=="string"?n:typeof n=="object"||typeof n=="function"?n.toString():typeof n=="number"?String(n):typeof n=="boolean"&&n?"true":n:"",m=(n,...c)=>n.reduce((d,$,o)=>`${d}${$}${Vn(c[o])}`,"").trim();var Fr=(n,c)=>`
    ${n!=null?n:""}
    
    render(typeof ${c} !== "undefined" ? <${c} /> : <></>);
    `;var br=n=>n==null?void 0:n.replace(/(\/\* hidden-start \*\/)(.|\n)*?(\/\* hidden-end \*\/)/g,"");var Ln=/^[a-zA-Z_$][a-zA-Z_$0-9]*$/,Te=n=>Ln.test(n);function Cr(n,c,k){let d=`${n}`;return c&&(d+="?.",Te(c)?d+=c:d+=`['${c}']`),k&&(d+="?.",Te(k)?d+=k:d+=`['${k}']`),d}function En(n,c,k){return(k!=null?k:[]).map($=>Cr(n,c,$))}var s=(n,c,k,d=' + " " + ')=>Array.isArray(k)?d?En(n,c,k).join(d):Cr(n,c,k[0]):Cr(n,c,k),R=(n,c,k)=>{let d=n;return c&&(d+=`.${c}`),k&&(Array.isArray(k)?d+=`.${k[0]}`:d+=`.${k}`),d};var P=n=>{let c=n.reduce((d,[$,o,l])=>(d[o]||(d[o]=[]),d[o].includes($)||(l?d[o].push([$,!0]):d[o].push($)),d),{});return Object.entries(c).map(([d,$])=>{let o=$.find(S=>Array.isArray(S)&&S[1]),l=$.filter(S=>typeof S=="string"),v=o?`${o[0]}`:"",I=l.length?`{ ${l.join(", ")} }`:"";return`import ${v}${v&&I?", ":""}${I} from "${d}";`})};var vr=E(require("pluralize")),_=(n="")=>vr.default.isSingular(n)?n:vr.default.singular(n);var Tr=E(require("pluralize")),Ae=(n="")=>Tr.default.isPlural(n)?n:Tr.default.plural(n);var Nn=/^id$/i,j=n=>Nn.test(n);var ee=n=>n.relationInfer&&n.relationInfer.accessor?Array.isArray(n.relationInfer.accessor)?`optionLabel: "${n.relationInfer.accessor[0]}",`:n.relationInfer.accessor!=="title"?`optionLabel: "${n.relationInfer.accessor}",`:"":"";var H=n=>{};var f=(n,c="")=>{if(Number.isInteger(Number(n)))return c.charAt(0).toLowerCase()+c.slice(1)+n;let d=n.replace(/^\d+/,"").replace(/[^a-zA-Z0-9]/g,"-"),$=g(d).replace(/ /g,"");return`${$.charAt(0).toLowerCase()+$.slice(1)}${c.charAt(0).toUpperCase()+c.slice(1)}`};var Or=n=>{let c=(k,d,$,o=c)=>{let l=n.map(I=>I(k,d,$,o));return gr(l)};return c};var Gr=n=>(k,d,$,o,l)=>n.reduce((v,I)=>I(v,d,$,o,l),k);var jr=(n,c,k,d)=>{let $=Array.isArray(c),o=Array.isArray(c)&&c.every(l=>typeof l=="string"||typeof l=="number");if($){if(!o){let v=d(n,c[0],k,d);return v?{...v,key:n,multiple:!0,priority:1}:!1}let l=d(n,c[0],k,d)||{type:"string"};return{key:n,multiple:!0,priority:1,type:l==null?void 0:l.type}}return!1};var qr=(n,c)=>typeof c=="boolean"?{key:n,type:"boolean"}:!1;var Ar=E(require("dayjs")),Rn=/(_at|_on|At|On|AT|ON)(\[\])?$/,zr=(n,c)=>{let k=Rn.test(n)&&(0,Ar.default)(c).isValid(),d=typeof c=="string"&&(0,Ar.default)(c).isValid();return k||d?{key:n,type:"date",priority:1}:!1};var Bn=/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,Ur=(n,c)=>typeof c=="string"&&Bn.test(c)?{key:n,type:"email",priority:1}:!1;var Mn=/\.(gif|jpe?g|tiff?|png|webp|bmp|svg)$/i,Wr=(n,c)=>typeof c=="string"&&Mn.test(c)?{key:n,type:"image",priority:2}:!1;var Kr=(n,c)=>typeof c>"u"||c===null?null:!1;var _r=(n,c)=>typeof c=="number"?{key:n,type:"number"}:!1;var Dn=/id$/i,Zr=(n,c,k,d)=>{let $=c!==null,o=!Array.isArray(c);if($&&o&&typeof c=="object"){if(Object.keys(c).length===1&&Dn.test(Object.keys(c)[0]))return{key:n,type:"relation",relation:!0,accessor:"id",priority:1};let I=yr(n,c);if((I?I.length>0:!1)&&I){let V=I&&Array.isArray(I)?I[0]:I,L=d(V,c[V],c,d);if(L){let A=Array.isArray(I)?I.map(p=>L.accessor?`${p}.${Array.isArray(L.accessor)?L.accessor[0]:L.accessor}`:p):L.accessor?Array.isArray(L.accessor)?`${I}.${L.accessor[0]}`:`${I}.${L.accessor}`:I;return{...L,fieldable:!0,key:n,accessor:A,priority:1}}}return{key:n,fieldable:!1,type:"object",priority:1}}return!1};var Xr=(n,c)=>typeof c=="string"&&c.length>100?{key:n,type:"richtext",priority:1}:!1;var Jr=(n,c)=>typeof c=="string"?{key:n,type:"text"}:!1;var Pn=/^(https?|ftp):\/\/(-\.)?([^\s/?\.#-]+\.?)+(\/[^\s]*)?$/i,Qr=(n,c)=>typeof c=="string"&&Pn.test(c)?{key:n,type:"url",priority:1}:!1;var Yr=[jr,qr,zr,Ur,Wr,Kr,_r,Zr,Dr,Xr,Jr,Qr];var en=(n,c,k,d)=>n.map(o=>{if(!o.relation&&(o.type==="text"||o.type==="richtext"||o.type==="number")&&!o.canRelation){let l=/^[a-z0-9-]+$/,v=A=>l.test(`${A}`),I=o.key.toLowerCase()!=="id";return!(!o.accessor||typeof o.accessor=="string")||!I?o:(o.multiple?d[o.key]:[d[o.key]]).every(A=>v(o.accessor?A[o.accessor]:A))?{...o,canRelation:!0}:o}return o});var Hn=/(image|photo|avatar|cover|thumbnail|icon)/i,rn=n=>n.map(k=>k.type==="url"&&Hn.test(k.key)?{...k,type:"image"}:k);var nn=(n,c)=>n.map(d=>{let $=xr(d,c);return $?{...d,relation:!0,type:"relation",resource:$}:d});var tn=(n,c,k,d,$)=>n.map(l=>{if(l.relation&&l.type==="relation"&&!l.resource){let v=l.accessor?d[l.key][l.accessor]:d[l.key],I=$(l.key,v,d,$);if(I&&I.type!=="relation")return{...l,fieldable:!0,relation:!1,type:I.type,canRelation:!0}}return l});var on=[rn,nn,tn,en];var ce=E(require("react")),je=E(require("prism-react-renderer")),an=E(require("prism-react-renderer/themes/vsDark")),se=({code:n,wrapperProps:c})=>((0,ce.useEffect)(()=>{var d;let k=document.createElement("style");k.appendChild(document.createTextNode(`
                #code-highlight::-webkit-scrollbar {-webkit-appearance: none; height: 7px; width: 7px;}
                #code-highlight::-webkit-scrollbar-thumb {background-color: #666b7a;}
                #code-highlight::-webkit-scrollbar-corner {background-color: rgb(30, 30, 30);}
                `)),(d=document.getElementById("code-highlight"))==null||d.appendChild(k)},[]),ce.default.createElement("div",{id:"code-highlight",style:{maxHeight:"75vh",height:"100%",overflow:"auto",backgroundColor:"rgb(30, 30, 30)"},...c!=null?c:{}},ce.default.createElement(je.default,{...je.defaultProps,theme:an.default,code:n,language:"tsx"},({className:k,style:d,tokens:$,getLineProps:o,getTokenProps:l})=>ce.default.createElement("pre",{className:k,style:{...d,padding:"14px 14px 14px 14px",margin:"0",width:"100%",boxSizing:"border-box"}},$.map((v,I)=>ce.default.createElement("div",{key:I,...o({line:v,key:I})},v.map((S,V)=>ce.default.createElement("span",{key:V,...l({token:S,key:V})}))))))));var J=E(require("react")),On=E(require("@pankod/refine-core")),qe=require("@aliemir/react-live");var sn=[["react","React",J.default],["@pankod/refine-core","RefineCore",On]],Sr=J.default.createContext({}),wr=({fetchError:n,code:c,additionalScope:k,errorComponent:d})=>{var v;let $=(v=c&&Ir(fr(c,[...sn,...k!=null?k:[]].reduce((I,[S,V,L,A])=>A?I:{...I,[S]:V},{}))))!=null?v:"",o=J.default.useMemo(()=>({React:J.default,...J.default,...[...sn,...k!=null?k:[]].reduce((I,[S,V,L])=>({...I,[V]:L}),{})}),[k]),l=J.default.useMemo(()=>()=>{let{error:S}=J.default.useContext(Sr);return d?J.default.createElement(d,{error:S!=null?S:n?"Something went wrong while fetching the resource data.":void 0}):null},[d,n]);return J.default.createElement(qe.LiveProvider,{Context:Sr,code:$,scope:o,noInline:!0},!n&&J.default.createElement(qe.LivePreview,{Context:Sr}),J.default.createElement(l,null))};var Se=E(require("react")),ze=require("@pankod/refine-core");var cn=(n,c,k)=>{let{resource:d,resourceName:$,id:o,resources:l}=(0,ze.useResource)({resourceNameOrRouteName:c}),v=k!=null?k:o,I=(0,ze.useDataProvider)(),[S,V]=Se.default.useState(void 0),[L,A]=Se.default.useState(!0),[p,N]=Se.default.useState(!1),y=Se.default.useCallback(async u=>{var T,t;let w=(T=Ne(d))!=null?T:kr($,void 0,l),a=I(w);N(!0);try{if(n==="list"||n==="create"){let r=await a.getList({resource:$});V((t=r.data)==null?void 0:t[0]),setTimeout(()=>{N(!1)},500)}if(n==="edit"||n==="show"&&u){let r=await a.getOne({resource:$,id:u});V(r.data),setTimeout(()=>{N(!1)},500)}}catch{setTimeout(()=>{N(!1)},500)}},[n,I,d,$,l]);return Se.default.useEffect(()=>{A(!1),!p&&!S&&y(v)},[y,v]),{data:S,loading:p,initial:L}};var we=E(require("react")),un=require("@pankod/refine-core");var Vr=require("lodash"),pn=({record:n,fields:c,infer:k})=>{let d=(0,un.useDataProvider)(),[$,o]=we.default.useState([]),[l,v]=we.default.useState(!0),[I,S]=we.default.useState(!1),V=we.default.useCallback(async L=>{console.groupCollapsed("@pankod/refine-inferencer is trying to detect relations");let A=[];S(!0);try{let p=L.map(async y=>{if(y&&(y.relation||y.canRelation)&&n){let u=Ne(y.resource),w=d(u),a=Array.isArray(y.accessor)?void 0:y.multiple?n[y.key].map(T=>y.accessor?(0,Vr.get)(T,y.accessor):T)[0]:y.accessor?(0,Vr.get)(n[y.key],y.accessor):n[y.key];if(a&&y.resource)try{let{data:T}=await w.getOne({resource:y.resource.name,id:a});A.push({status:"success",resource:y.resource.name,field:y.key});let t=k("__",T,{},k);return{...y,relationInfer:t}}catch{return A.push({status:"error",resource:y.resource.name,field:y.key}),{...y,relationInfer:null}}if(a){let T,t;try{let{data:e}=await w.getOne({resource:Ae(X(y.key)),id:a});A.push({status:"success",resource:Ae(X(y.key)),field:y.key}),T=e,t=!0}catch{A.push({status:"error",resource:Ae(X(y.key)),field:y.key});try{let{data:h}=await w.getOne({resource:_(X(y.key)),id:a});A.push({status:"success",resource:_(X(y.key)),field:y.key}),T=h,t=!1}catch{return A.push({status:"error",resource:_(X(y.key)),field:y.key}),{...y,relationInfer:null}}}let r=k("__",T,{},k),i=X(y.key);return{...y,relation:!0,type:"relation",resource:{name:t?Ae(i):_(i)},fieldable:!1,canRelation:void 0,relationInfer:r}}return{...y,relationInfer:null}}return y}),N=await Promise.all(p);o(N.filter(y=>y)),setTimeout(()=>{S(!1)},500)}catch{setTimeout(()=>{S(!1)},500)}setTimeout(()=>{console.log(`Tried to detect relations with ${A.length} attempts and succeeded with ${A.filter(p=>p.status==="success").length} attempts.`),console.groupEnd()},500)},[d,n]);return we.default.useEffect(()=>{v(!1),!I&&c&&c.length>0&&$.length===0&&V(c)},[V,I,c,V]),{fields:$,loading:I,initial:l}};var D=({type:n,additionalScope:c=[],customElements:k=[],fieldTransformers:d=[],renderer:$,loadingComponent:o,errorComponent:l,codeViewerComponent:v})=>{let I=Or([...Yr,...k]),S=Gr([...on,...d]),V=({resourceName:A,fieldTransformer:p,id:N})=>{var x;let{resource:y,resources:u}=(0,Lr.useResource)({resourceNameOrRouteName:A}),{resource:w}=(0,Lr.useResource)(),{data:a,loading:T,initial:t}=cn(n,A!=null?A:y==null?void 0:y.name,N),r=re.default.useMemo(()=>{if(a){let C=Object.keys(a).map(W=>{let K=a[W];return I(W,K,a,I)}).filter(Boolean),b=S(C,u,y,a,I);return p?b.flatMap(W=>{let K=p(W);return K?[K]:[]}):b}return[]},[a,u,y,p]),{fields:i,loading:e}=pn({record:a,fields:r,infer:I}),h=re.default.useMemo(()=>!T&&!e&&!t?$({resource:y,resources:u,fields:i,infer:I,isCustomPage:y.name!==w.name,id:N}):"",[y,u,i,T,e]);return re.default.createElement(re.default.Fragment,null,o&&(T||e)&&re.default.createElement(o,null),!T&&!e&&re.default.createElement(re.default.Fragment,null,re.default.createElement(wr,{fetchError:!T&&!t&&!a,code:Fr(h,M((x=y.label)!=null?x:y.name,n)),errorComponent:l,additionalScope:c}),v&&re.default.createElement(v,{code:br(h),loading:T||e})))};return({name:A,resource:p,fieldTransformer:N,id:y})=>re.default.createElement(V,{fieldTransformer:N,resourceName:p!=null?p:A,key:p!=null?p:A,id:y})};var Be=E(require("react")),dn=require("@pankod/refine-core");var Gn=E(require("@pankod/refine-antd"));var Ue=E(require("react")),Ve=require("@pankod/refine-antd"),ue=({error:n})=>n?Ue.default.createElement(Ve.Row,{justify:"center",align:"middle",style:{minHeight:"200px"}},Ue.default.createElement(Ve.Col,null,Ue.default.createElement(Ve.Alert,{message:"Error",description:n!=null?n:"",type:"error",showIcon:!0}))):null;var We=E(require("react")),Le=require("@pankod/refine-antd"),pe=()=>We.default.createElement(Le.Row,{justify:"center",align:"middle",style:{minHeight:"300px"}},We.default.createElement(Le.Col,null,We.default.createElement(Le.Spin,{size:"large"})));var z=E(require("react")),Z=require("@pankod/refine-antd");var me=({code:n,loading:c})=>{let k=z.default.useMemo(()=>ae(n!=null?n:""),[n]),d=z.default.useRef(null),[$,o]=z.default.useState(!1),[l,v]=z.default.useState(!1);if(c)return null;let I=()=>{var S,V,L,A;(S=d==null?void 0:d.current)==null||S.select(),(V=d==null?void 0:d.current)==null||V.setSelectionRange(0,Number.MAX_SAFE_INTEGER),typeof navigator<"u"&&(navigator.clipboard.writeText((A=(L=d==null?void 0:d.current)==null?void 0:L.value)!=null?A:""),v(!0),setTimeout(()=>{v(!1)},1e3))};return k?z.default.createElement(z.default.Fragment,null,z.default.createElement("div",{style:{position:"fixed",bottom:12,right:60}},z.default.createElement(Z.Space,null,z.default.createElement(Z.Button,{type:"primary",onClick:()=>o(!0),icon:z.default.createElement(Z.Icons.CodeOutlined,null)},"Show Code"),z.default.createElement(Z.Button,{type:"default",target:"_blank",href:"https://github.com/refinedev/refine/discussions/3046",icon:z.default.createElement(Z.Icons.MessageOutlined,null)},"Give Feedback"))),z.default.createElement("div",{style:{height:"50px"}}),z.default.createElement(Z.Modal,{width:800,visible:$,onCancel:()=>o(!1),closeIcon:z.default.createElement(Z.Icons.CloseOutlined,{style:{color:"#666b7a"}}),footer:null,bodyStyle:{padding:0}},z.default.createElement(se,{code:k}),z.default.createElement("textarea",{ref:d,defaultValue:(k!=null?k:"").replace(/\\n/g,`\r
`),id:"code-input","aria-hidden":"true",style:{padding:0,margin:0,width:0,height:0,opacity:0,border:"none",display:"block"}}),z.default.createElement(Z.Button,{style:{position:"absolute",bottom:12,right:12,width:100},key:"copy",type:"default",icon:l?z.default.createElement(Z.Icons.CheckOutlined,{style:{color:"green"}}):z.default.createElement(Z.Icons.CopyOutlined,null),onClick:I},l?"Copied":"Copy"))):null};var Ke=D({type:"show",additionalScope:[["@pankod/refine-antd","RefineAntd",Gn]],codeViewerComponent:me,loadingComponent:pe,errorComponent:ue,renderer:({resource:n,fields:c,isCustomPage:k,id:d})=>{var T;let $=M((T=n.label)!=null?T:n.name,"show"),o="record",l=[["React","react",!0],["IResourceComponentsProps","@pankod/refine-core"],["useShow","@pankod/refine-core"],["Show","@pankod/refine-antd"],["Typography","@pankod/refine-antd"]],I=c.filter(t=>(t==null?void 0:t.relation)&&!(t!=null&&t.fieldable)&&(t==null?void 0:t.resource)).filter(Boolean).map(t=>{if((t==null?void 0:t.relation)&&!t.fieldable&&t.resource){if(t.multiple){l.push(["useMany","@pankod/refine-core"]);let r=s(o,t.key);return t.accessor&&(r=`${s(o,t.key)}?.map((item: any) => ${s("item",void 0,t.accessor)})`),`
                    const { data: ${f(t.key,"Data")}, isLoading: ${f(t.key,"IsLoading")} } =
                    useMany({
                        resource: "${t.resource.name}",
                        ids: ${r} || [],
                        queryOptions: {
                            enabled: !!${o},
                        },
                    });
                    `}return l.push(["useOne","@pankod/refine-core"]),`
                    const { data: ${f(t.key,"Data")}, isLoading: ${f(t.key,"IsLoading")} } =
                    useOne({
                        resource: "${t.resource.name}",
                        id: ${s(o,t.key,t.accessor,!1)} || "",
                        queryOptions: {
                            enabled: !!${o},
                        },
                    });
                `}}).filter(Boolean),S=t=>{if(t.relation&&t.resource){let r=f(t.key,"Data"),i=f(t.key,"IsLoading");return t.multiple?(l.push(["TagField","@pankod/refine-antd"]),m`
                    <Title level={5}>${g(t.key)}</Title>
                    {${i} ? <>Loading...</> : (
                        <>
                        ${(()=>{var e,h;if(t.relationInfer)if((e=t.relationInfer)!=null&&e.accessor){if(Array.isArray(t.relationInfer.accessor))return"Not Handled.";{let x=_((h=t.resource)==null?void 0:h.name),C=s(x,void 0,t.relationInfer.accessor);return`{${r}?.data?.map((${x}: any) => <TagField key={${C}} value={${C}} />)}`}}else return;else return})()}
                        </>
                    )}
                    `):m`
                    <Title level={5}>${g(t.key)}</Title>
                    {${i} ? <>Loading...</> : (
                        <>
                        ${(()=>{var e;return t.relationInfer?(e=t.relationInfer)!=null&&e.accessor?Array.isArray(t.relationInfer.accessor)?`{${s(`${r}?.data`,void 0,t.relationInfer.accessor,' + " " + ')}}`:`{${r}?.data?.${t.relationInfer.accessor}}`:`{${r}?.data}`:`{${r}?.data?.id}`})()}
                        </>
                    )}
                    
                    `}},V=t=>{if(t.type==="text"){if(l.push(["TagField","@pankod/refine-antd"],["TextField","@pankod/refine-antd"]),t.multiple){let r=s("item",void 0,t.accessor);return m`
                    <Title level={5}>${g(t.key)}</Title>
                    {${s(o,t.key)}?.map((item: any) => (
                        <TagField value={${r}} key={${r}} />
                    ))}
                `}return m`
                    <Title level={5}>${g(t.key)}</Title>
                    <TextField value={${s(o,t.key,t.accessor)}} />
                `}},L=t=>{if(t.type==="image"){if(l.push(["ImageField","@pankod/refine-antd"]),t.multiple){let r=s("item",void 0,t.accessor);return m`
                    <Title level={5}>${g(t.key)}</Title>
                    {${s(o,t.key)}?.map((item: any) => (
                        <ImageField style={{ maxWidth: 200 }} value={${r}} key={${r}} />
                    ))}
                `}return m`
                    <Title level={5}>${g(t.key)}</Title>
                    <ImageField style={{ maxWidth: 200 }} value={${s(o,t.key,t.accessor," + ")}} />
                    `}},A=t=>{if(t.type==="email"){if(l.push(["TagField","@pankod/refine-antd"],["EmailField","@pankod/refine-antd"]),t.multiple){let r=s("item",void 0,t.accessor);return m`
                    <Title level={5}>${g(t.key)}</Title>
                    <>
                    {${s(o,t.key)}?.map((item: any) => (
                        <TagField value={${r}} key={${r}} />
                    ))}
                    </>
                `}return m`
                    <Title level={5}>${g(t.key)}</Title>
                    <EmailField value={${s(o,t.key,t.accessor," + ")}} />
                `}},p=t=>{if(t.type==="url"){if(l.push(["TagField","@pankod/refine-antd"],["UrlField","@pankod/refine-antd"]),t.multiple){let r=s("item",void 0,t.accessor);return m`
                    <Title level={5}>${g(t.key)}</Title>
                    {${s(o,t.key)}?.map((item: any) => (
                        <TagField value={${r}} key={${r}} />
                    ))}
                `}return m`
                    <Title level={5}>${g(t.key)}</Title>
                    <UrlField value={${s(o,t.key,t.accessor," + ")}} />
                `}},N=t=>{if(t.type==="boolean"){if(l.push(["TagField","@pankod/refine-antd"],["BooleanField","@pankod/refine-antd"]),t.multiple){let r=s("item",void 0,t.accessor);return m`
                    <Title level={5}>${g(t.key)}</Title>
                    {(${s(o,t.key)} as any[])?.map((item, index) => (
                        <BooleanField value={${r}} key={index} />
                    ))}
                `}return m`
                    <Title level={5}>${g(t.key)}</Title>
                    <BooleanField value={${s(o,t.key,t.accessor," && ")}} />
                `}},y=t=>{if(t.type==="date"){if(l.push(["DateField","@pankod/refine-antd"]),t.multiple){let r=s("item",void 0,t.accessor);return m`
                    <Title level={5}>${g(t.key)}</Title>
                    {${s(o,t.key)}?.map((item: any) => (
                        <DateField value={${r}} key={${r}} />
                    ))}
                `}return m`
                    <Title level={5}>${g(t.key)}</Title>
                    <DateField value={${s(o,t.key,t.accessor," + ' ' + ")}} />
                `}},u=t=>{if(t.type==="richtext")return l.push(["MarkdownField","@pankod/refine-antd"]),m`
                    <Title level={5}>${g(t.key)}</Title>
                    <MarkdownField value={${s(o,t.key,t.accessor,' + " " + ')}} />
                `},w=t=>{if(t.type==="number"){if(l.push(["NumberField","@pankod/refine-antd"]),t.multiple){let r=s("item",void 0,t.accessor);return m`
                    <Title level={5}>${g(t.key)}</Title>
                    {${s(o,t.key)}?.map((item: any) => (
                        <TagField value={${r}} key={${r}} />
                    ))}
                `}return m`
                    <Title level={5}>${g(t.key)}</Title>
                    <NumberField value={${s(o,t.key,t.accessor,' + " " + ')} ?? ""} />
                `}},a=c.map(t=>{switch(t==null?void 0:t.type){case"text":return V(t);case"number":return w(t);case"richtext":return u(t);case"email":return A(t);case"image":return L(t);case"date":return y(t);case"boolean":return N(t);case"url":return p(t);case"relation":return S(t);default:return}});return H(l),m`
        ${P(l)}
        
        const { Title } = Typography;

        export const ${$}: React.FC<IResourceComponentsProps> = () => {
            const { queryResult } = useShow(${k?`{ 
                        resource: "${n.name}", 
                        id: ${d}
                    }`:""});
            const { data, isLoading } = queryResult;
        
            const ${o} = data?.data;

            ${I}

            return (
                <Show isLoading={isLoading}>
                    ${a.join("")}
                </Show>
            );
        };
        `}});var jn=E(require("@pankod/refine-antd"));var _e=D({type:"list",additionalScope:[["@pankod/refine-antd","RefineAntd",jn]],codeViewerComponent:me,loadingComponent:pe,errorComponent:ue,renderer:({resource:n,fields:c,isCustomPage:k})=>{var r;let d=M((r=n.label)!=null?r:n.name,"list"),$="tableProps?.dataSource",o=[["React","react",!0],["IResourceComponentsProps","@pankod/refine-core"],["BaseRecord","@pankod/refine-core"],["useTable","@pankod/refine-antd"],["List","@pankod/refine-antd"],["Table","@pankod/refine-antd"],["Space","@pankod/refine-antd"]],v=c.filter(i=>(i==null?void 0:i.relation)&&!(i!=null&&i.fieldable)&&(i==null?void 0:i.resource)).filter(Boolean).map(i=>{if((i==null?void 0:i.relation)&&!i.fieldable&&i.resource){o.push(["useMany","@pankod/refine-core"]);let e="";return i.multiple?e=`[].concat(...(${$}?.map((item) => ${s("item",i.key,i.accessor,!1)}) ?? []))`:e=`${$}?.map((item) => ${s("item",i.key,i.accessor,!1)}) ?? []`,`
                    const { data: ${f(i.key,"Data")}, isLoading: ${f(i.key,"IsLoading")} } =
                    useMany({
                        resource: "${i.resource.name}",
                        ids: ${e},
                        queryOptions: {
                            enabled: !!${$},
                        },
                    });
                    `}}).filter(Boolean),I=i=>{var e,h;if(i.relation&&i.resource){let C=`${f(i.key,"Data")}?.data`,b=f(i.key,"IsLoading");if(Array.isArray(i.accessor))return;let B=`${b} ? <>Loading...</> : `,W=i.multiple?`dataIndex="${i.key}"`:`dataIndex={["${i.key}", ${i.accessor?`"${i.accessor}"`:""}]}`,K=`title="${g(i.key)}"`,ve="";if(i.multiple){o.push(["TagField","@pankod/refine-antd"]);let oe="item";if(i!=null&&i.relationInfer){let ie=`${C}?.find((resourceItems) => resourceItems.id === ${s("item",void 0,i.accessor)})`;oe=s(ie,void 0,(e=i==null?void 0:i.relationInfer)==null?void 0:e.accessor)}ve=`render={(value: any[]) => ${B} (
                        <>
                            {${s("value",void 0)}?.map((item, index) => (
                                <TagField key={index} value={${oe}} />
                            ))}
                        </>
                    )}`}else if(i!=null&&i.relationInfer){let oe=`${C}?.find((item) => item.id === value)`,ie=s(oe,void 0,(h=i==null?void 0:i.relationInfer)==null?void 0:h.accessor);ve=`render={(value) => ${B} ${ie}}`}else ve="";return m`<Table.Column ${W} ${K} ${ve} />`}},S=i=>{if(i.type==="image"){o.push(["ImageField","@pankod/refine-antd"]);let e=Array.isArray(i.accessor)||i.multiple?`dataIndex="${i.key}"`:`dataIndex={["${i.key}", ${i.accessor?`"${i.accessor}"`:""}]}`,h=`title="${g(i.key)}"`,x=m`render={(value: any) => <ImageField style={{ maxWidth: "100px" }} value={${s("value",void 0,Array.isArray(i.accessor)?i.accessor:void 0," + ")}} />}`;if(i.multiple){let C=s("item",void 0,i.accessor," + ");x=m`render={(value: any[]) => (<>{value?.map((item, index) => (
                        <ImageField style={{ maxWidth: "100px" }} value={${C}} key={index} />
                    ))}</>)}`}return m`<Table.Column ${e} ${h} ${x} />`}},V=i=>{if(i.type==="email"){o.push(["TagField","@pankod/refine-antd"],["EmailField","@pankod/refine-antd"]);let e=Array.isArray(i.accessor)||i.multiple?`dataIndex="${i.key}"`:`dataIndex={["${i.key}", ${i.accessor?`"${i.accessor}"`:""}]}`,h=`title="${g(i.key)}"`,x=m`render={(value: any) => <EmailField value={${s("value",void 0,Array.isArray(i.accessor)?i.accessor:void 0,' + " " + ')}} />}`;if(i.multiple){let C=s("item",void 0,i.accessor,' + " " + ');x=m`render={(value: any[]) => (<>{value?.map((item, index) => (
                        <TagField value={${C}} key={index} />
                    ))}</>)}`}return m`<Table.Column ${e} ${h} ${x} />`}},L=i=>{if(i.type==="url"){o.push(["UrlField","@pankod/refine-antd"],["TagField","@pankod/refine-antd"]);let e=Array.isArray(i.accessor)||i.multiple?`dataIndex="${i.key}"`:`dataIndex={["${i.key}", ${i.accessor?`"${i.accessor}"`:""}]}`,h=`title="${g(i.key)}"`,x=m`render={(value: any) => <UrlField value={${s("value",void 0,Array.isArray(i.accessor)?i.accessor:void 0," + ")}} />}`;if(i.multiple){let C=s("item",void 0,i.accessor," + ");x=m`render={(value: any[]) => (<>{value?.map((item, index) => (
                        <TagField value={${C}} key={index} />
                    ))}</>)}`}return m`<Table.Column ${e} ${h} ${x} />`}},A=i=>{if((i==null?void 0:i.type)==="boolean"){o.push(["BooleanField","@pankod/refine-antd"]);let e=Array.isArray(i.accessor)||i.multiple?`dataIndex="${i.key}"`:`dataIndex={["${i.key}", ${i.accessor?`"${i.accessor}"`:""}]}`,h=`title="${g(i.key)}"`,x=m`render={(value: any) => <BooleanField value={${s("value",void 0,Array.isArray(i.accessor)?i.accessor:void 0," && ")}} />}`;if(i.multiple){let C=s("item",void 0,i.accessor," && ");x=m`render={(value: any[]) => (<>{value?.map((item, index) => (
                        <BooleanField value={${C}} key={index} />
                    ))}</>)}`}return m`<Table.Column ${e} ${h} ${x} />`}},p=i=>{if(i.type==="date"){o.push(["DateField","@pankod/refine-antd"]);let e=Array.isArray(i.accessor)||i.multiple?`dataIndex="${i.key}"`:`dataIndex={["${i.key}", ${i.accessor?`"${i.accessor}"`:""}]}`,h=`title="${g(i.key)}"`,x=m`render={(value: any) => <DateField value={${s("value",void 0,Array.isArray(i.accessor)?i.accessor:void 0,' + " " + ')}} />}`;if(i.multiple){let C=s("item",void 0,i.accessor,' + " " + ');x=m`render={(value: any[]) => (<>{value?.map((item, index) => (
                        <DateField value={${C}} key={index} />
                    ))}</>)}`}return m`<Table.Column ${e} ${h} ${x} />`}},N=i=>{if((i==null?void 0:i.type)==="richtext"){o.push(["MarkdownField","@pankod/refine-antd"]);let e=Array.isArray(i.accessor)||i.multiple?`dataIndex="${i.key}"`:i.accessor?`dataIndex={["${i.key}", "${i.accessor}"]}`:`dataIndex="${i.key}"`,h=`title="${g(i.key)}"`,x=m`render={(value: any) => <MarkdownField value={(${s("value",void 0,Array.isArray(i.accessor)?i.accessor:void 0)}).slice(0, 80) + "..."} />}`;if(i.multiple){let C=s("item",void 0,i.accessor,' + " " + ');x=m`render={(value: any[]) => (<>{value?.map((item, index) => (
                        <MarkdownField value={(${C})?.slice(0, 80) + "..."} key={index} />
                    ))}</>)}`}return m`<Table.Column ${e} ${h} ${x} />`}},y=i=>{if(i&&(i.type==="text"||i.type==="number")){let e=i.accessor&&!Array.isArray(i.accessor)&&!i.multiple?`dataIndex={["${i.key}", "${i.accessor}"]}`:`dataIndex="${i.key}"`,h=`title="${g(i.key)}"`,x="";if(i.multiple){o.push(["TagField","@pankod/refine-antd"]);let C=s("item",void 0,i.accessor,' + " " + ');x=`render={(value: any[]) => (<>{value?.map((item) => (
                        <TagField value={${C}} key={${C}} />
                    ))}</>)}`}return!i.multiple&&Array.isArray(i.accessor)&&(x=`render={(value: any) => (<>{${s("value",void 0,i.accessor)}}</>)}`),`<Table.Column ${e} ${h} ${x} />`}},{canEdit:u,canShow:w,canDelete:a}=n!=null?n:{};u&&o.push(["EditButton","@pankod/refine-antd"]),w&&o.push(["ShowButton","@pankod/refine-antd"]),a&&o.push(["DeleteButton","@pankod/refine-antd"]);let T=u||w||a?m`
                <Table.Column
                    title="Actions"
                    dataIndex="actions"
                    render={(_, record: BaseRecord) => (
                        <Space>
                        ${u?m`
                            <EditButton
                                hideText
                                size="small"
                                recordItemId={record.id}
                            />
                            `:""}
                        ${w?m`
                            <ShowButton
                                hideText
                                size="small"
                                recordItemId={record.id}
                            />
                            `:""}
                        ${a?m`
                            <DeleteButton
                                hideText
                                size="small"
                                recordItemId={record.id}
                            />
                            `:""}
                        </Space>
                    )}
                />
            `:"",t=c.map(i=>{switch(i==null?void 0:i.type){case"text":case"number":return y(i);case"richtext":return N(i);case"email":return V(i);case"image":return S(i);case"date":return p(i);case"boolean":return A(i);case"url":return L(i);case"relation":return I(i);default:return}});return H(o),m`
        ${P(o)}
        
        export const ${d}: React.FC<IResourceComponentsProps> = () => {
            const { tableProps } = useTable({
                syncWithLocation: true,
                ${k?` resource: "${n.name}",`:""}
            });
        
            ${v}

            return (
                <List>
                    <Table {...tableProps} rowKey="id">
                        ${t.join(`\r
`)}
                        ${T}
                    </Table>
                </List>
            );
        };
        `}});var qn=E(require("@pankod/refine-antd")),mn=E(require("dayjs"));var Ze=D({type:"create",additionalScope:[["@pankod/refine-antd","RefineAntd",qn],["dayjs","dayjs",mn.default,!0]],codeViewerComponent:me,loadingComponent:pe,errorComponent:ue,renderer:({resource:n,fields:c,isCustomPage:k})=>{var N;let d=M((N=n.label)!=null?N:n.name,"create"),$=[["React","react",!0],["IResourceComponentsProps","@pankod/refine-core"],["Create","@pankod/refine-antd"],["Form","@pankod/refine-antd"],["useForm","@pankod/refine-antd"],["Input","@pankod/refine-antd"]],l=c.filter(y=>(y==null?void 0:y.relation)&&!(y!=null&&y.fieldable)&&(y==null?void 0:y.resource)).filter(Boolean).map(y=>{if((y==null?void 0:y.relation)&&!y.fieldable&&y.resource)return $.push(["useSelect","@pankod/refine-antd"]),`
                    const { selectProps: ${f(y.key,"SelectProps")} } =
                    useSelect({
                        resource: "${y.resource.name}",
                        ${ee(y)}
                    });
                `}).filter(Boolean),v=y=>{if(y.relation&&y.resource){$.push(["Select","@pankod/refine-antd"]);let u=f(y.key,"SelectProps"),w=y.accessor?y.multiple?`"${y.key}"`:`["${y.key}", "${y.accessor}"]`:`"${y.key}"`,a="",T="";return y.accessor&&y.multiple&&(T=`getValueFromEvent={(selected: string[]) => {
                        return selected?.map((item) => ({ ${Te(`${y.accessor}`)?y.accessor:`["${y.accessor}"]`}: item }));
                    }}`,a=`getValueProps={(value: any[]) => {
                        return {
                            value: value?.map((item) => ${s("item",void 0,y.accessor)}),
                        };
                    }}`),m`
                    <Form.Item
                        label="${g(y.key)}"
                        name={${w}}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                        ${a}
                        ${T}
                    >
                        <Select ${y.multiple?'mode="multiple"':""} {...${u}} />
                    </Form.Item>             
                    `}},I=y=>{if(y.type==="text"||y.type==="url"||y.type==="email"||y.type==="number")return j(y.key)||y.multiple?void 0:m`
                    <Form.Item
                        label="${g(y.key)}"
                        name={["${y.key}"${y.accessor?', "'+y.accessor+'"':""}]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                    >
                        <Input  />
                    </Form.Item>
                `},S=y=>{if(y.type==="image"){$.push(["Upload","@pankod/refine-antd"],["getValueFromEvent","@pankod/refine-antd"]);let u='valuePropName="fileList"';return y.multiple&&!y.accessor&&(u="getValueProps={(value) => ({ fileList: value?.map((item: any) => ({ url: item, name: item, uid: item }))})}"),y.multiple||(y.accessor?u="getValueProps={(value) => ({ fileList: value ? [value] : [] })}":u="getValueProps={(value) => ({ fileList: [{ url: value, name: value, uid: value }]})}"),m`
                    <Form.Item label="${g(y.key)}">
                        <Form.Item
                            name="${y.key}"
                            ${u}
                            getValueFromEvent={getValueFromEvent}
                            noStyle
                            rules={[
                                {
                                    required: true,
                                },
                            ]}
                        >
                            <Upload.Dragger
                                listType="picture"
                                ${y.multiple?"multiple":""}
                                beforeUpload={() => false}
                            >
                                <p className="ant-upload-text">
                                    Drag & drop a file in this area
                                </p>
                            </Upload.Dragger>
                        </Form.Item>
                    </Form.Item>
                    `}},V=y=>{if(y.type==="boolean")return $.push(["Checkbox","@pankod/refine-antd"]),y.multiple?void 0:m`
                    <Form.Item
                        label="${g(y.key)}"
                        valuePropName="checked"
                        name={["${y.key}"${y.accessor?', "'+y.accessor+'"':""}]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                    >
                        <Checkbox>${g(y.key)}</Checkbox>
                    </Form.Item>
                `},L=y=>{if(y.type==="date")return $.push(["DatePicker","@pankod/refine-antd"],["dayjs","dayjs",!0]),y.multiple?void 0:m`
                    <Form.Item
                        label="${g(y.key)}"
                        name={["${y.key}"${y.accessor?', "'+y.accessor+'"':""}]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                        getValueProps={(value) => ({ value: value ? dayjs(value) : undefined })}
                    >
                        <DatePicker />
                    </Form.Item>
                `},A=y=>{if(y.type==="richtext")return m`
                <Form.Item
                    label="${g(y.key)}"
                    name="${y.key}"
                    rules={[
                        {
                            required: true,
                        },
                    ]}
                >
                    <Input.TextArea rows={5} />
                </Form.Item>
                `},p=c.map(y=>{switch(y==null?void 0:y.type){case"text":case"number":case"email":case"url":return I(y);case"richtext":return A(y);case"image":return S(y);case"date":return L(y);case"boolean":return V(y);case"relation":return v(y);default:return}});return H($),m`
        ${P($)}
        
        export const ${d}: React.FC<IResourceComponentsProps> = () => {
            const { formProps, saveButtonProps, queryResult } = useForm(${k?`{
                          resource: "${n.name}",
                          action: "create",
                      }`:""});
        
            ${l}

            return (
                <Create saveButtonProps={saveButtonProps}>
                    <Form {...formProps} layout="vertical">
                        ${p.join("")}
                    </Form>
                </Create>
            );
        };
        `}});var zn=E(require("@pankod/refine-antd")),ln=E(require("dayjs"));var Xe=D({type:"edit",additionalScope:[["@pankod/refine-antd","RefineAntd",zn],["dayjs","dayjs",ln.default,!0]],codeViewerComponent:me,loadingComponent:pe,errorComponent:ue,renderer:({resource:n,fields:c,isCustomPage:k,id:d})=>{var u,w;let $=M((u=n.label)!=null?u:n.name,"edit"),o=f((w=n.label)!=null?w:n.name,"Data"),l=[["React","react",!0],["IResourceComponentsProps","@pankod/refine-core"],["Edit","@pankod/refine-antd"],["Form","@pankod/refine-antd"],["useForm","@pankod/refine-antd"],["Input","@pankod/refine-antd"]],I=c.filter(a=>(a==null?void 0:a.relation)&&!(a!=null&&a.fieldable)&&(a==null?void 0:a.resource)).filter(Boolean).map(a=>{if((a==null?void 0:a.relation)&&!a.fieldable&&a.resource){l.push(["useSelect","@pankod/refine-antd"]);let T=s(o,a.key,a.accessor,!1);return a.multiple&&a.accessor&&(T=`${s(o,a.key)}?.map((item: any) => ${s("item",void 0,a.accessor)})`),`
                    const { selectProps: ${f(a.key,"SelectProps")} } =
                    useSelect({
                        resource: "${a.resource.name}",
                        defaultValue: ${T},
                        ${a.relationInfer&&a.relationInfer.accessor?typeof a.relationInfer.accessor=="string"?a.relationInfer.accessor!=="title"?`optionLabel: "${a.relationInfer.accessor}",`:"":`optionLabel: "${a.relationInfer.accessor[0]}",`:""}
                    });
                `}}).filter(Boolean),S=a=>{if(a.relation&&a.resource){l.push(["Select","@pankod/refine-antd"]);let T=f(a.key,"SelectProps"),t=a.accessor?a.multiple?`"${a.key}"`:`["${a.key}", "${a.accessor}"]`:`"${a.key}"`,r="",i="";return a.accessor&&a.multiple&&(i=`getValueFromEvent={(selected: string[]) => {
                        return selected?.map((item) => ({ ${Te(`${a.accessor}`)?a.accessor:`["${a.accessor}"]`}: item }));
                    }}`,r=`getValueProps={(value: any[]) => {
                        return {
                            value: value?.map((item) => ${s("item",void 0,a.accessor)}),
                        };
                    }}`),m`
                    <Form.Item
                        label="${g(a.key)}"
                        name={${t}}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                        ${r}
                        ${i}
                    >
                        <Select ${a.multiple?'mode="multiple"':""} {...${T}} />
                    </Form.Item>             
                    `}},V=a=>{if(a.type==="text"||a.type==="url"||a.type==="email"||a.type==="number"){if(a.multiple){let T=s(a.key,"index",a.accessor).split("?.").map(t=>`"${t}"`).join(", ").replace(/"index"/,"index");return`
                        <>
                            {(${s(o,a.key)} as any[])?.map((item, index) => (
                                <Form.Item
                                    key={index}
                                    label={\`${g(a.key)} \${index+1}\`}
                                    name={[${T}]}
                                >
                                    <Input
                                        type="${a.type}"
                                    ${j(a.key)?"readOnly disabled":""} />
                                </Form.Item>
                            ))}
                        </>
                    `}return m`
                    <Form.Item
                        label="${g(a.key)}"
                        name={["${a.key}"${a.accessor?', "'+a.accessor+'"':""}]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                    >
                        <Input ${j(a.key)?"readOnly disabled":""} />
                    </Form.Item>
                `}},L=a=>{if(a.type==="image"){l.push(["Upload","@pankod/refine-antd"],["getValueFromEvent","@pankod/refine-antd"]);let T='valuePropName="fileList"';return a.multiple&&!a.accessor&&(T="getValueProps={(value) => ({ fileList: value?.map((item: any) => ({ url: item, name: item, uid: item }))})}"),a.multiple||(a.accessor?T="getValueProps={(value) => ({ fileList: value ? [value] : [] })}":T="getValueProps={(value) => ({ fileList: [{ url: value, name: value, uid: value }]})}"),m`
                    <Form.Item label="${g(a.key)}">
                        <Form.Item
                            name="${a.key}"
                            ${T}
                            getValueFromEvent={getValueFromEvent}
                            noStyle
                            rules={[
                                {
                                    required: true,
                                },
                            ]}
                        >
                            <Upload.Dragger
                                listType="picture"
                                ${a.multiple?"multiple":""}
                                beforeUpload={() => false}
                            >
                                <p className="ant-upload-text">
                                    Drag & drop a file in this area
                                </p>
                            </Upload.Dragger>
                        </Form.Item>
                    </Form.Item>
                    `}},A=a=>{if(a.type==="boolean"){if(l.push(["Checkbox","@pankod/refine-antd"]),a.multiple){let T=s(a.key,"index",a.accessor).split("?.").map(t=>`"${t}"`).join(", ").replace(/"index"/,"index");return`
                        <>
                            {(${s(o,a.key)} as any[])?.map((item, index) => (
                                <Form.Item
                                    key={index}
                                    valuePropName="checked"
                                    label={\`${g(a.key)} \${index+1}\`}
                                    name={[${T}]}
                                >
                                    <Checkbox>${g(a.key)}</Checkbox>
                                </Form.Item>
                            ))}
                        </>
                    `}return m`
                    <Form.Item
                        label="${g(a.key)}"
                        valuePropName="checked"
                        name={["${a.key}"${a.accessor?', "'+a.accessor+'"':""}]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                    >
                        <Checkbox>${g(a.key)}</Checkbox>
                    </Form.Item>
                `}},p=a=>{if(a.type==="date"){if(l.push(["DatePicker","@pankod/refine-antd"],["dayjs","dayjs",!0]),a.multiple){let T=s(a.key,"index",a.accessor).split("?.").map(t=>`"${t}"`).join(", ").replace(/"index"/,"index");return m`
                        <>
                            {(${s(o,a.key)} as any[])?.map((item, index) => (
                                <Form.Item
                                    key={index}
                                    label={\`${g(a.key)} \${index+1}\`}
                                    name={[${T}]}
                                    getValueProps={(value) => ({ value: value ? dayjs(value) : undefined })}
                                >
                                    <DatePicker />
                                </Form.Item>
                            ))}
                        </>
                    `}return m`
                    <Form.Item
                        label="${g(a.key)}"
                        name={["${a.key}"${a.accessor?', "'+a.accessor+'"':""}]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                        getValueProps={(value) => ({ value: value ? dayjs(value) : undefined })}
                    >
                        <DatePicker />
                    </Form.Item>
                `}},N=a=>{if(a.type==="richtext")return m`
                <Form.Item
                    label="${g(a.key)}"
                    name="${a.key}"
                    rules={[
                        {
                            required: true,
                        },
                    ]}
                >
                    <Input.TextArea rows={5} />
                </Form.Item>
                `},y=c.map(a=>{switch(a==null?void 0:a.type){case"text":case"number":case"email":case"url":return V(a);case"richtext":return N(a);case"image":return L(a);case"date":return p(a);case"boolean":return A(a);case"relation":return S(a);default:return}});return H(l),m`
        ${P(l)}
        
        export const ${$}: React.FC<IResourceComponentsProps> = () => {
            const { formProps, saveButtonProps, queryResult } = useForm(${k?`{
                          resource: "${n.name}",
                          id: ${d},
                          action: "edit",
                      }`:""});
        
            const ${o} = queryResult?.data?.data;
        
            ${I}

            return (
                <Edit saveButtonProps={saveButtonProps}>
                    <Form {...formProps} layout="vertical">
                        ${y.join("")}
                    </Form>
                </Edit>
            );
        };
        `}});var Un=({action:n,id:c,...k})=>{let{useParams:d}=(0,dn.useRouterContext)(),{action:$,id:o}=d();switch(n!=null?n:$){case"show":return Be.default.createElement(Ke,{...k,id:c!=null?c:o});case"create":return Be.default.createElement(Ze,{...k,id:c!=null?c:o});case"edit":return Be.default.createElement(Xe,{...k,id:c!=null?c:o});default:return Be.default.createElement(_e,{...k,id:c!=null?c:o})}};var Me=E(require("react")),$n=require("@pankod/refine-core");var Wn=E(require("@pankod/refine-mantine"));var Je=E(require("react")),Qe=require("@pankod/refine-mantine"),yn=require("@tabler/icons"),le=({error:n})=>n?Je.default.createElement(Qe.Center,{style:{minHeight:300}},Je.default.createElement(Qe.Alert,{title:"Error",color:"red",icon:Je.default.createElement(yn.IconAlertCircle,null)},n!=null?n:"")):null;var kn=E(require("react")),gn=require("@pankod/refine-mantine"),de=()=>kn.default.createElement(gn.LoadingOverlay,{visible:!0});var q=E(require("react")),Q=require("@pankod/refine-mantine"),ne=require("@tabler/icons");var ye=({code:n,loading:c})=>{let k=q.default.useMemo(()=>ae(n!=null?n:""),[n]),d=q.default.useRef(null),[$,o]=q.default.useState(!1),[l,v]=q.default.useState(!1);if(c)return null;let I=()=>{var S,V,L,A;(S=d==null?void 0:d.current)==null||S.select(),(V=d==null?void 0:d.current)==null||V.setSelectionRange(0,Number.MAX_SAFE_INTEGER),typeof navigator<"u"&&(navigator.clipboard.writeText((A=(L=d==null?void 0:d.current)==null?void 0:L.value)!=null?A:""),v(!0),setTimeout(()=>{v(!1)},1e3))};return k?q.default.createElement(q.default.Fragment,null,q.default.createElement(Q.Affix,{position:{bottom:12,right:60}},q.default.createElement(Q.Group,{spacing:"xs"},q.default.createElement(Q.Button,{leftIcon:q.default.createElement(ne.IconCode,{size:18}),variant:"filled",color:"primary",onClick:()=>o(!0)},"Show Code"),q.default.createElement(Q.Button,{leftIcon:q.default.createElement(ne.IconMessageCircle,{size:18}),variant:"default",component:"a",target:"_blank",href:"https://github.com/refinedev/refine/discussions/3046"},"Give Feedback"))),q.default.createElement("div",{style:{height:"50px"}}),q.default.createElement(Q.Modal,{size:800,opened:$,onClose:()=>o(!1),styles:{modal:{padding:"0px !important",borderRadius:0},header:{display:"none"}}},q.default.createElement(Q.ActionIcon,{onClick:()=>o(!1),variant:"transparent",sx:{position:"absolute",right:0,top:0,height:56,width:56}},q.default.createElement(ne.IconX,{size:18,style:{color:"#666b7a"}})),q.default.createElement(se,{code:k}),q.default.createElement("textarea",{ref:d,defaultValue:(k!=null?k:"").replace(/\\n/g,`\r
`),id:"code-input",style:{padding:0,margin:0,width:0,height:0,opacity:0,border:"none",display:"block"}}),q.default.createElement(Q.Button,{sx:{position:"absolute",bottom:12,right:12,width:115},key:"copy",variant:"default",leftIcon:l?q.default.createElement(ne.IconCheck,{size:18,style:{color:"green"}}):q.default.createElement(ne.IconCopy,{size:18}),onClick:I},l?"Copied":"Copy"))):null};var Ye=D({type:"show",additionalScope:[["@pankod/refine-mantine","RefineMantine",Wn]],codeViewerComponent:ye,loadingComponent:de,errorComponent:le,renderer:({resource:n,fields:c,isCustomPage:k,id:d})=>{var t;let $=M((t=n.label)!=null?t:n.name,"show"),o="record",l=[["useShow","@pankod/refine-core"],["Show","@pankod/refine-mantine"],["Title","@pankod/refine-mantine"]],I=c.filter(r=>(r==null?void 0:r.relation)&&!(r!=null&&r.fieldable)&&(r==null?void 0:r.resource)).filter(Boolean).map(r=>{if((r==null?void 0:r.relation)&&!r.fieldable&&r.resource){if(r.multiple){l.push(["useMany","@pankod/refine-core"]);let i=s(o,r.key);return r.accessor&&(i=`${s(o,r.key)}?.map((item: any) => ${s("item",void 0,r.accessor)})`),`
                    const { data: ${f(r.key,"Data")}, isLoading: ${f(r.key,"IsLoading")} } =
                    useMany({
                        resource: "${r.resource.name}",
                        ids: ${i} || [],
                        queryOptions: {
                            enabled: !!${o},
                        },
                    });
                    `}return l.push(["useOne","@pankod/refine-core"]),`
                    const { data: ${f(r.key,"Data")}, isLoading: ${f(r.key,"IsLoading")} } =
                    useOne({
                        resource: "${r.resource.name}",
                        id: ${s(o,r.key,r.accessor,!1)} || "",
                        queryOptions: {
                            enabled: !!${o},
                        },
                    });
                `}}).filter(Boolean),S=r=>{if(r.relation&&r.resource){let i=f(r.key,"Data"),e=f(r.key,"IsLoading");return r.multiple?(l.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]),m`
                    <Title my="xs" order={5}>${g(r.key)}</Title>
                    {${e} ? <>Loading...</> : (
                        <>
                        ${(()=>{var h,x;if(r.relationInfer)if((h=r.relationInfer)!=null&&h.accessor){if(Array.isArray(r.relationInfer.accessor))return"Not Handled.";{let C=_((x=r.resource)==null?void 0:x.name),b=s(C,void 0,r.relationInfer.accessor);return`
                                        <Group spacing="xs">
                                            {${i}?.data?.map((${C}: any) => <TagField key={${b}} value={${b}} />)}
                                        </Group>
                                        `}}else return"Not Handled.";else return"not-handled - relation with multiple but no resource"})()}
                        </>
                    )}
                    `):m`
                    <Title my="xs" order={5}>${g(r.key)}</Title>
                    {${e} ? <>Loading...</> : (
                        <>
                        ${(()=>{var h;return r.relationInfer?(h=r.relationInfer)!=null&&h.accessor?Array.isArray(r.relationInfer.accessor)?`{${s(`${i}?.data`,void 0,r.relationInfer.accessor,' + " " + ')}}`:`{${i}?.data?.${r.relationInfer.accessor}}`:`{${i}?.data}`:`{${i}?.data?.id}`})()}
                        </>
                    )}
                    
                    `}},V=r=>{if(r.type==="text"){if(l.push(["TextField","@pankod/refine-mantine"]),r.multiple){l.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]);let i=s("item",void 0,r.accessor);return m`
                    <Title my="xs" order={5}>${g(r.key)}</Title>
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any) => (
                            <TagField value={${i}} key={${i}} />
                        ))}
                    </Group>
                `}return m`
                    <Title my="xs" order={5}>${g(r.key)}</Title>
                    <TextField value={${s(o,r.key,r.accessor)}} />
                `}},L=r=>{if(r.type==="image"){if(l.push(["Image","@pankod/refine-mantine"]),r.multiple){l.push(["Group","@pankod/refine-mantine"]);let i=s("item",void 0,r.accessor);return m`
                    <Title my="xs" order={5}>${g(r.key)}</Title>
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any) => (
                            <Image sx={{ maxWidth: 200 }} src={${i}} key={${i}} />
                        ))}
                    </Group>
                `}return m`
                    <Title my="xs" order={5}>${g(r.key)}</Title>
                    <Image sx={{ maxWidth: 200 }} src={${s(o,r.key,r.accessor," + ")}} />
                    `}},A=r=>{if(r.type==="email"){if(l.push(["EmailField","@pankod/refine-mantine"]),r.multiple){l.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]);let i=s("item",void 0,r.accessor);return m`
                    <Title my="xs" order={5}>${g(r.key)}</Title>
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any) => (
                            <TagField value={${i}} key={${i}} />
                        ))}
                    </Group>
                `}return m`
                    <Title my="xs" order={5}>${g(r.key)}</Title>
                    <EmailField value={${s(o,r.key,r.accessor," + ")}} />
                `}},p=r=>{if(r.type==="url"){if(l.push(["UrlField","@pankod/refine-mantine"]),r.multiple){l.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]);let i=s("item",void 0,r.accessor);return m`
                    <Title my="xs" order={5}>${g(r.key)}</Title>
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any) => (
                            <TagField value={${i}} key={${i}} />
                        ))}
                    </Group>
                `}return m`
                    <Title my="xs" order={5}>${g(r.key)}</Title>
                    <UrlField value={${s(o,r.key,r.accessor," + ")}} />
                `}},N=r=>{if(r.type==="boolean"){if(l.push(["BooleanField","@pankod/refine-mantine"]),r.multiple){l.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]);let i=s("item",void 0,r.accessor);return m`
                    <Title my="xs" order={5}>${g(r.key)}</Title>
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any, index: number) => (
                            <TagField value={${i}} key={index} />
                        ))}
                    </Group>
                `}return m`
                    <Title my="xs" order={5}>${g(r.key)}</Title>
                    <BooleanField value={${s(o,r.key,r.accessor," && ")}} />
                `}},y=r=>{if(r.type==="date"){if(l.push(["DateField","@pankod/refine-mantine"]),r.multiple){l.push(["Group","@pankod/refine-mantine"]);let i=s("item",void 0,r.accessor);return m`
                    <Title my="xs" order={5}>${g(r.key)}</Title>
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any) => (
                            <DateField value={${i}} key={${i}} />
                        ))}
                    </Group>
                `}return m`
                    <Title my="xs" order={5}>${g(r.key)}</Title>
                    <DateField value={${s(o,r.key,r.accessor," + ' ' + ")}} />
                `}},u=r=>{if(r.type==="richtext")return l.push(["MarkdownField","@pankod/refine-mantine"]),m`
                    <Title mt="xs" order={5}>${g(r.key)}</Title>
                    <MarkdownField value={${s(o,r.key,r.accessor,' + " " + ')}} />
                `},w=r=>{if(r.type==="number"){if(l.push(["NumberField","@pankod/refine-mantine"]),r.multiple){l.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]);let i=s("item",void 0,r.accessor);return m`
                    <Title my="xs" order={5}>${g(r.key)}</Title>
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any) => (
                            <TagField value={${i}} key={${i}} />
                        ))}
                    </Group>
                `}return m`
                    <Title my="xs" order={5}>${g(r.key)}</Title>
                    <NumberField value={${s(o,r.key,r.accessor,' + " " + ')} ?? ""} />
                `}},a=r=>{if(r)return m`
                    ${r}
                `},T=c.map(r=>{switch(r==null?void 0:r.type){case"text":return a(V(r));case"number":return a(w(r));case"richtext":return a(u(r));case"email":return a(A(r));case"image":return a(L(r));case"date":return a(y(r));case"boolean":return a(N(r));case"url":return a(p(r));case"relation":return a(S(r));default:return}});return H(l),m`
        ${P(l)}
        
        export const ${$} = () => {
            const { queryResult } = useShow(${k?`{ 
                        resource: "${n.name}", 
                        id: ${d}
                    }`:""});
            const { data, isLoading } = queryResult;
        
            const ${o} = data?.data;
        
            ${I}

            return (
                <Show isLoading={isLoading}>
                    ${T.join("")}
                </Show>
            );
        };
        `}});var Kn=E(require("@pankod/refine-mantine")),_n=E(require("@pankod/refine-react-table"));var ke=n=>Array.isArray(n.accessor)||n.multiple?`accessorKey: "${n.key}"`:n.accessor?`accessorKey: "${R(n.key,void 0,n.accessor)}"`:`accessorKey: "${n.key}"`,er=D({type:"list",additionalScope:[["@pankod/refine-mantine","RefineMantine",Kn],["@pankod/refine-react-table","RefineReactTable",_n]],codeViewerComponent:ye,loadingComponent:de,errorComponent:le,renderer:({resource:n,fields:c,isCustomPage:k})=>{var i;let d=M((i=n.label)!=null?i:n.name,"list"),$="tableData?.data",o=[["IResourceComponentsProps","@pankod/refine-core"],["useTable","@pankod/refine-react-table"],["ColumnDef","@pankod/refine-react-table"],["flexRender","@pankod/refine-react-table"],["ScrollArea","@pankod/refine-mantine"],["List","@pankod/refine-mantine"],["Table","@pankod/refine-mantine"],["Pagination","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"],["EditButton","@pankod/refine-mantine"],["ShowButton","@pankod/refine-mantine"],["DeleteButton","@pankod/refine-mantine"]],l=c.filter(e=>(e==null?void 0:e.relation)&&!(e!=null&&e.fieldable)&&(e==null?void 0:e.resource)),v=l.filter(Boolean).map(e=>{if((e==null?void 0:e.relation)&&!e.fieldable&&e.resource){o.push(["GetManyResponse","@pankod/refine-core"]),o.push(["useMany","@pankod/refine-core"]);let h="";return e.multiple?h=`[].concat(...(${$}?.map((item) => ${s("item",e.key,e.accessor,!1)}) ?? []))`:h=`${$}?.map((item) => ${s("item",e.key,e.accessor,!1)}) ?? []`,`
                    const { data: ${f(e.key,"Data")} } =
                    useMany({
                        resource: "${e.resource.name}",
                        ids: ${h},
                        queryOptions: {
                            enabled: !!${$},
                        },
                    });
                    `}}).filter(Boolean),I=l==null?void 0:l.map(e=>{if(e&&e.resource)return f(e.key,"Data")}).filter(Boolean),S=e=>{var h;if(e.relation&&e.resource){let x=`${f(e.key,"Data")}?.data`;if(Array.isArray(e.accessor))return;let C=`id: "${e.key}"`,b=`header: "${g(e.key)}"`,B=ke(e),W="";if(e.multiple){o.push(["TagField","@pankod/refine-mantine"]);let K="item";e!=null&&e.relationInfer&&(K=s("item",void 0,e.relationInfer.accessor)),W=`cell: function render({ getValue, table }) {
                        const meta = table.options.meta as {
                            ${f(e.key,"Data")}: GetManyResponse;
                        };

                        const ${f(e.key)} = getValue<any[]>()?.map((item) => {
                            return meta.${f(e.key,"Data")}?.data?.find(
                                (resourceItems) => resourceItems.id === ${s("item",void 0,e.accessor)}
                            );
                        })

                        return (
                            <Group spacing="xs">
                                {${f(e.key)}?.map((item, index) => (
                                    <TagField key={index} value={${K}} />
                                ))}
                            </Group>
                        )
                    }
                `}else e!=null&&e.relationInfer?W=`cell: function render({ getValue, table }) {
                            const meta = table.options.meta as {
                                ${f(e.key,"Data")}: GetManyResponse;
                            };

                            const ${f(e.key)} = meta.${x}?.find(
                                (item) => item.id === getValue<any>(),
                            );

                            return ${s(f(e.key),void 0,(h=e==null?void 0:e.relationInfer)==null?void 0:h.accessor)} ?? "Loading...";
                        },`:W="";return`
                    {
                        ${C},
                        ${b},
                        ${B},
                        ${W}
                    }
                `}},V=e=>{if(e.type==="image"){o.push(["Image","@pankod/refine-mantine"]);let h=`id: "${e.key}"`,x=ke(e),C=`header: "${g(e.key)}"`,b=m`
                    cell: function render({ getValue }) {
                        ${e!=null&&e.accessor?"try {":""}
                            return <Image sx={{ maxWidth: "100px" }} src={${s("getValue<any>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0," + ")}} />
                        ${e!=null&&e.accessor?" } catch (error) { return null; }":""}
                       
                    }
                `;if(e.multiple){let B=s("item",void 0,e.accessor," + ");b=`
                        cell: function render({ getValue }) {
                            ${e!=null&&e.accessor?"try {":""}
                                return (
                                    <Group spacing="xs">
                                        {getValue<any[]>()?.map((item, index) => (
                                            <Image src={${B}} key={index} sx={{ maxWidth: "100px" }} />
                                        ))}
                                    </Group>
                                )
                            ${e!=null&&e.accessor?" } catch (error) { return null; }":""}
                        }
                    `}return`
                    {
                        ${h},
                        ${x},
                        ${C},
                        ${b}
                    }
                `}},L=e=>{if(e.type==="email"){o.push(["EmailField","@pankod/refine-mantine"]);let h=`id: "${e.key}"`,x=ke(e),C=`header: "${g(e.key)}"`,b=m`
                    cell: function render({ getValue }) {
                        return <EmailField value={${s("getValue<any>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0,' + " " + ')}} />
                    }
                `;return e.multiple&&(o.push(["TagField","@pankod/refine-mantine"]),b=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any>()?.map((item, index) => (
                                        <TagField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${C},
                        ${b}
                    }
                `}},A=e=>{if(e.type==="url"){o.push(["UrlField","@pankod/refine-mantine"]);let h=`id: "${e.key}"`,x=ke(e),C=`header: "${g(e.key)}"`,b=m`
                    cell: function render({ getValue }) {
                        return <UrlField value={${s("getValue<any>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0," + ")}} />
                    }
                `;return e.multiple&&(o.push(["TagField","@pankod/refine-mantine"]),b=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${C},
                        ${b}
                    }
                `}},p=e=>{if((e==null?void 0:e.type)==="boolean"){o.push(["BooleanField","@pankod/refine-mantine"]);let h=`id: "${e.key}"`,x=ke(e),C=`header: "${g(e.key)}"`,b=m`
                    cell: function render({ getValue }) {
                        return <BooleanField value={${s("getValue<any>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0," + ")}} />
                    }
                `;return e.multiple&&(b=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any[]>()?.map((item, index) => (
                                        <BooleanField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${C},
                        ${b}
                    }
                `}},N=e=>{if(e.type==="date"){o.push(["DateField","@pankod/refine-mantine"]);let h=`id: "${e.key}"`,x=ke(e),C=`header: "${g(e.key)}"`,b=m`
                    cell: function render({ getValue }) {
                        return <DateField value={${s("getValue<any>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0,' + " " + ')}} />
                    }
                `;return e.multiple&&(b=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any[]>()?.map((item, index) => (
                                        <DateField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${C},
                        ${b}
                    }
                `}},y=e=>{if((e==null?void 0:e.type)==="richtext"){o.push(["MarkdownField","@pankod/refine-mantine"]);let h=`id: "${e.key}"`,x=ke(e),C=`header: "${g(e.key)}"`,b=m`
                    cell: function render({ getValue }) {
                        return <MarkdownField value={(${s("getValue<string>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0)})?.slice(0, 80) + "..." } />
                    }
                `;return e.multiple&&(b=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<string[]>()?.map((item, index) => (
                                        <MarkdownField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${C},
                        ${b}
                    }
                `}},u=e=>{if(e&&(e.type==="text"||e.type==="number")){let h=`id: "${e.key}"`,x=ke(e),C=`header: "${g(e.key)}"`,b="";return e.multiple&&(o.push(["TagField","@pankod/refine-mantine"]),b=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${s("item",void 0,e.accessor,' + " " + ')}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),!e.multiple&&Array.isArray(e.accessor)&&(b=`
                        cell: function render({ getValue }) {
                            return (
                                <>{${s("getValue()",e.key,e.accessor)}}</>
                            );
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${C},
                        ${b}
                    }
                `}},{canEdit:w,canShow:a,canDelete:T}=n!=null?n:{};w&&o.push(["EditButton","@pankod/refine-mantine"]),a&&o.push(["ShowButton","@pankod/refine-mantine"]),T&&o.push(["DeleteButton","@pankod/refine-mantine"]);let t=w||a||T?m`
        {
            id: "actions",
            accessorKey: "id",
            header: "Actions",
            cell: function render({ getValue }) {
                return (
                    <Group spacing="xs" noWrap>
                        ${a?m`
                        <ShowButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `:""}
                        ${w?m`<EditButton
                            hideText
                            recordItemId={getValue() as string}
                        />`:""}
                        ${T?m`
                        <DeleteButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `:""}
                    </Group>
                );
            },
        },
            `:"",r=c.map(e=>{switch(e==null?void 0:e.type){case"text":case"number":return u(e);case"richtext":return y(e);case"email":return L(e);case"image":return V(e);case"date":return N(e);case"boolean":return p(e);case"url":return A(e);case"relation":return S(e);default:return}});return H(o),m`
        import React from "react";
        ${P(o)}
        
        export const ${d}: React.FC<IResourceComponentsProps> = () => {
            const columns = React.useMemo<ColumnDef<any>[]>(() => [
                ${[...r,t].filter(Boolean).join(",")}
            ], []);

            const {
                getHeaderGroups,
                getRowModel,
                setOptions,
                refineCore: {
                    setCurrent,
                    pageCount,
                    current,
                    tableQueryResult: { data: tableData },
                },
            } = useTable({
                columns,
                ${k?`
                refineCoreProps: {
                    resource: "${n.name}",
                }
                `:""}
            });

            ${v}

            setOptions((prev) => ({
                ...prev,
                meta: {
                    ...prev.meta,
                    ${I.join(", ")}
                },
            }));

            return (
                <List>
                    <ScrollArea>
                        <Table highlightOnHover>
                            <thead>
                                {getHeaderGroups().map((headerGroup) => (
                                    <tr key={headerGroup.id}>
                                        {headerGroup.headers.map((header) => {
                                            return (
                                                <th key={header.id}>
                                                    {!header.isPlaceholder && (
                                                        flexRender(
                                                            header.column
                                                                .columnDef
                                                                .header,
                                                            header.getContext(),
                                                        )
                                                    )}
                                                </th>
                                            );
                                        })}
                                    </tr>
                                ))}
                            </thead>
                            <tbody>
                                {getRowModel().rows.map((row) => {
                                    return (
                                        <tr key={row.id}>
                                            {row.getVisibleCells().map((cell) => {
                                                return (
                                                    <td key={cell.id}>
                                                        {flexRender(
                                                            cell.column.columnDef.cell,
                                                            cell.getContext(),
                                                        )}
                                                    </td>
                                                );
                                            })}
                                        </tr>
                                    );
                                })}
                            </tbody>
                        </Table>
                    </ScrollArea>    
                    <br />
                    <Pagination
                        position="right"
                        total={pageCount}
                        page={current}
                        onChange={setCurrent}
                    />
                </List>
            );
        };
        `}});var Zn=E(require("@pankod/refine-mantine"));var rr=D({type:"create",additionalScope:[["@pankod/refine-mantine","RefineMantine",Zn]],codeViewerComponent:ye,loadingComponent:de,errorComponent:le,renderer:({resource:n,fields:c,isCustomPage:k})=>{var w;let d=M((w=n.label)!=null?w:n.name,"create"),$=[["Create","@pankod/refine-mantine"],["useForm","@pankod/refine-mantine"]],o={},v=c.filter(a=>(a==null?void 0:a.relation)&&!(a!=null&&a.fieldable)&&(a==null?void 0:a.resource)).filter(Boolean).map(a=>{if((a==null?void 0:a.relation)&&!a.fieldable&&a.resource)return $.push(["useSelect","@pankod/refine-mantine"]),`
                    const { selectProps: ${f(a.key,"SelectProps")} } =
                    useSelect({
                        resource: "${a.resource.name}",
                        ${ee(a)}
                    });
                `}).filter(Boolean),I=a=>{if(a.relation&&a.resource){o={...o,[a.key]:a.multiple?[]:a.accessor?{[typeof a.accessor=="string"?a.accessor:a.accessor[0]]:""}:""};let T=f(a.key,"SelectProps");return a.multiple?($.push(["MultiSelect","@pankod/refine-mantine"]),m`
                        <MultiSelect mt="sm" label="${g(a.key)}" {...getInputProps("${R(a.key,void 0)}")} {...${T}} filterDataOnExactSearchMatch={undefined} />
                    `):($.push(["Select","@pankod/refine-mantine"]),m`
                    <Select mt="sm" label="${g(a.key)}" {...getInputProps("${R(a.key,void 0,a.accessor)}")} {...${T}} />
                `)}},S=a=>{if(a.type==="text"||a.type==="email"||a.type==="date"||a.type==="url")return j(a.key)||($.push(["TextInput","@pankod/refine-mantine"]),o={...o,[a.key]:a.multiple?[]:""},a.multiple)?void 0:m`
                    <TextInput mt="sm" label="${g(a.key)}" {...getInputProps("${R(a.key,void 0,a.accessor)}")} />
                `},V=a=>{if(a.type==="image")return m`
                {/* 
                    Dropzone component is not included in "@pankod/refine-mantine" package.
                    To use a <Dropzone> component, you can follow the official documentation for Mantine.
                    
                    Docs: https://mantine.dev/others/dropzone/
                */}
                `},L=a=>{if(a.type==="boolean")return $.push(["Checkbox","@pankod/refine-mantine"]),o={...o,[a.key]:a.multiple?[]:""},a.multiple?void 0:m`
                    <Checkbox mt="sm" label="${g(a.key)}" {...getInputProps("${R(a.key,void 0,a.accessor)}", { type: 'checkbox' })} />
                `},A=a=>{if(a.type==="date"){let T=S(a);return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-mantine" package.
                        To use a <DatePicker> component, you can follow the official documentation for Mantine.
                        
                        Docs: https://mantine.dev/dates/date-picker/
                    */}
                    ${T!=null?T:""}
                `}},p=a=>{if(a.type==="richtext")return $.push(["Textarea","@pankod/refine-mantine"]),o={...o,[a.key]:a.multiple?[]:""},a.multiple?void 0:m`
                    <Textarea mt="sm" label="${g(a.key)}" autosize {...getInputProps("${R(a.key,void 0,a.accessor)}")} />
                `},N=a=>{if(a.type==="number")return j(a.key)||($.push(["NumberInput","@pankod/refine-mantine"]),o={...o,[a.key]:a.multiple?[]:""},a.multiple)?void 0:m`
                    <NumberInput mt="sm" label="${g(a.key)}" {...getInputProps("${R(a.key,void 0,a.accessor)}")}/>
                `},y=a=>{if(a)return m`
                    ${a}
            `},u=c.map(a=>{switch(a==null?void 0:a.type){case"url":case"text":case"email":return y(S(a));case"number":return y(N(a));case"richtext":return y(p(a));case"image":return y(V(a));case"date":return y(A(a));case"boolean":return y(L(a));case"relation":return y(I(a));default:return}});return H($),m`
        ${P($)}
        
        export const ${d} = () => {
            const { getInputProps, saveButtonProps, setFieldValue, refineCore: { formLoading } } = useForm({
                initialValues: ${JSON.stringify(o)},
                ${k?`refineCoreProps: {
                            resource: "${n.name}",
                            action: "create",  
                        }`:""}
            });
        
            ${v}

            return (
                <Create isLoading={formLoading} saveButtonProps={saveButtonProps}>
                    ${u.join("")}
                </Create>
            );
        };
        `}});var Xn=E(require("@pankod/refine-mantine"));var nr=D({type:"edit",additionalScope:[["@pankod/refine-mantine","RefineMantine",Xn]],codeViewerComponent:ye,loadingComponent:de,errorComponent:le,renderer:({resource:n,fields:c,isCustomPage:k,id:d})=>{var T,t;let $=M((T=n.label)!=null?T:n.name,"edit"),o=f((t=n.label)!=null?t:n.name,"Data"),l=[["Edit","@pankod/refine-mantine"],["useForm","@pankod/refine-mantine"]],v={},S=c.filter(r=>(r==null?void 0:r.relation)&&!(r!=null&&r.fieldable)&&(r==null?void 0:r.resource)).filter(Boolean).map(r=>{if((r==null?void 0:r.relation)&&!r.fieldable&&r.resource){l.push(["useSelect","@pankod/refine-mantine"]);let i=s(o,r.key,r.accessor,!1);r.multiple&&r.accessor&&(i=`${s(o,r.key)}?.map((item: any) => ${s("item",void 0,r.accessor)})`);let e="";return r.multiple&&r.accessor&&(e=`React.useEffect(() => {
                            setFieldValue("${r.key}", ${i});
                        }, [${o}]);`),`
                    const { selectProps: ${f(r.key,"SelectProps")} } =
                    useSelect({
                        resource: "${r.resource.name}",
                        defaultValue: ${i},
                        ${ee(r)}
                    });

                    ${e}
                `}}).filter(Boolean),V=r=>{if(r.relation&&r.resource){v={...v,[r.key]:r.multiple?[]:r.accessor?{[typeof r.accessor=="string"?r.accessor:r.accessor[0]]:""}:""};let i=f(r.key,"SelectProps");return r.multiple?(l.push(["MultiSelect","@pankod/refine-mantine"]),m`
                        <MultiSelect mt="sm" label="${g(r.key)}" {...getInputProps("${R(r.key,void 0)}")} {...${i}} filterDataOnExactSearchMatch={undefined} />
                    `):(l.push(["Select","@pankod/refine-mantine"]),m`
                    <Select mt="sm" label="${g(r.key)}" {...getInputProps("${R(r.key,void 0,r.accessor)}")} {...${i}} />
                `)}},L=r=>{if(r.type==="text"||r.type==="email"||r.type==="date"||r.type==="url"){if(l.push(["TextInput","@pankod/refine-mantine"]),v={...v,[r.key]:r.multiple?[]:""},r.multiple){l.push(["Group","@pankod/refine-mantine"]);let i=R(r.key,"${index}",r.accessor);return`
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any, index: number) => (
                            <TextInput mt="sm" key={index} label={\`${g(r.key)} \${index + 1} \`} {...getInputProps(\`${i}\`)} />
                        ))}
                    </Group>
                    `}return m`
                    <TextInput mt="sm" ${j(r.key)?"disabled":""} label="${g(r.key)}" {...getInputProps("${R(r.key,void 0,r.accessor)}")} />
                `}},A=r=>{if(r.type==="image")return m`
                {/* 
                    Dropzone component is not included in "@pankod/refine-mantine" package.
                    To use a <Dropzone> component, you can follow the official documentation for Mantine.
                    
                    Docs: https://mantine.dev/others/dropzone/
                */}
                `},p=r=>{if(r.type==="boolean"){if(l.push(["Checkbox","@pankod/refine-mantine"]),v={...v,[r.key]:r.multiple?[]:""},r.multiple){l.push(["Group","@pankod/refine-mantine"]);let i=R(r.key,"${index}",r.accessor);return`
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any, index: number) => (
                            <Checkbox mt="sm" key={index} label={\`${g(r.key)} \${index + 1} \`} {...getInputProps(\`${i}\`, { type: 'checkbox' })} />
                        ))}
                    </Group>
                    `}return m`
                    <Checkbox mt="sm" label="${g(r.key)}" {...getInputProps("${R(r.key,void 0,r.accessor)}", { type: 'checkbox' })} />
                `}},N=r=>{if(r.type==="date")return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-mantine" package.
                        To use a <DatePicker> component, you can follow the official documentation for Mantine.
                        
                        Docs: https://mantine.dev/dates/date-picker/
                    */}
                    ${L(r)}
                `},y=r=>{if(r.type==="richtext"){if(l.push(["Textarea","@pankod/refine-mantine"]),v={...v,[r.key]:r.multiple?[]:""},r.multiple){l.push(["Group","@pankod/refine-mantine"]);let i=R(r.key,"${index}",r.accessor);return`
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any, index: number) => (
                            <Textarea mt="sm" key={index} label={\`${g(r.key)} \${index + 1} \`} {...getInputProps(\`${i}\`)} />
                        ))}
                    </Group>
                    `}return m`
                    <Textarea mt="sm" label="${g(r.key)}" autosize {...getInputProps("${R(r.key,void 0,r.accessor)}")} />
                `}},u=r=>{if(r.type==="number"){if(l.push(["NumberInput","@pankod/refine-mantine"]),v={...v,[r.key]:r.multiple?[]:""},r.multiple){l.push(["Group","@pankod/refine-mantine"]);let i=R(r.key,"${index}",r.accessor);return`
                    <Group spacing="xs">
                        {${s(o,r.key)}?.map((item: any, index: number) => (
                            <NumberInput mt="sm" key={index} label={\`${g(r.key)} \${index + 1} \`} {...getInputProps(\`${i}\`)} />
                        ))}
                    </Group>
                    `}return m`
                    <NumberInput mt="sm" ${j(r.key)?"disabled":""} label="${g(r.key)}" {...getInputProps("${R(r.key,void 0,r.accessor)}")}/>
                `}},w=r=>{if(r)return m`
                    ${r}
            `},a=c.map(r=>{switch(r==null?void 0:r.type){case"url":case"text":case"email":return w(L(r));case"number":return w(u(r));case"richtext":return w(y(r));case"image":return w(A(r));case"date":return w(N(r));case"boolean":return w(p(r));case"relation":return w(V(r));default:return}});return H(l),m`
        ${P(l)}
        
        export const ${$} = () => {
            const { getInputProps, saveButtonProps, setFieldValue, refineCore: { queryResult } } = useForm({
                initialValues: ${JSON.stringify(v)},
                ${k?`refineCoreProps: {
                            resource: "${n.name}",
                            id: ${d},
                            action: "edit",  
                        }`:""}
            });
        
            const ${o} = queryResult?.data?.data;
        
            ${S}

            return (
                <Edit saveButtonProps={saveButtonProps}>
                    ${a.join("")}
                </Edit>
            );
        };
        `}});var Jn=({action:n,id:c,...k})=>{let{useParams:d}=(0,$n.useRouterContext)(),{action:$,id:o}=d();switch(n!=null?n:$){case"show":return Me.default.createElement(Ye,{...k,id:c!=null?c:o});case"create":return Me.default.createElement(rr,{...k,id:c!=null?c:o});case"edit":return Me.default.createElement(nr,{...k,id:c!=null?c:o});default:return Me.default.createElement(er,{...k,id:c!=null?c:o})}};var De=E(require("react")),hn=require("@pankod/refine-core");var Qn=E(require("@pankod/refine-mui"));var tr=E(require("react")),Ee=require("@pankod/refine-mui"),ge=({error:n})=>n?tr.default.createElement(Ee.Box,{style:{display:"flex",justifyContent:"center",alignItems:"center",minHeight:"200px"}},tr.default.createElement(Ee.Alert,{severity:"error"},tr.default.createElement(Ee.AlertTitle,null,"Error"),n!=null?n:"")):null;var Er=E(require("react")),or=require("@pankod/refine-mui"),$e=()=>Er.default.createElement(or.Box,{style:{display:"flex",justifyContent:"center",alignItems:"center",minHeight:"300px"}},Er.default.createElement(or.CircularProgress,{size:"large"}));var O=E(require("react")),Y=require("@pankod/refine-mui"),te=require("@tabler/icons");var he=({code:n,loading:c})=>{let k=O.default.useMemo(()=>ae(n!=null?n:""),[n]),d=O.default.useRef(null),[$,o]=O.default.useState(!1),[l,v]=O.default.useState(!1);if(c)return null;let I=()=>{var S,V,L,A;(S=d==null?void 0:d.current)==null||S.select(),(V=d==null?void 0:d.current)==null||V.setSelectionRange(0,99999),typeof navigator<"u"&&(navigator.clipboard.writeText((A=(L=d==null?void 0:d.current)==null?void 0:L.value)!=null?A:""),v(!0),setTimeout(()=>{v(!1)},1e3))};return k?O.default.createElement(O.default.Fragment,null,O.default.createElement(Y.Box,{sx:{position:"fixed",bottom:12,right:60}},O.default.createElement(Y.Box,{sx:{display:"flex",gap:2}},O.default.createElement(Y.Button,{variant:"contained",color:"primary",onClick:()=>o(!0),startIcon:O.default.createElement(te.IconCode,{size:18}),sx:{whiteSpace:"nowrap"}},"Show Code"),O.default.createElement(Y.Button,{variant:"contained",color:"secondary",component:"a",target:"_blank",href:"https://github.com/refinedev/refine/discussions/3046",startIcon:O.default.createElement(te.IconMessageCircle,{size:18}),sx:{whiteSpace:"nowrap"}},"Give Feedback"))),O.default.createElement("div",{style:{height:"50px"}}),O.default.createElement(Y.Modal,{open:$,onClose:()=>o(!1)},O.default.createElement(Y.Box,{sx:{position:"absolute",maxWidth:"800px",width:"100%",top:"50%",left:"50%",transform:"translate(-50%, -50%)",boxShadow:24}},O.default.createElement(Y.IconButton,{onClick:()=>o(!1),disableRipple:!0,sx:{position:"absolute",right:0,top:0,height:56,width:56,display:"flex",alignItems:"center",justifyContent:"center"}},O.default.createElement(te.IconX,{size:18,style:{color:"#666b7a"}})),O.default.createElement(se,{code:k}),O.default.createElement("textarea",{ref:d,defaultValue:(k!=null?k:"").replace(/\\n/g,`\r
`),id:"code-input",style:{display:"none"}}),O.default.createElement("div",{style:{position:"absolute",bottom:12,right:12}},O.default.createElement(Y.Button,{key:"copy",sx:{textTransform:"capitalize",width:100},startIcon:l?O.default.createElement(te.IconCheck,{size:18,style:{color:"green"}}):O.default.createElement(te.IconCopy,{size:18}),color:"inherit",variant:"contained",onClick:I},l?"Copied":"Copy"))))):null};var ar=D({type:"show",additionalScope:[["@pankod/refine-mui","RefineMui",Qn]],codeViewerComponent:he,loadingComponent:$e,errorComponent:ge,renderer:({resource:n,fields:c,isCustomPage:k,id:d})=>{var T;let $=M((T=n.label)!=null?T:n.name,"show"),o="record",l=[["useShow","@pankod/refine-core"],["Show","@pankod/refine-mui"],["Typography","@pankod/refine-mui"],["Stack","@pankod/refine-mui"]],I=c.filter(t=>(t==null?void 0:t.relation)&&!(t!=null&&t.fieldable)&&(t==null?void 0:t.resource)).filter(Boolean).map(t=>{if((t==null?void 0:t.relation)&&!t.fieldable&&t.resource){if(t.multiple){l.push(["useMany","@pankod/refine-core"]);let r=s(o,t.key);return t.accessor&&(r=`${s(o,t.key)}?.map((item: any) => ${s("item",void 0,t.accessor)})`),`
                    const { data: ${f(t.key,"Data")}, isLoading: ${f(t.key,"IsLoading")} } =
                    useMany({
                        resource: "${t.resource.name}",
                        ids: ${r} || [],
                        queryOptions: {
                            enabled: !!${o},
                        },
                    });
                    `}return l.push(["useOne","@pankod/refine-core"]),`
                    const { data: ${f(t.key,"Data")}, isLoading: ${f(t.key,"IsLoading")} } =
                    useOne({
                        resource: "${t.resource.name}",
                        id: ${s(o,t.key,t.accessor,!1)} || "",
                        queryOptions: {
                            enabled: !!${o},
                        },
                    });
                `}}).filter(Boolean),S=t=>{if(t.relation&&t.resource){let r=f(t.key,"Data"),i=f(t.key,"IsLoading");return t.multiple?(l.push(["TagField","@pankod/refine-mui"]),m`
                    <Typography variant="body1" fontWeight="bold">
                        ${g(t.key)}
                    </Typography>
                    {${i} ? (
                        <>
                            Loading...
                        </>
                        ) : (
                        <>
                        ${(()=>{var e;if(t.relationInfer)if((e=t.relationInfer)!=null&&e.accessor){if(Array.isArray(t.relationInfer.accessor))return"Not Handled.";{let h=f(t.key),x=s(h,void 0,t.relationInfer.accessor);return`
                                            <Stack direction="row" spacing={1}>
                                                {${r}?.data?.map((${h}: any) => (
                                                    <TagField key={${x}} value={${x}} />
                                                ))}
                                            </Stack>
                                        `}}else return"Not Handled.";else return"not-handled - relation with multiple but no resource"})()}
                        </>
                    )}
                    `):m`
                    <Typography variant="body1" fontWeight="bold">
                        ${g(t.key)}
                    </Typography>

                    {${i} ? (
                        <>Loading...</>
                    ) : (
                        <>
                        ${(()=>{var e;return t.relationInfer?(e=t.relationInfer)!=null&&e.accessor?Array.isArray(t.relationInfer.accessor)?`{${s(`${r}?.data`,void 0,t.relationInfer.accessor,' + " " + ')}}`:`{${r}?.data?.${t.relationInfer.accessor}}`:`{${r}?.data}`:`{${r}?.data?.id}`})()}
                        </>
                    )}
                    `}},V=t=>{if(t.type==="text"){if(l.push(["TextFieldComponent as TextField","@pankod/refine-mui"]),t.multiple){l.push(["TagField","@pankod/refine-mui"]);let r=s("item",void 0,t.accessor);return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${g(t.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${s(o,t.key)}?.map((item: any) => (
                        <TagField value={${r}} key={${r}} />
                    ))}
                    </Stack>
                `}return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${g(t.key)}
                    </Typography>
                    <TextField value={${s(o,t.key,t.accessor)}} />
                `}},L=t=>{if(t.type==="image"){if(t.multiple){let r=s("item",void 0,t.accessor);return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${g(t.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${s(o,t.key)}?.map((item: any) => (
                        <img style={{ maxWidth: 200, width: "100%", height: 200 }} src={${r}} key={${r}} />
                    ))}
                    </Stack>
                `}return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${g(t.key)}
                    </Typography>
                    <img style={{ maxWidth: 200, width: "100%", height: 200 }} src={${s(o,t.key,t.accessor," + ")}} />
                    `}},A=t=>{if(t.type==="email"){if(l.push(["EmailField","@pankod/refine-mui"]),t.multiple){l.push(["TagField","@pankod/refine-mui"]);let r=s("item",void 0,t.accessor);return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${g(t.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${s(o,t.key)}?.map((item: any) => (
                        <TagField value={${r}} key={${r}} />
                    ))}
                    </Stack>
                `}return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${g(t.key)}
                    </Typography>
                    <EmailField value={${s(o,t.key,t.accessor," + ")}} />
                `}},p=t=>{if(t.type==="url"){if(l.push(["UrlField","@pankod/refine-mui"]),t.multiple){l.push(["TagField","@pankod/refine-mui"]);let r=s("item",void 0,t.accessor);return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${g(t.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${s(o,t.key)}?.map((item: any) => (
                        <TagField value={${r}} key={${r}} />
                    ))}
                    </Stack>
                `}return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${g(t.key)}
                    </Typography>
                    <UrlField value={${s(o,t.key,t.accessor," + ")}} />
                `}},N=t=>{if(t.type==="boolean"){if(l.push(["BooleanField","@pankod/refine-mui"]),t.multiple){l.push(["TagField","@pankod/refine-mui"]);let r=s("item",void 0,t.accessor);return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${g(t.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${s(o,t.key)}?.map((item: any, index: number) => (
                        <TagField value={${r}} key={index} />
                    ))}
                    </Stack>
                `}return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${g(t.key)}
                    </Typography>
                    <BooleanField value={${s(o,t.key,t.accessor," && ")}} />
                `}},y=t=>{if(t.type==="date"){if(l.push(["DateField","@pankod/refine-mui"]),t.multiple){let r=s("item",void 0,t.accessor);return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${g(t.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${s(o,t.key)}?.map((item: any) => (
                        <DateField value={${r}} key={${r}} />
                    ))}
                    </Stack>
                `}return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${g(t.key)}
                    </Typography>
                    <DateField value={${s(o,t.key,t.accessor," + ' ' + ")}} />
                `}},u=t=>{if(t.type==="richtext")return l.push(["MarkdownField","@pankod/refine-mui"]),m`
                    <Typography variant="body1" fontWeight="bold">
                        ${g(t.key)}
                    </Typography>
                    <MarkdownField value={${s(o,t.key,t.accessor,' + " " + ')}} />
                `},w=t=>{if(t.type==="number"){if(l.push(["NumberField","@pankod/refine-mui"]),t.multiple){l.push(["TagField","@pankod/refine-mui"]);let r=s("item",void 0,t.accessor);return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${g(t.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${s(o,t.key)}?.map((item: any) => (
                        <TagField value={${r}} key={${r}} />
                    ))}
                    </Stack>
                `}return m`
                    <Typography variant="body1" fontWeight="bold">
                        ${g(t.key)}
                    </Typography>
                    <NumberField value={${s(o,t.key,t.accessor,' + " " + ')} ?? ""} />
                `}},a=c.map(t=>{switch(t==null?void 0:t.type){case"text":return V(t);case"number":return w(t);case"richtext":return u(t);case"email":return A(t);case"image":return L(t);case"date":return y(t);case"boolean":return N(t);case"url":return p(t);case"relation":return S(t);default:return}});return H(l),m`
        ${P(l)}

        export const ${$} = () => {
            const { queryResult } = useShow(${k?`{ 
                        resource: "${n.name}", 
                        id: ${d}
                    }`:""});
            const { data, isLoading } = queryResult;
        
            const ${o} = data?.data;
        
            ${I}

            return (
                <Show isLoading={isLoading}>
                    <Stack gap={1}>
                        ${a.join("")}
                    </Stack>
                </Show>
            );
        };
        `}});var Yn=E(require("@pankod/refine-mui"));var sr=D({type:"list",additionalScope:[["@pankod/refine-mui","RefineMui",Yn]],codeViewerComponent:he,loadingComponent:$e,errorComponent:ge,renderer:({resource:n,fields:c,isCustomPage:k})=>{var i;let d=M((i=n.label)!=null?i:n.name,"list"),$="dataGridProps?.rows",o=[["React","react",!0],["useDataGrid","@pankod/refine-mui"],["DataGrid","@pankod/refine-mui"],["GridColumns","@pankod/refine-mui"],["EditButton","@pankod/refine-mui"],["ShowButton","@pankod/refine-mui"],["DeleteButton","@pankod/refine-mui"],["List","@pankod/refine-mui"]],l=c.filter(e=>(e==null?void 0:e.relation)&&!(e!=null&&e.fieldable)&&(e==null?void 0:e.resource)),v=l.filter(Boolean).map(e=>{if((e==null?void 0:e.relation)&&!e.fieldable&&e.resource){o.push(["useMany","@pankod/refine-core"]);let h="";return e.multiple?h=`[].concat(...(${$}?.map((item: any) => ${s("item",e.key,e.accessor,!1)}) ?? []))`:h=`${$}?.map((item: any) => ${s("item",e.key,e.accessor,!1)}) ?? []`,`
                    const { data: ${f(e.key,"Data")}, isLoading: ${f(e.key,"IsLoading")} } =
                    useMany({
                        resource: "${e.resource.name}",
                        ids: ${h},
                        queryOptions: {
                            enabled: !!${$},
                        },
                    });
                    `}}).filter(Boolean),I=l==null?void 0:l.map(e=>{if(e&&e.resource)return`${f(e.key,"Data")}?.data`}).filter(Boolean),S=e=>{var h,x;if(e.relation&&e.resource){let C=`${f(e.key,"Data")}?.data`,b=f(e.key,"IsLoading");if(Array.isArray(e.accessor))return;let B=`${b} ? <>Loading...</> : `,W=`field: "${e.key}"`,K=e.accessor&&!e.multiple&&!Array.isArray(e.accessor)?`valueGetter: ({ row }) => {
                const value = ${s("row",e.key,e.accessor,!1)};

                return value;
            },`:"",ve=`headerName: "${g(e.key)}"`,oe="";if(e.multiple){o.push(["TagField","@pankod/refine-mui"]);let ie="item";if(e!=null&&e.relationInfer){let Oe=`${C}?.find((resourceItems) => resourceItems.id === ${s("item",void 0,e.accessor)})`;ie=s(Oe,void 0,(h=e==null?void 0:e.relationInfer)==null?void 0:h.accessor)}oe=`
                    renderCell: function render({ value }) {
                        return ${B} (
                            <>
                                {${s("value",void 0)}?.map((item: any, index: number) => (
                                    <TagField key={index} value={${ie}} />
                                ))}
                            </>
                        )
                    }
                    `}else if(e!=null&&e.relationInfer){let ie=`${C}?.find((item) => item.id === value)`,Oe=s(ie,void 0,(x=e==null?void 0:e.relationInfer)==null?void 0:x.accessor);oe=`
                        renderCell: function render({ value }) {
                            return ${B} ${Oe};
                        }
                        `}else oe="";return`
                    {
                        ${W},
                        ${ve},${K}
                        minWidth: 300,${oe}
                    }
                `}},V=e=>{if(e.type==="image"){let h=`field: "${e.key}"`,x=`headerName: "${g(e.key)}"`,C=e.accessor&&!e.multiple&&!Array.isArray(e.accessor)?`valueGetter: ({ row }) => {
                const value = ${s("row",e.key,e.accessor,!1)};

                return value;
            },`:"",b=`
                    renderCell: function render({ value }) {
                        return (
                            <img src={${s("value",void 0,Array.isArray(e.accessor)?e.accessor:void 0," + ")}} style={{ height: "50px", maxWidth: "100px" }} />
                        )
                    }
                `;return e.multiple&&(b=`
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item: any, index: number) => (
                                    <img src={${s("item",void 0,e.accessor," + ")}} key={index} style={{ height: "50px", maxWidth: "100px" }} />
                                ))}
                                </>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},${C}
                        minWidth: 100,${b}
                    }
                `}},L=e=>{if(e.type==="email"){o.push(["EmailField","@pankod/refine-mui"]);let h=`field: "${e.key}"`,x=`headerName: "${g(e.key)}"`,C=e.accessor&&!e.multiple&&!Array.isArray(e.accessor)?`valueGetter: ({ row }) => {
                const value = ${s("row",e.key,e.accessor,!1)};

                return value;
            },`:"",b=`
                renderCell: function render({ value }) {
                    return (
                        <EmailField value={${s("value",void 0,Array.isArray(e.accessor)?e.accessor:void 0," + ")}} />
                    )
                }
            `;return e.multiple&&(o.push(["TagField","@pankod/refine-mui"]),b=`
                    renderCell: function render({ value }) {
                        return (
                            <>
                            {value?.map((item: any, index: number) => (
                                <TagField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                            ))}
                            </>
                        )
                    }
                `),`
                {
                    ${h},
                    ${x},${C}
                    minWidth: 250,${b}
                }
            `}},A=e=>{if(e.type==="url"){o.push(["UrlField","@pankod/refine-mui"]);let h=`field: "${e.key}"`,x=`headerName: "${g(e.key)}"`,C=e.accessor&&!e.multiple&&!Array.isArray(e.accessor)?`valueGetter: ({ row }) => {
                const value = ${s("row",e.key,e.accessor,!1)};

                return value;
            },`:"",b=`
                    renderCell: function render({ value }) {
                        return (
                            <UrlField value={${s("value",void 0,Array.isArray(e.accessor)?e.accessor:void 0," + ")}} />
                        )
                    }
                `;return e.multiple&&(o.push(["TagField","@pankod/refine-mui"]),b=`
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item: any, index: any) => (
                                    <TagField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                ))}
                                </>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},${C}
                        minWidth: 250,${b}
                    }
                `}},p=e=>{if(e!=null&&e.type){o.push(["Checkbox","@pankod/refine-mui"]);let h=`field: "${e.key}"`,x=`headerName: "${g(e.key)}"`,C=e.accessor&&!e.multiple&&!Array.isArray(e.accessor)?`valueGetter: ({ row }) => {
                const value = ${s("row",e.key,e.accessor,!1)};

                return value;
            },`:"",b=`
                    renderCell: function render({ value }) {
                        return (
                            <Checkbox checked={!!${s("value",void 0,Array.isArray(e.accessor)?e.accessor:void 0," && ")}} />
                        );
                    }
                `;return e.multiple&&(b=`
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item: any, index: number) => (
                                    <Checkbox checked={!!${s("item",void 0,e.accessor," && ")}} key={index} />
                                ))}
                                </>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},${C}
                        minWidth: 250,${b}
                    }
                `}},N=e=>{if(e.type==="date"){o.push(["DateField","@pankod/refine-mui"]);let h=`field: "${e.key}"`,x=`headerName: "${g(e.key)}"`,C=e.accessor&&!e.multiple&&!Array.isArray(e.accessor)?`valueGetter: ({ row }) => {
                const value = ${s("row",e.key,e.accessor,!1)};

                return value;
            },`:"",b=`
                    renderCell: function render({ value }) {
                        return <DateField value={value} />;
                    }
                `;return e.multiple&&(b=`
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item: any, index: number) => (
                                    <DateField value={${s("item",void 0,e.accessor,' + " " + ')}} key={index} />
                                ))}
                                </>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},${C}
                        minWidth: 250,${b}
                    }
                `}},y=e=>{if((e==null?void 0:e.type)==="richtext"){o.push(["MarkdownField","@pankod/refine-mui"]);let h=`field: "${e.key}"`,x=`headerName: "${g(e.key)}"`,C=e.accessor&&!e.multiple&&!Array.isArray(e.accessor)?`valueGetter: ({ row }) => {
                const value = ${s("row",e.key,e.accessor,!1)};

                return value;
            },`:"",b=`
                    renderCell: function render({ value }) {
                        return <MarkdownField value={(value ?? "").slice(0, 80) + "..."} />;
                    }
                `;return e.multiple&&(b=`
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item: any, index: number) => (
                                    <MarkdownField value={(${s("item",void 0,e.accessor,' + " " + ')}).slice(0, 80) + "..."} key={index} />
                                ))}
                                </>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},${C}
                        minWidth: 250,${b}
                    }
                `}},u=e=>{if(e&&(e.type==="text"||e.type==="number")){let h=`field: "${e.key}"`,x=`headerName: "${g(e.key)}"`,C=e.accessor&&!e.multiple&&!Array.isArray(e.accessor)?`valueGetter: ({ row }) => {
                const value = ${s("row",e.key,e.accessor,!1)};

                return value;
            },`:"",b="";if(e.multiple){o.push(["TagField","@pankod/refine-mui"]);let B=s("item",void 0,e.accessor,' + " " + ');b=`
                    renderCell: function render({ row }) {
                        return (
                            <>
                                {(${s("row",e.key)})?.map((item: any) => (
                                    <TagField value={${B}} key={${B}} />
                                ))}
                            </>
                        )
                    }
                    `}return!e.multiple&&Array.isArray(e.accessor)&&(b=`
                    renderCell: function render({ row }) {
                        return 
                            <>{${s("row",e.key,e.accessor)}}</>
                        );
                    }
                    `),`
                {
                    ${h},
                    ${x},${C}${e.type==="number"?"type: 'number',":""}
                    minWidth: ${j(e.key)?50:200},${b}
                }
                `}},{canEdit:w,canShow:a,canDelete:T}=n!=null?n:{};w&&o.push(["EditButton","@pankod/refine-mui"]),a&&o.push(["ShowButton","@pankod/refine-mui"]),T&&o.push(["DeleteButton","@pankod/refine-mui"]);let t=w||a||T?m`
                {
                    field: "actions",
                    headerName: "Actions",
                    renderCell: function render({ row }) {
                        return (
                            <>
                                ${w?m`<EditButton hideText recordItemId={row.id} />`:""}
                                ${a?m`<ShowButton hideText recordItemId={row.id} />`:""}
                                ${T?m`<DeleteButton hideText recordItemId={row.id} />`:""}
                            </>
                        );
                    },
                    align: "center",
                    headerAlign: "center",
                    minWidth: 80,
                },
        `:"",r=c.map(e=>{switch(e==null?void 0:e.type){case"text":case"number":return u(e);case"richtext":return y(e);case"email":return L(e);case"image":return V(e);case"date":return N(e);case"boolean":return p(e);case"url":return A(e);case"relation":return S(e);default:return}});return H(o),m`
        ${P(o)}
        
        export const ${d} = () => {
            const { dataGridProps } = useDataGrid(
                ${k?`{ resource: "${n.name}" }`:""} 
            );
        
            ${v}

            const columns = React.useMemo<GridColumns<any>>(() => [
                ${[...r,t].filter(Boolean).join(`,\r
`)}
            ], [${I.join(",")}]);

            return (
                <List>
                    <DataGrid {...dataGridProps} columns={columns} autoHeight />
                </List>
            );
        };
        `}});var et=E(require("@pankod/refine-mui")),rt=E(require("@pankod/refine-react-hook-form"));var ir=D({type:"create",additionalScope:[["@pankod/refine-mui","RefineMui",et],["@pankod/refine-react-hook-form","RefineReactHookForm",rt]],codeViewerComponent:he,loadingComponent:$e,errorComponent:ge,renderer:({resource:n,fields:c,isCustomPage:k})=>{var A;let d=M((A=n.label)!=null?A:n.name,"create"),$=[["Create","@pankod/refine-mui"],["Box","@pankod/refine-mui"],["useForm","@pankod/refine-react-hook-form"]],l=c.filter(p=>(p==null?void 0:p.relation)&&!(p!=null&&p.fieldable)&&(p==null?void 0:p.resource)).filter(Boolean).map(p=>{if((p==null?void 0:p.relation)&&!p.fieldable&&p.resource)return $.push(["useAutocomplete","@pankod/refine-mui"]),`
                    const { autocompleteProps: ${f(p.key,"AutocompleteProps")} } =
                    useAutocomplete({
                        resource: "${p.resource.name}",
                    });
                `}).filter(Boolean),v=p=>{if(p.relation&&p.resource){$.push(["Autocomplete","@pankod/refine-mui"],["Controller","@pankod/refine-react-hook-form"]);let N=f(p.key,"AutocompleteProps"),y=p.relationInfer&&p.relationInfer.accessor?typeof p.relationInfer.accessor=="string"?p.relationInfer.accessor:p.relationInfer.accessor[0]:"title",u=y.includes(".")||y.includes("[")||y.includes("]")||y.includes("-"),w=p.accessor?s("item",void 0,p.accessor,!1):"(item?.id ?? item)",a=p.accessor?s("value",void 0,p.accessor,!1):"(value?.id ?? value)",T=p.accessor?"value":p.multiple?"value?.map((item: any) => item?.id ?? item)":"value?.id ?? value";return m`
                    <Controller
                        control={control}
                        name="${R(p.key,void 0)}"
                        rules={{ required: "This field is required" }}
                        // eslint-disable-next-line
                        ${p.multiple?"defaultValue={[] as any}":"defaultValue={null as any}"} 
                        render={({ field }) => (
                            <Autocomplete
                                {...${N}}
                                {...field}
                                ${p.multiple?"multiple":""}
                                onChange={(_, value) => {
                                    field.onChange(${T});
                                }}
                                getOptionLabel={(item) => {
                                    return (
                                        ${N}?.options?.find(
                                            (p) =>
                                                p?.id?.toString() ===
                                                ${w}?.toString(),
                                        )?.${u?`["${y}"]`:y} ?? ""
                                    );
                                }}
                                isOptionEqualToValue={(option, value) =>
                                    value === undefined ||
                                    option.id.toString() === ${a}?.toString()
                                }
                                renderInput={(params) => (
                                    <TextField
                                        {...params}
                                        label="${g(p.key)}"
                                        margin="normal"
                                        variant="outlined"
                                        error={!!${s("(errors as any)",p.key,p.accessor,!1)}}
                                        helperText={${s("(errors as any)",p.key,p.accessor,!1)}?.message}
                                        required
                                    />
                                )}
                            />
                        )}
                    />
                `}},I=p=>{if(p.type==="text"||p.type==="url"||p.type==="email"||p.type==="number"||p.type==="date"||p.type==="richtext")return j(p.key)||($.push(["TextField","@pankod/refine-mui"]),p.multiple)?void 0:m`
                    <TextField
                        {...register("${R(p.key,void 0,p.accessor)}", {
                            required: "This field is required",
                        })}
                        error={!!${s("(errors as any)",p.key,p.accessor,!1)}}
                        helperText={${s("(errors as any)",p.key,p.accessor,!1)}?.message}
                        margin="normal"
                        fullWidth
                        InputLabelProps={{ shrink: true }}
                        ${p.type!=="date"&&p.type!=="richtext"?`type="${p.type}"`:""}
                        ${p.type==="richtext"?"multiline":""}
                        label="${g(p.key)}"
                        name="${R(p.key,void 0,p.accessor)}"
                    />
                `},S=p=>{if(p.type==="boolean")return $.push(["Checkbox","@pankod/refine-mui"],["FormControlLabel","@pankod/refine-mui"],["Controller","@pankod/refine-react-hook-form"]),p.multiple?void 0:m`
                    <Controller
                        control={control}
                        name="${R(p.key,void 0,p.accessor)}"
                        // eslint-disable-next-line
                        defaultValue={null as any}
                        render={({ field }) => (
                            <FormControlLabel label="${g(p.key)}" control={
                                <Checkbox
                                    {...field}
                                    checked={field.value}
                                    onChange={(event) => {
                                        field.onChange(event.target.checked);
                                    }}
                                />
                            } />
                        )}
                    />
                `},V=p=>{if(p.type==="date"){let N=I(p);return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-mui" package.
                        To use a <DatePicker> component, you can follow the official documentation for Material UI.
                        
                        Docs: https://mui.com/x/react-date-pickers/date-picker/#basic-usage
                    */}
                    ${N!=null?N:""}
                    `}},L=c.map(p=>{switch(p==null?void 0:p.type){case"text":case"number":case"email":case"url":case"richtext":return I(p);case"date":return V(p);case"boolean":return S(p);case"relation":return v(p);default:return}});return H($),m`
        ${P($)}
        
        export const ${d} = () => {
            const {
                saveButtonProps,
                refineCore: { formLoading },
                register,
                control,
                formState: { errors },
            } = useForm(
                ${k?`{ 
                    refineCoreProps: {
                        resource: "${n.name}",
                        action: "create",
                    }
                }`:""}
            );
        
            ${l}

            return (
                <Create isLoading={formLoading} saveButtonProps={saveButtonProps}>
                    <Box
                        component="form"
                        sx={{ display: "flex", flexDirection: "column" }}
                        autoComplete="off"
                    >
                        ${L.join("")}
                    </Box>
                </Create>
            );
        };
        `}});var nt=E(require("@pankod/refine-mui")),tt=E(require("@pankod/refine-react-hook-form"));var cr=D({type:"edit",additionalScope:[["@pankod/refine-mui","RefineMui",nt],["@pankod/refine-react-hook-form","RefineReactHookForm",tt]],codeViewerComponent:he,loadingComponent:$e,errorComponent:ge,renderer:({resource:n,fields:c,isCustomPage:k,id:d})=>{var N,y;let $=M((N=n.label)!=null?N:n.name,"edit"),o=f((y=n.label)!=null?y:n.name,"Data"),l=[["Edit","@pankod/refine-mui"],["Box","@pankod/refine-mui"],["useForm","@pankod/refine-react-hook-form"]],I=c.filter(u=>(u==null?void 0:u.relation)&&!(u!=null&&u.fieldable)&&(u==null?void 0:u.resource)).filter(Boolean).map(u=>{if((u==null?void 0:u.relation)&&!u.fieldable&&u.resource){l.push(["useAutocomplete","@pankod/refine-mui"]);let w=s(o,u.key,u.accessor,!1);return u.multiple&&u.accessor&&(w=`${s(o,u.key)}?.map((item: any) => ${s("item",void 0,u.accessor)})`),`
                    const { autocompleteProps: ${f(u.key,"AutocompleteProps")} } =
                    useAutocomplete({
                        resource: "${u.resource.name}",
                        defaultValue: ${w},
                    });
                `}}).filter(Boolean),S=u=>{if(u.relation&&u.resource){l.push(["Autocomplete","@pankod/refine-mui"],["Controller","@pankod/refine-react-hook-form"]);let w=f(u.key,"AutocompleteProps"),a=u.relationInfer&&u.relationInfer.accessor?typeof u.relationInfer.accessor=="string"?u.relationInfer.accessor:u.relationInfer.accessor[0]:"title",T=a.includes(".")||a.includes("[")||a.includes("]")||a.includes("-"),t=u.accessor?s("item",void 0,u.accessor,!1):"(item?.id ?? item)",r=u.accessor?s("value",void 0,u.accessor,!1):"(value?.id ?? value)",i=u.accessor?"value":u.multiple?"value?.map((item: any) => item?.id ?? item)":"value?.id ?? value";return m`
                    <Controller
                        control={control}
                        name="${R(u.key,void 0)}"
                        rules={{ required: "This field is required" }}
                        // eslint-disable-next-line
                        ${u.multiple?"defaultValue={[] as any}":"defaultValue={null as any}"} 
                        render={({ field }) => (
                            <Autocomplete
                                {...${w}}
                                {...field}
                                ${u.multiple?"multiple":""}
                                onChange={(_, value) => {
                                    field.onChange(${i});
                                }}
                                getOptionLabel={(item) => {
                                    return (
                                        ${w}?.options?.find(
                                            (p) =>
                                                p?.id?.toString() ===
                                                ${t}?.toString(),
                                        )?.${T?`["${a}"]`:a} ?? ""
                                    );
                                }}
                                isOptionEqualToValue={(option, value) =>
                                    value === undefined ||
                                    option?.id?.toString() === ${r}?.toString()
                                }
                                renderInput={(params) => (
                                    <TextField
                                        {...params}
                                        label="${g(u.key)}"
                                        margin="normal"
                                        variant="outlined"
                                        error={!!${s("(errors as any)",u.key,u.accessor,!1)}}
                                        helperText={${s("(errors as any)",u.key,u.accessor,!1)}?.message}
                                        required
                                    />
                                )}
                            />
                        )}
                    />
                `}},V=u=>{if(u.type==="text"||u.type==="url"||u.type==="email"||u.type==="number"||u.type==="date"||u.type==="richtext"){if(l.push(["TextField","@pankod/refine-mui"]),u.multiple){l.push(["Box","@pankod/refine-mui"]);let w=R(u.key,"${index}",u.accessor),a=s("(errors as any)",u.key,void 0,!1)+"?.[index]";return`
                        <Box sx={{display: "flex", gap: 1}}>
                            {${s(o,u.key)}?.map((item: any, index: number) => (
                                <TextField
                                    key={index}
                                    {...register(\`${w}\`, {
                                        required: "This field is required",
                                    })}
                                    error={!!${a}}
                                    helperText={${a}?.message as string}
                                    margin="normal"
                                    fullWidth
                                    InputLabelProps={{ shrink: true }}
                                    ${u.type!=="date"&&u.type!=="richtext"?`type="${u.type}"`:""}
                                    ${u.type==="richtext"?"multiline":""}
                                    label={\`${g(u.key)} \${index+1}\`}
                                    name={\`${R(u.key,void 0,u.accessor)}.\${index}\`}
                                />
                            ))}
                        </Box>
                    `}return m`
                    <TextField
                        {...register("${R(u.key,void 0,u.accessor)}", {
                            required: "This field is required",
                        })}
                        error={!!${s("(errors as any)",u.key,u.accessor,!1)}}
                        helperText={${s("(errors as any)",u.key,u.accessor,!1)}?.message}
                        margin="normal"
                        fullWidth
                        InputLabelProps={{ shrink: true }}
                        ${u.type!=="date"&&u.type!=="richtext"?`type="${u.type}"`:""}
                        ${u.type==="richtext"?"multiline":""}
                        label="${g(u.key)}"
                        name="${R(u.key,void 0,u.accessor)}"
                        ${j(u.key)?"disabled":""}
                    />
                `}},L=u=>{if(u.type==="boolean"){if(l.push(["Checkbox","@pankod/refine-mui"],["FormControlLabel","@pankod/refine-mui"],["Controller","@pankod/refine-react-hook-form"]),u.multiple){l.push(["Box","@pankod/refine-mui"]);let w=R(u.key,"${index}",u.accessor);return`
                        <Box sx={{display: "flex", gap: 1}}>
                            {${s(o,u.key)}?.map((item: any, index: number) => (
                                <Controller
                                    key={index}
                                    control={control}
                                    name={\`${w}\`}
                                    // eslint-disable-next-line
                                    defaultValue={null as any}
                                    render={({ field }) => (
                                        <FormControlLabel label={\`${g(u.key)} \${index+1}\`} control={
                                            <Checkbox
                                                {...field}
                                                checked={field.value}
                                                onChange={(event) => {
                                                    field.onChange(event.target.checked);
                                                }}
                                            />
                                        } />
                                    )}
                                />
                            ))}
                        </Box>
                    `}return m`
                    <Controller
                        control={control}
                        name="${R(u.key,void 0,u.accessor)}"
                        // eslint-disable-next-line
                        defaultValue={null as any}
                        render={({ field }) => (
                            <FormControlLabel label="${g(u.key)}" control={
                                <Checkbox
                                    {...field}
                                    checked={field.value}
                                    onChange={(event) => {
                                        field.onChange(event.target.checked);
                                    }}
                                />
                            } />
                        )}
                    />
                `}},A=u=>{if(u.type==="date")return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-mui" package.
                        To use a <DatePicker> component, you can follow the official documentation for Material UI.
                        
                        Docs: https://mui.com/x/react-date-pickers/date-picker/#basic-usage
                    */}
                    ${V(u)}
                    `},p=c.map(u=>{switch(u==null?void 0:u.type){case"text":case"number":case"email":case"url":case"richtext":return V(u);case"date":return A(u);case"boolean":return L(u);case"relation":return S(u);default:return}});return H(l),m`
        ${P(l)}
        
        export const ${$} = () => {
            const {
                saveButtonProps,
                refineCore: { queryResult },
                register,
                control,
                formState: { errors },
            } = useForm(
                ${k?`{ 
                    refineCoreProps: {
                        resource: "${n.name}",
                        id: ${d},
                        action: "edit",
                    }
                }`:""}
            );
        
            const ${o} = queryResult?.data?.data;
        
            ${I}

            return (
                <Edit saveButtonProps={saveButtonProps}>
                    <Box
                        component="form"
                        sx={{ display: "flex", flexDirection: "column" }}
                        autoComplete="off"
                    >
                        ${p.join("")}
                    </Box>
                </Edit>
            );
        };
        `}});var ot=({action:n,id:c,...k})=>{let{useParams:d}=(0,hn.useRouterContext)(),{action:$,id:o}=d();switch(n!=null?n:$){case"show":return De.default.createElement(ar,{...k,id:c!=null?c:o});case"create":return De.default.createElement(ir,{...k,id:c!=null?c:o});case"edit":return De.default.createElement(cr,{...k,id:c!=null?c:o});default:return De.default.createElement(sr,{...k,id:c!=null?c:o})}};var He=E(require("react")),In=require("@pankod/refine-core");var at=E(require("@pankod/refine-chakra-ui"));var Pe=E(require("react")),Ie=require("@pankod/refine-chakra-ui"),fe=({error:n})=>n?Pe.default.createElement(Ie.Center,{style:{minHeight:300}},Pe.default.createElement(Ie.Alert,{status:"error",variant:"subtle",flexDirection:"column",alignItems:"center",justifyContent:"center",textAlign:"center",height:"200px"},Pe.default.createElement(Ie.AlertIcon,{boxSize:"40px",mr:0}),Pe.default.createElement(Ie.AlertDescription,{mt:4,maxWidth:"sm"},n!=null?n:""))):null;var Nr=E(require("react")),ur=require("@pankod/refine-chakra-ui"),xe=()=>Nr.default.createElement(ur.Box,{position:"relative",bg:"chakra-body-bg",minH:120},Nr.default.createElement(ur.Spinner,{position:"absolute",top:"50%",left:"50%",transform:"translate(-50%, -50%)"}));var G=E(require("react")),U=require("@pankod/refine-chakra-ui"),Fe=require("@tabler/icons");var be=({code:n,loading:c})=>{let k=G.default.useMemo(()=>ae(n!=null?n:""),[n]),d=G.default.useRef(null),[$,o]=G.default.useState(!1),{isOpen:l,onOpen:v,onClose:I}=(0,U.useDisclosure)();if(c)return null;let S=()=>{var V,L,A,p;(V=d==null?void 0:d.current)==null||V.select(),(L=d==null?void 0:d.current)==null||L.setSelectionRange(0,Number.MAX_SAFE_INTEGER),typeof navigator<"u"&&(navigator.clipboard.writeText((p=(A=d==null?void 0:d.current)==null?void 0:A.value)!=null?p:""),o(!0),setTimeout(()=>{o(!1)},1e3))};return k?G.default.createElement(G.default.Fragment,null,G.default.createElement(U.Box,{position:"fixed",bottom:"12px",right:"60px"},G.default.createElement(U.HStack,{spacing:4},G.default.createElement(U.Button,{variant:"solid",colorScheme:"green",leftIcon:G.default.createElement(Fe.IconCode,{size:18}),onClick:v},"Show Code"),G.default.createElement(U.Button,{colorScheme:"gray",variant:"solid",leftIcon:G.default.createElement(Fe.IconMessageCircle,{size:18}),as:"a",target:"_blank",href:"https://github.com/refinedev/refine/discussions/3046"},"Give Feedback"))),G.default.createElement("div",{style:{height:"50px"}}),G.default.createElement(U.Modal,{size:"4xl",isOpen:l,onClose:I},G.default.createElement(U.ModalOverlay,null),G.default.createElement(U.ModalContent,null,G.default.createElement(U.ModalCloseButton,{sx:{position:"absolute",right:0,top:0,height:"56px",width:"56px",color:"#666b7a"}}),G.default.createElement(U.ModalBody,{sx:{padding:0}},G.default.createElement(se,{code:k}),G.default.createElement("textarea",{ref:d,defaultValue:(k!=null?k:"").replace(/\\n/g,`\r
`),id:"code-input",style:{padding:0,margin:0,width:0,height:0,opacity:0,border:"none",display:"block"}})),G.default.createElement(U.Button,{sx:{position:"absolute",bottom:"12px",right:"12px",width:100},key:"copy",leftIcon:$?G.default.createElement(Fe.IconCheck,{size:18,style:{color:"green"}}):G.default.createElement(Fe.IconCopy,{size:18}),onClick:S},$?"Copied":"Copy")))):null};var pr=D({type:"show",additionalScope:[["@pankod/refine-chakra-ui","RefineChakraUI",at]],codeViewerComponent:be,loadingComponent:xe,errorComponent:fe,renderer:({resource:n,fields:c,isCustomPage:k,id:d})=>{var t;let $=M((t=n.label)!=null?t:n.name,"show"),o="record",l=[["useShow","@pankod/refine-core"],["Show","@pankod/refine-chakra-ui"],["Heading","@pankod/refine-chakra-ui"]],I=c.filter(r=>(r==null?void 0:r.relation)&&!(r!=null&&r.fieldable)&&(r==null?void 0:r.resource)).filter(Boolean).map(r=>{if((r==null?void 0:r.relation)&&!r.fieldable&&r.resource){if(r.multiple){l.push(["useMany","@pankod/refine-core"]);let i=s(o,r.key);return r.accessor&&(i=`${s(o,r.key)}?.map((item: any) => ${s("item",void 0,r.accessor)})`),`
                    const { data: ${f(r.key,"Data")}, isLoading: ${f(r.key,"IsLoading")} } =
                    useMany({
                        resource: "${r.resource.name}",
                        ids: ${i} || [],
                        queryOptions: {
                            enabled: !!${o},
                        },
                    });
                    `}return l.push(["useOne","@pankod/refine-core"]),`
                    const { data: ${f(r.key,"Data")}, isLoading: ${f(r.key,"IsLoading")} } =
                    useOne({
                        resource: "${r.resource.name}",
                        id: ${s(o,r.key,r.accessor,!1)} || "",
                        queryOptions: {
                            enabled: !!${o},
                        },
                    });
                `}}).filter(Boolean),S=r=>{if(r.relation&&r.resource){let i=f(r.key,"Data"),e=f(r.key,"IsLoading");return r.multiple?(l.push(["TagField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),m`
                    <Heading as="h5" size="sm" mt={4} >${g(r.key)}</Heading>
                    {${e} ? <>Loading...</> : (
                        <>
                        ${(()=>{var h;if(r.relationInfer)if((h=r.relationInfer)!=null&&h.accessor){if(Array.isArray(r.relationInfer.accessor))return"Not Handled.";{let x=f(r.key),C=s(x,void 0,r.relationInfer.accessor);return`<HStack spacing="12px">{${i}?.data?.map((${x}: any) => <TagField key={${C}} value={${C}} />)}</HStack>`}}else return"Not Handled.";else return"not-handled - relation with multiple but no resource"})()}
                        </>
                    )}
                    `):m`
                    <Heading as="h5" size="sm" mt={4} >${g(r.key)}</Heading>
                    {${e} ? <>Loading...</> : (
                        <>
                        ${(()=>{var h;return r.relationInfer?(h=r.relationInfer)!=null&&h.accessor?Array.isArray(r.relationInfer.accessor)?`{${s(`${i}?.data`,void 0,r.relationInfer.accessor,' + " " + ')}}`:`{${i}?.data?.${r.relationInfer.accessor}}`:`{${i}?.data}`:`{${i}?.data?.id}`})()}
                        </>
                    )}
                    
                    `}},V=r=>{if(r.type==="text"){if(l.push(["TagField","@pankod/refine-chakra-ui"],["TextField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),r.multiple){let i=s("item",void 0,r.accessor);return m`
                    <Heading as="h5" size="sm" mt={4} >${g(r.key)}</Heading>
                    <HStack spacing="12px">
                    {${s(o,r.key)}?.map((item: any) => (
                        <TagField value={${i}} key={${i}} />
                    ))}
                    </HStack>
                `}return m`
                    <Heading as="h5" size="sm" mt={4} >${g(r.key)}</Heading>
                    <TextField value={${s(o,r.key,r.accessor)}} />
                `}},L=r=>{if(r.type==="image"){if(l.push(["Image","@pankod/refine-chakra-ui"]),r.multiple){let i=s("item",void 0,r.accessor);return m`
                    <Heading as="h5" size="sm" mt={4} >${g(r.key)}</Heading>
                    {${s(o,r.key)}?.map((item: any) => (
                        <Image sx={{ maxWidth: 200 }} src={${i}} key={${i}} />
                    ))}
                `}return m`
                    <Heading as="h5" size="sm" mt={4} >${g(r.key)}</Heading>
                    <Image sx={{ maxWidth: 200 }} src={${s(o,r.key,r.accessor," + ")}} />
                    `}},A=r=>{if(r.type==="email"){if(l.push(["TagField","@pankod/refine-chakra-ui"],["EmailField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),r.multiple){let i=s("item",void 0,r.accessor);return m`
                    <Heading as="h5" size="sm" mt={4} >${g(r.key)}</Heading>
                    <HStack spacing="12px">
                    {${s(o,r.key)}?.map((item: any) => (
                        <TagField value={${i}} key={${i}} />
                    ))}
                    </HStack>
                `}return m`
                    <Heading as="h5" size="sm" mt={4} >${g(r.key)}</Heading>
                    <EmailField value={${s(o,r.key,r.accessor," + ")}} />
                `}},p=r=>{if(r.type==="url"){if(l.push(["TagField","@pankod/refine-chakra-ui"],["UrlField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),r.multiple){let i=s("item",void 0,r.accessor);return m`
                    <Heading as="h5" size="sm" mt={4} >${g(r.key)}</Heading>
                    <HStack spacing="12px">
                    {${s(o,r.key)}?.map((item: any) => (
                        <TagField value={${i}} key={${i}} />
                    ))}
                    </HStack>
                `}return m`
                    <Heading as="h5" size="sm" mt={4} >${g(r.key)}</Heading>
                    <UrlField value={${s(o,r.key,r.accessor," + ")}} />
                `}},N=r=>{if(r.type==="boolean"){if(l.push(["TagField","@pankod/refine-chakra-ui"],["BooleanField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),r.multiple){let i=s("item",void 0,r.accessor);return m`
                    <Heading as="h5" size="sm" mt={4} >${g(r.key)}</Heading>
                    <HStack spacing="12px">
                    {${s(o,r.key)}?.map((item: any) => (
                        <TagField value={${i}} key={${i}} />
                    ))}
                    </HStack>
                `}return m`
                    <Heading as="h5" size="sm" mt={4} >${g(r.key)}</Heading>
                    <BooleanField value={${s(o,r.key,r.accessor," && ")}} />
                `}},y=r=>{if(r.type==="date"){if(l.push(["DateField","@pankod/refine-chakra-ui"]),r.multiple){let i=s("item",void 0,r.accessor);return m`
                    <Heading as="h5" size="sm" mt={4} >${g(r.key)}</Heading>
                    {${s(o,r.key)}?.map((item: any) => (
                        <DateField value={${i}} key={${i}} />
                    ))}
                `}return m`
                    <Heading as="h5" size="sm" mt={4} >${g(r.key)}</Heading>
                    <DateField value={${s(o,r.key,r.accessor," + ' ' + ")}} />
                `}},u=r=>{if(r.type==="richtext")return l.push(["MarkdownField","@pankod/refine-chakra-ui"]),m`
                    <Heading as="h5" size="sm" mt={4}>${g(r.key)}</Heading>
                    <MarkdownField value={${s(o,r.key,r.accessor,' + " " + ')}} />
                `},w=r=>{if(r.type==="number"){if(l.push(["NumberField","@pankod/refine-chakra-ui"],["TagField","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"]),r.multiple){let i=s("item",void 0,r.accessor);return m`
                    <Heading as="h5" size="sm" mt={4} >${g(r.key)}</Heading>
                    <HStack spacing="12px">
                    {${s(o,r.key)}?.map((item: any) => (
                        <TagField value={${i}} key={${i}} />
                    ))}
                    </HStack>
                `}return m`
                    <Heading as="h5" size="sm" mt={4} >${g(r.key)}</Heading>
                    <NumberField value={${s(o,r.key,r.accessor,' + " " + ')} ?? ""} />
                `}},a=r=>{if(r)return m`
                    ${r}
                `},T=c.map(r=>{switch(r==null?void 0:r.type){case"text":return a(V(r));case"number":return a(w(r));case"richtext":return a(u(r));case"email":return a(A(r));case"image":return a(L(r));case"date":return a(y(r));case"boolean":return a(N(r));case"url":return a(p(r));case"relation":return a(S(r));default:return}});return H(l),m`
        ${P(l)}
        
        export const ${$} = () => {
            const { queryResult } = useShow(${k?`{ 
                        resource: "${n.name}", 
                        id: ${d}
                    }`:""});
            const { data, isLoading } = queryResult;

            const ${o} = data?.data;
        
            ${I}

            return (
                <Show isLoading={isLoading}>
                    ${T.join("")}
                </Show>
            );
        };
        `}});var st=E(require("@pankod/refine-chakra-ui")),it=E(require("@pankod/refine-react-table")),ct=E(require("@tabler/icons"));var Ce=n=>Array.isArray(n.accessor)||n.multiple?`accessorKey: "${n.key}"`:n.accessor?`accessorKey: "${R(n.key,void 0,n.accessor)}"`:`accessorKey: "${n.key}"`,mr=D({type:"list",additionalScope:[["@pankod/refine-chakra-ui","RefineChakraUI",st],["@pankod/refine-react-table","RefineReactTable",it],["@tabler/icons","TablerIcons",ct]],codeViewerComponent:be,loadingComponent:xe,errorComponent:fe,renderer:({resource:n,fields:c,isCustomPage:k})=>{var i;let d=M((i=n.label)!=null?i:n.name,"list"),$="tableData?.data",o=[["IResourceComponentsProps","@pankod/refine-core"],["useTable","@pankod/refine-react-table"],["ColumnDef","@pankod/refine-react-table"],["flexRender","@pankod/refine-react-table"],["List","@pankod/refine-chakra-ui"],["TableContainer","@pankod/refine-chakra-ui"],["Table","@pankod/refine-chakra-ui"],["Thead","@pankod/refine-chakra-ui"],["Tr","@pankod/refine-chakra-ui"],["Th","@pankod/refine-chakra-ui"],["Tbody","@pankod/refine-chakra-ui"],["Tr","@pankod/refine-chakra-ui"],["Td","@pankod/refine-chakra-ui"],["HStack","@pankod/refine-chakra-ui"],["Button","@pankod/refine-chakra-ui"],["IconButton","@pankod/refine-chakra-ui"],["usePagination","@pankod/refine-chakra-ui"],["Box","@pankod/refine-chakra-ui"],["IconChevronRight","@tabler/icons"],["IconChevronLeft","@tabler/icons"]],l=c.filter(e=>(e==null?void 0:e.relation)&&!(e!=null&&e.fieldable)&&(e==null?void 0:e.resource)),v=l.filter(Boolean).map(e=>{if((e==null?void 0:e.relation)&&!e.fieldable&&e.resource){o.push(["GetManyResponse","@pankod/refine-core"]),o.push(["useMany","@pankod/refine-core"]);let h="";return e.multiple?h=`[].concat(...(${$}?.map((item) => ${s("item",e.key,e.accessor,!1)}) ?? []))`:h=`${$}?.map((item) => ${s("item",e.key,e.accessor,!1)}) ?? []`,`
                    const { data: ${f(e.key,"Data")} } =
                    useMany({
                        resource: "${e.resource.name}",
                        ids: ${h},
                        queryOptions: {
                            enabled: !!${$},
                        },
                    });
                    `}}).filter(Boolean),I=l==null?void 0:l.map(e=>{if(e&&e.resource)return f(e.key,"Data")}).filter(Boolean),S=e=>{var h;if(e.relation&&e.resource){let x=`${f(e.key,"Data")}?.data`;if(Array.isArray(e.accessor))return;let C=`id: "${e.key}"`,b=`header: "${g(e.key)}"`,B=Ce(e),W="";if(e.multiple){o.push(["TagField","@pankod/refine-chakra-ui"]);let K="item";e!=null&&e.relationInfer&&(K=s("item",void 0,e.relationInfer.accessor)),W=`cell: function render({ getValue, table }) {
                        const meta = table.options.meta as {
                            ${f(e.key,"Data")}: GetManyResponse;
                        };

                        const ${f(e.key,"")} = getValue<any[]>()?.map((item) => {
                            return meta.${f(e.key,"Data")}?.data?.find(
                                (resourceItems) => resourceItems.id === ${s("item",void 0,e.accessor)}
                            );
                        })


                        return (
                            <HStack>
                                {${f(e.key,"")}?.map((item, index) => (
                                    <TagField key={index} value={${K}} />
                                ))}
                            </HStack>
                        )
                    }
                `}else e!=null&&e.relationInfer?W=`cell: function render({ getValue, table }) {
                            const meta = table.options.meta as {
                                ${f(e.key,"Data")}: GetManyResponse;
                            };

                            const ${f(e.key,"")} = meta.${x}?.find(
                                (item) => item.id === getValue<any>(),
                            );

                            return ${s(f(e.key),void 0,(h=e==null?void 0:e.relationInfer)==null?void 0:h.accessor)} ?? "Loading...";
                        },`:W="";return`
                    {
                        ${C},
                        ${b},
                        ${B},
                        ${W}
                    }
                `}},V=e=>{if(e.type==="image"){o.push(["Image","@pankod/refine-chakra-ui"]);let h=`id: "${e.key}"`,x=Ce(e),C=`header: "${g(e.key)}"`,b=m`
                    cell: function render({ getValue }) {
                        ${e!=null&&e.accessor?"try {":""}
                            return <Image sx={{ maxWidth: "100px" }} src={${s("getValue<any>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0," + ")}} />
                        ${e!=null&&e.accessor?" } catch (error) { return null; }":""}
                    }
                `;if(e.multiple){let B=s("item",void 0,e.accessor," + ");b=`
                        cell: function render({ getValue }) {
                            ${e!=null&&e.accessor?"try {":""}
                                return (
                                    <HStack>
                                        {getValue<any[]>()?.map((item, index) => (
                                            <Image src={${B}} key={index} sx={{ height: "50px", maxWidth: "100px" }} />
                                        ))}
                                    </HStack>
                                )
                            ${e!=null&&e.accessor?" } catch (error) { return null; }":""}
                        }
                    `}return`
                    {
                        ${h},
                        ${x},
                        ${C},
                        ${b}
                    }
                `}},L=e=>{if(e.type==="email"){o.push(["EmailField","@pankod/refine-chakra-ui"]);let h=`id: "${e.key}"`,x=Ce(e),C=`header: "${g(e.key)}"`,b=m`
                    cell: function render({ getValue }) {
                        return <EmailField value={${s("getValue<any>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0,' + " " + ')}} />
                    }
                `;return e.multiple&&(o.push(["TagField","@pankod/refine-chakra-ui"]),b=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${C},
                        ${b}
                    }
                `}},A=e=>{if(e.type==="url"){o.push(["UrlField","@pankod/refine-chakra-ui"]);let h=`id: "${e.key}"`,x=Ce(e),C=`header: "${g(e.key)}"`,b=m`
                    cell: function render({ getValue }) {
                        return <UrlField value={${s("getValue<any>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0," + ")}} />
                    }
                `;return e.multiple&&(o.push(["TagField","@pankod/refine-chakra-ui"]),b=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${C},
                        ${b}
                    }
                `}},p=e=>{if((e==null?void 0:e.type)==="boolean"){o.push(["BooleanField","@pankod/refine-chakra-ui"]);let h=`id: "${e.key}"`,x=Ce(e),C=`header: "${g(e.key)}"`,b=m`
                    cell: function render({ getValue }) {
                        return <BooleanField value={${s("getValue<any>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0," + ")}} />
                    }
                `;return e.multiple&&(b=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <BooleanField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${C},
                        ${b}
                    }
                `}},N=e=>{if(e.type==="date"){o.push(["DateField","@pankod/refine-chakra-ui"]);let h=`id: "${e.key}"`,x=Ce(e),C=`header: "${g(e.key)}"`,b=m`
                    cell: function render({ getValue }) {
                        return <DateField value={${s("getValue<any>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0,' + " " + ')}} />
                    }
                `;return e.multiple&&(b=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <DateField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${C},
                        ${b}
                    }
                `}},y=e=>{if((e==null?void 0:e.type)==="richtext"){o.push(["MarkdownField","@pankod/refine-chakra-ui"]);let h=`id: "${e.key}"`,x=Ce(e),C=`header: "${g(e.key)}"`,b=m`
                    cell: function render({ getValue }) {
                        return <MarkdownField value={(${s("getValue<string>()",void 0,Array.isArray(e.accessor)?e.accessor:void 0)})?.slice(0, 80) + "..." } />
                    }
                `;return e.multiple&&(b=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<string[]>()?.map((item, index) => (
                                        <MarkdownField value={${s("item",void 0,e.accessor," + ")}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${C},
                        ${b}
                    }
                `}},u=e=>{if(e&&(e.type==="text"||e.type==="number")){let h=`id: "${e.key}"`,x=Ce(e),C=`header: "${g(e.key)}"`,b="";return e.multiple&&(o.push(["TagField","@pankod/refine-chakra-ui"]),b=`
                        cell: function render({ getValue }) {
                            return (
                                <HStack>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${s("item",void 0,e.accessor,' + " " + ')}} key={index} />
                                    ))}
                                </HStack>
                            )
                        }
                    `),!e.multiple&&Array.isArray(e.accessor)&&(b=`
                        cell: function render({ getValue }) {
                            return (
                                <>{${s("getValue<any>()",e.key,e.accessor)}}</>
                            );
                        }
                    `),`
                    {
                        ${h},
                        ${x},
                        ${C},
                        ${b}
                    }
                `}},{canEdit:w,canShow:a,canDelete:T}=n!=null?n:{};w&&o.push(["EditButton","@pankod/refine-chakra-ui"]),a&&o.push(["ShowButton","@pankod/refine-chakra-ui"]),T&&o.push(["DeleteButton","@pankod/refine-chakra-ui"]);let t=w||a||T?m`
        {
            id: "actions",
            accessorKey: "id",
            header: "Actions",
            cell: function render({ getValue }) {
                return (
                    <HStack>
                    ${a?m`
                        <ShowButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `:""}
                        ${w?m`
                        <EditButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `:""}
                        ${T?m`
                        <DeleteButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `:""}
                    </HStack>
                );
            },
        },
            `:"",r=c.map(e=>{switch(e==null?void 0:e.type){case"text":case"number":return u(e);case"richtext":return y(e);case"email":return L(e);case"image":return V(e);case"date":return N(e);case"boolean":return p(e);case"url":return A(e);case"relation":return S(e);default:return}});return H(o),m`
        import React from "react";
        ${P(o)}
        
        export const ${d}: React.FC<IResourceComponentsProps> = () => {
            const columns = React.useMemo<ColumnDef<any>[]>(() => [
                ${[...r,t].filter(Boolean).join(",")}
            ], []);

            const {
                getHeaderGroups,
                getRowModel,
                setOptions,
                refineCore: {
                    setCurrent,
                    pageCount,
                    current,
                    tableQueryResult: { data: tableData },
                },
            } = useTable({
                columns,
                ${k?`
                refineCoreProps: {
                    resource: "${n.name}",
                }
                `:""}
                
            });

            ${v}

            setOptions((prev) => ({
                ...prev,
                meta: {
                    ...prev.meta,
                    ${I.join(", ")}
                },
            }));

            return (
                <List>
                    <TableContainer whiteSpace="pre-line">
                        <Table variant="simple">
                            <Thead>
                                {getHeaderGroups().map((headerGroup) => (
                                    <Tr key={headerGroup.id}>
                                        {headerGroup.headers.map((header) => (
                                            <Th key={header.id}>
                                                {!header.isPlaceholder && (
                                                    flexRender(
                                                        header.column.columnDef
                                                            .header,
                                                        header.getContext(),
                                                    )
                                                )}
                                            </Th>
                                        ))}
                                    </Tr>
                                ))}
                            </Thead>
                            <Tbody>
                                {getRowModel().rows.map((row) => (
                                    <Tr key={row.id}>
                                        {row.getVisibleCells().map((cell) => (
                                            <Td key={cell.id}>
                                                {flexRender(
                                                    cell.column.columnDef.cell,
                                                    cell.getContext(),
                                                )}
                                            </Td>
                                        ))}
                                    </Tr>
                                ))}
                            </Tbody>
                        </Table>
                    </TableContainer>
                    <Pagination
                        current={current}
                        pageCount={pageCount}
                        setCurrent={setCurrent}
                    />
                </List>   
            );
        };

        type PaginationProps = {
            current: number;
            pageCount: number;
            setCurrent: (page: number) => void;
        };

        const Pagination: React.FC<PaginationProps> = ({
            current,
            pageCount,
            setCurrent,
        }) => {
            const pagination = usePagination({
                current,
                pageCount,
            });
        
            return (
                <Box display="flex" justifyContent="flex-end">
                    <HStack my="3" spacing="1">
                        {pagination?.prev && (
                            <IconButton
                                aria-label="previous page"
                                onClick={() => setCurrent(current - 1)}
                                disabled={!pagination?.prev}
                                variant="outline"
                            >
                                <IconChevronLeft size="18" />
                            </IconButton>
                        )}
        
                        {pagination?.items.map((page) => {
                            if (typeof page === "string")
                                return <span key={page}>...</span>;
        
                            return (
                                <Button
                                    key={page}
                                    onClick={() => setCurrent(page)}
                                    variant={page === current ? "solid" : "outline"}
                                >
                                    {page}
                                </Button>
                            );
                        })}
                        {pagination?.next && (
                            <IconButton
                                aria-label="next page"
                                onClick={() => setCurrent(current + 1)}
                                variant="outline"
                            >
                                <IconChevronRight size="18" />
                            </IconButton>
                        )}
                    </HStack>
                </Box>
            );
        };
        `}});var ut=E(require("@pankod/refine-core")),pt=E(require("@pankod/refine-chakra-ui")),mt=E(require("@pankod/refine-react-hook-form"));var lr=D({type:"create",additionalScope:[["@pankod/refine-core","RefineCore",ut],["@pankod/refine-chakra-ui","RefineChakraUI",pt],["@pankod/refine-react-hook-form","RefineReactHookForm",mt]],codeViewerComponent:be,loadingComponent:xe,errorComponent:fe,renderer:({resource:n,fields:c,isCustomPage:k})=>{var A;let d=M((A=n.label)!=null?A:n.name,"create"),$=[["Create","@pankod/refine-chakra-ui"],["FormControl","@pankod/refine-chakra-ui"],["FormLabel","@pankod/refine-chakra-ui"],["FormErrorMessage","@pankod/refine-chakra-ui"],["useForm","@pankod/refine-react-hook-form"]],l=c.filter(p=>(p==null?void 0:p.relation)&&!(p!=null&&p.fieldable)&&(p==null?void 0:p.resource)).filter(Boolean).map(p=>{if((p==null?void 0:p.relation)&&!p.fieldable&&p.resource)return $.push(["useSelect","@pankod/refine-core"]),`
                    const { options: ${f(p.key,"Options")} } =
                    useSelect({
                        resource: "${p.resource.name}",
                        ${ee(p)}
                    });
                `}).filter(Boolean),v=p=>{if(p.relation&&p.resource){$.push(["useSelect","@pankod/refine-core"]),$.push(["Select","@pankod/refine-chakra-ui"]);let N=f(p.key,"Options");return m`
                <FormControl mb="3" isInvalid={!!errors?.${R(p.key,void 0)}}>
                    <FormLabel>${g(p.key)}</FormLabel>
                    <Select
                        placeholder="Select ${_(p.resource.name)}"
                        {...register("${R(p.key,void 0,p.accessor)}", {
                            required: "This field is required",
                        })}
                    >
                        {${N}?.map((option) => (
                            <option value={option.value} key={option.value}>
                                {option.label}
                            </option>
                        ))}
                    </Select>
                    <FormErrorMessage>
                        {${s("(errors as any)",p.key,p.accessor,!1)}?.message as string}
                    </FormErrorMessage>
                </FormControl>
                `}},I=p=>{if(p.type==="text"||p.type==="url"||p.type==="email"||p.type==="number"||p.type==="date"||p.type==="richtext")return j(p.key)||($.push(["Input","@pankod/refine-chakra-ui"]),p.multiple)?void 0:m`
                    <FormControl mb="3" isInvalid={!!${s("(errors as any)",p.key,p.accessor,!1)}}>
                        <FormLabel>${g(p.key)}</FormLabel>
                        <Input
                            ${p.type!=="date"&&p.type!=="richtext"?`type="${p.type}"`:""}
                            {...register("${R(p.key,void 0,p.accessor)}", {
                                required: "This field is required",
                            })}
                        />
                        <FormErrorMessage>
                            {${s("(errors as any)",p.key,p.accessor,!1)}?.message as string}
                        </FormErrorMessage>
                    </FormControl>
                `},S=p=>{if(p.type==="boolean")return $.push(["Checkbox","@pankod/refine-chakra-ui"]),p.multiple?void 0:m`
                    <FormControl mb="3" isInvalid={!!${s("errors",p.key,p.accessor,!1)}}>
                        <FormLabel>${g(p.key)}</FormLabel>
                        <Checkbox
                            {...register("${R(p.key,void 0,p.accessor)}", {
                                required: "This field is required",
                            })}
                        />
                        <FormErrorMessage>
                            {${s("errors",p.key,p.accessor,!1)}?.message as string}
                        </FormErrorMessage>
                    </FormControl>
                   
                `},V=p=>{if(p.type==="date"){let N=I(p);return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-chakra-ui" package.
                        To use a <DatePicker> component, you can examine the following links:
                        
                        - https://github.com/aboveyunhai/chakra-dayzed-datepicker
                        - https://github.com/wojtekmaj/react-date-picker
                    */}
                    ${N!=null?N:""}
                    `}},L=c.map(p=>{switch(p==null?void 0:p.type){case"text":case"number":case"email":case"url":case"richtext":return I(p);case"date":return V(p);case"boolean":return S(p);case"relation":return v(p);default:return}});return H($),m`
        ${P($)}
        
        export const ${d} = () => {
            const {
                refineCore: { formLoading },
                saveButtonProps,
                register,
                formState: { errors },
            } = useForm(
                ${k?`
                { 
                    refineCoreProps: {
                        resource: "${n.name}",
                        action: "create",
                    }
                }`:""}
            );
        
            ${l}

            return (
                <Create isLoading={formLoading} saveButtonProps={saveButtonProps}>
                    ${L.join("")}
                </Create>
            );
        };
        `}});var lt=E(require("@pankod/refine-core")),dt=E(require("@pankod/refine-chakra-ui")),yt=E(require("@pankod/refine-react-hook-form"));var dr=D({type:"edit",additionalScope:[["@pankod/refine-core","RefineCore",lt],["@pankod/refine-chakra-ui","RefineChakraUI",dt],["@pankod/refine-react-hook-form","RefineReactHookForm",yt]],codeViewerComponent:be,loadingComponent:xe,errorComponent:fe,renderer:({resource:n,fields:c,isCustomPage:k,id:d})=>{var N,y;let $=M((N=n.label)!=null?N:n.name,"edit"),o=f((y=n.label)!=null?y:n.name,"Data"),l=[["Edit","@pankod/refine-chakra-ui"],["FormControl","@pankod/refine-chakra-ui"],["FormLabel","@pankod/refine-chakra-ui"],["FormErrorMessage","@pankod/refine-chakra-ui"],["useForm","@pankod/refine-react-hook-form"]],I=c.filter(u=>(u==null?void 0:u.relation)&&!(u!=null&&u.fieldable)&&(u==null?void 0:u.resource)).filter(Boolean).map(u=>{if((u==null?void 0:u.relation)&&!u.fieldable&&u.resource){l.push(["useSelect","@pankod/refine-core"]);let w=s(o,u.key,u.accessor,!1);return u.multiple&&u.accessor&&(w=`${s(o,u.key)}?.map((item: any) => ${s("item",void 0,u.accessor)})`),`
                    const { options: ${f(u.key,"Options")} } =
                    useSelect({
                        resource: "${u.resource.name}",
                        defaultValue: ${w},
                        ${ee(u)}
                    });
                `}}).filter(Boolean),S=u=>{if(u.relation&&u.resource){l.push(["useSelect","@pankod/refine-core"]),l.push(["Select","@pankod/refine-chakra-ui"]);let w=f(u.key,"Options");return m`
                <FormControl mb="3" isInvalid={!!errors?.${R(u.key,void 0)}}>
                    <FormLabel>${g(u.key)}</FormLabel>
                    <Select
                        placeholder="Select ${_(u.resource.name)}"
                        {...register("${R(u.key,void 0,u.accessor)}", {
                            required: "This field is required",
                        })}
                    >
                        {${w}?.map((option) => (
                            <option value={option.value} key={option.value}>
                                {option.label}
                            </option>
                        ))}
                    </Select>
                    <FormErrorMessage>
                        {${s("(errors as any)",u.key,u.accessor,!1)}?.message as string}
                    </FormErrorMessage>
                </FormControl>
                `}},V=u=>{if(u.type==="text"||u.type==="url"||u.type==="email"||u.type==="number"||u.type==="date"||u.type==="richtext"){if(l.push(["Input","@pankod/refine-chakra-ui"]),u.multiple){let w=R(u.key,"${index}",u.accessor),a=s(s("(errors as any)",u.key,void 0,!1)+"?.[index]",void 0,u.accessor);return`
                        <>
                            {${s(o,u.key)}?.map((item: any, index: number) => (
                                <FormControl key={index} mb="3" isInvalid={!!${a}}>
                                    <FormLabel>${g(u.key)} #{index + 1}</FormLabel>
                                    <Input
                                        {...register(\`${w}\`, {
                                            required: "This field is required",
                                        })}
                                    />
                                    <FormErrorMessage>
                                        {${s(a,"message")} as string}
                                    </FormErrorMessage>
                                </FormControl>
                            ))}
                        </>
                    `}return m`
                    <FormControl mb="3" isInvalid={!!${s("(errors as any)",u.key,u.accessor,!1)}}>
                        <FormLabel>${g(u.key)}</FormLabel>
                        <Input
                            ${j(u.key)?"disabled":""}
                            ${u.type!=="date"&&u.type!=="richtext"?`type="${u.type}"`:""}
                            {...register("${R(u.key,void 0,u.accessor)}", {
                                required: "This field is required",
                            })}
                        />
                        <FormErrorMessage>
                            {${s("(errors as any)",u.key,u.accessor,!1)}?.message as string}
                        </FormErrorMessage>
                    </FormControl>
                `}},L=u=>{if(u.type==="boolean"){if(l.push(["Checkbox","@pankod/refine-chakra-ui"]),u.multiple){let w=R(u.key,void 0,u.accessor),a=s("(errors as any)",u.key,void 0,!1)+"?.[index]";return`
                        <>
                            {${s(o,u.key)}?.map((item: any, index: number) => (
                                <FormControl key={index} mb="3" isInvalid={!!${a}}>
                                    <FormLabel>${g(u.key)} #{index + 1}</FormLabel>
                                    <Checkbox
                                        {...register(\`${w}.\${index}\`, {
                                            required: "This field is required",
                                        })}
                                    />
                                    <FormErrorMessage>
                                        {${a}?.message as string}
                                    </FormErrorMessage>
                                </FormControl>
                            ))}
                        </>
                    `}return m`
                    <FormControl mb="3" isInvalid={!!${s("errors",u.key,u.accessor,!1)}}>
                        <FormLabel>${g(u.key)}</FormLabel>
                        <Checkbox
                            {...register("${R(u.key,void 0,u.accessor)}", {
                                required: "This field is required",
                            })}
                        />
                        <FormErrorMessage>
                            {${s("errors",u.key,u.accessor,!1)}?.message as string}
                        </FormErrorMessage>
                    </FormControl>
                   
                `}},A=u=>{if(u.type==="date")return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-chakra-ui" package.
                        To use a <DatePicker> component, you can examine the following links:
                        
                        - https://github.com/aboveyunhai/chakra-dayzed-datepicker
                        - https://github.com/wojtekmaj/react-date-picker
                    */}
                    ${V(u)}
                    `},p=c.map(u=>{switch(u==null?void 0:u.type){case"text":case"number":case"email":case"url":case"richtext":return V(u);case"date":return A(u);case"boolean":return L(u);case"relation":return S(u);default:return}});return H(l),m`
        ${P(l)}
        
        export const ${$} = () => {
            const {
                refineCore: { formLoading, queryResult },
                saveButtonProps,
                register,
                formState: { errors },
            } = useForm(
                ${k?`
                { 
                    refineCoreProps: {
                        resource: "${n.name}",
                        id: ${d},
                        action: "edit",
                    }
                }`:""}
            );
        
            const ${o} = queryResult?.data?.data;
        
            ${I}

            return (
                <Edit isLoading={formLoading} saveButtonProps={saveButtonProps}>
                    ${p.join("")}
                </Edit>
            );
        };
        `}});var kt=({action:n,id:c,...k})=>{let{useParams:d}=(0,In.useRouterContext)(),{action:$,id:o}=d();switch(n!=null?n:$){case"show":return He.default.createElement(pr,{...k,id:c!=null?c:o});case"create":return He.default.createElement(lr,{...k,id:c!=null?c:o});case"edit":return He.default.createElement(dr,{...k,id:c!=null?c:o});default:return He.default.createElement(mr,{...k,id:c!=null?c:o})}};
