// @pankod/refine-inferencer@1.6.1/dist/mantine.js
var $r=Object.create;var ce=Object.defineProperty;var xr=Object.getOwnPropertyDescriptor;var hr=Object.getOwnPropertyNames;var br=Object.getPrototypeOf,Fr=Object.prototype.hasOwnProperty;var Cr=(r,t)=>{for(var i in t)ce(r,i,{get:t[i],enumerable:!0})},Ee=(r,t,i,s)=>{if(t&&typeof t=="object"||typeof t=="function")for(let c of hr(t))!Fr.call(r,c)&&c!==i&&ce(r,c,{get:()=>t[c],enumerable:!(s=xr(t,c))||s.enumerable});return r};var R=(r,t,i)=>(i=r!=null?$r(br(r)):{},Ee(t||!r||!r.__esModule?ce(i,"default",{value:r,enumerable:!0}):i,r)),Tr=r=>Ee(ce({},"__esModule",{value:!0}),r);var Ur={};Cr(Ur,{MantineCreateInferencer:()=>ge,MantineEditInferencer:()=>Ie,MantineInferencer:()=>Hr,MantineListInferencer:()=>fe,MantineShowInferencer:()=>ye});module.exports=Tr(Ur);var ie=R(require("react")),kr=require("@pankod/refine-core");var Or=R(require("@pankod/refine-mantine"));var O=R(require("react")),Me=require("@pankod/refine-core");var Ar=/(image|photo|avatar|url|logo|cover|thumbnail|icon)/i,ke=(r,t)=>{let i=["name","label","title","count","content","username","nickname","login","firstName","lastName","url"];Ar.test(r)&&i.unshift("url");let s=i.find(c=>c in t&&!!t[c]);if(s)return s==="firstName"&&"lastName"in t?["firstName","lastName"]:s};function $e(r,t,i){var c;if(t)return t;let s=i==null?void 0:i.find(o=>o.name===r);return(c=s==null?void 0:s.options)!=null&&c.dataProviderName?s.options.dataProviderName:"default"}var oe=r=>{var t;return(t=r==null?void 0:r.options)==null?void 0:t.dataProviderName};var xe=r=>{let t=[...r].filter(Boolean).sort((i,s)=>(s.priority||0)-(i.priority||0))[0];return t||null};var Be=R(require("prettier/standalone")),Oe=R(require("prettier/parser-typescript")),he=r=>{try{return Be.default.format(r,{parser:"typescript",plugins:[Oe.default],arrowParens:"always",trailingComma:"all",semi:!0,tabWidth:4,printWidth:80})}catch(t){return console.warn("Formatting error in inferencer",t),r}};var be=r=>r.replace(/ {2,}/g," ").replace(/\n{2,}/g,`
`);var Fe=/(-id|-ids|_id|_ids|Id|Ids|ID|IDs)(\[\])?$/,ze=(r,t)=>{let i=Fe.test(r),s=typeof t=="string"||typeof t=="number",c=Array.isArray(t)&&t.every(o=>typeof o=="string"||typeof o=="number");return i&&s||i&&c?{key:r,relation:!0,multiple:!!c,type:"relation",priority:1}:!1};var D=r=>r.replace(Fe,"");var I=r=>D(r).replace(/([a-z])([A-Z])/g,"$1 $2").replace(/_/g," ").replace(/-/g," ").split(" ").map(u=>u.charAt(0).toUpperCase()+u.slice(1)).join(" ");var Ce=r=>r.replace(/export\s+(const|let|var|type|interface|function|class)\s+(\w+)\s*(=|:)\s*/g,"$1 $2 = ").split(`
`).filter(c=>!c.trim().startsWith("export default")).join(`
`);var je=/import(?:(?:(?:[ \n\t]+([^ *\n\t\{\},]+)[ \n\t]*(?:,|[ \n\t]+))?([ \n\t]*\{(?:[ \n\t]*[^ \n\t"'\{\}]+[ \n\t]*,?)+\})?[ \n\t]*)|[ \n\t]*\*[ \n\t]*as[ \n\t]+([^ \n\t\{\}]+)[ \n\t]+)from[ \n\t]*(?:['"])([^'"\n]+)(?:['"])(?:;?)/g,vr=/import[ \n\t](?:['"])([^'"\n]+)(?:['"])(?:;?)/g,Rr=/((?:\w|\s|_)*)( as )((?:\w|\s|_)*)( |,)?/g,Te=(r,t={})=>{let i=r.matchAll(je),s=new Set;for(let c of i){let[,o,u,g,m]=c;if(m in t){let k=t[m];if(m==="react")continue;o&&s.add(`const { default: ${o} } = ${k};`),u&&s.add(`const${u.replace(Rr,"$1: $3$4")} = ${k};`),g&&s.add(`const ${g} = ${k};`)}}return be(`
    ${Array.from(s).join(`
`)}
    ${r.replace(je,"").replace(vr,"")}
    `)};var ae=R(require("pluralize"));var Ae=(r,t)=>r?t.find(s=>{let c=D(r.key);return s.name===c||s.name===(ae.default.isPlural(c)?c:ae.default.plural(c))||s.name===(ae.default.isSingular(c)?c:ae.default.singular(c))}):void 0;var Ke=R(require("pluralize"));var B=(r,t)=>{let i=r.replace(/[^a-zA-Z0-9]/g," "),s=Ke.default.singular(i),c=I(s).replace(/ /g,""),o=t.charAt(0).toUpperCase()+t.slice(1);return/^[a-zA-Z]/.test(r)?`${c}${o}`:`${o}${c}`};var wr=r=>r?Array.isArray(r)?r.join(""):typeof r=="string"?r:typeof r=="object"||typeof r=="function"?r.toString():typeof r=="number"?String(r):typeof r=="boolean"&&r?"true":r:"",f=(r,...t)=>r.reduce((s,c,o)=>`${s}${c}${wr(t[o])}`,"").trim();var ve=(r,t)=>`
    ${r!=null?r:""}
    
    render(typeof ${t} !== "undefined" ? <${t} /> : <></>);
    `;var Re=r=>r==null?void 0:r.replace(/(\/\* hidden-start \*\/)(.|\n)*?(\/\* hidden-end \*\/)/g,"");var Sr=/^[a-zA-Z_$][a-zA-Z_$0-9]*$/,_e=r=>Sr.test(r);function we(r,t,i){let s=`${r}`;return t&&(s+="?.",_e(t)?s+=t:s+=`['${t}']`),i&&(s+="?.",_e(i)?s+=i:s+=`['${i}']`),s}function Nr(r,t,i){return(i!=null?i:[]).map(c=>we(r,t,c))}var l=(r,t,i,s=' + " " + ')=>Array.isArray(i)?s?Nr(r,t,i).join(s):we(r,t,i[0]):we(r,t,i),N=(r,t,i)=>{let s=r;return t&&(s+=`.${t}`),i&&(Array.isArray(i)?s+=`.${i[0]}`:s+=`.${i}`),s};var _=r=>{let t=r.reduce((s,[c,o,u])=>(s[o]||(s[o]=[]),s[o].includes(c)||(u?s[o].push([c,!0]):s[o].push(c)),s),{});return Object.entries(t).map(([s,c])=>{let o=c.find(k=>Array.isArray(k)&&k[1]),u=c.filter(k=>typeof k=="string"),g=o?`${o[0]}`:"",m=u.length?`{ ${u.join(", ")} }`:"";return`import ${g}${g&&m?", ":""}${m} from "${s}";`})};var Se=R(require("pluralize")),U=(r="")=>Se.default.isSingular(r)?r:Se.default.singular(r);var Ne=R(require("pluralize")),re=(r="")=>Ne.default.isPlural(r)?r:Ne.default.plural(r);var Vr=/^id$/i,ee=r=>Vr.test(r);var se=r=>r.relationInfer&&r.relationInfer.accessor?Array.isArray(r.relationInfer.accessor)?`optionLabel: "${r.relationInfer.accessor[0]}",`:r.relationInfer.accessor!=="title"?`optionLabel: "${r.relationInfer.accessor}",`:"":"";var H=r=>{};var A=(r,t="")=>{if(Number.isInteger(Number(r)))return t.charAt(0).toLowerCase()+t.slice(1)+r;let s=r.replace(/^\d+/,"").replace(/[^a-zA-Z0-9]/g,"-"),c=I(s).replace(/ /g,"");return`${c.charAt(0).toLowerCase()+c.slice(1)}${t.charAt(0).toUpperCase()+t.slice(1)}`};var He=r=>{let t=(i,s,c,o=t)=>{let u=r.map(m=>m(i,s,c,o));return xe(u)};return t};var Ue=r=>(i,s,c,o,u)=>r.reduce((g,m)=>m(g,s,c,o,u),i);var We=(r,t,i,s)=>{let c=Array.isArray(t),o=Array.isArray(t)&&t.every(u=>typeof u=="string"||typeof u=="number");if(c){if(!o){let g=s(r,t[0],i,s);return g?{...g,key:r,multiple:!0,priority:1}:!1}let u=s(r,t[0],i,s)||{type:"string"};return{key:r,multiple:!0,priority:1,type:u==null?void 0:u.type}}return!1};var qe=(r,t)=>typeof t=="boolean"?{key:r,type:"boolean"}:!1;var Ve=R(require("dayjs")),Lr=/(_at|_on|At|On|AT|ON)(\[\])?$/,Ze=(r,t)=>{let i=Lr.test(r)&&(0,Ve.default)(t).isValid(),s=typeof t=="string"&&(0,Ve.default)(t).isValid();return i||s?{key:r,type:"date",priority:1}:!1};var Gr=/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,Je=(r,t)=>typeof t=="string"&&Gr.test(t)?{key:r,type:"email",priority:1}:!1;var Pr=/\.(gif|jpe?g|tiff?|png|webp|bmp|svg)$/i,Xe=(r,t)=>typeof t=="string"&&Pr.test(t)?{key:r,type:"image",priority:2}:!1;var Qe=(r,t)=>typeof t>"u"||t===null?null:!1;var Ye=(r,t)=>typeof t=="number"?{key:r,type:"number"}:!1;var Dr=/id$/i,er=(r,t,i,s)=>{let c=t!==null,o=!Array.isArray(t);if(c&&o&&typeof t=="object"){if(Object.keys(t).length===1&&Dr.test(Object.keys(t)[0]))return{key:r,type:"relation",relation:!0,accessor:"id",priority:1};let m=ke(r,t);if((m?m.length>0:!1)&&m){let F=m&&Array.isArray(m)?m[0]:m,b=s(F,t[F],t,s);if(b){let $=Array.isArray(m)?m.map(C=>b.accessor?`${C}.${Array.isArray(b.accessor)?b.accessor[0]:b.accessor}`:C):b.accessor?Array.isArray(b.accessor)?`${m}.${b.accessor[0]}`:`${m}.${b.accessor}`:m;return{...b,fieldable:!0,key:r,accessor:$,priority:1}}}return{key:r,fieldable:!1,type:"object",priority:1}}return!1};var rr=(r,t)=>typeof t=="string"&&t.length>100?{key:r,type:"richtext",priority:1}:!1;var nr=(r,t)=>typeof t=="string"?{key:r,type:"text"}:!1;var Mr=/^(https?|ftp):\/\/(-\.)?([^\s/?\.#-]+\.?)+(\/[^\s]*)?$/i,tr=(r,t)=>typeof t=="string"&&Mr.test(t)?{key:r,type:"url",priority:1}:!1;var or=[We,qe,Ze,Je,Xe,Qe,Ye,er,ze,rr,nr,tr];var ar=(r,t,i,s)=>r.map(o=>{if(!o.relation&&(o.type==="text"||o.type==="richtext"||o.type==="number")&&!o.canRelation){let u=/^[a-z0-9-]+$/,g=$=>u.test(`${$}`),m=o.key.toLowerCase()!=="id";return!(!o.accessor||typeof o.accessor=="string")||!m?o:(o.multiple?s[o.key]:[s[o.key]]).every($=>g(o.accessor?$[o.accessor]:$))?{...o,canRelation:!0}:o}return o});var Er=/(image|photo|avatar|cover|thumbnail|icon)/i,sr=r=>r.map(i=>i.type==="url"&&Er.test(i.key)?{...i,type:"image"}:i);var ir=(r,t)=>r.map(s=>{let c=Ae(s,t);return c?{...s,relation:!0,type:"relation",resource:c}:s});var cr=(r,t,i,s,c)=>r.map(u=>{if(u.relation&&u.type==="relation"&&!u.resource){let g=u.accessor?s[u.key][u.accessor]:s[u.key],m=c(u.key,g,s,c);if(m&&m.type!=="relation")return{...u,fieldable:!0,relation:!1,type:m.type,canRelation:!0}}return u});var ur=[sr,ir,cr,ar];var W=R(require("react")),ue=R(require("prism-react-renderer")),pr=R(require("prism-react-renderer/themes/vsDark")),Le=({code:r,wrapperProps:t})=>((0,W.useEffect)(()=>{var s;let i=document.createElement("style");i.appendChild(document.createTextNode(`
                #code-highlight::-webkit-scrollbar {-webkit-appearance: none; height: 7px; width: 7px;}
                #code-highlight::-webkit-scrollbar-thumb {background-color: #666b7a;}
                #code-highlight::-webkit-scrollbar-corner {background-color: rgb(30, 30, 30);}
                `)),(s=document.getElementById("code-highlight"))==null||s.appendChild(i)},[]),W.default.createElement("div",{id:"code-highlight",style:{maxHeight:"75vh",height:"100%",overflow:"auto",backgroundColor:"rgb(30, 30, 30)"},...t!=null?t:{}},W.default.createElement(ue.default,{...ue.defaultProps,theme:pr.default,code:r,language:"tsx"},({className:i,style:s,tokens:c,getLineProps:o,getTokenProps:u})=>W.default.createElement("pre",{className:i,style:{...s,padding:"14px 14px 14px 14px",margin:"0",width:"100%",boxSizing:"border-box"}},c.map((g,m)=>W.default.createElement("div",{key:m,...o({line:g,key:m})},g.map((k,F)=>W.default.createElement("span",{key:F,...u({token:k,key:F})}))))))));var M=R(require("react")),Br=R(require("@pankod/refine-core")),pe=require("@aliemir/react-live");var mr=[["react","React",M.default],["@pankod/refine-core","RefineCore",Br]],Ge=M.default.createContext({}),Pe=({fetchError:r,code:t,additionalScope:i,errorComponent:s})=>{var g;let c=(g=t&&Ce(Te(t,[...mr,...i!=null?i:[]].reduce((m,[k,F,b,$])=>$?m:{...m,[k]:F},{}))))!=null?g:"",o=M.default.useMemo(()=>({React:M.default,...M.default,...[...mr,...i!=null?i:[]].reduce((m,[k,F,b])=>({...m,[F]:b}),{})}),[i]),u=M.default.useMemo(()=>()=>{let{error:k}=M.default.useContext(Ge);return s?M.default.createElement(s,{error:k!=null?k:r?"Something went wrong while fetching the resource data.":void 0}):null},[s,r]);return M.default.createElement(pe.LiveProvider,{Context:Ge,code:c,scope:o,noInline:!0},!r&&M.default.createElement(pe.LivePreview,{Context:Ge}),M.default.createElement(u,null))};var ne=R(require("react")),me=require("@pankod/refine-core");var lr=(r,t,i)=>{let{resource:s,resourceName:c,id:o,resources:u}=(0,me.useResource)({resourceNameOrRouteName:t}),g=i!=null?i:o,m=(0,me.useDataProvider)(),[k,F]=ne.default.useState(void 0),[b,$]=ne.default.useState(!0),[C,L]=ne.default.useState(!1),p=ne.default.useCallback(async P=>{var x,G;let v=(x=oe(s))!=null?x:$e(c,void 0,u),a=m(v);L(!0);try{if(r==="list"||r==="create"){let e=await a.getList({resource:c});F((G=e.data)==null?void 0:G[0]),setTimeout(()=>{L(!1)},500)}if(r==="edit"||r==="show"&&P){let e=await a.getOne({resource:c,id:P});F(e.data),setTimeout(()=>{L(!1)},500)}}catch{setTimeout(()=>{L(!1)},500)}},[r,m,s,c,u]);return ne.default.useEffect(()=>{$(!1),!C&&!k&&p(g)},[p,g]),{data:k,loading:C,initial:b}};var te=R(require("react")),dr=require("@pankod/refine-core");var De=require("lodash"),yr=({record:r,fields:t,infer:i})=>{let s=(0,dr.useDataProvider)(),[c,o]=te.default.useState([]),[u,g]=te.default.useState(!0),[m,k]=te.default.useState(!1),F=te.default.useCallback(async b=>{console.groupCollapsed("@pankod/refine-inferencer is trying to detect relations");let $=[];k(!0);try{let C=b.map(async p=>{if(p&&(p.relation||p.canRelation)&&r){let P=oe(p.resource),v=s(P),a=Array.isArray(p.accessor)?void 0:p.multiple?r[p.key].map(x=>p.accessor?(0,De.get)(x,p.accessor):x)[0]:p.accessor?(0,De.get)(r[p.key],p.accessor):r[p.key];if(a&&p.resource)try{let{data:x}=await v.getOne({resource:p.resource.name,id:a});$.push({status:"success",resource:p.resource.name,field:p.key});let G=i("__",x,{},i);return{...p,relationInfer:G}}catch{return $.push({status:"error",resource:p.resource.name,field:p.key}),{...p,relationInfer:null}}if(a){let x,G;try{let{data:n}=await v.getOne({resource:re(D(p.key)),id:a});$.push({status:"success",resource:re(D(p.key)),field:p.key}),x=n,G=!0}catch{$.push({status:"error",resource:re(D(p.key)),field:p.key});try{let{data:h}=await v.getOne({resource:U(D(p.key)),id:a});$.push({status:"success",resource:U(D(p.key)),field:p.key}),x=h,G=!1}catch{return $.push({status:"error",resource:U(D(p.key)),field:p.key}),{...p,relationInfer:null}}}let e=i("__",x,{},i),y=D(p.key);return{...p,relation:!0,type:"relation",resource:{name:G?re(y):U(y)},fieldable:!1,canRelation:void 0,relationInfer:e}}return{...p,relationInfer:null}}return p}),L=await Promise.all(C);o(L.filter(p=>p)),setTimeout(()=>{k(!1)},500)}catch{setTimeout(()=>{k(!1)},500)}setTimeout(()=>{console.log(`Tried to detect relations with ${$.length} attempts and succeeded with ${$.filter(C=>C.status==="success").length} attempts.`),console.groupEnd()},500)},[s,r]);return te.default.useEffect(()=>{g(!1),!m&&t&&t.length>0&&c.length===0&&F(t)},[F,m,t,F]),{fields:c,loading:m,initial:u}};var q=({type:r,additionalScope:t=[],customElements:i=[],fieldTransformers:s=[],renderer:c,loadingComponent:o,errorComponent:u,codeViewerComponent:g})=>{let m=He([...or,...i]),k=Ue([...ur,...s]),F=({resourceName:$,fieldTransformer:C,id:L})=>{var S;let{resource:p,resources:P}=(0,Me.useResource)({resourceNameOrRouteName:$}),{resource:v}=(0,Me.useResource)(),{data:a,loading:x,initial:G}=lr(r,$!=null?$:p==null?void 0:p.name,L),e=O.default.useMemo(()=>{if(a){let V=Object.keys(a).map(K=>{let Y=a[K];return m(K,Y,a,m)}).filter(Boolean),T=k(V,P,p,a,m);return C?T.flatMap(K=>{let Y=C(K);return Y?[Y]:[]}):T}return[]},[a,P,p,C]),{fields:y,loading:n}=yr({record:a,fields:e,infer:m}),h=O.default.useMemo(()=>!x&&!n&&!G?c({resource:p,resources:P,fields:y,infer:m,isCustomPage:p.name!==v.name,id:L}):"",[p,P,y,x,n]);return O.default.createElement(O.default.Fragment,null,o&&(x||n)&&O.default.createElement(o,null),!x&&!n&&O.default.createElement(O.default.Fragment,null,O.default.createElement(Pe,{fetchError:!x&&!G&&!a,code:ve(h,B((S=p.label)!=null?S:p.name,r)),errorComponent:u,additionalScope:t}),g&&O.default.createElement(g,{code:Re(h),loading:x||n})))};return({name:$,resource:C,fieldTransformer:L,id:p})=>O.default.createElement(F,{fieldTransformer:L,resourceName:C!=null?C:$,key:C!=null?C:$,id:p})};var le=R(require("react")),de=require("@pankod/refine-mantine"),fr=require("@tabler/icons"),Z=({error:r})=>r?le.default.createElement(de.Center,{style:{minHeight:300}},le.default.createElement(de.Alert,{title:"Error",color:"red",icon:le.default.createElement(fr.IconAlertCircle,null)},r!=null?r:"")):null;var gr=R(require("react")),Ir=require("@pankod/refine-mantine"),J=()=>gr.default.createElement(Ir.LoadingOverlay,{visible:!0});var w=R(require("react")),E=require("@pankod/refine-mantine"),j=require("@tabler/icons");var X=({code:r,loading:t})=>{let i=w.default.useMemo(()=>he(r!=null?r:""),[r]),s=w.default.useRef(null),[c,o]=w.default.useState(!1),[u,g]=w.default.useState(!1);if(t)return null;let m=()=>{var k,F,b,$;(k=s==null?void 0:s.current)==null||k.select(),(F=s==null?void 0:s.current)==null||F.setSelectionRange(0,Number.MAX_SAFE_INTEGER),typeof navigator<"u"&&(navigator.clipboard.writeText(($=(b=s==null?void 0:s.current)==null?void 0:b.value)!=null?$:""),g(!0),setTimeout(()=>{g(!1)},1e3))};return i?w.default.createElement(w.default.Fragment,null,w.default.createElement(E.Affix,{position:{bottom:12,right:60}},w.default.createElement(E.Group,{spacing:"xs"},w.default.createElement(E.Button,{leftIcon:w.default.createElement(j.IconCode,{size:18}),variant:"filled",color:"primary",onClick:()=>o(!0)},"Show Code"),w.default.createElement(E.Button,{leftIcon:w.default.createElement(j.IconMessageCircle,{size:18}),variant:"default",component:"a",target:"_blank",href:"https://github.com/refinedev/refine/discussions/3046"},"Give Feedback"))),w.default.createElement("div",{style:{height:"50px"}}),w.default.createElement(E.Modal,{size:800,opened:c,onClose:()=>o(!1),styles:{modal:{padding:"0px !important",borderRadius:0},header:{display:"none"}}},w.default.createElement(E.ActionIcon,{onClick:()=>o(!1),variant:"transparent",sx:{position:"absolute",right:0,top:0,height:56,width:56}},w.default.createElement(j.IconX,{size:18,style:{color:"#666b7a"}})),w.default.createElement(Le,{code:i}),w.default.createElement("textarea",{ref:s,defaultValue:(i!=null?i:"").replace(/\\n/g,`\r
`),id:"code-input",style:{padding:0,margin:0,width:0,height:0,opacity:0,border:"none",display:"block"}}),w.default.createElement(E.Button,{sx:{position:"absolute",bottom:12,right:12,width:115},key:"copy",variant:"default",leftIcon:u?w.default.createElement(j.IconCheck,{size:18,style:{color:"green"}}):w.default.createElement(j.IconCopy,{size:18}),onClick:m},u?"Copied":"Copy"))):null};var ye=q({type:"show",additionalScope:[["@pankod/refine-mantine","RefineMantine",Or]],codeViewerComponent:X,loadingComponent:J,errorComponent:Z,renderer:({resource:r,fields:t,isCustomPage:i,id:s})=>{var G;let c=B((G=r.label)!=null?G:r.name,"show"),o="record",u=[["useShow","@pankod/refine-core"],["Show","@pankod/refine-mantine"],["Title","@pankod/refine-mantine"]],m=t.filter(e=>(e==null?void 0:e.relation)&&!(e!=null&&e.fieldable)&&(e==null?void 0:e.resource)).filter(Boolean).map(e=>{if((e==null?void 0:e.relation)&&!e.fieldable&&e.resource){if(e.multiple){u.push(["useMany","@pankod/refine-core"]);let y=l(o,e.key);return e.accessor&&(y=`${l(o,e.key)}?.map((item: any) => ${l("item",void 0,e.accessor)})`),`
                    const { data: ${A(e.key,"Data")}, isLoading: ${A(e.key,"IsLoading")} } =
                    useMany({
                        resource: "${e.resource.name}",
                        ids: ${y} || [],
                        queryOptions: {
                            enabled: !!${o},
                        },
                    });
                    `}return u.push(["useOne","@pankod/refine-core"]),`
                    const { data: ${A(e.key,"Data")}, isLoading: ${A(e.key,"IsLoading")} } =
                    useOne({
                        resource: "${e.resource.name}",
                        id: ${l(o,e.key,e.accessor,!1)} || "",
                        queryOptions: {
                            enabled: !!${o},
                        },
                    });
                `}}).filter(Boolean),k=e=>{if(e.relation&&e.resource){let y=A(e.key,"Data"),n=A(e.key,"IsLoading");return e.multiple?(u.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]),f`
                    <Title my="xs" order={5}>${I(e.key)}</Title>
                    {${n} ? <>Loading...</> : (
                        <>
                        ${(()=>{var h,S;if(e.relationInfer)if((h=e.relationInfer)!=null&&h.accessor){if(Array.isArray(e.relationInfer.accessor))return"Not Handled.";{let V=U((S=e.resource)==null?void 0:S.name),T=l(V,void 0,e.relationInfer.accessor);return`
                                        <Group spacing="xs">
                                            {${y}?.data?.map((${V}: any) => <TagField key={${T}} value={${T}} />)}
                                        </Group>
                                        `}}else return"Not Handled.";else return"not-handled - relation with multiple but no resource"})()}
                        </>
                    )}
                    `):f`
                    <Title my="xs" order={5}>${I(e.key)}</Title>
                    {${n} ? <>Loading...</> : (
                        <>
                        ${(()=>{var h;return e.relationInfer?(h=e.relationInfer)!=null&&h.accessor?Array.isArray(e.relationInfer.accessor)?`{${l(`${y}?.data`,void 0,e.relationInfer.accessor,' + " " + ')}}`:`{${y}?.data?.${e.relationInfer.accessor}}`:`{${y}?.data}`:`{${y}?.data?.id}`})()}
                        </>
                    )}
                    
                    `}},F=e=>{if(e.type==="text"){if(u.push(["TextField","@pankod/refine-mantine"]),e.multiple){u.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]);let y=l("item",void 0,e.accessor);return f`
                    <Title my="xs" order={5}>${I(e.key)}</Title>
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any) => (
                            <TagField value={${y}} key={${y}} />
                        ))}
                    </Group>
                `}return f`
                    <Title my="xs" order={5}>${I(e.key)}</Title>
                    <TextField value={${l(o,e.key,e.accessor)}} />
                `}},b=e=>{if(e.type==="image"){if(u.push(["Image","@pankod/refine-mantine"]),e.multiple){u.push(["Group","@pankod/refine-mantine"]);let y=l("item",void 0,e.accessor);return f`
                    <Title my="xs" order={5}>${I(e.key)}</Title>
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any) => (
                            <Image sx={{ maxWidth: 200 }} src={${y}} key={${y}} />
                        ))}
                    </Group>
                `}return f`
                    <Title my="xs" order={5}>${I(e.key)}</Title>
                    <Image sx={{ maxWidth: 200 }} src={${l(o,e.key,e.accessor," + ")}} />
                    `}},$=e=>{if(e.type==="email"){if(u.push(["EmailField","@pankod/refine-mantine"]),e.multiple){u.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]);let y=l("item",void 0,e.accessor);return f`
                    <Title my="xs" order={5}>${I(e.key)}</Title>
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any) => (
                            <TagField value={${y}} key={${y}} />
                        ))}
                    </Group>
                `}return f`
                    <Title my="xs" order={5}>${I(e.key)}</Title>
                    <EmailField value={${l(o,e.key,e.accessor," + ")}} />
                `}},C=e=>{if(e.type==="url"){if(u.push(["UrlField","@pankod/refine-mantine"]),e.multiple){u.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]);let y=l("item",void 0,e.accessor);return f`
                    <Title my="xs" order={5}>${I(e.key)}</Title>
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any) => (
                            <TagField value={${y}} key={${y}} />
                        ))}
                    </Group>
                `}return f`
                    <Title my="xs" order={5}>${I(e.key)}</Title>
                    <UrlField value={${l(o,e.key,e.accessor," + ")}} />
                `}},L=e=>{if(e.type==="boolean"){if(u.push(["BooleanField","@pankod/refine-mantine"]),e.multiple){u.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]);let y=l("item",void 0,e.accessor);return f`
                    <Title my="xs" order={5}>${I(e.key)}</Title>
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any, index: number) => (
                            <TagField value={${y}} key={index} />
                        ))}
                    </Group>
                `}return f`
                    <Title my="xs" order={5}>${I(e.key)}</Title>
                    <BooleanField value={${l(o,e.key,e.accessor," && ")}} />
                `}},p=e=>{if(e.type==="date"){if(u.push(["DateField","@pankod/refine-mantine"]),e.multiple){u.push(["Group","@pankod/refine-mantine"]);let y=l("item",void 0,e.accessor);return f`
                    <Title my="xs" order={5}>${I(e.key)}</Title>
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any) => (
                            <DateField value={${y}} key={${y}} />
                        ))}
                    </Group>
                `}return f`
                    <Title my="xs" order={5}>${I(e.key)}</Title>
                    <DateField value={${l(o,e.key,e.accessor," + ' ' + ")}} />
                `}},P=e=>{if(e.type==="richtext")return u.push(["MarkdownField","@pankod/refine-mantine"]),f`
                    <Title mt="xs" order={5}>${I(e.key)}</Title>
                    <MarkdownField value={${l(o,e.key,e.accessor,' + " " + ')}} />
                `},v=e=>{if(e.type==="number"){if(u.push(["NumberField","@pankod/refine-mantine"]),e.multiple){u.push(["TagField","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"]);let y=l("item",void 0,e.accessor);return f`
                    <Title my="xs" order={5}>${I(e.key)}</Title>
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any) => (
                            <TagField value={${y}} key={${y}} />
                        ))}
                    </Group>
                `}return f`
                    <Title my="xs" order={5}>${I(e.key)}</Title>
                    <NumberField value={${l(o,e.key,e.accessor,' + " " + ')} ?? ""} />
                `}},a=e=>{if(e)return f`
                    ${e}
                `},x=t.map(e=>{switch(e==null?void 0:e.type){case"text":return a(F(e));case"number":return a(v(e));case"richtext":return a(P(e));case"email":return a($(e));case"image":return a(b(e));case"date":return a(p(e));case"boolean":return a(L(e));case"url":return a(C(e));case"relation":return a(k(e));default:return}});return H(u),f`
        ${_(u)}
        
        export const ${c} = () => {
            const { queryResult } = useShow(${i?`{ 
                        resource: "${r.name}", 
                        id: ${s}
                    }`:""});
            const { data, isLoading } = queryResult;
        
            const ${o} = data?.data;
        
            ${m}

            return (
                <Show isLoading={isLoading}>
                    ${x.join("")}
                </Show>
            );
        };
        `}});var zr=R(require("@pankod/refine-mantine")),jr=R(require("@pankod/refine-react-table"));var Q=r=>Array.isArray(r.accessor)||r.multiple?`accessorKey: "${r.key}"`:r.accessor?`accessorKey: "${N(r.key,void 0,r.accessor)}"`:`accessorKey: "${r.key}"`,fe=q({type:"list",additionalScope:[["@pankod/refine-mantine","RefineMantine",zr],["@pankod/refine-react-table","RefineReactTable",jr]],codeViewerComponent:X,loadingComponent:J,errorComponent:Z,renderer:({resource:r,fields:t,isCustomPage:i})=>{var y;let s=B((y=r.label)!=null?y:r.name,"list"),c="tableData?.data",o=[["IResourceComponentsProps","@pankod/refine-core"],["useTable","@pankod/refine-react-table"],["ColumnDef","@pankod/refine-react-table"],["flexRender","@pankod/refine-react-table"],["ScrollArea","@pankod/refine-mantine"],["List","@pankod/refine-mantine"],["Table","@pankod/refine-mantine"],["Pagination","@pankod/refine-mantine"],["Group","@pankod/refine-mantine"],["EditButton","@pankod/refine-mantine"],["ShowButton","@pankod/refine-mantine"],["DeleteButton","@pankod/refine-mantine"]],u=t.filter(n=>(n==null?void 0:n.relation)&&!(n!=null&&n.fieldable)&&(n==null?void 0:n.resource)),g=u.filter(Boolean).map(n=>{if((n==null?void 0:n.relation)&&!n.fieldable&&n.resource){o.push(["GetManyResponse","@pankod/refine-core"]),o.push(["useMany","@pankod/refine-core"]);let h="";return n.multiple?h=`[].concat(...(${c}?.map((item) => ${l("item",n.key,n.accessor,!1)}) ?? []))`:h=`${c}?.map((item) => ${l("item",n.key,n.accessor,!1)}) ?? []`,`
                    const { data: ${A(n.key,"Data")} } =
                    useMany({
                        resource: "${n.resource.name}",
                        ids: ${h},
                        queryOptions: {
                            enabled: !!${c},
                        },
                    });
                    `}}).filter(Boolean),m=u==null?void 0:u.map(n=>{if(n&&n.resource)return A(n.key,"Data")}).filter(Boolean),k=n=>{var h;if(n.relation&&n.resource){let S=`${A(n.key,"Data")}?.data`;if(Array.isArray(n.accessor))return;let V=`id: "${n.key}"`,T=`header: "${I(n.key)}"`,z=Q(n),K="";if(n.multiple){o.push(["TagField","@pankod/refine-mantine"]);let Y="item";n!=null&&n.relationInfer&&(Y=l("item",void 0,n.relationInfer.accessor)),K=`cell: function render({ getValue, table }) {
                        const meta = table.options.meta as {
                            ${A(n.key,"Data")}: GetManyResponse;
                        };

                        const ${A(n.key)} = getValue<any[]>()?.map((item) => {
                            return meta.${A(n.key,"Data")}?.data?.find(
                                (resourceItems) => resourceItems.id === ${l("item",void 0,n.accessor)}
                            );
                        })

                        return (
                            <Group spacing="xs">
                                {${A(n.key)}?.map((item, index) => (
                                    <TagField key={index} value={${Y}} />
                                ))}
                            </Group>
                        )
                    }
                `}else n!=null&&n.relationInfer?K=`cell: function render({ getValue, table }) {
                            const meta = table.options.meta as {
                                ${A(n.key,"Data")}: GetManyResponse;
                            };

                            const ${A(n.key)} = meta.${S}?.find(
                                (item) => item.id === getValue<any>(),
                            );

                            return ${l(A(n.key),void 0,(h=n==null?void 0:n.relationInfer)==null?void 0:h.accessor)} ?? "Loading...";
                        },`:K="";return`
                    {
                        ${V},
                        ${T},
                        ${z},
                        ${K}
                    }
                `}},F=n=>{if(n.type==="image"){o.push(["Image","@pankod/refine-mantine"]);let h=`id: "${n.key}"`,S=Q(n),V=`header: "${I(n.key)}"`,T=f`
                    cell: function render({ getValue }) {
                        ${n!=null&&n.accessor?"try {":""}
                            return <Image sx={{ maxWidth: "100px" }} src={${l("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0," + ")}} />
                        ${n!=null&&n.accessor?" } catch (error) { return null; }":""}
                       
                    }
                `;if(n.multiple){let z=l("item",void 0,n.accessor," + ");T=`
                        cell: function render({ getValue }) {
                            ${n!=null&&n.accessor?"try {":""}
                                return (
                                    <Group spacing="xs">
                                        {getValue<any[]>()?.map((item, index) => (
                                            <Image src={${z}} key={index} sx={{ maxWidth: "100px" }} />
                                        ))}
                                    </Group>
                                )
                            ${n!=null&&n.accessor?" } catch (error) { return null; }":""}
                        }
                    `}return`
                    {
                        ${h},
                        ${S},
                        ${V},
                        ${T}
                    }
                `}},b=n=>{if(n.type==="email"){o.push(["EmailField","@pankod/refine-mantine"]);let h=`id: "${n.key}"`,S=Q(n),V=`header: "${I(n.key)}"`,T=f`
                    cell: function render({ getValue }) {
                        return <EmailField value={${l("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0,' + " " + ')}} />
                    }
                `;return n.multiple&&(o.push(["TagField","@pankod/refine-mantine"]),T=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any>()?.map((item, index) => (
                                        <TagField value={${l("item",void 0,n.accessor," + ")}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${S},
                        ${V},
                        ${T}
                    }
                `}},$=n=>{if(n.type==="url"){o.push(["UrlField","@pankod/refine-mantine"]);let h=`id: "${n.key}"`,S=Q(n),V=`header: "${I(n.key)}"`,T=f`
                    cell: function render({ getValue }) {
                        return <UrlField value={${l("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0," + ")}} />
                    }
                `;return n.multiple&&(o.push(["TagField","@pankod/refine-mantine"]),T=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${l("item",void 0,n.accessor," + ")}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${S},
                        ${V},
                        ${T}
                    }
                `}},C=n=>{if((n==null?void 0:n.type)==="boolean"){o.push(["BooleanField","@pankod/refine-mantine"]);let h=`id: "${n.key}"`,S=Q(n),V=`header: "${I(n.key)}"`,T=f`
                    cell: function render({ getValue }) {
                        return <BooleanField value={${l("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0," + ")}} />
                    }
                `;return n.multiple&&(T=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any[]>()?.map((item, index) => (
                                        <BooleanField value={${l("item",void 0,n.accessor," + ")}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${S},
                        ${V},
                        ${T}
                    }
                `}},L=n=>{if(n.type==="date"){o.push(["DateField","@pankod/refine-mantine"]);let h=`id: "${n.key}"`,S=Q(n),V=`header: "${I(n.key)}"`,T=f`
                    cell: function render({ getValue }) {
                        return <DateField value={${l("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0,' + " " + ')}} />
                    }
                `;return n.multiple&&(T=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any[]>()?.map((item, index) => (
                                        <DateField value={${l("item",void 0,n.accessor," + ")}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${S},
                        ${V},
                        ${T}
                    }
                `}},p=n=>{if((n==null?void 0:n.type)==="richtext"){o.push(["MarkdownField","@pankod/refine-mantine"]);let h=`id: "${n.key}"`,S=Q(n),V=`header: "${I(n.key)}"`,T=f`
                    cell: function render({ getValue }) {
                        return <MarkdownField value={(${l("getValue<string>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0)})?.slice(0, 80) + "..." } />
                    }
                `;return n.multiple&&(T=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<string[]>()?.map((item, index) => (
                                        <MarkdownField value={${l("item",void 0,n.accessor," + ")}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),`
                    {
                        ${h},
                        ${S},
                        ${V},
                        ${T}
                    }
                `}},P=n=>{if(n&&(n.type==="text"||n.type==="number")){let h=`id: "${n.key}"`,S=Q(n),V=`header: "${I(n.key)}"`,T="";return n.multiple&&(o.push(["TagField","@pankod/refine-mantine"]),T=`
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${l("item",void 0,n.accessor,' + " " + ')}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `),!n.multiple&&Array.isArray(n.accessor)&&(T=`
                        cell: function render({ getValue }) {
                            return (
                                <>{${l("getValue()",n.key,n.accessor)}}</>
                            );
                        }
                    `),`
                    {
                        ${h},
                        ${S},
                        ${V},
                        ${T}
                    }
                `}},{canEdit:v,canShow:a,canDelete:x}=r!=null?r:{};v&&o.push(["EditButton","@pankod/refine-mantine"]),a&&o.push(["ShowButton","@pankod/refine-mantine"]),x&&o.push(["DeleteButton","@pankod/refine-mantine"]);let G=v||a||x?f`
        {
            id: "actions",
            accessorKey: "id",
            header: "Actions",
            cell: function render({ getValue }) {
                return (
                    <Group spacing="xs" noWrap>
                        ${a?f`
                        <ShowButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `:""}
                        ${v?f`<EditButton
                            hideText
                            recordItemId={getValue() as string}
                        />`:""}
                        ${x?f`
                        <DeleteButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `:""}
                    </Group>
                );
            },
        },
            `:"",e=t.map(n=>{switch(n==null?void 0:n.type){case"text":case"number":return P(n);case"richtext":return p(n);case"email":return b(n);case"image":return F(n);case"date":return L(n);case"boolean":return C(n);case"url":return $(n);case"relation":return k(n);default:return}});return H(o),f`
        import React from "react";
        ${_(o)}
        
        export const ${s}: React.FC<IResourceComponentsProps> = () => {
            const columns = React.useMemo<ColumnDef<any>[]>(() => [
                ${[...e,G].filter(Boolean).join(",")}
            ], []);

            const {
                getHeaderGroups,
                getRowModel,
                setOptions,
                refineCore: {
                    setCurrent,
                    pageCount,
                    current,
                    tableQueryResult: { data: tableData },
                },
            } = useTable({
                columns,
                ${i?`
                refineCoreProps: {
                    resource: "${r.name}",
                }
                `:""}
            });

            ${g}

            setOptions((prev) => ({
                ...prev,
                meta: {
                    ...prev.meta,
                    ${m.join(", ")}
                },
            }));

            return (
                <List>
                    <ScrollArea>
                        <Table highlightOnHover>
                            <thead>
                                {getHeaderGroups().map((headerGroup) => (
                                    <tr key={headerGroup.id}>
                                        {headerGroup.headers.map((header) => {
                                            return (
                                                <th key={header.id}>
                                                    {!header.isPlaceholder && (
                                                        flexRender(
                                                            header.column
                                                                .columnDef
                                                                .header,
                                                            header.getContext(),
                                                        )
                                                    )}
                                                </th>
                                            );
                                        })}
                                    </tr>
                                ))}
                            </thead>
                            <tbody>
                                {getRowModel().rows.map((row) => {
                                    return (
                                        <tr key={row.id}>
                                            {row.getVisibleCells().map((cell) => {
                                                return (
                                                    <td key={cell.id}>
                                                        {flexRender(
                                                            cell.column.columnDef.cell,
                                                            cell.getContext(),
                                                        )}
                                                    </td>
                                                );
                                            })}
                                        </tr>
                                    );
                                })}
                            </tbody>
                        </Table>
                    </ScrollArea>    
                    <br />
                    <Pagination
                        position="right"
                        total={pageCount}
                        page={current}
                        onChange={setCurrent}
                    />
                </List>
            );
        };
        `}});var Kr=R(require("@pankod/refine-mantine"));var ge=q({type:"create",additionalScope:[["@pankod/refine-mantine","RefineMantine",Kr]],codeViewerComponent:X,loadingComponent:J,errorComponent:Z,renderer:({resource:r,fields:t,isCustomPage:i})=>{var v;let s=B((v=r.label)!=null?v:r.name,"create"),c=[["Create","@pankod/refine-mantine"],["useForm","@pankod/refine-mantine"]],o={},g=t.filter(a=>(a==null?void 0:a.relation)&&!(a!=null&&a.fieldable)&&(a==null?void 0:a.resource)).filter(Boolean).map(a=>{if((a==null?void 0:a.relation)&&!a.fieldable&&a.resource)return c.push(["useSelect","@pankod/refine-mantine"]),`
                    const { selectProps: ${A(a.key,"SelectProps")} } =
                    useSelect({
                        resource: "${a.resource.name}",
                        ${se(a)}
                    });
                `}).filter(Boolean),m=a=>{if(a.relation&&a.resource){o={...o,[a.key]:a.multiple?[]:a.accessor?{[typeof a.accessor=="string"?a.accessor:a.accessor[0]]:""}:""};let x=A(a.key,"SelectProps");return a.multiple?(c.push(["MultiSelect","@pankod/refine-mantine"]),f`
                        <MultiSelect mt="sm" label="${I(a.key)}" {...getInputProps("${N(a.key,void 0)}")} {...${x}} filterDataOnExactSearchMatch={undefined} />
                    `):(c.push(["Select","@pankod/refine-mantine"]),f`
                    <Select mt="sm" label="${I(a.key)}" {...getInputProps("${N(a.key,void 0,a.accessor)}")} {...${x}} />
                `)}},k=a=>{if(a.type==="text"||a.type==="email"||a.type==="date"||a.type==="url")return ee(a.key)||(c.push(["TextInput","@pankod/refine-mantine"]),o={...o,[a.key]:a.multiple?[]:""},a.multiple)?void 0:f`
                    <TextInput mt="sm" label="${I(a.key)}" {...getInputProps("${N(a.key,void 0,a.accessor)}")} />
                `},F=a=>{if(a.type==="image")return f`
                {/* 
                    Dropzone component is not included in "@pankod/refine-mantine" package.
                    To use a <Dropzone> component, you can follow the official documentation for Mantine.
                    
                    Docs: https://mantine.dev/others/dropzone/
                */}
                `},b=a=>{if(a.type==="boolean")return c.push(["Checkbox","@pankod/refine-mantine"]),o={...o,[a.key]:a.multiple?[]:""},a.multiple?void 0:f`
                    <Checkbox mt="sm" label="${I(a.key)}" {...getInputProps("${N(a.key,void 0,a.accessor)}", { type: 'checkbox' })} />
                `},$=a=>{if(a.type==="date"){let x=k(a);return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-mantine" package.
                        To use a <DatePicker> component, you can follow the official documentation for Mantine.
                        
                        Docs: https://mantine.dev/dates/date-picker/
                    */}
                    ${x!=null?x:""}
                `}},C=a=>{if(a.type==="richtext")return c.push(["Textarea","@pankod/refine-mantine"]),o={...o,[a.key]:a.multiple?[]:""},a.multiple?void 0:f`
                    <Textarea mt="sm" label="${I(a.key)}" autosize {...getInputProps("${N(a.key,void 0,a.accessor)}")} />
                `},L=a=>{if(a.type==="number")return ee(a.key)||(c.push(["NumberInput","@pankod/refine-mantine"]),o={...o,[a.key]:a.multiple?[]:""},a.multiple)?void 0:f`
                    <NumberInput mt="sm" label="${I(a.key)}" {...getInputProps("${N(a.key,void 0,a.accessor)}")}/>
                `},p=a=>{if(a)return f`
                    ${a}
            `},P=t.map(a=>{switch(a==null?void 0:a.type){case"url":case"text":case"email":return p(k(a));case"number":return p(L(a));case"richtext":return p(C(a));case"image":return p(F(a));case"date":return p($(a));case"boolean":return p(b(a));case"relation":return p(m(a));default:return}});return H(c),f`
        ${_(c)}
        
        export const ${s} = () => {
            const { getInputProps, saveButtonProps, setFieldValue, refineCore: { formLoading } } = useForm({
                initialValues: ${JSON.stringify(o)},
                ${i?`refineCoreProps: {
                            resource: "${r.name}",
                            action: "create",  
                        }`:""}
            });
        
            ${g}

            return (
                <Create isLoading={formLoading} saveButtonProps={saveButtonProps}>
                    ${P.join("")}
                </Create>
            );
        };
        `}});var _r=R(require("@pankod/refine-mantine"));var Ie=q({type:"edit",additionalScope:[["@pankod/refine-mantine","RefineMantine",_r]],codeViewerComponent:X,loadingComponent:J,errorComponent:Z,renderer:({resource:r,fields:t,isCustomPage:i,id:s})=>{var x,G;let c=B((x=r.label)!=null?x:r.name,"edit"),o=A((G=r.label)!=null?G:r.name,"Data"),u=[["Edit","@pankod/refine-mantine"],["useForm","@pankod/refine-mantine"]],g={},k=t.filter(e=>(e==null?void 0:e.relation)&&!(e!=null&&e.fieldable)&&(e==null?void 0:e.resource)).filter(Boolean).map(e=>{if((e==null?void 0:e.relation)&&!e.fieldable&&e.resource){u.push(["useSelect","@pankod/refine-mantine"]);let y=l(o,e.key,e.accessor,!1);e.multiple&&e.accessor&&(y=`${l(o,e.key)}?.map((item: any) => ${l("item",void 0,e.accessor)})`);let n="";return e.multiple&&e.accessor&&(n=`React.useEffect(() => {
                            setFieldValue("${e.key}", ${y});
                        }, [${o}]);`),`
                    const { selectProps: ${A(e.key,"SelectProps")} } =
                    useSelect({
                        resource: "${e.resource.name}",
                        defaultValue: ${y},
                        ${se(e)}
                    });

                    ${n}
                `}}).filter(Boolean),F=e=>{if(e.relation&&e.resource){g={...g,[e.key]:e.multiple?[]:e.accessor?{[typeof e.accessor=="string"?e.accessor:e.accessor[0]]:""}:""};let y=A(e.key,"SelectProps");return e.multiple?(u.push(["MultiSelect","@pankod/refine-mantine"]),f`
                        <MultiSelect mt="sm" label="${I(e.key)}" {...getInputProps("${N(e.key,void 0)}")} {...${y}} filterDataOnExactSearchMatch={undefined} />
                    `):(u.push(["Select","@pankod/refine-mantine"]),f`
                    <Select mt="sm" label="${I(e.key)}" {...getInputProps("${N(e.key,void 0,e.accessor)}")} {...${y}} />
                `)}},b=e=>{if(e.type==="text"||e.type==="email"||e.type==="date"||e.type==="url"){if(u.push(["TextInput","@pankod/refine-mantine"]),g={...g,[e.key]:e.multiple?[]:""},e.multiple){u.push(["Group","@pankod/refine-mantine"]);let y=N(e.key,"${index}",e.accessor);return`
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any, index: number) => (
                            <TextInput mt="sm" key={index} label={\`${I(e.key)} \${index + 1} \`} {...getInputProps(\`${y}\`)} />
                        ))}
                    </Group>
                    `}return f`
                    <TextInput mt="sm" ${ee(e.key)?"disabled":""} label="${I(e.key)}" {...getInputProps("${N(e.key,void 0,e.accessor)}")} />
                `}},$=e=>{if(e.type==="image")return f`
                {/* 
                    Dropzone component is not included in "@pankod/refine-mantine" package.
                    To use a <Dropzone> component, you can follow the official documentation for Mantine.
                    
                    Docs: https://mantine.dev/others/dropzone/
                */}
                `},C=e=>{if(e.type==="boolean"){if(u.push(["Checkbox","@pankod/refine-mantine"]),g={...g,[e.key]:e.multiple?[]:""},e.multiple){u.push(["Group","@pankod/refine-mantine"]);let y=N(e.key,"${index}",e.accessor);return`
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any, index: number) => (
                            <Checkbox mt="sm" key={index} label={\`${I(e.key)} \${index + 1} \`} {...getInputProps(\`${y}\`, { type: 'checkbox' })} />
                        ))}
                    </Group>
                    `}return f`
                    <Checkbox mt="sm" label="${I(e.key)}" {...getInputProps("${N(e.key,void 0,e.accessor)}", { type: 'checkbox' })} />
                `}},L=e=>{if(e.type==="date")return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-mantine" package.
                        To use a <DatePicker> component, you can follow the official documentation for Mantine.
                        
                        Docs: https://mantine.dev/dates/date-picker/
                    */}
                    ${b(e)}
                `},p=e=>{if(e.type==="richtext"){if(u.push(["Textarea","@pankod/refine-mantine"]),g={...g,[e.key]:e.multiple?[]:""},e.multiple){u.push(["Group","@pankod/refine-mantine"]);let y=N(e.key,"${index}",e.accessor);return`
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any, index: number) => (
                            <Textarea mt="sm" key={index} label={\`${I(e.key)} \${index + 1} \`} {...getInputProps(\`${y}\`)} />
                        ))}
                    </Group>
                    `}return f`
                    <Textarea mt="sm" label="${I(e.key)}" autosize {...getInputProps("${N(e.key,void 0,e.accessor)}")} />
                `}},P=e=>{if(e.type==="number"){if(u.push(["NumberInput","@pankod/refine-mantine"]),g={...g,[e.key]:e.multiple?[]:""},e.multiple){u.push(["Group","@pankod/refine-mantine"]);let y=N(e.key,"${index}",e.accessor);return`
                    <Group spacing="xs">
                        {${l(o,e.key)}?.map((item: any, index: number) => (
                            <NumberInput mt="sm" key={index} label={\`${I(e.key)} \${index + 1} \`} {...getInputProps(\`${y}\`)} />
                        ))}
                    </Group>
                    `}return f`
                    <NumberInput mt="sm" ${ee(e.key)?"disabled":""} label="${I(e.key)}" {...getInputProps("${N(e.key,void 0,e.accessor)}")}/>
                `}},v=e=>{if(e)return f`
                    ${e}
            `},a=t.map(e=>{switch(e==null?void 0:e.type){case"url":case"text":case"email":return v(b(e));case"number":return v(P(e));case"richtext":return v(p(e));case"image":return v($(e));case"date":return v(L(e));case"boolean":return v(C(e));case"relation":return v(F(e));default:return}});return H(u),f`
        ${_(u)}
        
        export const ${c} = () => {
            const { getInputProps, saveButtonProps, setFieldValue, refineCore: { queryResult } } = useForm({
                initialValues: ${JSON.stringify(g)},
                ${i?`refineCoreProps: {
                            resource: "${r.name}",
                            id: ${s},
                            action: "edit",  
                        }`:""}
            });
        
            const ${o} = queryResult?.data?.data;
        
            ${k}

            return (
                <Edit saveButtonProps={saveButtonProps}>
                    ${a.join("")}
                </Edit>
            );
        };
        `}});var Hr=({action:r,id:t,...i})=>{let{useParams:s}=(0,kr.useRouterContext)(),{action:c,id:o}=s();switch(r!=null?r:c){case"show":return ie.default.createElement(ye,{...i,id:t!=null?t:o});case"create":return ie.default.createElement(ge,{...i,id:t!=null?t:o});case"edit":return ie.default.createElement(Ie,{...i,id:t!=null?t:o});default:return ie.default.createElement(fe,{...i,id:t!=null?t:o})}};
