// @pankod/refine-inferencer@1.6.1/dist/mui.js
var $r=Object.create;var ue=Object.defineProperty;var xr=Object.getOwnPropertyDescriptor;var br=Object.getOwnPropertyNames;var Cr=Object.getPrototypeOf,Fr=Object.prototype.hasOwnProperty;var vr=(e,t)=>{for(var a in t)ue(e,a,{get:t[a],enumerable:!0})},Ge=(e,t,a,o)=>{if(t&&typeof t=="object"||typeof t=="function")for(let u of br(t))!Fr.call(e,u)&&u!==a&&ue(e,u,{get:()=>t[u],enumerable:!(o=xr(t,u))||o.enumerable});return e};var R=(e,t,a)=>(a=e!=null?$r(Cr(e)):{},Ge(t||!e||!e.__esModule?ue(a,"default",{value:e,enumerable:!0}):a,e)),Ar=e=>Ge(ue({},"__esModule",{value:!0}),e);var Kr={};vr(Kr,{MuiCreateInferencer:()=>he,MuiEditInferencer:()=>ke,MuiInferencer:()=>_r,MuiListInferencer:()=>ge,MuiShowInferencer:()=>fe});module.exports=Ar(Kr);var se=R(require("react")),Ir=require("@pankod/refine-core");var jr=R(require("@pankod/refine-mui"));var j=R(require("react")),We=require("@pankod/refine-core");var Tr=/(image|photo|avatar|url|logo|cover|thumbnail|icon)/i,$e=(e,t)=>{let a=["name","label","title","count","content","username","nickname","login","firstName","lastName","url"];Tr.test(e)&&a.unshift("url");let o=a.find(u=>u in t&&!!t[u]);if(o)return o==="firstName"&&"lastName"in t?["firstName","lastName"]:o};function xe(e,t,a){var u;if(t)return t;let o=a==null?void 0:a.find(i=>i.name===e);return(u=o==null?void 0:o.options)!=null&&u.dataProviderName?o.options.dataProviderName:"default"}var oe=e=>{var t;return(t=e==null?void 0:e.options)==null?void 0:t.dataProviderName};var be=e=>{let t=[...e].filter(Boolean).sort((a,o)=>(o.priority||0)-(a.priority||0))[0];return t||null};var ze=R(require("prettier/standalone")),He=R(require("prettier/parser-typescript")),Ce=e=>{try{return ze.default.format(e,{parser:"typescript",plugins:[He.default],arrowParens:"always",trailingComma:"all",semi:!0,tabWidth:4,printWidth:80})}catch(t){return console.warn("Formatting error in inferencer",t),e}};var Fe=e=>e.replace(/ {2,}/g," ").replace(/\n{2,}/g,`
`);var ve=/(-id|-ids|_id|_ids|Id|Ids|ID|IDs)(\[\])?$/,qe=(e,t)=>{let a=ve.test(e),o=typeof t=="string"||typeof t=="number",u=Array.isArray(t)&&t.every(i=>typeof i=="string"||typeof i=="number");return a&&o||a&&u?{key:e,relation:!0,multiple:!!u,type:"relation",priority:1}:!1};var V=e=>e.replace(ve,"");var f=e=>V(e).replace(/([a-z])([A-Z])/g,"$1 $2").replace(/_/g," ").replace(/-/g," ").split(" ").map(p=>p.charAt(0).toUpperCase()+p.slice(1)).join(" ");var Ae=e=>e.replace(/export\s+(const|let|var|type|interface|function|class)\s+(\w+)\s*(=|:)\s*/g,"$1 $2 = ").split(`
`).filter(u=>!u.trim().startsWith("export default")).join(`
`);var Ue=/import(?:(?:(?:[ \n\t]+([^ *\n\t\{\},]+)[ \n\t]*(?:,|[ \n\t]+))?([ \n\t]*\{(?:[ \n\t]*[^ \n\t"'\{\}]+[ \n\t]*,?)+\})?[ \n\t]*)|[ \n\t]*\*[ \n\t]*as[ \n\t]+([^ \n\t\{\}]+)[ \n\t]+)from[ \n\t]*(?:['"])([^'"\n]+)(?:['"])(?:;?)/g,wr=/import[ \n\t](?:['"])([^'"\n]+)(?:['"])(?:;?)/g,Rr=/((?:\w|\s|_)*)( as )((?:\w|\s|_)*)( |,)?/g,Te=(e,t={})=>{let a=e.matchAll(Ue),o=new Set;for(let u of a){let[,i,p,h,d]=u;if(d in t){let k=t[d];if(d==="react")continue;i&&o.add(`const { default: ${i} } = ${k};`),p&&o.add(`const${p.replace(Rr,"$1: $3$4")} = ${k};`),h&&o.add(`const ${h} = ${k};`)}}return Fe(`
    ${Array.from(o).join(`
`)}
    ${e.replace(Ue,"").replace(wr,"")}
    `)};var ae=R(require("pluralize"));var we=(e,t)=>e?t.find(o=>{let u=V(e.key);return o.name===u||o.name===(ae.default.isPlural(u)?u:ae.default.plural(u))||o.name===(ae.default.isSingular(u)?u:ae.default.singular(u))}):void 0;var _e=R(require("pluralize"));var O=(e,t)=>{let a=e.replace(/[^a-zA-Z0-9]/g," "),o=_e.default.singular(a),u=f(o).replace(/ /g,""),i=t.charAt(0).toUpperCase()+t.slice(1);return/^[a-zA-Z]/.test(e)?`${u}${i}`:`${i}${u}`};var Sr=e=>e?Array.isArray(e)?e.join(""):typeof e=="string"?e:typeof e=="object"||typeof e=="function"?e.toString():typeof e=="number"?String(e):typeof e=="boolean"&&e?"true":e:"",$=(e,...t)=>e.reduce((o,u,i)=>`${o}${u}${Sr(t[i])}`,"").trim();var Re=(e,t)=>`
    ${e!=null?e:""}
    
    render(typeof ${t} !== "undefined" ? <${t} /> : <></>);
    `;var Se=e=>e==null?void 0:e.replace(/(\/\* hidden-start \*\/)(.|\n)*?(\/\* hidden-end \*\/)/g,"");var Lr=/^[a-zA-Z_$][a-zA-Z_$0-9]*$/,Ke=e=>Lr.test(e);function Le(e,t,a){let o=`${e}`;return t&&(o+="?.",Ke(t)?o+=t:o+=`['${t}']`),a&&(o+="?.",Ke(a)?o+=a:o+=`['${a}']`),o}function Nr(e,t,a){return(a!=null?a:[]).map(u=>Le(e,t,u))}var l=(e,t,a,o=' + " " + ')=>Array.isArray(a)?o?Nr(e,t,a).join(o):Le(e,t,a[0]):Le(e,t,a),B=(e,t,a)=>{let o=e;return t&&(o+=`.${t}`),a&&(Array.isArray(a)?o+=`.${a[0]}`:o+=`.${a}`),o};var z=e=>{let t=e.reduce((o,[u,i,p])=>(o[i]||(o[i]=[]),o[i].includes(u)||(p?o[i].push([u,!0]):o[i].push(u)),o),{});return Object.entries(t).map(([o,u])=>{let i=u.find(k=>Array.isArray(k)&&k[1]),p=u.filter(k=>typeof k=="string"),h=i?`${i[0]}`:"",d=p.length?`{ ${p.join(", ")} }`:"";return`import ${h}${h&&d?", ":""}${d} from "${o}";`})};var Ne=R(require("pluralize")),Y=(e="")=>Ne.default.isSingular(e)?e:Ne.default.singular(e);var Pe=R(require("pluralize")),ee=(e="")=>Pe.default.isPlural(e)?e:Pe.default.plural(e);var Pr=/^id$/i,X=e=>Pr.test(e);var H=e=>{};var N=(e,t="")=>{if(Number.isInteger(Number(e)))return t.charAt(0).toLowerCase()+t.slice(1)+e;let o=e.replace(/^\d+/,"").replace(/[^a-zA-Z0-9]/g,"-"),u=f(o).replace(/ /g,"");return`${u.charAt(0).toLowerCase()+u.slice(1)}${t.charAt(0).toUpperCase()+t.slice(1)}`};var Ze=e=>{let t=(a,o,u,i=t)=>{let p=e.map(d=>d(a,o,u,i));return be(p)};return t};var Xe=e=>(a,o,u,i,p)=>e.reduce((h,d)=>d(h,o,u,i,p),a);var Je=(e,t,a,o)=>{let u=Array.isArray(t),i=Array.isArray(t)&&t.every(p=>typeof p=="string"||typeof p=="number");if(u){if(!i){let h=o(e,t[0],a,o);return h?{...h,key:e,multiple:!0,priority:1}:!1}let p=o(e,t[0],a,o)||{type:"string"};return{key:e,multiple:!0,priority:1,type:p==null?void 0:p.type}}return!1};var Qe=(e,t)=>typeof t=="boolean"?{key:e,type:"boolean"}:!1;var Be=R(require("dayjs")),Br=/(_at|_on|At|On|AT|ON)(\[\])?$/,Ye=(e,t)=>{let a=Br.test(e)&&(0,Be.default)(t).isValid(),o=typeof t=="string"&&(0,Be.default)(t).isValid();return a||o?{key:e,type:"date",priority:1}:!1};var Er=/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,er=(e,t)=>typeof t=="string"&&Er.test(t)?{key:e,type:"email",priority:1}:!1;var Vr=/\.(gif|jpe?g|tiff?|png|webp|bmp|svg)$/i,rr=(e,t)=>typeof t=="string"&&Vr.test(t)?{key:e,type:"image",priority:2}:!1;var tr=(e,t)=>typeof t>"u"||t===null?null:!1;var nr=(e,t)=>typeof t=="number"?{key:e,type:"number"}:!1;var Dr=/id$/i,or=(e,t,a,o)=>{let u=t!==null,i=!Array.isArray(t);if(u&&i&&typeof t=="object"){if(Object.keys(t).length===1&&Dr.test(Object.keys(t)[0]))return{key:e,type:"relation",relation:!0,accessor:"id",priority:1};let d=$e(e,t);if((d?d.length>0:!1)&&d){let b=d&&Array.isArray(d)?d[0]:d,C=o(b,t[b],t,o);if(C){let I=Array.isArray(d)?d.map(c=>C.accessor?`${c}.${Array.isArray(C.accessor)?C.accessor[0]:C.accessor}`:c):C.accessor?Array.isArray(C.accessor)?`${d}.${C.accessor[0]}`:`${d}.${C.accessor}`:d;return{...C,fieldable:!0,key:e,accessor:I,priority:1}}}return{key:e,fieldable:!1,type:"object",priority:1}}return!1};var ar=(e,t)=>typeof t=="string"&&t.length>100?{key:e,type:"richtext",priority:1}:!1;var sr=(e,t)=>typeof t=="string"?{key:e,type:"text"}:!1;var Mr=/^(https?|ftp):\/\/(-\.)?([^\s/?\.#-]+\.?)+(\/[^\s]*)?$/i,ir=(e,t)=>typeof t=="string"&&Mr.test(t)?{key:e,type:"url",priority:1}:!1;var cr=[Je,Qe,Ye,er,rr,tr,nr,or,qe,ar,sr,ir];var ur=(e,t,a,o)=>e.map(i=>{if(!i.relation&&(i.type==="text"||i.type==="richtext"||i.type==="number")&&!i.canRelation){let p=/^[a-z0-9-]+$/,h=I=>p.test(`${I}`),d=i.key.toLowerCase()!=="id";return!(!i.accessor||typeof i.accessor=="string")||!d?i:(i.multiple?o[i.key]:[o[i.key]]).every(I=>h(i.accessor?I[i.accessor]:I))?{...i,canRelation:!0}:i}return i});var Wr=/(image|photo|avatar|cover|thumbnail|icon)/i,pr=e=>e.map(a=>a.type==="url"&&Wr.test(a.key)?{...a,type:"image"}:a);var lr=(e,t)=>e.map(o=>{let u=we(o,t);return u?{...o,relation:!0,type:"relation",resource:u}:o});var mr=(e,t,a,o,u)=>e.map(p=>{if(p.relation&&p.type==="relation"&&!p.resource){let h=p.accessor?o[p.key][p.accessor]:o[p.key],d=u(p.key,h,o,u);if(d&&d.type!=="relation")return{...p,fieldable:!0,relation:!1,type:d.type,canRelation:!0}}return p});var dr=[pr,lr,mr,ur];var q=R(require("react")),pe=R(require("prism-react-renderer")),yr=R(require("prism-react-renderer/themes/vsDark")),Ee=({code:e,wrapperProps:t})=>((0,q.useEffect)(()=>{var o;let a=document.createElement("style");a.appendChild(document.createTextNode(`
                #code-highlight::-webkit-scrollbar {-webkit-appearance: none; height: 7px; width: 7px;}
                #code-highlight::-webkit-scrollbar-thumb {background-color: #666b7a;}
                #code-highlight::-webkit-scrollbar-corner {background-color: rgb(30, 30, 30);}
                `)),(o=document.getElementById("code-highlight"))==null||o.appendChild(a)},[]),q.default.createElement("div",{id:"code-highlight",style:{maxHeight:"75vh",height:"100%",overflow:"auto",backgroundColor:"rgb(30, 30, 30)"},...t!=null?t:{}},q.default.createElement(pe.default,{...pe.defaultProps,theme:yr.default,code:e,language:"tsx"},({className:a,style:o,tokens:u,getLineProps:i,getTokenProps:p})=>q.default.createElement("pre",{className:a,style:{...o,padding:"14px 14px 14px 14px",margin:"0",width:"100%",boxSizing:"border-box"}},u.map((h,d)=>q.default.createElement("div",{key:d,...i({line:h,key:d})},h.map((k,b)=>q.default.createElement("span",{key:b,...p({token:k,key:b})}))))))));var M=R(require("react")),Or=R(require("@pankod/refine-core")),le=require("@aliemir/react-live");var fr=[["react","React",M.default],["@pankod/refine-core","RefineCore",Or]],Ve=M.default.createContext({}),De=({fetchError:e,code:t,additionalScope:a,errorComponent:o})=>{var h;let u=(h=t&&Ae(Te(t,[...fr,...a!=null?a:[]].reduce((d,[k,b,C,I])=>I?d:{...d,[k]:b},{}))))!=null?h:"",i=M.default.useMemo(()=>({React:M.default,...M.default,...[...fr,...a!=null?a:[]].reduce((d,[k,b,C])=>({...d,[b]:C}),{})}),[a]),p=M.default.useMemo(()=>()=>{let{error:k}=M.default.useContext(Ve);return o?M.default.createElement(o,{error:k!=null?k:e?"Something went wrong while fetching the resource data.":void 0}):null},[o,e]);return M.default.createElement(le.LiveProvider,{Context:Ve,code:u,scope:i,noInline:!0},!e&&M.default.createElement(le.LivePreview,{Context:Ve}),M.default.createElement(p,null))};var re=R(require("react")),me=require("@pankod/refine-core");var gr=(e,t,a)=>{let{resource:o,resourceName:u,id:i,resources:p}=(0,me.useResource)({resourceNameOrRouteName:t}),h=a!=null?a:i,d=(0,me.useDataProvider)(),[k,b]=re.default.useState(void 0),[C,I]=re.default.useState(!0),[c,w]=re.default.useState(!1),m=re.default.useCallback(async s=>{var F,n;let A=(F=oe(o))!=null?F:xe(u,void 0,p),x=d(A);w(!0);try{if(e==="list"||e==="create"){let g=await x.getList({resource:u});b((n=g.data)==null?void 0:n[0]),setTimeout(()=>{w(!1)},500)}if(e==="edit"||e==="show"&&s){let g=await x.getOne({resource:u,id:s});b(g.data),setTimeout(()=>{w(!1)},500)}}catch{setTimeout(()=>{w(!1)},500)}},[e,d,o,u,p]);return re.default.useEffect(()=>{I(!1),!c&&!k&&m(h)},[m,h]),{data:k,loading:c,initial:C}};var te=R(require("react")),hr=require("@pankod/refine-core");var Me=require("lodash"),kr=({record:e,fields:t,infer:a})=>{let o=(0,hr.useDataProvider)(),[u,i]=te.default.useState([]),[p,h]=te.default.useState(!0),[d,k]=te.default.useState(!1),b=te.default.useCallback(async C=>{console.groupCollapsed("@pankod/refine-inferencer is trying to detect relations");let I=[];k(!0);try{let c=C.map(async m=>{if(m&&(m.relation||m.canRelation)&&e){let s=oe(m.resource),A=o(s),x=Array.isArray(m.accessor)?void 0:m.multiple?e[m.key].map(F=>m.accessor?(0,Me.get)(F,m.accessor):F)[0]:m.accessor?(0,Me.get)(e[m.key],m.accessor):e[m.key];if(x&&m.resource)try{let{data:F}=await A.getOne({resource:m.resource.name,id:x});I.push({status:"success",resource:m.resource.name,field:m.key});let n=a("__",F,{},a);return{...m,relationInfer:n}}catch{return I.push({status:"error",resource:m.resource.name,field:m.key}),{...m,relationInfer:null}}if(x){let F,n;try{let{data:r}=await A.getOne({resource:ee(V(m.key)),id:x});I.push({status:"success",resource:ee(V(m.key)),field:m.key}),F=r,n=!0}catch{I.push({status:"error",resource:ee(V(m.key)),field:m.key});try{let{data:v}=await A.getOne({resource:Y(V(m.key)),id:x});I.push({status:"success",resource:Y(V(m.key)),field:m.key}),F=v,n=!1}catch{return I.push({status:"error",resource:Y(V(m.key)),field:m.key}),{...m,relationInfer:null}}}let g=a("__",F,{},a),E=V(m.key);return{...m,relation:!0,type:"relation",resource:{name:n?ee(E):Y(E)},fieldable:!1,canRelation:void 0,relationInfer:g}}return{...m,relationInfer:null}}return m}),w=await Promise.all(c);i(w.filter(m=>m)),setTimeout(()=>{k(!1)},500)}catch{setTimeout(()=>{k(!1)},500)}setTimeout(()=>{console.log(`Tried to detect relations with ${I.length} attempts and succeeded with ${I.filter(c=>c.status==="success").length} attempts.`),console.groupEnd()},500)},[o,e]);return te.default.useEffect(()=>{h(!1),!d&&t&&t.length>0&&u.length===0&&b(t)},[b,d,t,b]),{fields:u,loading:d,initial:p}};var U=({type:e,additionalScope:t=[],customElements:a=[],fieldTransformers:o=[],renderer:u,loadingComponent:i,errorComponent:p,codeViewerComponent:h})=>{let d=Ze([...cr,...a]),k=Xe([...dr,...o]),b=({resourceName:I,fieldTransformer:c,id:w})=>{var L;let{resource:m,resources:s}=(0,We.useResource)({resourceNameOrRouteName:I}),{resource:A}=(0,We.useResource)(),{data:x,loading:F,initial:n}=gr(e,I!=null?I:m==null?void 0:m.name,w),g=j.default.useMemo(()=>{if(x){let P=Object.keys(x).map(J=>{let Q=x[J];return d(J,Q,x,d)}).filter(Boolean),T=k(P,s,m,x,d);return c?T.flatMap(J=>{let Q=c(J);return Q?[Q]:[]}):T}return[]},[x,s,m,c]),{fields:E,loading:r}=kr({record:x,fields:g,infer:d}),v=j.default.useMemo(()=>!F&&!r&&!n?u({resource:m,resources:s,fields:E,infer:d,isCustomPage:m.name!==A.name,id:w}):"",[m,s,E,F,r]);return j.default.createElement(j.default.Fragment,null,i&&(F||r)&&j.default.createElement(i,null),!F&&!r&&j.default.createElement(j.default.Fragment,null,j.default.createElement(De,{fetchError:!F&&!n&&!x,code:Re(v,O((L=m.label)!=null?L:m.name,e)),errorComponent:p,additionalScope:t}),h&&j.default.createElement(h,{code:Se(v),loading:F||r})))};return({name:I,resource:c,fieldTransformer:w,id:m})=>j.default.createElement(b,{fieldTransformer:w,resourceName:c!=null?c:I,key:c!=null?c:I,id:m})};var de=R(require("react")),ne=require("@pankod/refine-mui"),_=({error:e})=>e?de.default.createElement(ne.Box,{style:{display:"flex",justifyContent:"center",alignItems:"center",minHeight:"200px"}},de.default.createElement(ne.Alert,{severity:"error"},de.default.createElement(ne.AlertTitle,null,"Error"),e!=null?e:"")):null;var Oe=R(require("react")),ye=require("@pankod/refine-mui"),K=()=>Oe.default.createElement(ye.Box,{style:{display:"flex",justifyContent:"center",alignItems:"center",minHeight:"300px"}},Oe.default.createElement(ye.CircularProgress,{size:"large"}));var S=R(require("react")),W=require("@pankod/refine-mui"),G=require("@tabler/icons");var Z=({code:e,loading:t})=>{let a=S.default.useMemo(()=>Ce(e!=null?e:""),[e]),o=S.default.useRef(null),[u,i]=S.default.useState(!1),[p,h]=S.default.useState(!1);if(t)return null;let d=()=>{var k,b,C,I;(k=o==null?void 0:o.current)==null||k.select(),(b=o==null?void 0:o.current)==null||b.setSelectionRange(0,99999),typeof navigator<"u"&&(navigator.clipboard.writeText((I=(C=o==null?void 0:o.current)==null?void 0:C.value)!=null?I:""),h(!0),setTimeout(()=>{h(!1)},1e3))};return a?S.default.createElement(S.default.Fragment,null,S.default.createElement(W.Box,{sx:{position:"fixed",bottom:12,right:60}},S.default.createElement(W.Box,{sx:{display:"flex",gap:2}},S.default.createElement(W.Button,{variant:"contained",color:"primary",onClick:()=>i(!0),startIcon:S.default.createElement(G.IconCode,{size:18}),sx:{whiteSpace:"nowrap"}},"Show Code"),S.default.createElement(W.Button,{variant:"contained",color:"secondary",component:"a",target:"_blank",href:"https://github.com/refinedev/refine/discussions/3046",startIcon:S.default.createElement(G.IconMessageCircle,{size:18}),sx:{whiteSpace:"nowrap"}},"Give Feedback"))),S.default.createElement("div",{style:{height:"50px"}}),S.default.createElement(W.Modal,{open:u,onClose:()=>i(!1)},S.default.createElement(W.Box,{sx:{position:"absolute",maxWidth:"800px",width:"100%",top:"50%",left:"50%",transform:"translate(-50%, -50%)",boxShadow:24}},S.default.createElement(W.IconButton,{onClick:()=>i(!1),disableRipple:!0,sx:{position:"absolute",right:0,top:0,height:56,width:56,display:"flex",alignItems:"center",justifyContent:"center"}},S.default.createElement(G.IconX,{size:18,style:{color:"#666b7a"}})),S.default.createElement(Ee,{code:a}),S.default.createElement("textarea",{ref:o,defaultValue:(a!=null?a:"").replace(/\\n/g,`\r
`),id:"code-input",style:{display:"none"}}),S.default.createElement("div",{style:{position:"absolute",bottom:12,right:12}},S.default.createElement(W.Button,{key:"copy",sx:{textTransform:"capitalize",width:100},startIcon:p?S.default.createElement(G.IconCheck,{size:18,style:{color:"green"}}):S.default.createElement(G.IconCopy,{size:18}),color:"inherit",variant:"contained",onClick:d},p?"Copied":"Copy"))))):null};var fe=U({type:"show",additionalScope:[["@pankod/refine-mui","RefineMui",jr]],codeViewerComponent:Z,loadingComponent:K,errorComponent:_,renderer:({resource:e,fields:t,isCustomPage:a,id:o})=>{var F;let u=O((F=e.label)!=null?F:e.name,"show"),i="record",p=[["useShow","@pankod/refine-core"],["Show","@pankod/refine-mui"],["Typography","@pankod/refine-mui"],["Stack","@pankod/refine-mui"]],d=t.filter(n=>(n==null?void 0:n.relation)&&!(n!=null&&n.fieldable)&&(n==null?void 0:n.resource)).filter(Boolean).map(n=>{if((n==null?void 0:n.relation)&&!n.fieldable&&n.resource){if(n.multiple){p.push(["useMany","@pankod/refine-core"]);let g=l(i,n.key);return n.accessor&&(g=`${l(i,n.key)}?.map((item: any) => ${l("item",void 0,n.accessor)})`),`
                    const { data: ${N(n.key,"Data")}, isLoading: ${N(n.key,"IsLoading")} } =
                    useMany({
                        resource: "${n.resource.name}",
                        ids: ${g} || [],
                        queryOptions: {
                            enabled: !!${i},
                        },
                    });
                    `}return p.push(["useOne","@pankod/refine-core"]),`
                    const { data: ${N(n.key,"Data")}, isLoading: ${N(n.key,"IsLoading")} } =
                    useOne({
                        resource: "${n.resource.name}",
                        id: ${l(i,n.key,n.accessor,!1)} || "",
                        queryOptions: {
                            enabled: !!${i},
                        },
                    });
                `}}).filter(Boolean),k=n=>{if(n.relation&&n.resource){let g=N(n.key,"Data"),E=N(n.key,"IsLoading");return n.multiple?(p.push(["TagField","@pankod/refine-mui"]),$`
                    <Typography variant="body1" fontWeight="bold">
                        ${f(n.key)}
                    </Typography>
                    {${E} ? (
                        <>
                            Loading...
                        </>
                        ) : (
                        <>
                        ${(()=>{var r;if(n.relationInfer)if((r=n.relationInfer)!=null&&r.accessor){if(Array.isArray(n.relationInfer.accessor))return"Not Handled.";{let v=N(n.key),L=l(v,void 0,n.relationInfer.accessor);return`
                                            <Stack direction="row" spacing={1}>
                                                {${g}?.data?.map((${v}: any) => (
                                                    <TagField key={${L}} value={${L}} />
                                                ))}
                                            </Stack>
                                        `}}else return"Not Handled.";else return"not-handled - relation with multiple but no resource"})()}
                        </>
                    )}
                    `):$`
                    <Typography variant="body1" fontWeight="bold">
                        ${f(n.key)}
                    </Typography>

                    {${E} ? (
                        <>Loading...</>
                    ) : (
                        <>
                        ${(()=>{var r;return n.relationInfer?(r=n.relationInfer)!=null&&r.accessor?Array.isArray(n.relationInfer.accessor)?`{${l(`${g}?.data`,void 0,n.relationInfer.accessor,' + " " + ')}}`:`{${g}?.data?.${n.relationInfer.accessor}}`:`{${g}?.data}`:`{${g}?.data?.id}`})()}
                        </>
                    )}
                    `}},b=n=>{if(n.type==="text"){if(p.push(["TextFieldComponent as TextField","@pankod/refine-mui"]),n.multiple){p.push(["TagField","@pankod/refine-mui"]);let g=l("item",void 0,n.accessor);return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${f(n.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${l(i,n.key)}?.map((item: any) => (
                        <TagField value={${g}} key={${g}} />
                    ))}
                    </Stack>
                `}return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${f(n.key)}
                    </Typography>
                    <TextField value={${l(i,n.key,n.accessor)}} />
                `}},C=n=>{if(n.type==="image"){if(n.multiple){let g=l("item",void 0,n.accessor);return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${f(n.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${l(i,n.key)}?.map((item: any) => (
                        <img style={{ maxWidth: 200, width: "100%", height: 200 }} src={${g}} key={${g}} />
                    ))}
                    </Stack>
                `}return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${f(n.key)}
                    </Typography>
                    <img style={{ maxWidth: 200, width: "100%", height: 200 }} src={${l(i,n.key,n.accessor," + ")}} />
                    `}},I=n=>{if(n.type==="email"){if(p.push(["EmailField","@pankod/refine-mui"]),n.multiple){p.push(["TagField","@pankod/refine-mui"]);let g=l("item",void 0,n.accessor);return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${f(n.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${l(i,n.key)}?.map((item: any) => (
                        <TagField value={${g}} key={${g}} />
                    ))}
                    </Stack>
                `}return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${f(n.key)}
                    </Typography>
                    <EmailField value={${l(i,n.key,n.accessor," + ")}} />
                `}},c=n=>{if(n.type==="url"){if(p.push(["UrlField","@pankod/refine-mui"]),n.multiple){p.push(["TagField","@pankod/refine-mui"]);let g=l("item",void 0,n.accessor);return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${f(n.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${l(i,n.key)}?.map((item: any) => (
                        <TagField value={${g}} key={${g}} />
                    ))}
                    </Stack>
                `}return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${f(n.key)}
                    </Typography>
                    <UrlField value={${l(i,n.key,n.accessor," + ")}} />
                `}},w=n=>{if(n.type==="boolean"){if(p.push(["BooleanField","@pankod/refine-mui"]),n.multiple){p.push(["TagField","@pankod/refine-mui"]);let g=l("item",void 0,n.accessor);return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${f(n.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${l(i,n.key)}?.map((item: any, index: number) => (
                        <TagField value={${g}} key={index} />
                    ))}
                    </Stack>
                `}return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${f(n.key)}
                    </Typography>
                    <BooleanField value={${l(i,n.key,n.accessor," && ")}} />
                `}},m=n=>{if(n.type==="date"){if(p.push(["DateField","@pankod/refine-mui"]),n.multiple){let g=l("item",void 0,n.accessor);return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${f(n.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${l(i,n.key)}?.map((item: any) => (
                        <DateField value={${g}} key={${g}} />
                    ))}
                    </Stack>
                `}return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${f(n.key)}
                    </Typography>
                    <DateField value={${l(i,n.key,n.accessor," + ' ' + ")}} />
                `}},s=n=>{if(n.type==="richtext")return p.push(["MarkdownField","@pankod/refine-mui"]),$`
                    <Typography variant="body1" fontWeight="bold">
                        ${f(n.key)}
                    </Typography>
                    <MarkdownField value={${l(i,n.key,n.accessor,' + " " + ')}} />
                `},A=n=>{if(n.type==="number"){if(p.push(["NumberField","@pankod/refine-mui"]),n.multiple){p.push(["TagField","@pankod/refine-mui"]);let g=l("item",void 0,n.accessor);return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${f(n.key)}
                    </Typography>
                    <Stack direction="row" spacing={1}>
                    {${l(i,n.key)}?.map((item: any) => (
                        <TagField value={${g}} key={${g}} />
                    ))}
                    </Stack>
                `}return $`
                    <Typography variant="body1" fontWeight="bold">
                        ${f(n.key)}
                    </Typography>
                    <NumberField value={${l(i,n.key,n.accessor,' + " " + ')} ?? ""} />
                `}},x=t.map(n=>{switch(n==null?void 0:n.type){case"text":return b(n);case"number":return A(n);case"richtext":return s(n);case"email":return I(n);case"image":return C(n);case"date":return m(n);case"boolean":return w(n);case"url":return c(n);case"relation":return k(n);default:return}});return H(p),$`
        ${z(p)}

        export const ${u} = () => {
            const { queryResult } = useShow(${a?`{ 
                        resource: "${e.name}", 
                        id: ${o}
                    }`:""});
            const { data, isLoading } = queryResult;
        
            const ${i} = data?.data;
        
            ${d}

            return (
                <Show isLoading={isLoading}>
                    <Stack gap={1}>
                        ${x.join("")}
                    </Stack>
                </Show>
            );
        };
        `}});var Gr=R(require("@pankod/refine-mui"));var ge=U({type:"list",additionalScope:[["@pankod/refine-mui","RefineMui",Gr]],codeViewerComponent:Z,loadingComponent:K,errorComponent:_,renderer:({resource:e,fields:t,isCustomPage:a})=>{var E;let o=O((E=e.label)!=null?E:e.name,"list"),u="dataGridProps?.rows",i=[["React","react",!0],["useDataGrid","@pankod/refine-mui"],["DataGrid","@pankod/refine-mui"],["GridColumns","@pankod/refine-mui"],["EditButton","@pankod/refine-mui"],["ShowButton","@pankod/refine-mui"],["DeleteButton","@pankod/refine-mui"],["List","@pankod/refine-mui"]],p=t.filter(r=>(r==null?void 0:r.relation)&&!(r!=null&&r.fieldable)&&(r==null?void 0:r.resource)),h=p.filter(Boolean).map(r=>{if((r==null?void 0:r.relation)&&!r.fieldable&&r.resource){i.push(["useMany","@pankod/refine-core"]);let v="";return r.multiple?v=`[].concat(...(${u}?.map((item: any) => ${l("item",r.key,r.accessor,!1)}) ?? []))`:v=`${u}?.map((item: any) => ${l("item",r.key,r.accessor,!1)}) ?? []`,`
                    const { data: ${N(r.key,"Data")}, isLoading: ${N(r.key,"IsLoading")} } =
                    useMany({
                        resource: "${r.resource.name}",
                        ids: ${v},
                        queryOptions: {
                            enabled: !!${u},
                        },
                    });
                    `}}).filter(Boolean),d=p==null?void 0:p.map(r=>{if(r&&r.resource)return`${N(r.key,"Data")}?.data`}).filter(Boolean),k=r=>{var v,L;if(r.relation&&r.resource){let P=`${N(r.key,"Data")}?.data`,T=N(r.key,"IsLoading");if(Array.isArray(r.accessor))return;let D=`${T} ? <>Loading...</> : `,J=`field: "${r.key}"`,Q=r.accessor&&!r.multiple&&!Array.isArray(r.accessor)?`valueGetter: ({ row }) => {
                const value = ${l("row",r.key,r.accessor,!1)};

                return value;
            },`:"",je=`headerName: "${f(r.key)}"`,ie="";if(r.multiple){i.push(["TagField","@pankod/refine-mui"]);let ce="item";if(r!=null&&r.relationInfer){let Ie=`${P}?.find((resourceItems) => resourceItems.id === ${l("item",void 0,r.accessor)})`;ce=l(Ie,void 0,(v=r==null?void 0:r.relationInfer)==null?void 0:v.accessor)}ie=`
                    renderCell: function render({ value }) {
                        return ${D} (
                            <>
                                {${l("value",void 0)}?.map((item: any, index: number) => (
                                    <TagField key={index} value={${ce}} />
                                ))}
                            </>
                        )
                    }
                    `}else if(r!=null&&r.relationInfer){let ce=`${P}?.find((item) => item.id === value)`,Ie=l(ce,void 0,(L=r==null?void 0:r.relationInfer)==null?void 0:L.accessor);ie=`
                        renderCell: function render({ value }) {
                            return ${D} ${Ie};
                        }
                        `}else ie="";return`
                    {
                        ${J},
                        ${je},${Q}
                        minWidth: 300,${ie}
                    }
                `}},b=r=>{if(r.type==="image"){let v=`field: "${r.key}"`,L=`headerName: "${f(r.key)}"`,P=r.accessor&&!r.multiple&&!Array.isArray(r.accessor)?`valueGetter: ({ row }) => {
                const value = ${l("row",r.key,r.accessor,!1)};

                return value;
            },`:"",T=`
                    renderCell: function render({ value }) {
                        return (
                            <img src={${l("value",void 0,Array.isArray(r.accessor)?r.accessor:void 0," + ")}} style={{ height: "50px", maxWidth: "100px" }} />
                        )
                    }
                `;return r.multiple&&(T=`
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item: any, index: number) => (
                                    <img src={${l("item",void 0,r.accessor," + ")}} key={index} style={{ height: "50px", maxWidth: "100px" }} />
                                ))}
                                </>
                            )
                        }
                    `),`
                    {
                        ${v},
                        ${L},${P}
                        minWidth: 100,${T}
                    }
                `}},C=r=>{if(r.type==="email"){i.push(["EmailField","@pankod/refine-mui"]);let v=`field: "${r.key}"`,L=`headerName: "${f(r.key)}"`,P=r.accessor&&!r.multiple&&!Array.isArray(r.accessor)?`valueGetter: ({ row }) => {
                const value = ${l("row",r.key,r.accessor,!1)};

                return value;
            },`:"",T=`
                renderCell: function render({ value }) {
                    return (
                        <EmailField value={${l("value",void 0,Array.isArray(r.accessor)?r.accessor:void 0," + ")}} />
                    )
                }
            `;return r.multiple&&(i.push(["TagField","@pankod/refine-mui"]),T=`
                    renderCell: function render({ value }) {
                        return (
                            <>
                            {value?.map((item: any, index: number) => (
                                <TagField value={${l("item",void 0,r.accessor," + ")}} key={index} />
                            ))}
                            </>
                        )
                    }
                `),`
                {
                    ${v},
                    ${L},${P}
                    minWidth: 250,${T}
                }
            `}},I=r=>{if(r.type==="url"){i.push(["UrlField","@pankod/refine-mui"]);let v=`field: "${r.key}"`,L=`headerName: "${f(r.key)}"`,P=r.accessor&&!r.multiple&&!Array.isArray(r.accessor)?`valueGetter: ({ row }) => {
                const value = ${l("row",r.key,r.accessor,!1)};

                return value;
            },`:"",T=`
                    renderCell: function render({ value }) {
                        return (
                            <UrlField value={${l("value",void 0,Array.isArray(r.accessor)?r.accessor:void 0," + ")}} />
                        )
                    }
                `;return r.multiple&&(i.push(["TagField","@pankod/refine-mui"]),T=`
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item: any, index: any) => (
                                    <TagField value={${l("item",void 0,r.accessor," + ")}} key={index} />
                                ))}
                                </>
                            )
                        }
                    `),`
                    {
                        ${v},
                        ${L},${P}
                        minWidth: 250,${T}
                    }
                `}},c=r=>{if(r!=null&&r.type){i.push(["Checkbox","@pankod/refine-mui"]);let v=`field: "${r.key}"`,L=`headerName: "${f(r.key)}"`,P=r.accessor&&!r.multiple&&!Array.isArray(r.accessor)?`valueGetter: ({ row }) => {
                const value = ${l("row",r.key,r.accessor,!1)};

                return value;
            },`:"",T=`
                    renderCell: function render({ value }) {
                        return (
                            <Checkbox checked={!!${l("value",void 0,Array.isArray(r.accessor)?r.accessor:void 0," && ")}} />
                        );
                    }
                `;return r.multiple&&(T=`
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item: any, index: number) => (
                                    <Checkbox checked={!!${l("item",void 0,r.accessor," && ")}} key={index} />
                                ))}
                                </>
                            )
                        }
                    `),`
                    {
                        ${v},
                        ${L},${P}
                        minWidth: 250,${T}
                    }
                `}},w=r=>{if(r.type==="date"){i.push(["DateField","@pankod/refine-mui"]);let v=`field: "${r.key}"`,L=`headerName: "${f(r.key)}"`,P=r.accessor&&!r.multiple&&!Array.isArray(r.accessor)?`valueGetter: ({ row }) => {
                const value = ${l("row",r.key,r.accessor,!1)};

                return value;
            },`:"",T=`
                    renderCell: function render({ value }) {
                        return <DateField value={value} />;
                    }
                `;return r.multiple&&(T=`
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item: any, index: number) => (
                                    <DateField value={${l("item",void 0,r.accessor,' + " " + ')}} key={index} />
                                ))}
                                </>
                            )
                        }
                    `),`
                    {
                        ${v},
                        ${L},${P}
                        minWidth: 250,${T}
                    }
                `}},m=r=>{if((r==null?void 0:r.type)==="richtext"){i.push(["MarkdownField","@pankod/refine-mui"]);let v=`field: "${r.key}"`,L=`headerName: "${f(r.key)}"`,P=r.accessor&&!r.multiple&&!Array.isArray(r.accessor)?`valueGetter: ({ row }) => {
                const value = ${l("row",r.key,r.accessor,!1)};

                return value;
            },`:"",T=`
                    renderCell: function render({ value }) {
                        return <MarkdownField value={(value ?? "").slice(0, 80) + "..."} />;
                    }
                `;return r.multiple&&(T=`
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item: any, index: number) => (
                                    <MarkdownField value={(${l("item",void 0,r.accessor,' + " " + ')}).slice(0, 80) + "..."} key={index} />
                                ))}
                                </>
                            )
                        }
                    `),`
                    {
                        ${v},
                        ${L},${P}
                        minWidth: 250,${T}
                    }
                `}},s=r=>{if(r&&(r.type==="text"||r.type==="number")){let v=`field: "${r.key}"`,L=`headerName: "${f(r.key)}"`,P=r.accessor&&!r.multiple&&!Array.isArray(r.accessor)?`valueGetter: ({ row }) => {
                const value = ${l("row",r.key,r.accessor,!1)};

                return value;
            },`:"",T="";if(r.multiple){i.push(["TagField","@pankod/refine-mui"]);let D=l("item",void 0,r.accessor,' + " " + ');T=`
                    renderCell: function render({ row }) {
                        return (
                            <>
                                {(${l("row",r.key)})?.map((item: any) => (
                                    <TagField value={${D}} key={${D}} />
                                ))}
                            </>
                        )
                    }
                    `}return!r.multiple&&Array.isArray(r.accessor)&&(T=`
                    renderCell: function render({ row }) {
                        return 
                            <>{${l("row",r.key,r.accessor)}}</>
                        );
                    }
                    `),`
                {
                    ${v},
                    ${L},${P}${r.type==="number"?"type: 'number',":""}
                    minWidth: ${X(r.key)?50:200},${T}
                }
                `}},{canEdit:A,canShow:x,canDelete:F}=e!=null?e:{};A&&i.push(["EditButton","@pankod/refine-mui"]),x&&i.push(["ShowButton","@pankod/refine-mui"]),F&&i.push(["DeleteButton","@pankod/refine-mui"]);let n=A||x||F?$`
                {
                    field: "actions",
                    headerName: "Actions",
                    renderCell: function render({ row }) {
                        return (
                            <>
                                ${A?$`<EditButton hideText recordItemId={row.id} />`:""}
                                ${x?$`<ShowButton hideText recordItemId={row.id} />`:""}
                                ${F?$`<DeleteButton hideText recordItemId={row.id} />`:""}
                            </>
                        );
                    },
                    align: "center",
                    headerAlign: "center",
                    minWidth: 80,
                },
        `:"",g=t.map(r=>{switch(r==null?void 0:r.type){case"text":case"number":return s(r);case"richtext":return m(r);case"email":return C(r);case"image":return b(r);case"date":return w(r);case"boolean":return c(r);case"url":return I(r);case"relation":return k(r);default:return}});return H(i),$`
        ${z(i)}
        
        export const ${o} = () => {
            const { dataGridProps } = useDataGrid(
                ${a?`{ resource: "${e.name}" }`:""} 
            );
        
            ${h}

            const columns = React.useMemo<GridColumns<any>>(() => [
                ${[...g,n].filter(Boolean).join(`,\r
`)}
            ], [${d.join(",")}]);

            return (
                <List>
                    <DataGrid {...dataGridProps} columns={columns} autoHeight />
                </List>
            );
        };
        `}});var zr=R(require("@pankod/refine-mui")),Hr=R(require("@pankod/refine-react-hook-form"));var he=U({type:"create",additionalScope:[["@pankod/refine-mui","RefineMui",zr],["@pankod/refine-react-hook-form","RefineReactHookForm",Hr]],codeViewerComponent:Z,loadingComponent:K,errorComponent:_,renderer:({resource:e,fields:t,isCustomPage:a})=>{var I;let o=O((I=e.label)!=null?I:e.name,"create"),u=[["Create","@pankod/refine-mui"],["Box","@pankod/refine-mui"],["useForm","@pankod/refine-react-hook-form"]],p=t.filter(c=>(c==null?void 0:c.relation)&&!(c!=null&&c.fieldable)&&(c==null?void 0:c.resource)).filter(Boolean).map(c=>{if((c==null?void 0:c.relation)&&!c.fieldable&&c.resource)return u.push(["useAutocomplete","@pankod/refine-mui"]),`
                    const { autocompleteProps: ${N(c.key,"AutocompleteProps")} } =
                    useAutocomplete({
                        resource: "${c.resource.name}",
                    });
                `}).filter(Boolean),h=c=>{if(c.relation&&c.resource){u.push(["Autocomplete","@pankod/refine-mui"],["Controller","@pankod/refine-react-hook-form"]);let w=N(c.key,"AutocompleteProps"),m=c.relationInfer&&c.relationInfer.accessor?typeof c.relationInfer.accessor=="string"?c.relationInfer.accessor:c.relationInfer.accessor[0]:"title",s=m.includes(".")||m.includes("[")||m.includes("]")||m.includes("-"),A=c.accessor?l("item",void 0,c.accessor,!1):"(item?.id ?? item)",x=c.accessor?l("value",void 0,c.accessor,!1):"(value?.id ?? value)",F=c.accessor?"value":c.multiple?"value?.map((item: any) => item?.id ?? item)":"value?.id ?? value";return $`
                    <Controller
                        control={control}
                        name="${B(c.key,void 0)}"
                        rules={{ required: "This field is required" }}
                        // eslint-disable-next-line
                        ${c.multiple?"defaultValue={[] as any}":"defaultValue={null as any}"} 
                        render={({ field }) => (
                            <Autocomplete
                                {...${w}}
                                {...field}
                                ${c.multiple?"multiple":""}
                                onChange={(_, value) => {
                                    field.onChange(${F});
                                }}
                                getOptionLabel={(item) => {
                                    return (
                                        ${w}?.options?.find(
                                            (p) =>
                                                p?.id?.toString() ===
                                                ${A}?.toString(),
                                        )?.${s?`["${m}"]`:m} ?? ""
                                    );
                                }}
                                isOptionEqualToValue={(option, value) =>
                                    value === undefined ||
                                    option.id.toString() === ${x}?.toString()
                                }
                                renderInput={(params) => (
                                    <TextField
                                        {...params}
                                        label="${f(c.key)}"
                                        margin="normal"
                                        variant="outlined"
                                        error={!!${l("(errors as any)",c.key,c.accessor,!1)}}
                                        helperText={${l("(errors as any)",c.key,c.accessor,!1)}?.message}
                                        required
                                    />
                                )}
                            />
                        )}
                    />
                `}},d=c=>{if(c.type==="text"||c.type==="url"||c.type==="email"||c.type==="number"||c.type==="date"||c.type==="richtext")return X(c.key)||(u.push(["TextField","@pankod/refine-mui"]),c.multiple)?void 0:$`
                    <TextField
                        {...register("${B(c.key,void 0,c.accessor)}", {
                            required: "This field is required",
                        })}
                        error={!!${l("(errors as any)",c.key,c.accessor,!1)}}
                        helperText={${l("(errors as any)",c.key,c.accessor,!1)}?.message}
                        margin="normal"
                        fullWidth
                        InputLabelProps={{ shrink: true }}
                        ${c.type!=="date"&&c.type!=="richtext"?`type="${c.type}"`:""}
                        ${c.type==="richtext"?"multiline":""}
                        label="${f(c.key)}"
                        name="${B(c.key,void 0,c.accessor)}"
                    />
                `},k=c=>{if(c.type==="boolean")return u.push(["Checkbox","@pankod/refine-mui"],["FormControlLabel","@pankod/refine-mui"],["Controller","@pankod/refine-react-hook-form"]),c.multiple?void 0:$`
                    <Controller
                        control={control}
                        name="${B(c.key,void 0,c.accessor)}"
                        // eslint-disable-next-line
                        defaultValue={null as any}
                        render={({ field }) => (
                            <FormControlLabel label="${f(c.key)}" control={
                                <Checkbox
                                    {...field}
                                    checked={field.value}
                                    onChange={(event) => {
                                        field.onChange(event.target.checked);
                                    }}
                                />
                            } />
                        )}
                    />
                `},b=c=>{if(c.type==="date"){let w=d(c);return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-mui" package.
                        To use a <DatePicker> component, you can follow the official documentation for Material UI.
                        
                        Docs: https://mui.com/x/react-date-pickers/date-picker/#basic-usage
                    */}
                    ${w!=null?w:""}
                    `}},C=t.map(c=>{switch(c==null?void 0:c.type){case"text":case"number":case"email":case"url":case"richtext":return d(c);case"date":return b(c);case"boolean":return k(c);case"relation":return h(c);default:return}});return H(u),$`
        ${z(u)}
        
        export const ${o} = () => {
            const {
                saveButtonProps,
                refineCore: { formLoading },
                register,
                control,
                formState: { errors },
            } = useForm(
                ${a?`{ 
                    refineCoreProps: {
                        resource: "${e.name}",
                        action: "create",
                    }
                }`:""}
            );
        
            ${p}

            return (
                <Create isLoading={formLoading} saveButtonProps={saveButtonProps}>
                    <Box
                        component="form"
                        sx={{ display: "flex", flexDirection: "column" }}
                        autoComplete="off"
                    >
                        ${C.join("")}
                    </Box>
                </Create>
            );
        };
        `}});var qr=R(require("@pankod/refine-mui")),Ur=R(require("@pankod/refine-react-hook-form"));var ke=U({type:"edit",additionalScope:[["@pankod/refine-mui","RefineMui",qr],["@pankod/refine-react-hook-form","RefineReactHookForm",Ur]],codeViewerComponent:Z,loadingComponent:K,errorComponent:_,renderer:({resource:e,fields:t,isCustomPage:a,id:o})=>{var w,m;let u=O((w=e.label)!=null?w:e.name,"edit"),i=N((m=e.label)!=null?m:e.name,"Data"),p=[["Edit","@pankod/refine-mui"],["Box","@pankod/refine-mui"],["useForm","@pankod/refine-react-hook-form"]],d=t.filter(s=>(s==null?void 0:s.relation)&&!(s!=null&&s.fieldable)&&(s==null?void 0:s.resource)).filter(Boolean).map(s=>{if((s==null?void 0:s.relation)&&!s.fieldable&&s.resource){p.push(["useAutocomplete","@pankod/refine-mui"]);let A=l(i,s.key,s.accessor,!1);return s.multiple&&s.accessor&&(A=`${l(i,s.key)}?.map((item: any) => ${l("item",void 0,s.accessor)})`),`
                    const { autocompleteProps: ${N(s.key,"AutocompleteProps")} } =
                    useAutocomplete({
                        resource: "${s.resource.name}",
                        defaultValue: ${A},
                    });
                `}}).filter(Boolean),k=s=>{if(s.relation&&s.resource){p.push(["Autocomplete","@pankod/refine-mui"],["Controller","@pankod/refine-react-hook-form"]);let A=N(s.key,"AutocompleteProps"),x=s.relationInfer&&s.relationInfer.accessor?typeof s.relationInfer.accessor=="string"?s.relationInfer.accessor:s.relationInfer.accessor[0]:"title",F=x.includes(".")||x.includes("[")||x.includes("]")||x.includes("-"),n=s.accessor?l("item",void 0,s.accessor,!1):"(item?.id ?? item)",g=s.accessor?l("value",void 0,s.accessor,!1):"(value?.id ?? value)",E=s.accessor?"value":s.multiple?"value?.map((item: any) => item?.id ?? item)":"value?.id ?? value";return $`
                    <Controller
                        control={control}
                        name="${B(s.key,void 0)}"
                        rules={{ required: "This field is required" }}
                        // eslint-disable-next-line
                        ${s.multiple?"defaultValue={[] as any}":"defaultValue={null as any}"} 
                        render={({ field }) => (
                            <Autocomplete
                                {...${A}}
                                {...field}
                                ${s.multiple?"multiple":""}
                                onChange={(_, value) => {
                                    field.onChange(${E});
                                }}
                                getOptionLabel={(item) => {
                                    return (
                                        ${A}?.options?.find(
                                            (p) =>
                                                p?.id?.toString() ===
                                                ${n}?.toString(),
                                        )?.${F?`["${x}"]`:x} ?? ""
                                    );
                                }}
                                isOptionEqualToValue={(option, value) =>
                                    value === undefined ||
                                    option?.id?.toString() === ${g}?.toString()
                                }
                                renderInput={(params) => (
                                    <TextField
                                        {...params}
                                        label="${f(s.key)}"
                                        margin="normal"
                                        variant="outlined"
                                        error={!!${l("(errors as any)",s.key,s.accessor,!1)}}
                                        helperText={${l("(errors as any)",s.key,s.accessor,!1)}?.message}
                                        required
                                    />
                                )}
                            />
                        )}
                    />
                `}},b=s=>{if(s.type==="text"||s.type==="url"||s.type==="email"||s.type==="number"||s.type==="date"||s.type==="richtext"){if(p.push(["TextField","@pankod/refine-mui"]),s.multiple){p.push(["Box","@pankod/refine-mui"]);let A=B(s.key,"${index}",s.accessor),x=l("(errors as any)",s.key,void 0,!1)+"?.[index]";return`
                        <Box sx={{display: "flex", gap: 1}}>
                            {${l(i,s.key)}?.map((item: any, index: number) => (
                                <TextField
                                    key={index}
                                    {...register(\`${A}\`, {
                                        required: "This field is required",
                                    })}
                                    error={!!${x}}
                                    helperText={${x}?.message as string}
                                    margin="normal"
                                    fullWidth
                                    InputLabelProps={{ shrink: true }}
                                    ${s.type!=="date"&&s.type!=="richtext"?`type="${s.type}"`:""}
                                    ${s.type==="richtext"?"multiline":""}
                                    label={\`${f(s.key)} \${index+1}\`}
                                    name={\`${B(s.key,void 0,s.accessor)}.\${index}\`}
                                />
                            ))}
                        </Box>
                    `}return $`
                    <TextField
                        {...register("${B(s.key,void 0,s.accessor)}", {
                            required: "This field is required",
                        })}
                        error={!!${l("(errors as any)",s.key,s.accessor,!1)}}
                        helperText={${l("(errors as any)",s.key,s.accessor,!1)}?.message}
                        margin="normal"
                        fullWidth
                        InputLabelProps={{ shrink: true }}
                        ${s.type!=="date"&&s.type!=="richtext"?`type="${s.type}"`:""}
                        ${s.type==="richtext"?"multiline":""}
                        label="${f(s.key)}"
                        name="${B(s.key,void 0,s.accessor)}"
                        ${X(s.key)?"disabled":""}
                    />
                `}},C=s=>{if(s.type==="boolean"){if(p.push(["Checkbox","@pankod/refine-mui"],["FormControlLabel","@pankod/refine-mui"],["Controller","@pankod/refine-react-hook-form"]),s.multiple){p.push(["Box","@pankod/refine-mui"]);let A=B(s.key,"${index}",s.accessor);return`
                        <Box sx={{display: "flex", gap: 1}}>
                            {${l(i,s.key)}?.map((item: any, index: number) => (
                                <Controller
                                    key={index}
                                    control={control}
                                    name={\`${A}\`}
                                    // eslint-disable-next-line
                                    defaultValue={null as any}
                                    render={({ field }) => (
                                        <FormControlLabel label={\`${f(s.key)} \${index+1}\`} control={
                                            <Checkbox
                                                {...field}
                                                checked={field.value}
                                                onChange={(event) => {
                                                    field.onChange(event.target.checked);
                                                }}
                                            />
                                        } />
                                    )}
                                />
                            ))}
                        </Box>
                    `}return $`
                    <Controller
                        control={control}
                        name="${B(s.key,void 0,s.accessor)}"
                        // eslint-disable-next-line
                        defaultValue={null as any}
                        render={({ field }) => (
                            <FormControlLabel label="${f(s.key)}" control={
                                <Checkbox
                                    {...field}
                                    checked={field.value}
                                    onChange={(event) => {
                                        field.onChange(event.target.checked);
                                    }}
                                />
                            } />
                        )}
                    />
                `}},I=s=>{if(s.type==="date")return`
                    {/* 
                        DatePicker component is not included in "@pankod/refine-mui" package.
                        To use a <DatePicker> component, you can follow the official documentation for Material UI.
                        
                        Docs: https://mui.com/x/react-date-pickers/date-picker/#basic-usage
                    */}
                    ${b(s)}
                    `},c=t.map(s=>{switch(s==null?void 0:s.type){case"text":case"number":case"email":case"url":case"richtext":return b(s);case"date":return I(s);case"boolean":return C(s);case"relation":return k(s);default:return}});return H(p),$`
        ${z(p)}
        
        export const ${u} = () => {
            const {
                saveButtonProps,
                refineCore: { queryResult },
                register,
                control,
                formState: { errors },
            } = useForm(
                ${a?`{ 
                    refineCoreProps: {
                        resource: "${e.name}",
                        id: ${o},
                        action: "edit",
                    }
                }`:""}
            );
        
            const ${i} = queryResult?.data?.data;
        
            ${d}

            return (
                <Edit saveButtonProps={saveButtonProps}>
                    <Box
                        component="form"
                        sx={{ display: "flex", flexDirection: "column" }}
                        autoComplete="off"
                    >
                        ${c.join("")}
                    </Box>
                </Edit>
            );
        };
        `}});var _r=({action:e,id:t,...a})=>{let{useParams:o}=(0,Ir.useRouterContext)(),{action:u,id:i}=o();switch(e!=null?e:u){case"show":return se.default.createElement(fe,{...a,id:t!=null?t:i});case"create":return se.default.createElement(he,{...a,id:t!=null?t:i});case"edit":return se.default.createElement(ke,{...a,id:t!=null?t:i});default:return se.default.createElement(ge,{...a,id:t!=null?t:i})}};
