// @refinedev/cli@2.16.50/dist/index.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  PackageManagerTypes: () => PackageManagerTypes,
  ProjectTypes: () => ProjectTypes,
  UIFrameworks: () => UIFrameworks,
  appendAfterImports: () => appendAfterImports,
  getFileContent: () => getFileContent,
  getImports: () => getImports,
  getNameChangeInImport: () => getNameChangeInImport
});
module.exports = __toCommonJS(src_exports);

// src/definitions/projectTypes.ts
var ProjectTypes = /* @__PURE__ */ ((ProjectTypes2) => {
  ProjectTypes2["REACT_SCRIPT"] = "react-scripts";
  ProjectTypes2["REMIX"] = "remix";
  ProjectTypes2["REMIX_VITE"] = "remix-vite";
  ProjectTypes2["REMIX_SPA"] = "remix-spa";
  ProjectTypes2["NEXTJS"] = "nextjs";
  ProjectTypes2["VITE"] = "vite";
  ProjectTypes2["CRACO"] = "craco";
  ProjectTypes2["PARCEL"] = "parcel";
  ProjectTypes2["UNKNOWN"] = "unknown";
  return ProjectTypes2;
})(ProjectTypes || {});

// src/definitions/uiFrameworks.ts
var UIFrameworks = /* @__PURE__ */ ((UIFrameworks2) => {
  UIFrameworks2["ANTD"] = "antd";
  UIFrameworks2["MUI"] = "mui";
  UIFrameworks2["MANTINE"] = "mantine";
  UIFrameworks2["CHAKRA"] = "chakra-ui";
  return UIFrameworks2;
})(UIFrameworks || {});

// src/definitions/package.ts
var PackageManagerTypes = /* @__PURE__ */ ((PackageManagerTypes2) => {
  PackageManagerTypes2["NPM"] = "npm";
  PackageManagerTypes2["YARN"] = "yarn";
  PackageManagerTypes2["PNPM"] = "pnpm";
  return PackageManagerTypes2;
})(PackageManagerTypes || {});

// src/utils/swizzle/import.ts
var packageRegex = /import(?:\s+(type))?\s*(?:([^\s\{\},]+)\s*(?:,\s*)?)?(\{[^}]+\})?\s*(?:\*\s*as\s+([^\s\{\}]+)\s*)?from\s*['"]([^'"]+)['"];?/g;
var nameChangeRegex = /((?:\w|\s|_)*)( as )((?:\w|\s|_)*)( |,)?/g;
var getImports = (content) => {
  const matches = content.matchAll(packageRegex);
  const imports = [];
  for (const match of matches) {
    const [
      statement,
      typePrefix,
      defaultImport,
      namedImports,
      namespaceImport,
      importPath
    ] = match;
    imports.push({
      isType: typePrefix === "type",
      statement,
      importPath,
      ...defaultImport && { defaultImport },
      ...namedImports && { namedImports },
      ...namespaceImport && { namespaceImport }
    });
  }
  return imports == null ? void 0 : imports.filter(Boolean);
};
var getNameChangeInImport = (namedImportString) => {
  const matches = namedImportString.matchAll(nameChangeRegex);
  const nameChanges = [];
  for (const match of matches) {
    const [statement, fromName, _as, toName, afterCharacter] = match;
    nameChanges.push({
      statement,
      fromName: fromName.trim(),
      toName: toName.trim(),
      afterCharacter
    });
  }
  return nameChanges;
};

// src/utils/swizzle/appendAfterImports.ts
var appendAfterImports = (content, append) => {
  const imports = getImports(content);
  const lastImport = imports[imports.length - 1];
  const lastImportIndex = lastImport ? content.indexOf(lastImport.statement) : content.length - 1;
  return `${content.slice(
    0,
    lastImportIndex + (lastImport == null ? void 0 : lastImport.statement.length)
  )}
${append}
${content.slice(
    lastImportIndex + (lastImport == null ? void 0 : lastImport.statement.length)
  )}`;
};

// src/utils/swizzle/getFileContent.ts
var import_fs_extra = require("fs-extra");
var import_path = require("path");
function getFileContent(path) {
  if (!(this == null ? void 0 : this.absolutePackageDir)) {
    return void 0;
  }
  try {
    return (0, import_fs_extra.readFileSync)((0, import_path.join)(this.absolutePackageDir, path)).toString();
  } catch (err) {
    return void 0;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  PackageManagerTypes,
  ProjectTypes,
  UIFrameworks,
  appendAfterImports,
  getFileContent,
  getImports,
  getNameChangeInImport
});
